---
title:
  '01 Analysis - `r unlist(strsplit(getwd(), "/"))[length(unlist(strsplit(getwd(), "/"))) - 2]`'
author:
  - name: "Emir Turkes [emir.turkes@eturkes.com]"
  - name: "UK Dementia Research Institute at UCL"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(
    inputFile,
    encoding = encoding,
    output_file = file.path(
      "..", "..", "results",
      unlist(strsplit(getwd(), "/"))[length(unlist(strsplit(getwd(), "/")))], "01-analysis.html"
    )
  )})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of the [Tau Vulnerability Project](https://github.com/eturkes/tau-vulnerability).*

The data here is derived from @`r unlist(strsplit(getwd(), "/"))[length(unlist(strsplit(getwd(), "/")))]` and will be referenced using the name ``r unlist(strsplit(getwd(), "/"))[length(unlist(strsplit(getwd(), "/")))]``.

```{r, boilerplate}
# Load in necessary boilerplate and libraries.
# --------------------------------------------

#    This file is part of tau-vulnerability.
#    Copyright (C) 2019-2020  Emir Turkes, UK DRI at UCL, Columbia University Medical Center
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

# These should be checked per document.
# -------------------------------------
options(stringsAsFactors = FALSE)
packages <- c(
  "conflicted", "scales", "RColorBrewer", "Seurat", "dplyr", "ggplot2", "ggrepel", "GSEABase",
  "biomaRt", "Biostrings", "SingleCellExperiment", "SCnorm", "viridis", "scater", "GSVA", "limma",
  "DT", "dendextend", "ComplexHeatmap", "GO.db", "tm", "reshape2", "igraph", "qgraph", "pals",
  "cluster", "future", "edgeR", "circlize", "factoextra", "GEOquery"
)
invisible(suppressPackageStartupMessages(lapply(packages, library, character.only = TRUE)))
conflict_prefer("strsplit", "Biostrings", quiet = TRUE)
conflict_prefer("intersect", "GSEABase", quiet = TRUE)
conflict_prefer("select", "AnnotationDbi", quiet = TRUE)
conflict_prefer("degree", "igraph", quiet = TRUE)
conflict_prefer("strsplit", "Biostrings", quiet = TRUE)
conflict_prefer("cpm", "edgeR", quiet = TRUE)
conflict_prefer("groups", "igraph", quiet = TRUE)
source(file.path("..", "utils.R"))

`%notin%` <- Negate(`%in%`)

analysis_no <- 1
color <- colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100)
gene_set_list <- "GO.BP"
# -------------------------------------

# Everything else in this chunk remains generally unchanged.
# ----------------------------------------------------------
data_name <- unlist(strsplit(getwd(), "/"))[6] # Name of dataset.
assets_dir <- file.path("..", "..", "assets") # Backed up data.

# Unique cache directory for each analysis number.
# ------------------------------------------------
cache_dir <- file.path("..", "..", "cache", data_name, paste0("0", analysis_no))
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}
# ------------------------------------------------

knitr::opts_chunk$set(dpi = 150)
# ----------------------------------------------------------
```

# All Samples

First, we run a standard pipeline on the raw data for QC and normalization.

```{r}
section_no <- 1
gene_set_name <- gene_set_list[section_no]

geo_data <- getGEO("GSE151460", GSEMatrix = TRUE)
filenames <- pData(phenoData(geo_data[[1]]))$supplementary_file_1

for (i in seq_along(filenames)) {
  if (i == 1) {
    gene_mat <- read.delim(
      file.path(
        assets_dir, data_name,
        sub("^(.*)[.].*", "\\1", strsplit(filenames, "/")[[i]][9])
      ),
      header = FALSE
    )
    rownames(gene_mat) <- gene_mat$V1
  } else {
    tmp <- read.delim(
      file.path(
        assets_dir, data_name,
        sub("^(.*)[.].*", "\\1", strsplit(filenames, "/")[[i]][9])
      ),
      header = FALSE
    )
    gene_mat <- cbind(gene_mat, tmp$V2)
  }
}
gene_mat <- gene_mat[ , -1]
colnames(gene_mat) <- pData(phenoData(geo_data[[1]]))$geo_accession

geo_keep <- c(
  grep("ECII_5months_344N1-Rasgrp2", pData(phenoData(geo_data[[1]]))$description),
  grep("CA1_5months_126C5-Sstr4-#7", pData(phenoData(geo_data[[1]]))$description),
  grep("CA3_5months_Gprin3", pData(phenoData(geo_data[[1]]))$description)
)
gene_mat <- gene_mat[ , geo_keep]

meta <- pData(phenoData(geo_data[[1]]))["cell type:ch1"][[1]]
meta <- meta[geo_keep]
meta <- sub("ECII", "EC_L2_Lat_EX_Reln", meta)
meta <- sub("CA1", "HIP_CA1_EX_Wfs1", meta)
meta <- sub("CA3", "HIP_CA3_EX_Prkcd", meta)
meta <- factor(meta, levels = unique(meta))

keep <- filterByExpr(gene_mat, group = meta, min.total.count = 10)
gene_mat <- gene_mat[keep, ]
gene_mat <- DGEList(gene_mat)
gene_mat <- calcNormFactors(gene_mat)
gene_mat <- cpm(gene_mat, log = TRUE, prior.count = 3)

gene_sets_orig <- getGmt(
  file.path(
      "..", "..", "assets", "gene-sets", paste0("mmusculus", ".", gene_set_name, ".ENSG.gmt")
  )
)
rds <- file.path(cache_dir, paste0(gene_set_name, "_gene_sets.rds"))
if (file.exists(rds)) {
  gene_sets <- readRDS(rds)
} else {
  gene_sets <- gene_sets_orig
  gene_sets <- filterGeneSets(gene_sets, 3, 500)
  for (i in seq(length(gene_sets@.Data))) {
    go_id <- gene_sets[[i]]@setName
    suppressWarnings(gene_sets[[i]]@setName <- gene_sets[[i]]@shortDescription)
    suppressWarnings(gene_sets[[i]]@shortDescription <- go_id)
  }
  saveRDS(gene_sets, rds)
}

rds <- file.path(cache_dir, paste0(gene_set_name, "_gene_anno.rds"))
if (file.exists(rds)) {
  gene_anno <- readRDS(rds)
} else {
  mart <- useEnsembl("ensembl", paste0("mmusculus", "_gene_ensembl"))
  attributes <- c("external_gene_name", "ensembl_gene_id", "chromosome_name")
  gene_anno <- getBM(attributes, "ensembl_gene_id", rownames(gene_mat), mart)
  gene_anno <- gene_anno[gene_anno$chromosome_name %in% c(1:22), ]
  gene_anno <- gene_anno[gene_anno$ensembl_gene_id %in% unique(unlist(geneIds(gene_sets))), ]
  saveRDS(gene_anno, rds)
}

keep <- lapply(
  geneIds(gene_sets), function(x, y) na.omit(fastmatch::fmatch(x, y)), gene_anno$ensembl_gene_id
)
keep <- filterGeneSets(keep, 3, 500)
gene_sets <- gene_sets[names(gene_sets) %in% names(keep)]

gene_anno <- gene_anno[gene_anno$ensembl_gene_id %in% unique(unlist(geneIds(gene_sets))), ]
gene_mat <- gene_mat[rownames(gene_mat) %in% gene_anno$ensembl_gene_id, ]

hist(gene_mat, main = "Final pseudobulk logCPM distribution")

# We use a modified overlap coefficient on the gene sets which divides the number of genes in the
# larger set by the number of shared genes with the smaller set.
# We perform this twice, once to sort the sets so that more general terms are towards the top, and
# again to remove sets with identical genes, keeping the more general term.
# ------------------------------------------------------------------------------------------------
rds <- file.path(cache_dir, paste0(gene_set_name, "_gene_set_overlap.rds"))
if (file.exists(rds)) {
  overlap <- readRDS(rds)
} else {
  overlap <- computeGeneSetsOverlapMax(gene_sets, unique(unlist(geneIds(gene_sets))))
  saveRDS(overlap, rds)
}
keep <- rowSums(overlap)
keep <- keep[order(keep, decreasing = TRUE)]
gene_sets_sorted <- gene_sets[match(names(keep), names(gene_sets))]
rds <- file.path(cache_dir, paste0(gene_set_name, "_gene_set_overlap_sorted.rds"))
if (file.exists(rds)) {
  overlap <- readRDS(rds)
} else {
  overlap <- computeGeneSetsOverlapMax(gene_sets_sorted, unique(unlist(geneIds(gene_sets_sorted))))
  saveRDS(overlap, rds)
}

overlap[upper.tri(overlap)] <- 0
diag(overlap) <- 0
keep <- apply(overlap, 1, max)
keep <- keep[keep < 1]
gene_sets <- gene_sets[names(gene_sets) %in% names(keep)]

gene_mat <- gene_mat[rownames(gene_mat) %in% unique(unlist(geneIds(gene_sets))), ]
gene_anno <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(gene_mat), ]

keep <- lapply(
  geneIds(gene_sets), function(x, y) na.omit(fastmatch::fmatch(x, y)), gene_anno$ensembl_gene_id
)
keep <- filterGeneSets(keep, 3, 500)
gene_sets <- gene_sets[names(gene_sets) %in% names(keep)]

gene_mat <- gene_mat[rownames(gene_mat) %in% unique(unlist(geneIds(gene_sets))), ]
gene_anno <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(gene_mat), ]

print("Original gene sets")
gene_sets_orig
print("Filtered gene sets")
gene_sets

pca <- prcomp(t(gene_mat), scale = FALSE)
percent_var <- round(100 * pca$sdev ^ 2 / sum(pca$sdev ^ 2), 1)
gg_df <- data.frame(
  pc1 = pca$x[ , 1],
  pc2 = pca$x[ , 2],
  Description = pData(phenoData(geo_data[[1]]))$description[geo_keep],
  Donor = colnames(gene_mat)
)
gg_df$Donor <- factor(gg_df$Donor)
ggplot(gg_df, aes(pc1, pc2)) +
  geom_point(aes(shape = Description, colour = Donor)) +
  ggtitle("PCA") +
  xlab(paste0("PC1, VarExp: ", percent_var[1], "%")) +
  ylab(paste0("PC2, VarExp: ", percent_var[2], "%")) +
  scale_shape_manual(values = 1:nlevels(gg_df$Donor)) +
  theme(plot.title = element_text(hjust = 0.5))
# ------------------------------------------------------------------------------------------------
```

## DE Gene Sets

ssGSEA from the GSVA package was used to compute a sample x gene set matrix.
We then fit four similar but distinct models using limma to find gene sets that vary with the Braak stage at which each cell type develops pathology.
Importantly, we vary the adjusted p-value cutoff for each model so that the number of DE terms obtained are roughly linear between models, as seen below.
Please collapse the code block for details.

```{r}
# Run GSVA on pseudo-bulk counts.
# -------------------------------
rds <- file.path(cache_dir, paste0(gene_set_name, "_gsva_pseudobulk.rds"))
if (file.exists(rds)) {
  gsva <- readRDS(rds)
} else {
  gsva <- gsva(gene_mat, gene_sets, method = "ssgsea", ssgsea.norm = FALSE)
  saveRDS(gsva, rds)
}
# -------------------------------

# Fit model on pseudo-bulk GSVA results.
# --------------------------------------
design <- model.matrix(~ 0 + meta)
colnames(design) <- unique(meta)
fit <- lmFit(gsva, design)
contrast_mat <- makeContrasts(
  EC_L2_Lat_EX_Reln-HIP_CA3_EX_Prkcd, EC_L2_Lat_EX_Reln-HIP_CA1_EX_Wfs1,
  HIP_CA1_EX_Wfs1-HIP_CA3_EX_Prkcd, levels = design
)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat))
tests <- decideTests(cont_fit, "nestedF", p.value = 1e-5)
tests_up <- tests[tests[ , 1] == 1 & tests[ , 2] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, ]))
if (length(discard) > 0) {
  tests_up <- tests_up[-discard, ]
}
tests_down <- tests[tests[ , 1] == -1 & tests[ , 2] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, ]))
if (length(discard) > 0) {
  tests_down <- tests_down[-discard, ]
}
results <- topTable(cont_fit, number = Inf)

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$F), ]
}
# idx <- ifelse(nrow(all) < 200, nrow(all), 200)
# all <- all[1:idx, ]
# up <- up[rownames(up) %in% rownames(all), ]
# down <- down[rownames(down) %in% rownames(all), ]
gsva_list_pseudobulk <- list(all, up, down)
names(gsva_list_pseudobulk) <- c(
  "All DE", paste0("Upregulated in ", unique(meta)[1]),
  paste0("Downregulated in ", unique(meta)[1])
)
# --------------------------------------

# Fit a strict model on pseudo-bulk GSVA results.
# -----------------------------------------------
design <- model.matrix(~ 0 + meta)
colnames(design) <- unique(meta)
fit <- lmFit(gsva, design)
contrast_mat <- makeContrasts(
  EC_L2_Lat_EX_Reln-HIP_CA3_EX_Prkcd, HIP_CA1_EX_Wfs1-HIP_CA3_EX_Prkcd,
  EC_L2_Lat_EX_Reln-HIP_CA1_EX_Wfs1, levels = design
)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat))
tests <- decideTests(cont_fit, "nestedF", p.value = 1e-5)
tests_up <- tests[tests[ , 1] == 1 & tests[ , 2] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, ]))
if (length(discard) > 0) {
  tests_up <- tests_up[-discard, ]
}
tests_down <- tests[tests[ , 1] == -1 & tests[ , 2] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, ]))
if (length(discard) > 0) {
  tests_down <- tests_down[-discard, ]
}
results <- topTable(cont_fit, number = Inf)

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$F), ]
}
# idx <- ifelse(nrow(all) < 200, nrow(all), 200)
# all <- all[1:idx, ]
# up <- up[rownames(up) %in% rownames(all), ]
# down <- down[rownames(down) %in% rownames(all), ]
gsva_list_pseudobulk_strict <- list(all, up, down)
names(gsva_list_pseudobulk_strict) <- c(
  "All DE", paste0("Upregulated in ", unique(meta)[1]),
  paste0("Downregulated in ", unique(meta)[1])
)
# -----------------------------------------------

# Fit the strictest model on pseudo-bulk GSVA results.
# ----------------------------------------------------
design <- model.matrix(~ 0 + meta)
colnames(design) <- unique(meta)
fit <- lmFit(gsva, design)
contrast_mat <- makeContrasts(
  EC_L2_Lat_EX_Reln-HIP_CA1_EX_Wfs1, HIP_CA1_EX_Wfs1-HIP_CA3_EX_Prkcd, levels = design
)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat))
tests <- decideTests(cont_fit, "nestedF", p.value = 1e-5)
tests_up <- tests[tests[ , 1] == 1 & tests[ , 2] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, ]))
if (length(discard) > 0) {
  tests_up <- tests_up[-discard, ]
}
tests_down <- tests[tests[ , 1] == -1 & tests[ , 2] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, ]))
if (length(discard) > 0) {
  tests_down <- tests_down[-discard, ]
}
results <- topTable(cont_fit, number = Inf)

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$F), ]
}
# idx <- ifelse(nrow(all) < 200, nrow(all), 200)
# all <- all[1:idx, ]
# up <- up[rownames(up) %in% rownames(all), ]
# down <- down[rownames(down) %in% rownames(all), ]
gsva_list_pseudobulk_strictest <- list(all, up, down)
names(gsva_list_pseudobulk_strictest) <- c(
  "All DE", paste0("Upregulated in ", unique(meta)[1]),
  paste0("Downregulated in ", unique(meta)[1])
)
# ----------------------------------------------------

# Fit model on pseudo-bulk GSVA results.
# --------------------------------------
design <- model.matrix(~ 0 + meta)
colnames(design) <- unique(meta)
fit <- lmFit(gsva, design)
contrast_mat <- makeContrasts(
  EC_L2_Lat_EX_Reln-HIP_CA3_EX_Prkcd, EC_L2_Lat_EX_Reln-HIP_CA1_EX_Wfs1,
  HIP_CA1_EX_Wfs1-HIP_CA3_EX_Prkcd, levels = design
)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat))
tests <- decideTests(cont_fit, "nestedF", p.value = 1e-5)
tests_up <- tests[tests[ , 1] == 1 & tests[ , 2] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, ]))
if (length(discard) > 0) {
  tests_up <- tests_up[-discard, ]
}
tests_down <- tests[tests[ , 1] == -1 & tests[ , 2] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, ]))
if (length(discard) > 0) {
  tests_down <- tests_down[-discard, ]
}
results <- topTable(cont_fit, number = Inf)

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$F), ]
}
idx <- ifelse(
  nrow(all) < nrow(gsva_list_pseudobulk_strictest[[1]]), nrow(all),
  nrow(gsva_list_pseudobulk_strictest[[1]])
)
all <- all[1:idx, ]
up <- up[rownames(up) %in% rownames(all), ]
down <- down[rownames(down) %in% rownames(all), ]
gsva_list_pseudobulk <- list(all, up, down)
names(gsva_list_pseudobulk) <- c(
  "All DE", paste0("Upregulated in ", unique(meta)[1]),
  paste0("Downregulated in ", unique(meta)[1])
)
# --------------------------------------

# Fit a strict model on pseudo-bulk GSVA results.
# -----------------------------------------------
design <- model.matrix(~ 0 + meta)
colnames(design) <- unique(meta)
fit <- lmFit(gsva, design)
contrast_mat <- makeContrasts(
  EC_L2_Lat_EX_Reln-HIP_CA3_EX_Prkcd, HIP_CA1_EX_Wfs1-HIP_CA3_EX_Prkcd,
  EC_L2_Lat_EX_Reln-HIP_CA1_EX_Wfs1, levels = design
)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat))
tests <- decideTests(cont_fit, "nestedF", p.value = 1e-5)
tests_up <- tests[tests[ , 1] == 1 & tests[ , 2] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, ]))
if (length(discard) > 0) {
  tests_up <- tests_up[-discard, ]
}
tests_down <- tests[tests[ , 1] == -1 & tests[ , 2] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, ]))
if (length(discard) > 0) {
  tests_down <- tests_down[-discard, ]
}
results <- topTable(cont_fit, number = Inf)

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$F), ]
}
idx <- ifelse(
  nrow(all) < nrow(gsva_list_pseudobulk_strictest[[1]]), nrow(all),
  nrow(gsva_list_pseudobulk_strictest[[1]])
)
all <- all[1:idx, ]
up <- up[rownames(up) %in% rownames(all), ]
down <- down[rownames(down) %in% rownames(all), ]
gsva_list_pseudobulk_strict <- list(all, up, down)
names(gsva_list_pseudobulk_strict) <- c(
  "All DE", paste0("Upregulated in ", unique(meta)[1]),
  paste0("Downregulated in ", unique(meta)[1])
)
# -----------------------------------------------
```

**Upregulated in Direction of `r unique(meta)[1]` `r gene_set_name` Pseudobulk Initial Model**

```{r}
ncol <- ncol(gsva_list_pseudobulk[[2]])
datatable_download_exp(gsva_list_pseudobulk[[2]][(ncol-3):ncol])
```

**Upregulated in Direction of `r unique(meta)[1]` `r gene_set_name` Pseudobulk Strict Model**

```{r}
ncol <- ncol(gsva_list_pseudobulk_strict[[2]])
datatable_download_exp(gsva_list_pseudobulk_strict[[2]][(ncol-3):ncol])
```

**Upregulated in Direction of `r unique(meta)[1]` `r gene_set_name` Pseudobulk Strictest Model**

```{r}
ncol <- ncol(gsva_list_pseudobulk_strictest[[2]])
datatable_download_exp(gsva_list_pseudobulk_strictest[[2]][(ncol-3):ncol])
```

**Downregulated in Direction of `r unique(meta)[1]` `r gene_set_name` Pseudobulk Initial Model**

```{r}
ncol <- ncol(gsva_list_pseudobulk[[3]])
datatable_download_exp(gsva_list_pseudobulk[[3]][(ncol-3):ncol])
```

**Downregulated in Direction of `r unique(meta)[1]` `r gene_set_name` Pseudobulk Strict Model**

```{r}
ncol <- ncol(gsva_list_pseudobulk_strict[[3]])
datatable_download_exp(gsva_list_pseudobulk_strict[[3]][(ncol-3):ncol])
```

**Downregulated in Direction of `r unique(meta)[1]` `r gene_set_name` Pseudobulk Strictest Model**

```{r}
ncol <- ncol(gsva_list_pseudobulk_strictest[[3]])
datatable_download_exp(gsva_list_pseudobulk_strictest[[3]][(ncol-3):ncol])
```

**Number of DE Gene Sets Per Model**

```{r}
data <- data.frame(
  x = c(
    dim(gsva_list_pseudobulk[[1]])[1], dim(gsva_list_pseudobulk_strict[[1]])[1],
    dim(gsva_list_pseudobulk_strictest[[1]])[1]
  ),
  y = c(
    dim(gsva_list_pseudobulk[[1]])[1], dim(gsva_list_pseudobulk_strict[[1]])[1],
    dim(gsva_list_pseudobulk_strictest[[1]])[1]
  )
)
ggplot(data, aes(x, y)) +
  geom_point() +
  geom_smooth(method = lm) +
  geom_label_repel(
    aes(label = c("initial_model", "strict_model", "strictest_model"))
  ) +
  labs(x = "Number of DE gene sets", y = "Number of DE gene sets")
```

## DE Genes

We fit the same models as before, this time on the genes themselves.
Please collapse the code block for details.

```{r}
# Fit model on pseudo-bulk counts.
# --------------------------------
design <- model.matrix(~ 0 + meta)
colnames(design) <- unique(meta)
fit <- lmFit(gene_mat, design)
contrast_mat <- makeContrasts(
  EC_L2_Lat_EX_Reln-HIP_CA3_EX_Prkcd, EC_L2_Lat_EX_Reln-HIP_CA1_EX_Wfs1,
  HIP_CA1_EX_Wfs1-HIP_CA3_EX_Prkcd, levels = design
)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat), robust = TRUE, trend = TRUE)
tests <- decideTests(cont_fit, "nestedF", p.value = 1e-5)
tests_up <- tests[tests[ , 1] == 1 & tests[ , 2] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, ]))
if (length(discard) > 0) {
  tests_up <- tests_up[-discard, ]
}
tests_down <- tests[tests[ , 1] == -1 & tests[ , 2] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, ]))
if (length(discard) > 0) {
  tests_down <- tests_down[-discard, ]
}
results <- topTable(cont_fit, number = Inf)
# --------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$F), ]
}
# idx <- ifelse(nrow(all) < 300, nrow(all), 300)
# all <- all[1:idx, ]
# up <- up[rownames(up) %in% rownames(all), ]
# down <- down[rownames(down) %in% rownames(all), ]
markers_list_pseudobulk <- list(all, up, down)

# Fit a strict model on pseudo-bulk counts.
# -----------------------------------------
design <- model.matrix(~ 0 + meta)
colnames(design) <- unique(meta)
fit <- lmFit(gene_mat, design)
contrast_mat <- makeContrasts(
  EC_L2_Lat_EX_Reln-HIP_CA3_EX_Prkcd, HIP_CA1_EX_Wfs1-HIP_CA3_EX_Prkcd,
  EC_L2_Lat_EX_Reln-HIP_CA1_EX_Wfs1, levels = design
)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat), robust = TRUE, trend = TRUE)
tests <- decideTests(cont_fit, "nestedF", p.value = 1e-5)
tests_up <- tests[tests[ , 1] == 1 & tests[ , 2] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, ]))
if (length(discard) > 0) {
  tests_up <- tests_up[-discard, ]
}
tests_down <- tests[tests[ , 1] == -1 & tests[ , 2] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, ]))
if (length(discard) > 0) {
  tests_down <- tests_down[-discard, ]
}
results <- topTable(cont_fit, number = Inf)
# -----------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$F), ]
}
# idx <- ifelse(nrow(all) < 200, nrow(all), 200)
# all <- all[1:idx, ]
# up <- up[rownames(up) %in% rownames(all), ]
# down <- down[rownames(down) %in% rownames(all), ]
markers_list_pseudobulk_strict <- list(all, up, down)

# Fit the strictest model on pseudo-bulk counts.
# ----------------------------------------------
design <- model.matrix(~ 0 + meta)
colnames(design) <- unique(meta)
fit <- lmFit(gene_mat, design)
contrast_mat <- makeContrasts(
  EC_L2_Lat_EX_Reln-HIP_CA1_EX_Wfs1, HIP_CA1_EX_Wfs1-HIP_CA3_EX_Prkcd, levels = design
)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat), robust = TRUE, trend = TRUE)
tests <- decideTests(cont_fit, "nestedF", p.value = 1e-5)
tests_up <- tests[tests[ , 1] == 1 & tests[ , 2] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, ]))
if (length(discard) > 0) {
  tests_up <- tests_up[-discard, ]
}
tests_down <- tests[tests[ , 1] == -1 & tests[ , 2] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, ]))
if (length(discard) > 0) {
  tests_down <- tests_down[-discard, ]
}
results <- topTable(cont_fit, number = Inf)
# ----------------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$F), ]
}
# idx <- ifelse(nrow(all) < 100, nrow(all), 100)
# all <- all[1:idx, ]
# up <- up[rownames(up) %in% rownames(all), ]
# down <- down[rownames(down) %in% rownames(all), ]
markers_list_pseudobulk_strictest <- list(all, up, down)

# Fit model on pseudo-bulk counts.
# --------------------------------
design <- model.matrix(~ 0 + meta)
colnames(design) <- unique(meta)
fit <- lmFit(gene_mat, design)
contrast_mat <- makeContrasts(
  EC_L2_Lat_EX_Reln-HIP_CA3_EX_Prkcd, EC_L2_Lat_EX_Reln-HIP_CA1_EX_Wfs1,
  HIP_CA1_EX_Wfs1-HIP_CA3_EX_Prkcd, levels = design
)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat), robust = TRUE, trend = TRUE)
tests <- decideTests(cont_fit, "nestedF", p.value = 1e-5)
tests_up <- tests[tests[ , 1] == 1 & tests[ , 2] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, ]))
if (length(discard) > 0) {
  tests_up <- tests_up[-discard, ]
}
tests_down <- tests[tests[ , 1] == -1 & tests[ , 2] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, ]))
if (length(discard) > 0) {
  tests_down <- tests_down[-discard, ]
}
results <- topTable(cont_fit, number = Inf)
# --------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$F), ]
}
idx <- ifelse(
  nrow(all) < nrow(markers_list_pseudobulk_strictest[[1]]), nrow(all),
  nrow(markers_list_pseudobulk_strictest[[1]])
)
all <- all[1:idx, ]
up <- up[rownames(up) %in% rownames(all), ]
down <- down[rownames(down) %in% rownames(all), ]
markers_list_pseudobulk <- list(all, up, down)

# Fit a strict model on pseudo-bulk counts.
# -----------------------------------------
design <- model.matrix(~ 0 + meta)
colnames(design) <- unique(meta)
fit <- lmFit(gene_mat, design)
contrast_mat <- makeContrasts(
  EC_L2_Lat_EX_Reln-HIP_CA3_EX_Prkcd, HIP_CA1_EX_Wfs1-HIP_CA3_EX_Prkcd,
  EC_L2_Lat_EX_Reln-HIP_CA1_EX_Wfs1, levels = design
)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat), robust = TRUE, trend = TRUE)
tests <- decideTests(cont_fit, "nestedF", p.value = 1e-5)
tests_up <- tests[tests[ , 1] == 1 & tests[ , 2] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, ]))
if (length(discard) > 0) {
  tests_up <- tests_up[-discard, ]
}
tests_down <- tests[tests[ , 1] == -1 & tests[ , 2] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, ]))
if (length(discard) > 0) {
  tests_down <- tests_down[-discard, ]
}
results <- topTable(cont_fit, number = Inf)
# -----------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$F), ]
}
idx <- ifelse(
  nrow(all) < nrow(markers_list_pseudobulk_strictest[[1]]), nrow(all),
  nrow(markers_list_pseudobulk_strictest[[1]])
)
all <- all[1:idx, ]
up <- up[rownames(up) %in% rownames(all), ]
down <- down[rownames(down) %in% rownames(all), ]
markers_list_pseudobulk_strict <- list(all, up, down)

# Convert ENSEMBL IDs back to gene symbols.
# -----------------------------------------
genes <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(markers_list_pseudobulk[[2]]), ]
genes <- genes[order(match(genes$ensembl_gene_id, rownames(markers_list_pseudobulk[[2]]))), ]
up <- markers_list_pseudobulk[[2]]
rownames(up) <- genes$external_gene_name

genes <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(markers_list_pseudobulk[[3]]), ]
genes <- genes[order(match(genes$ensembl_gene_id, rownames(markers_list_pseudobulk[[3]]))), ]
down <- markers_list_pseudobulk[[3]]
rownames(down) <- genes$external_gene_name

all <- rbind(up, down)
all <- all[order(-all$F), ]
markers_list_symbols_pseudobulk <- list(all, up, down)

genes <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(markers_list_pseudobulk_strict[[2]]), ]
genes <- genes[order(match(genes$ensembl_gene_id, rownames(markers_list_pseudobulk_strict[[2]]))), ]
up <- markers_list_pseudobulk_strict[[2]]
rownames(up) <- genes$external_gene_name

genes <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(markers_list_pseudobulk_strict[[3]]), ]
genes <- genes[order(match(genes$ensembl_gene_id, rownames(markers_list_pseudobulk_strict[[3]]))), ]
down <- markers_list_pseudobulk_strict[[3]]
rownames(down) <- genes$external_gene_name

all <- rbind(up, down)
all <- all[order(-all$F), ]
markers_list_symbols_pseudobulk_strict <- list(all, up, down)

genes <- gene_anno[
  gene_anno$ensembl_gene_id %in% rownames(markers_list_pseudobulk_strictest[[2]]),
]
genes <- genes[
  order(match(genes$ensembl_gene_id, rownames(markers_list_pseudobulk_strictest[[2]]))),
]
up <- markers_list_pseudobulk_strictest[[2]]
rownames(up) <- genes$external_gene_name

genes <- gene_anno[
  gene_anno$ensembl_gene_id %in% rownames(markers_list_pseudobulk_strictest[[3]]),
]
genes <- genes[
  order(match(genes$ensembl_gene_id, rownames(markers_list_pseudobulk_strictest[[3]]))),
]
down <- markers_list_pseudobulk_strictest[[3]]
rownames(down) <- genes$external_gene_name

all <- rbind(up, down)
all <- all[order(-all$F), ]
markers_list_symbols_pseudobulk_strictest <- list(all, up, down)
# -----------------------------------------

# Create new Seurat object with gene symbols.
# -------------------------------------------
mat_symbols <- gene_mat
rownames(mat_symbols) <- gene_anno$external_gene_name
# -------------------------------------------
```

**Upregulated in Direction of `r unique(meta)[1]` `r gene_set_name` Pseudobulk Initial Model**

```{r}
ncol <- ncol(markers_list_symbols_pseudobulk[[2]])
datatable_download_exp(markers_list_symbols_pseudobulk[[2]][(ncol-3):ncol])
```

**Upregulated in Direction of `r unique(meta)[1]` `r gene_set_name` Pseudobulk Strict Model**

```{r}
ncol <- ncol(markers_list_symbols_pseudobulk_strict[[2]])
datatable_download_exp(markers_list_symbols_pseudobulk_strict[[2]][(ncol-3):ncol])
```

**Upregulated in Direction of `r unique(meta)[1]` `r gene_set_name` Pseudobulk Strictest Model**

```{r}
ncol <- ncol(markers_list_symbols_pseudobulk_strictest[[2]])
datatable_download_exp(markers_list_symbols_pseudobulk_strictest[[2]][(ncol-3):ncol])
```

**Downregulated in Direction of `r unique(meta)[1]` `r gene_set_name` Pseudobulk Initial Model**

```{r}
ncol <- ncol(markers_list_symbols_pseudobulk[[3]])
datatable_download_exp(markers_list_symbols_pseudobulk[[3]][(ncol-3):ncol])
```

**Downregulated in Direction of `r unique(meta)[1]` `r gene_set_name` Pseudobulk Strict Model**

```{r}
ncol <- ncol(markers_list_symbols_pseudobulk_strict[[3]])
datatable_download_exp(markers_list_symbols_pseudobulk_strict[[3]][(ncol-3):ncol])
```

**Downregulated in Direction of `r unique(meta)[1]` `r gene_set_name` Pseudobulk Strictest Model**

```{r}
ncol <- ncol(markers_list_symbols_pseudobulk_strictest[[3]])
datatable_download_exp(markers_list_symbols_pseudobulk_strictest[[3]][(ncol-3):ncol])
```

**Number of DE Genes Per Model**

```{r}
data <- data.frame(
  x = c(
    dim(markers_list_pseudobulk[[1]])[1], dim(markers_list_pseudobulk_strict[[1]])[1],
    dim(markers_list_pseudobulk_strictest[[1]])[1]
  ),
  y = c(
    dim(markers_list_pseudobulk[[1]])[1], dim(markers_list_pseudobulk_strict[[1]])[1],
    dim(markers_list_pseudobulk_strictest[[1]])[1]
  )
)
ggplot(data, aes(x, y)) +
  geom_point() +
  geom_smooth(method = lm) +
  geom_label_repel(
    aes(label = c("initial_model", "strict_model", "strictest_model"))
  ) +
  labs(x = "Number of DE genes", y = "Number of DE genes")
```

```{r, fig.height = 40, fig.width = 20}
# TODO: Comments and refactoring.
# -------------------------------

anno_color <- hue_pal()(length(unique(colnames(gene_mat))))
color3 <- colorRamp2(
  c(min(gsva), (max(gsva) + min(gsva)) / 2, max(gsva)),
  c(brewer.pal(3, "RdBu")[3], brewer.pal(3, "RdBu")[2], brewer.pal(3, "RdBu")[1])
)

gsva_list_pseudobulk_all <- list(
  c(
    rownames(gsva_list_pseudobulk[[1]]), rownames(gsva_list_pseudobulk_strict[[1]]),
    rownames(gsva_list_pseudobulk_strictest[[1]])
  ),
  c(
    rownames(gsva_list_pseudobulk[[2]]), rownames(gsva_list_pseudobulk_strict[[2]]),
    rownames(gsva_list_pseudobulk_strictest[[2]])
  ),
  c(
    rownames(gsva_list_pseudobulk[[3]]), rownames(gsva_list_pseudobulk_strict[[3]]),
    rownames(gsva_list_pseudobulk_strictest[[3]])
  )
)

markers_list_pseudobulk_all <- list(
  c(
    rownames(markers_list_pseudobulk[[1]]),
    rownames(markers_list_pseudobulk_strict[[1]]),
    rownames(markers_list_pseudobulk_strictest[[1]])
  ),
  c(
    rownames(markers_list_pseudobulk[[2]]),
    rownames(markers_list_pseudobulk_strict[[2]]),
    rownames(markers_list_pseudobulk_strictest[[2]])
  ),
  c(
    rownames(markers_list_pseudobulk[[3]]),
    rownames(markers_list_pseudobulk_strict[[3]]),
    rownames(markers_list_pseudobulk_strictest[[3]])
  )
)
markers_list_symbols_pseudobulk_all <- list(
  c(
    rownames(markers_list_symbols_pseudobulk[[1]]),
    rownames(markers_list_symbols_pseudobulk_strict[[1]]),
    rownames(markers_list_symbols_pseudobulk_strictest[[1]])
  ),
  c(
    rownames(markers_list_symbols_pseudobulk[[2]]),
    rownames(markers_list_symbols_pseudobulk_strict[[2]]),
    rownames(markers_list_symbols_pseudobulk_strictest[[2]])
  ),
  c(
    rownames(markers_list_symbols_pseudobulk[[3]]),
    rownames(markers_list_symbols_pseudobulk_strict[[3]]),
    rownames(markers_list_symbols_pseudobulk_strictest[[3]])
  )
)

mat <- gsva
mat <- mat[rownames(mat) %in% gsva_list_pseudobulk_all[[1]], ]
mat_up <- mat[rownames(mat) %in% gsva_list_pseudobulk_all[[2]], ]
mat_down <- mat[rownames(mat) %in% gsva_list_pseudobulk_all[[3]], ]
gene_sets_sub_up <- gene_sets[names(gene_sets) %in% rownames(mat_up)]
up_vector <- vector()
for (j in seq(length(gene_sets_sub_up@.Data))) {
  if (any(gene_sets_sub_up[[j]]@geneIds %in% markers_list_pseudobulk_all[[2]])) {
    up_vector <- append(up_vector, names(gene_sets_sub_up)[j])
  }
}
gene_sets_sub_down <- gene_sets[names(gene_sets) %in% rownames(mat_down)]
down_vector <- c()
for (j in seq(length(gene_sets_sub_down@.Data))) {
  if (any(gene_sets_sub_down[[j]]@geneIds %in% markers_list_pseudobulk_all[[3]])) {
    down_vector <- append(down_vector, names(gene_sets_sub_down)[j])
  }
}
all_vector <- append(up_vector, down_vector)
mat <- mat[rownames(mat) %in% all_vector, ]
gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(mat)]
gene_sets_sub_ids <- gene_sets_sub
for (j in seq(length(gene_sets_sub_ids@.Data))) {
  suppressWarnings(gene_sets_sub_ids[[j]]@setName <- gene_sets_sub_ids[[j]]@shortDescription)
}

discard <- computeGeneSetsOverlapMax(
  gene_sets_sub_ids, unique(unlist(geneIds(gene_sets_sub_ids)))
)
discard[!upper.tri(discard)] <- 0
discard <- colnames(discard)[colSums(discard) == 0]
discard <- which(names(gene_sets_sub_ids) %in% discard)
mat <- mat[-discard, ]
gene_sets_sub <- gene_sets_sub[-discard, ]
gene_sets_sub_ids <- gene_sets_sub_ids[-discard, ]

gene_set_overlap <- computeGeneSetsOverlapMax(
  gene_sets_sub_ids, unique(unlist(geneIds(gene_sets_sub_ids)))
)
clustering <- hclust(get_dist(gene_set_overlap, "spearman"), "ward.D2")

anno <- HeatmapAnnotation(
  donor = colnames(gene_mat),
  cell_type = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = levels(meta),
    labels_gp = gpar(col = "black", fontsize = 20)
  ),
  col = list(
    donor = c(
      "GSM4578978" = anno_color[1], "GSM4578979" = anno_color[2], "GSM4578980" = anno_color[3],
      "GSM4578866" = anno_color[4], "GSM4578867" = anno_color[5], "GSM4578868" = anno_color[6],
      "GSM4578872" = anno_color[7], "GSM4578873" = anno_color[8], "GSM4578874" = anno_color[9]
    )
  ),
  annotation_legend_param = list(donor = list(nrow = 1))
)

modules_up <- rownames(mat)[rownames(mat) %in% gsva_list_pseudobulk_all[[2]]]
modules_down <- rownames(mat)[rownames(mat) %in% gsva_list_pseudobulk_all[[3]]]

module_gene_sets_sub <- gene_sets[names(gene_sets) %in% modules_up]
frequent_genes_tmp <- as.data.frame(table(unlist(geneIds(module_gene_sets_sub))))
frequent_genes_tmp <- frequent_genes_tmp[
  frequent_genes_tmp$Var1 %in% gene_anno$ensembl_gene_id,
]
rownames(frequent_genes_tmp) <- NULL
frequent_genes_up <- frequent_genes_tmp

module_gene_sets_sub <- gene_sets[names(gene_sets) %in% modules_down]
frequent_genes_tmp <- as.data.frame(table(unlist(geneIds(module_gene_sets_sub))))
frequent_genes_tmp <- frequent_genes_tmp[
  frequent_genes_tmp$Var1 %in% gene_anno$ensembl_gene_id,
]
rownames(frequent_genes_tmp) <- NULL
frequent_genes_down <- frequent_genes_tmp
frequent_genes <- rbind(frequent_genes_up, frequent_genes_down)

node_data <- melt(geneIds(gene_sets_sub))
node_data <- node_data[node_data$value %in% frequent_genes$Var1, ]
names(node_data)[1] <- "ensembl_gene_id"
names(node_data)[2] <- "gene_set"
node_data$direction <- ifelse(
  node_data$gene_set %in% gsva_list_pseudobulk_all[[2]], "up", "down"
)

unique_direction <- as.data.frame.matrix(table(node_data[ , c(1, 3)]), )
unique_direction$down <- unique_direction$down * -1
module_membership_norm <- unique_direction
module_membership_norm <- module_membership_norm[
  abs(rowSums(module_membership_norm)) >= 2,
]

hub_gene_all_GO <- table(unlist(geneIds(gene_sets)))
hub_gene_all_GO <- hub_gene_all_GO[
  names(hub_gene_all_GO) %in% rownames(module_membership_norm)
]

module_membership_upper_quart <- median(
  (abs(rowSums(module_membership_norm)) / hub_gene_all_GO) *
    abs(rowSums(module_membership_norm))
)

modules_with_hubs <- vector("list", 26)
modules_with_hubs[[1]] <- 0
hub_gene_list <- list(gene_anno)
for (l in 2:26) {
  set.seed(1)
  heatmap <- Heatmap(
    mat, cluster_rows = clustering, cluster_columns = FALSE, show_column_names = FALSE,
    show_row_names = FALSE, show_heatmap_legend = FALSE, row_split = l,
  )
  modules <- suppressWarnings(row_order(heatmap))
  hub_gene_list <- vector("list", length(modules))
  for (m in seq_along(modules)) {
    module_sets <- rownames(mat)[modules[[m]]]
    modules_up <- module_sets[module_sets %in% gsva_list_pseudobulk_all[[2]]]
    modules_down <- module_sets[module_sets %in% gsva_list_pseudobulk_all[[3]]]
    module_hub_genes <- vector()

    if (length(modules_up) == 0) {
    } else {
      module_gene_sets_sub <- gene_sets[names(gene_sets) %in% modules_up]
      frequent_genes <- as.data.frame(table(unlist(geneIds(module_gene_sets_sub))))
      hub_gene_all_GO_sub <- hub_gene_all_GO[
        names(hub_gene_all_GO) %in% frequent_genes$Var1
      ]
      frequent_genes <- frequent_genes[frequent_genes$Var1 %in% names(hub_gene_all_GO), ]
      frequent_genes$Freq <-
        (frequent_genes$Freq / hub_gene_all_GO_sub) * frequent_genes$Freq
      frequent_genes <- frequent_genes[
        frequent_genes$Freq >= module_membership_upper_quart,
      ]
      module_hub_genes_tmp <- gene_anno[
        gene_anno$ensembl_gene_id %in% frequent_genes$Var1,
      ]
      module_hub_genes_tmp <- module_hub_genes_tmp[
        module_hub_genes_tmp$ensembl_gene_id %in% markers_list_pseudobulk_all[[2]],
      ]
      module_hub_genes <- module_hub_genes_tmp$ensembl_gene_id
    }

    if (length(modules_down) == 0) {
    } else {
      module_gene_sets_sub <- gene_sets[names(gene_sets) %in% modules_down]
      frequent_genes <- as.data.frame(table(unlist(geneIds(module_gene_sets_sub))))
      hub_gene_all_GO_sub <- hub_gene_all_GO[
        names(hub_gene_all_GO) %in% frequent_genes$Var1
      ]
      frequent_genes <- frequent_genes[frequent_genes$Var1 %in% names(hub_gene_all_GO), ]
      frequent_genes$Freq <-
        (frequent_genes$Freq / hub_gene_all_GO_sub) * frequent_genes$Freq
      frequent_genes <- frequent_genes[
        frequent_genes$Freq >= module_membership_upper_quart,
      ]
      module_hub_genes_tmp <- gene_anno[
        gene_anno$ensembl_gene_id %in% frequent_genes$Var1,
      ]
      module_hub_genes_tmp <- module_hub_genes_tmp[
        module_hub_genes_tmp$ensembl_gene_id %in% markers_list_pseudobulk_all[[3]],
      ]
      module_hub_genes <- append(module_hub_genes, module_hub_genes_tmp$ensembl_gene_id)
    }

    hub_gene_list[[m]] <- module_hub_genes
  }
  modules_with_hubs[[l]] <- length(which(lengths(hub_gene_list) != 0)) / l
}
optimal_k <- 27 - which.max(rev(modules_with_hubs))
module_colors <- cols25(optimal_k)
clustering <- color_branches(clustering, optimal_k, col = module_colors)
set.seed(1)
heatmap <- Heatmap(
  mat, cluster_rows = clustering, cluster_columns = FALSE, show_column_names = FALSE,
  show_row_names = FALSE, show_heatmap_legend = FALSE, row_split = optimal_k
)

modules <- suppressWarnings(row_order(heatmap))
module_names <- modules
go_anno <- select(GO.db, names(gene_sets_sub_ids), c("GOID", "TERM", "DEFINITION"), "GOID")
for (l in seq_along(module_names)) {
  module_names[[l]] <- paste0(
    rownames(mat)[module_names[[l]]], " ", go_anno$DEFINITION[module_names[[l]]]
  )
}
top_words_tmp <- data.frame(
  gene_sets = unlist(module_names),
  module = rep(seq_along(module_names), lapply(module_names, length))
)
top_words_all <- suppressWarnings(
  as.character(
    tapply(top_words_tmp$gene_sets, top_words_tmp$module, word_cloud)
  )
)
top_words <- suppressWarnings(
  as.character(
    tapply(top_words_tmp$gene_sets, top_words_tmp$module, word_cloud, width = 5)
  )
)
row_anno <-  HeatmapAnnotation(
  top_words = anno_block(
    gp = gpar(col = module_colors), labels = seq(optimal_k),
    labels_rot = 0, labels_gp = gpar(fontsize = 20)
  ),
  which = "row"
)

row_colors <- rownames(mat)
row_colors <- data.frame(
  row_colors = row_colors,
  DE_strict = row_colors %in% rownames(gsva_list_pseudobulk_strict[[1]]),
  DE_strictest = row_colors %in% rownames(gsva_list_pseudobulk_strictest[[1]])
) %>%
  mutate(
    type = case_when(
      DE_strictest == TRUE ~ "firebrick",
      DE_strict == TRUE ~ "darkgoldenrod",
      TRUE ~ "black"
    )
  )
row_colors <- row_colors$type

# idx <- which.max(nchar(rownames(mat)))
# rownames(mat)[idx] <-
#   "oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen..."

draw(
  Heatmap(
    t(
      apply(
        mat, 1,
        function (x) (x - min(x)) / (max(x) - min(x))
      )
    ),
    color,
    cluster_rows = clustering,
    cluster_columns = FALSE,
    row_names_max_width = max_text_width(rownames(mat), gpar(fontsize = 10)),
    show_column_names = FALSE,
    column_split = meta,
    top_annotation = anno,
    rect_gp = gpar(col = "lightgray"),
    column_title = " ",
    row_split = optimal_k,
    row_title = top_words,
    row_title_rot = 0,
    row_title_gp = gpar(fontsize = 10),
    row_names_gp = gpar(fontsize = 10, col = unlist(row_colors)),
    left_annotation = row_anno,
    row_gap = unit(2, "mm"),
    column_gap = unit(2, "mm"),
    heatmap_legend_param = list(
      title = "scaled ssGSEA score", direction = "horizontal", legend_width = unit(7.5, "cm")
    )
  ),
  heatmap_legend_side = "top",
  annotation_legend_side = "top"
)

hub_gene_data <- vector()
for (l in seq_along(modules)) {
  module_sets <- rownames(mat)[modules[[l]]]
  modules_up <- module_sets[module_sets %in% gsva_list_pseudobulk_all[[2]]]
  modules_down <- module_sets[module_sets %in% gsva_list_pseudobulk_all[[3]]]
  frequent_genes <- vector()
  frequent_genes_up <- vector()
  frequent_genes_down <- vector()

  if (length(modules_up) == 0) {
  } else {
    module_gene_sets_sub <- gene_sets[names(gene_sets) %in% modules_up]
    frequent_genes_tmp <- as.data.frame(table(unlist(geneIds(module_gene_sets_sub))))
    frequent_genes_tmp <- frequent_genes_tmp[
      frequent_genes_tmp$Var1 %in% gene_anno$ensembl_gene_id,
    ]
    rownames(frequent_genes_tmp) <- NULL
    if (any(frequent_genes_tmp$Var1 %in% markers_list_pseudobulk_all[[1]])) {
      discard <- which(
        frequent_genes_tmp$Var1 %in% markers_list_pseudobulk_all[[1]] &
          frequent_genes_tmp$Var1 %in% markers_list_pseudobulk_all[[3]]
      )
      if (length(discard) > 0) {
        frequent_genes_tmp <- frequent_genes_tmp[-discard, ]
      }
    }
    frequent_genes_up <- frequent_genes_tmp
  }

  if (length(modules_down) == 0) {
  } else {
    module_gene_sets_sub <- gene_sets[names(gene_sets) %in% modules_down]
    frequent_genes_tmp <- as.data.frame(table(unlist(geneIds(module_gene_sets_sub))))
    frequent_genes_tmp <- frequent_genes_tmp[
      frequent_genes_tmp$Var1 %in% gene_anno$ensembl_gene_id,
    ]
    rownames(frequent_genes_tmp) <- NULL
    if (any(frequent_genes_tmp$Var1 %in% markers_list_pseudobulk_all[[1]])) {
      discard <- which(
        frequent_genes_tmp$Var1 %in% markers_list_pseudobulk_all[[1]] &
          frequent_genes_tmp$Var1 %in% markers_list_pseudobulk_all[[2]]
      )
      if (length(discard) > 0) {
        frequent_genes_tmp <- frequent_genes_tmp[-discard, ]
      }
    }
    frequent_genes_down <- frequent_genes_tmp
  }

  frequent_genes <- rbind(frequent_genes_up, frequent_genes_down)
  if (dim(frequent_genes)[1] != 0) {
    module_gene_sets_sub <- gene_sets[names(gene_sets) %in% module_sets]
    node_data <- melt(geneIds(module_gene_sets_sub))
    node_data <- node_data[node_data$value %in% frequent_genes$Var1, ]
    names(node_data)[1] <- "ensembl_gene_id"
    names(node_data)[2] <- "gene_set"
    gene_anno_sub <- gene_anno[
      gene_anno$ensembl_gene_id %in% node_data$ensembl_gene_id, -3
    ]
    node_data <- node_data[node_data$ensembl_gene_id %in% gene_anno_sub$ensembl_gene_id, ]
    node_data <- left_join(node_data, gene_anno_sub, by = "ensembl_gene_id")
    node_data$direction <- ifelse(
      node_data$gene_set %in% gsva_list_pseudobulk_all[[2]], "up", "down"
    )
    node_data$DE <- ifelse(
      node_data$ensembl_gene_id %in% markers_list_pseudobulk_all[[1]], "yes", "no"
    )
    node_data$module <- l
    hub_gene_data <- rbind(hub_gene_data, node_data)
  }
}

unique_direction <- as.data.frame.matrix(table(hub_gene_data[ , c(3, 4)]), )
unique_direction$down <- unique_direction$down * -1
discard <- which(rowSums(unique_direction) == 0)
if (length(discard) > 0) {
  unique_direction <- unique_direction[-discard, ]
}
module_membership_norm <- unique_direction
unique_direction <- ifelse(rowSums(unique_direction) > 0, "up", "down")
hub_gene_data <- hub_gene_data[
  hub_gene_data$external_gene_name %in% names(unique_direction),
]

rownames(hub_gene_data) <- NULL
discard <- vector()
for (i_gene in seq_along(hub_gene_data$external_gene_name)) {
  sub_gene <- hub_gene_data$external_gene_name[i_gene]
  sub_unique_direction <- unique_direction[which(names(unique_direction) == sub_gene)]
  if (hub_gene_data$direction[i_gene] != sub_unique_direction[[1]]) {
    discard <- append(discard, i_gene)
  }
}
hub_gene_data <- hub_gene_data[-discard, ]

frequent_genes_tmp <- as.data.frame(table(hub_gene_data$external_gene_name))
frequent_genes <- frequent_genes_tmp
hub_gene_data <- hub_gene_data[
  hub_gene_data$external_gene_name %in% frequent_genes$Var1,
]

hub_gene_data_up <- hub_gene_data[which(hub_gene_data$direction == "up"), ]
hub_gene_data_down <- hub_gene_data[which(hub_gene_data$direction == "down"), ]

hub_mat <- as.data.frame.matrix(table(hub_gene_data_up[ , c(3, 2)]), )
hub_mat <- hub_mat == 1
co_mat <- hub_mat %*% t(hub_mat)
diag(co_mat) <- 0
graph_up <- graph_from_adjacency_matrix(co_mat, "upper", TRUE)

edges_up <- get.edgelist(graph_up, FALSE)
set.seed(1)
layout_up <- qgraph.layout.fruchtermanreingold(
  edges_up, vcount = vcount(graph_up), E(graph_up)$weight,
  area = vcount(graph_up) ^ 2, repulse.rad = vcount(graph_up) ^ 2
)

hub_gene_all_GO <- table(unlist(geneIds(gene_sets)))
hub_gene_all_GO <- hub_gene_all_GO[
  names(hub_gene_all_GO) %in% unique(hub_gene_data_up$ensembl_gene_id)
]
sub_gene_anno <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(hub_gene_all_GO), ]
rownames(sub_gene_anno) <- NULL
sub_gene_anno <- sub_gene_anno[
  match(rownames(hub_gene_all_GO), sub_gene_anno$ensembl_gene_id),
]
rownames(hub_gene_all_GO) <- sub_gene_anno$external_gene_name
hub_gene_all_GO_up <- hub_gene_all_GO[order(names(hub_gene_all_GO))]

module_membership_up <- as.data.frame.matrix(table(hub_gene_data_up[ , c(3, 6)]), )
module_membership_binary_up <- module_membership_up
module_membership_binary_up[module_membership_binary_up > 0] <- 1
module_membership_list_up <- lapply(
  split(module_membership_up, row.names(module_membership_up)), unlist
)
module_network_colors_up <- module_colors[unique(hub_gene_data_up$module)]
V(graph_up)$pie.color = list(module_network_colors_up)

module_membership_norm_up <- module_membership_norm[
  rownames(module_membership_norm) %in% rownames(module_membership_up),
]

V(graph_up)$vertex.pie = list(module_network_colors_up)
V(graph_up)$vertex.shape <- "pie"
V(graph_up)$vertex.color <- module_membership_list_up
V(graph_up)$vertex.label.family = "sans"

DE_up <- vector()
for (l in seq(length(V(graph_up)))) {
  if (V(graph_up)[l]$name %in% rownames(markers_list_symbols_pseudobulk[[2]])) {
    DE_up <- append(DE_up, "yes")
  } else {
    DE_up <- append(DE_up, "no")
  }
}

DE_opp_up <- vector()
for (l in seq(length(V(graph_up)))) {
  if (V(graph_up)[l]$name %in% markers_list_symbols_pseudobulk_all[[3]]) {
    DE_opp_up <- append(DE_opp_up, "yes")
  } else {
    DE_opp_up <- append(DE_opp_up, "no")
  }
}

DE_strict_up <- vector()
for (l in seq(length(V(graph_up)))) {
  if (V(graph_up)[l]$name %in% rownames(markers_list_symbols_pseudobulk_strict[[2]])) {
    DE_strict_up <- append(DE_strict_up, "yes")
  } else {
    DE_strict_up <- append(DE_strict_up, "no")
  }
}

DE_strictest_up <- vector()
for (l in seq(length(V(graph_up)))) {
  if (V(graph_up)[l]$name %in% rownames(markers_list_symbols_pseudobulk_strictest[[2]])) {
    DE_strictest_up <- append(DE_strictest_up, "yes")
  } else {
    DE_strictest_up <- append(DE_strictest_up, "no")
  }
}

direction_up <- vector()
for (l in seq(length(E(graph_up)))) {
  hub_gene_sub_up <- hub_gene_data_up[
    hub_gene_data_up$external_gene_name %in% head_of(graph_up, l)[[1]]$name,
    ]
  direction_up <- append(direction_up, unique(hub_gene_sub_up$direction))
}
E(graph_up)$edge.color <- ifelse(direction_up == "up", "firebrick1", "dodgerblue")

hub_mat <- as.data.frame.matrix(table(hub_gene_data_down[ , c(3, 2)]), )
hub_mat <- hub_mat == 1
co_mat <- hub_mat %*% t(hub_mat)
diag(co_mat) <- 0
graph_down <- graph_from_adjacency_matrix(co_mat, "upper", TRUE)

edges_down <- get.edgelist(graph_down, FALSE)
set.seed(1)
layout_down <- qgraph.layout.fruchtermanreingold(
  edges_down, vcount = vcount(graph_down), E(graph_down)$weight,
  area = vcount(graph_down) ^ 2, repulse.rad = vcount(graph_down) ^ 2
)

hub_gene_all_GO <- table(unlist(geneIds(gene_sets)))
hub_gene_all_GO <- hub_gene_all_GO[
  names(hub_gene_all_GO) %in% unique(hub_gene_data_down$ensembl_gene_id)
]
sub_gene_anno <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(hub_gene_all_GO), ]
rownames(sub_gene_anno) <- NULL
sub_gene_anno <- sub_gene_anno[
  match(rownames(hub_gene_all_GO), sub_gene_anno$ensembl_gene_id),
]
rownames(hub_gene_all_GO) <- sub_gene_anno$external_gene_name
hub_gene_all_GO_down <- hub_gene_all_GO[order(names(hub_gene_all_GO))]

module_membership_down <- as.data.frame.matrix(table(hub_gene_data_down[ , c(3, 6)]), )
module_membership_binary_down <- module_membership_down
module_membership_binary_down[module_membership_binary_down > 0] <- 1
module_membership_list_down <- lapply(
  split(module_membership_down, row.names(module_membership_down)), unlist
)
module_network_colors_down <- module_colors[unique(hub_gene_data_down$module)]
V(graph_down)$pie.color = list(module_network_colors_down)

module_membership_norm_down <- module_membership_norm[
  rownames(module_membership_norm) %in% rownames(module_membership_down),
]

V(graph_down)$pie.color = list(module_network_colors_down)
V(graph_down)$vertex.shape <- "pie"
V(graph_down)$vertex.color <- module_membership_list_down
V(graph_down)$vertex.label.family = "sans"

DE_down <- vector()
for (l in seq(length(V(graph_down)))) {
  if (V(graph_down)[l]$name %in% rownames(markers_list_symbols_pseudobulk[[3]])) {
    DE_down <- append(DE_down, "yes")
  } else {
    DE_down <- append(DE_down, "no")
  }
}

DE_opp_down <- vector()
for (l in seq(length(V(graph_down)))) {
  if (V(graph_down)[l]$name %in% markers_list_symbols_pseudobulk_all[[2]]) {
    DE_opp_down <- append(DE_opp_down, "yes")
  } else {
    DE_opp_down <- append(DE_opp_down, "no")
  }
}

DE_strict_down <- vector()
for (l in seq(length(V(graph_down)))) {
  if (V(graph_down)[l]$name %in% rownames(markers_list_symbols_pseudobulk_strict[[3]])) {
    DE_strict_down <- append(DE_strict_down, "yes")
  } else {
    DE_strict_down <- append(DE_strict_down, "no")
  }
}

DE_strictest_down <- vector()
for (l in seq(length(V(graph_down)))) {
  if (
    V(graph_down)[l]$name %in% rownames(markers_list_symbols_pseudobulk_strictest[[3]])
  ) {
    DE_strictest_down <- append(DE_strictest_down, "yes")
  } else {
    DE_strictest_down <- append(DE_strictest_down, "no")
  }
}

direction_down <- vector()
for (l in seq(length(E(graph_down)))) {
  hub_gene_sub_down <- hub_gene_data_down[
    hub_gene_data_down$external_gene_name %in% head_of(graph_down, l)[[1]]$name,
    ]
  direction_down <- append(direction_down, unique(hub_gene_sub_down$direction))
}
E(graph_down)$edge.color <- ifelse(direction_down == "up", "firebrick1", "dodgerblue")

hub_mat <- as.data.frame.matrix(table(hub_gene_data[ , c(3, 2)]), )
hub_mat <- hub_mat == 1
co_mat <- hub_mat %*% t(hub_mat)
diag(co_mat) <- 0
graph <- graph_from_adjacency_matrix(co_mat, "upper", TRUE)

edges <- get.edgelist(graph, FALSE)
set.seed(1)
layout <- qgraph.layout.fruchtermanreingold(
  edges, vcount = vcount(graph), E(graph)$weight,
  area = vcount(graph) ^ 2, repulse.rad = vcount(graph) ^ 2
)

hub_gene_all_GO <- table(unlist(geneIds(gene_sets)))
hub_gene_all_GO <- hub_gene_all_GO[
  names(hub_gene_all_GO) %in% unique(hub_gene_data$ensembl_gene_id)
]
sub_gene_anno <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(hub_gene_all_GO), ]
rownames(sub_gene_anno) <- NULL
sub_gene_anno <- sub_gene_anno[
  match(rownames(hub_gene_all_GO), sub_gene_anno$ensembl_gene_id),
]
rownames(hub_gene_all_GO) <- sub_gene_anno$external_gene_name
hub_gene_all_GO <- hub_gene_all_GO[order(names(hub_gene_all_GO))]

module_membership <- as.data.frame.matrix(table(hub_gene_data[ , c(3, 6)]), )
module_membership_binary <- module_membership
module_membership_binary[module_membership_binary > 0] <- 1
module_membership_list <- lapply(
  split(module_membership, row.names(module_membership)), unlist
)
module_network_colors <- module_colors
module_membership_norm <- module_membership_norm[
  rownames(module_membership_norm) %in% rownames(module_membership),
]

V(graph)$pie.color = list(module_network_colors)
V(graph)$vertex.shape <- "pie"
V(graph)$vertex.color <- module_membership_list
V(graph)$vertex.label.family = "sans"

module_max_genes <- apply(module_membership_norm, 2, function (x) which(x == max(x)))
module_max_genes <- rownames(module_membership_norm)[unlist(module_max_genes)]
module_max <- vector()
for (l in seq(length(V(graph)))) {
  if (V(graph)[l]$name %in% module_max_genes) {
    module_max <- append(module_max, "yes")
  } else {
    module_max <- append(module_max, "no")
  }
}

unique_markers <- markers_list_symbols_pseudobulk_all[[2]][
  markers_list_symbols_pseudobulk_all[[2]] %notin%
    hub_gene_data_down$external_gene_name
]
unique_markers <- append(
  unique_markers,
  markers_list_symbols_pseudobulk_all[[3]][
    markers_list_symbols_pseudobulk_all[[3]] %notin%
      hub_gene_data_up$external_gene_name
  ]
)
DE <- vector()
for (l in seq(length(V(graph)))) {
  if (V(graph)[l]$name %in% unique_markers) {
    DE <- append(DE, "yes")
  } else {
    DE <- append(DE, "no")
  }
}

direction <- vector()
for (l in seq(length(E(graph)))) {
  hub_gene_sub <- hub_gene_data[
    hub_gene_data$external_gene_name %in% head_of(graph, l)[[1]]$name,
  ]
  direction <- append(direction, unique(hub_gene_sub$direction))
}
E(graph)$edge.color <- ifelse(direction == "up", "firebrick1", "dodgerblue")
```

```{r}
# TODO: Comments, refactoring, and fix R Markdown plots.
# ------------------------------------------------------
DE_alt <- c(
  rownames(markers_list_symbols_pseudobulk[[1]]),
  rownames(markers_list_symbols_pseudobulk_strict[[1]]),
  rownames(markers_list_symbols_pseudobulk_strictest[[1]])
)

all_DE_match <- as.data.frame(
  table(V(graph)$name[match(DE_alt, V(graph)$name)]), stringsAsFactors = FALSE
)
all_DE_match_tmp <- data.frame(Var1 = V(graph)$name[V(graph)$name %notin% DE_alt], Freq = 0)
all_DE_match <- rbind(all_DE_match, all_DE_match_tmp)
all_DE_match <- all_DE_match[order(all_DE_match$Var1), ]

module_membership_median <- median(
  (abs(rowSums(module_membership_norm)) / hub_gene_all_GO) *
  abs(rowSums(module_membership_norm))
)

vertex_frame_color <- data.frame(DE = DE) %>%
  mutate(
    type = case_when(
      DE == "yes" ~ "white",
      TRUE ~ "darkgray"
    )
  )
vertex_frame_color <- vertex_frame_color$type

vertex_label_color <- data.frame(DE = DE) %>%
  mutate(
    type = case_when(
      DE == "yes" ~ "white",
      TRUE ~ "darkgray"
    )
  )
vertex_label_color <- vertex_label_color$type

vertex_label_font <- data.frame(DE = DE) %>%
  mutate(
    type = case_when(
      TRUE ~ 4
    )
  )
vertex_label_font <- vertex_label_font$type

vertex_label <- data.frame(DE = DE) %>%
  mutate(
    type = case_when(
      TRUE ~ V(graph)$name
    )
  )
vertex_label <- vertex_label$type
vertex_label[vertex_label == "NA"] <- NA

png(
  file.path("..", "..", "results", "roussarie-2020-RNAseq", "network.png"), 150,
  150, "in",
  res = 72
)
par(bg = "black", mar = c(5.1, 4.1, 4.1, 285.1), xpd = TRUE)
plot(
    graph,
    rescale = FALSE,
    xlim = c(-1, 1),
    ylim = c(-1, 1),
    layout = norm_coords(layout),
    vertex.size = sqrt(
      (abs(rowSums(module_membership_norm)) / hub_gene_all_GO) *
      abs(rowSums(module_membership_norm))
    ) * 5,
    vertex.label.color = vertex_label_color,
    edge.width = E(graph)$weight ^ 2 / 2,
    vertex.shape = "pie",
    vertex.pie = module_membership_list,
    edge.color = ifelse(direction == "up", "firebrick1", "dodgerblue"),
    vertex.label.family = "sans",
    vertex.label.font = vertex_label_font,
    vertex.label.cex = (all_DE_match$Freq + 1) * 2.5, 1,
    vertex.frame.color = vertex_frame_color,
    vertex.label = vertex_label
)
legend(
    "right",
    legend = paste0(
      "Module ", unique(hub_gene_data$module),
      ": ", gsub("\n", " ", top_words[unique(hub_gene_data$module)])
    ),
    pch = 20,
    col = module_network_colors,
    pt.cex = 15,
    cex = 10,
    bty = "n",
    text.col = "white",
    inset = c(-0.62, 0)
)
```

# References

This is the concluding section of the document. Here we output the `sessionInfo` and create a bibliography for works cited.

```{r}
sessionInfo()
```
