---
title: '01 CA1 Vulnerable AD vs. Control - `r unlist(strsplit(getwd(), "/"))[6]`'
author:
  - name: "Emir Turkes [emir.turkes@eturkes.com]"
  - name: "UK Dementia Research Institute at UCL"
date: '`r strftime(Sys.time(), "%B %d, %Y")`'
bibliography: '../../../../`r unlist(strsplit(getwd(), "/"))[4]`.bib'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_file = file.path(
    "..", "..", "..",
    "..", "results", unlist(strsplit(getwd(), "/"))[6],
    unlist(strsplit(getwd(), "/"))[7], unlist(strsplit(getwd(), "/"))[8], "01-CA1vuln-AD-CTRL.html"
  ))})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
.tocify-subheader .tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 35px; text-indent: 0;
}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of the [Tau Vulnerability Project](https://github.com/eturkes/tau-vulnerability).*

In this document we perform gene set enrichment and network analysis.
The data here is derived from @`r unlist(strsplit(getwd(), "/"))[6]` and will be referenced using the name ``r unlist(strsplit(getwd(), "/"))[6]``.

```{r}
#    This file is part of tau-vulnerability.
#    Copyright (C) 2019-2021  Emir Turkes, UK DRI at UCL, Columbia University Medical Center
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

# These should be checked per document.
# -------------------------------------
options(stringsAsFactors = FALSE)
packages <- c(
  "conflicted", "scales", "RColorBrewer", "Seurat", "dplyr", "ggplot2", "ggrepel", "GSEABase",
  "biomaRt", "Biostrings", "SingleCellExperiment", "SCnorm", "viridis", "scater", "GSVA", "limma",
  "DT", "GOSemSim", "factoextra", "dendextend", "ComplexHeatmap", "GO.db", "tm", "reshape2",
  "igraph", "qgraph", "pals", "cluster", "future", "edgeR", "circlize"
)
invisible(suppressPackageStartupMessages(lapply(packages, library, character.only = TRUE)))
conflict_prefer("strsplit", "Biostrings", quiet = TRUE)
conflict_prefer("intersect", "GSEABase", quiet = TRUE)
conflict_prefer("select", "AnnotationDbi", quiet = TRUE)
conflict_prefer("degree", "igraph", quiet = TRUE)
conflict_prefer("strsplit", "Biostrings", quiet = TRUE)
conflict_prefer("cpm", "edgeR", quiet = TRUE)
conflict_prefer("melt", "reshape2", quiet = TRUE)
source(file.path("..", "..", "..", "utils.R"))

nbclust_fun <- function(x, k) {
  list(cluster = cutree(hclust(get_dist((x), "spearman"), "ward.D2"), k))
}
`%notin%` <- Negate(`%in%`)

protocol <- c("human", "droplet", "single-nuc", "umis") # See `cluster_pipeline` in `utils.R`.
vars_to_regress <- NULL # See `cluster_pipeline` in `utils.R`.
parallel_override <- NULL # See `parallel_plan` in `utils.R`.

# Metadata to plot after dimensionality reduction and clustering.
# Values in list can include "no_legend and/or "no_label" to exclude those.
# -------------------------------------------------------------------------
metadata_to_plot <- vector("list", 7)
names(metadata_to_plot) <- c(
  "seurat_clusters", "orig_clusters", "subject", "gender", "library_date", "seq_platform", "Phase"
)
metadata_to_plot$subject <- "no_label"
metadata_to_plot$gender <- "no_label"
metadata_to_plot$library_date <- "no_label"
metadata_to_plot$seq_platform <- "no_label"
metadata_to_plot$Phase <- "no_label"
# -------------------------------------------------------------------------
# --------------------------------------------

gene_set_list <- "GO.BP"
iss_codebook_name <- "codebook_comb.txt" # Filename of in-situ sequencing codebook to reference.
subset_names <- "all"
color <- colorRampPalette(rev(brewer.pal(9, "RdBu")))(100)
color3 <- colorRampPalette(rev(brewer.pal(11, "RdYlBu")))(100)

# Clusters of interest.
# ---------------------
idents <- vector("list", 2)
idents[[1]] <- "seurat1_3"
idents[[2]] <- "seurat2_3"
names(idents) <- c("CA1_Braak56", "CA1_control")
# ---------------------
# -------------------------------------

# Everything else in this chunk remains generally unchanged.
# ----------------------------------------------------------
analysis_no <- 1
data_name <- unlist(strsplit(getwd(), "/"))[6] # Name of dataset.
assets_dir <- file.path("..", "..", "..", "..", "assets") # Backed up data.
results_dir <- file.path("..", "..", "..", "..", "results")

# Unique cache directory for each analysis number.
# ------------------------------------------------
cache_dir <- file.path(
  "..", "..", "..", "..", "cache",
  data_name, unlist(strsplit(getwd(), "/"))[7], "05-gsva", paste0("0", analysis_no)
)
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}
# ------------------------------------------------

if (protocol[1] == "human") {
  organism <- "hsapiens"
} else if (protocol[1] == "mouse") {
  organism <- "mmusculus"
}

knitr::opts_chunk$set(fig.width = 12, fig.height = 12)
# ----------------------------------------------------------
```

```{r}
seurat1 <- readRDS(
  file.path(cache_dir, "..", "..", "..", "AD", "02", paste0(subset_names[1], "_seurat.rds"))
)
seurat1

for (i in 1:length(metadata_to_plot)) {
  print(names(metadata_to_plot)[i])
  if ("no_legend" %in% metadata_to_plot[[i]] && "no_label" %in% metadata_to_plot[[i]]) {
    print(red_dim_plot(seurat1, "umap1", "umap2", names(metadata_to_plot)[i]) + NoLegend())
  } else if ("no_legend" %in% metadata_to_plot[[i]]) {
    print(red_dim_plot(seurat1, "umap1", "umap2", names(metadata_to_plot)[i], "cat") + NoLegend())
  } else if ("no_label" %in% metadata_to_plot[[i]]) {
    print(red_dim_plot(seurat1, "umap1", "umap2", names(metadata_to_plot)[i]))
  } else {
    print(red_dim_plot(seurat1, "umap1", "umap2", names(metadata_to_plot)[i], "cat"))
  }
}

seurat2 <- readRDS(
  file.path(cache_dir, "..", "..", "..", "control", "02", paste0(subset_names[1], "_seurat.rds"))
)
seurat2

for (i in 1:length(metadata_to_plot)) {
  print(names(metadata_to_plot)[i])
  if ("no_legend" %in% metadata_to_plot[[i]] && "no_label" %in% metadata_to_plot[[i]]) {
    print(red_dim_plot(seurat2, "umap1", "umap2", names(metadata_to_plot)[i]) + NoLegend())
  } else if ("no_legend" %in% metadata_to_plot[[i]]) {
    print(red_dim_plot(seurat2, "umap1", "umap2", names(metadata_to_plot)[i], "cat") + NoLegend())
  } else if ("no_label" %in% metadata_to_plot[[i]]) {
    print(red_dim_plot(seurat2, "umap1", "umap2", names(metadata_to_plot)[i]))
  } else {
    print(red_dim_plot(seurat2, "umap1", "umap2", names(metadata_to_plot)[i], "cat"))
  }
}
```

# Prep

```{r, fig.height = 5, fig.width = 7.5}
section_no <- 1
gene_set_name <- gene_set_list[section_no]

genes <- intersect(rownames(seurat1), rownames(seurat2))
seurat1 <- seurat1[genes, ]
seurat2 <- seurat2[genes, ]

seurat1$seurat_clusters <- paste0("seurat1_", seurat1$seurat_clusters)
seurat2$seurat_clusters <- paste0("seurat2_", seurat2$seurat_clusters)
seurat <- merge(seurat1, seurat2)
rm(seurat1, seurat2)
seurat

rds <- file.path(cache_dir, paste0(gene_set_name, "_gene_sets.rds"))
rds2 <- file.path(cache_dir, paste0(gene_set_name, "_gene_sets_filtered.rds"))
if (file.exists(rds) & file.exists(rds2)) {
  gene_sets <- readRDS(rds)
  gene_sets_filtered <- readRDS(rds2)
} else {
  gene_sets <- getGmt(
    file.path(assets_dir, "gene-sets", paste0(organism, ".", gene_set_name, ".ENSG.gmt"))
  )
  gene_sets <- filterGeneSets(gene_sets, 5, 500)
  gene_sets_filtered <- gene_sets # Put aside for further filtering later.
  for (i in seq(length(gene_sets@.Data))) {
    go_id <- gene_sets[[i]]@setName
    suppressWarnings(gene_sets[[i]]@setName <- gene_sets[[i]]@shortDescription)
    suppressWarnings(gene_sets[[i]]@shortDescription <- go_id)
  }

  filter_sets <- read.delim(file.path(assets_dir, "gene-sets", "genes-09-00593-s001-s5.tsv"))
  filter_sets <- filter_sets$GO.TERM
  gene_sets_filtered <- gene_sets_filtered[names(gene_sets_filtered) %in% filter_sets]
  for (i in seq(length(gene_sets_filtered@.Data))) {
    go_id <- gene_sets_filtered[[i]]@setName
    suppressWarnings(gene_sets_filtered[[i]]@setName <- gene_sets_filtered[[i]]@shortDescription)
    suppressWarnings(gene_sets_filtered[[i]]@shortDescription <- go_id)
  }

  saveRDS(gene_sets, rds)
  saveRDS(gene_sets_filtered, rds2)
}
gene_sets

iss_codebook <- read.csv(file.path(assets_dir, "gene-sets", iss_codebook_name), FALSE, sep = "\t")
print("In-situ sequencing (ISS) codebook")
head(iss_codebook)

open_targets_all <- read.csv(file.path(assets_dir, "gene-sets", "open-targets-all.csv"))
open_targets_all <- unique(open_targets_all$target.gene_info.symbol)

rds <- file.path(cache_dir, paste0(gene_set_name, "_open_targets_priority.rds"))
if (file.exists(rds)) {
  open_targets_priority <- readRDS(rds)
} else {
  open_targets_priority <- read.csv(
    file.path(assets_dir, "gene-sets", "open-targets-prioritization.csv")
  )
  if (organism == "hsapiens") {
    open_targets_priority <- human_to_mouse_genes(open_targets_priority$target.gene_info.symbol)
  } else if (organism == "mmusculus") {
    open_targets_priority <- open_targets_priority$target.gene_info.symbol
  }
  saveRDS(open_targets_priority, rds)
}

allen_mouse_Rorb_Wfs1_Bok_DE <- read.csv(
  file.path(results_dir, "allen-mouse-EC", "05-gsva", "01-Rorb-Wfs1-Bok-DE.csv"), FALSE
)
allen_mouse_Rorb_Wfs1_Bok_DE <- unique(allen_mouse_Rorb_Wfs1_Bok_DE$V1)
allen_mouse_Rorb_Wfs1_Bok_DE_strict <- read.csv(
  file.path(results_dir, "allen-mouse-EC", "05-gsva", "01-Rorb-Wfs1-Bok-DE-strict.csv"), FALSE
)
allen_mouse_Rorb_Wfs1_Bok_DE_strict <- unique(allen_mouse_Rorb_Wfs1_Bok_DE_strict$V1)
allen_mouse_Rorb_Wfs1_Bok_DE_stricter <- read.csv(
  file.path(results_dir, "allen-mouse-EC", "05-gsva", "01-Rorb-Wfs1-Bok-DE-stricter.csv"), FALSE
)
allen_mouse_Rorb_Wfs1_Bok_DE_stricter <- unique(allen_mouse_Rorb_Wfs1_Bok_DE_stricter$V1)
allen_mouse_Rorb_Wfs1_Bok_DE_strictest <- read.csv(
  file.path(results_dir, "allen-mouse-EC", "05-gsva", "01-Rorb-Wfs1-Bok-DE-strictest.csv"), FALSE
)
allen_mouse_Rorb_Wfs1_Bok_DE_strictest <- unique(allen_mouse_Rorb_Wfs1_Bok_DE_strictest$V1)
allen_all <- read.delim(file.path(results_dir, "allen-mouse-EC", "05-gsva", "allen-all.tsv"), FALSE)
allen_all <- unique(allen_all$V1)

# Create a Seurat object containing only clusters of interest with downsampling.
# Also convert gene symbols to ENSEMBL IDs.
# ------------------------------------------------------------------------------
rds <- file.path(cache_dir, paste0(gene_set_name, "_seurat_cleaned.rds"))
rds2 <- file.path(cache_dir, paste0(gene_set_name, "_gene_anno.rds"))
if (file.exists(rds) & file.exists(rds2)) {
  seurat <- readRDS(rds)
  gene_anno <- readRDS(rds2)
} else {

  # Remove obsolete data.
  # ---------------------
  DefaultAssay(seurat) <- "RNA"
  seurat[["SCT"]] <- NULL
  # ---------------------

  # Create subsets.
  # ---------------
  sample1_seurat <- subset(seurat, subset = subject == "S11/020")
  expr <- FetchData(sample1_seurat, "seurat_clusters")
  sample1_seurat <- sample1_seurat[ , which(expr == idents[[1]])]
  sample2_seurat <- subset(seurat, subset = subject == "S09/230")
  expr <- FetchData(sample2_seurat, "seurat_clusters")
  sample2_seurat <- sample2_seurat[ , which(expr == idents[[1]])]
  sample3_seurat <- subset(seurat, subset = subject == "S09/324")
  expr <- FetchData(sample3_seurat, "seurat_clusters")
  sample3_seurat <- sample3_seurat[ , which(expr == idents[[1]])]
  sample4_seurat <- subset(seurat, subset = subject == "S09/066")
  expr <- FetchData(sample4_seurat, "seurat_clusters")
  sample4_seurat <- sample4_seurat[ , which(expr == idents[[2]])]
  sample5_seurat <- subset(seurat, subset = subject == "S10/103")
  expr <- FetchData(sample5_seurat, "seurat_clusters")
  sample5_seurat <- sample5_seurat[ , which(expr == idents[[2]])]
  sample6_seurat <- subset(seurat, subset = subject == "S12/001")
  expr <- FetchData(sample6_seurat, "seurat_clusters")
  sample6_seurat <- sample6_seurat[ , which(expr == idents[[2]])]

  downsample <- min(
    dim(sample1_seurat)[2], dim(sample2_seurat)[2], dim(sample3_seurat)[2],
    dim(sample4_seurat)[2], dim(sample5_seurat)[2], dim(sample6_seurat)[2]
  )

  sample1_seurat <- subset(seurat, subset = subject == "S11/020")
  expr <- FetchData(sample1_seurat, "seurat_clusters")
  sample1_seurat <- sample1_seurat[ , which(expr == idents[[1]])]
  sample2_seurat <- subset(seurat, subset = subject == "S09/230")
  expr <- FetchData(sample2_seurat, "seurat_clusters")
  sample2_seurat <- sample2_seurat[ , which(expr == idents[[1]])]
  sample3_seurat <- subset(seurat, subset = subject == "S09/324")
  expr <- FetchData(sample3_seurat, "seurat_clusters")
  sample3_seurat <- sample3_seurat[ , which(expr == idents[[1]])]
  # downsample <- min(dim(sample1_seurat)[2], dim(sample2_seurat)[2], dim(sample3_seurat)[2])
  set.seed(1)
  sample1_seurat <- subset(sample1_seurat, cells = sample(Cells(sample1_seurat), downsample))
  set.seed(1)
  sample2_seurat <- subset(sample2_seurat, cells = sample(Cells(sample2_seurat), downsample))
  set.seed(1)
  sample3_seurat <- subset(sample3_seurat, cells = sample(Cells(sample3_seurat), downsample))
  condition1_seurat <- merge(sample1_seurat, c(sample2_seurat, sample3_seurat))

  sample1_seurat <- subset(seurat, subset = subject == "S09/066")
  expr <- FetchData(sample1_seurat, "seurat_clusters")
  sample1_seurat <- sample1_seurat[ , which(expr == idents[[2]])]
  sample2_seurat <- subset(seurat, subset = subject == "S10/103")
  expr <- FetchData(sample2_seurat, "seurat_clusters")
  sample2_seurat <- sample2_seurat[ , which(expr == idents[[2]])]
  sample3_seurat <- subset(seurat, subset = subject == "S12/001")
  expr <- FetchData(sample3_seurat, "seurat_clusters")
  sample3_seurat <- sample3_seurat[ , which(expr == idents[[2]])]
  # downsample <- min(dim(sample1_seurat)[2], dim(sample2_seurat)[2], dim(sample3_seurat)[2])
  set.seed(1)
  sample1_seurat <- subset(sample1_seurat, cells = sample(Cells(sample1_seurat), downsample))
  set.seed(1)
  sample2_seurat <- subset(sample2_seurat, cells = sample(Cells(sample2_seurat), downsample))
  set.seed(1)
  sample3_seurat <- subset(sample3_seurat, cells = sample(Cells(sample3_seurat), downsample))
  condition2_seurat <- merge(sample1_seurat, c(sample2_seurat, sample3_seurat))

  # downsample <- min(
  #   dim(condition1_seurat)[2], dim(condition2_seurat)[2], dim(condition3_seurat)[2],
  #   dim(condition4_seurat)[2], dim(condition5_seurat)[2]
  # )
  # set.seed(1)
  # condition1_seurat <- subset(
  #   condition1_seurat, cells = sample(Cells(condition1_seurat), downsample)
  # )
  # set.seed(1)
  # condition2_seurat <- subset(
  #   condition2_seurat, cells = sample(Cells(condition2_seurat), downsample)
  # )
  # set.seed(1)
  # condition3_seurat <- subset(
  #   condition3_seurat, cells = sample(Cells(condition3_seurat), downsample)
  # )

  seurat <- merge(condition1_seurat, condition2_seurat)
  rm(sample1_seurat, sample2_seurat, sample3_seurat, sample4_seurat, sample5_seurat, sample6_seurat)
  seurat <- seurat[rowSums(as.matrix(GetAssayData(seurat, slot = "counts")) > 0) >= 10, ]
  # ---------------

  # Get gene annotations.
  # ---------------------
  mart <- useEnsembl("ensembl", paste0(organism, "_gene_ensembl"))
  attributes <- c("external_gene_name", "ensembl_gene_id", "chromosome_name")
  gene_anno <- getBM(attributes, "external_gene_name", rownames(seurat), mart)
  gene_anno <- gene_anno[gene_anno$chromosome_name %in% c(1:22), ]
  gene_anno <- gene_anno[gene_anno$ensembl_gene_id %in% unique(unlist(geneIds(gene_sets))), ]

  tmp <- lapply(
    geneIds(gene_sets), function(x, y) na.omit(fastmatch::fmatch(x, y)), gene_anno$ensembl_gene_id
  )
  tmp <- filterGeneSets(tmp, 5, 500)
  gene_sets <- gene_sets[names(gene_sets) %in% names(tmp)]

  gene_anno <- gene_anno[gene_anno$ensembl_gene_id %in% unique(unlist(geneIds(gene_sets))), ]
  # ---------------------

  # For gene symbols with multiple ENSEMBL IDs, duplicate the gene symbol to have an identical row
  # for each ENSEMBL ID.
  # ----------------------------------------------------------------------------------------------
  dup <- gene_anno[duplicated(gene_anno$external_gene_name), ]
  if (nrow(dup) > 0) {
    for (i in 1:dim(dup)[1]) {
      for (j in 1:dim(gene_anno)[1]) {
        if (dup$ensembl_gene_id[i] == gene_anno$ensembl_gene_id[j]) {
          gene_anno$external_gene_name[j] <- paste0(gene_anno$external_gene_name[j], "-alt")
        }
      }
    }
    if (any(duplicated(gene_anno$external_gene_name))) {
      stop("Duplicates in gene_anno.")
    }
    seurat <- seurat[rownames(seurat) %in% gene_anno$external_gene_name, ]
    new_mat <- GetAssayData(seurat, "counts")
    for (i in 1:dim(dup)[1]) {
      for (j in 1:dim(seurat)[1]) {
        if (dup$external_gene_name[i] == rownames(seurat)[j]) {
          new_row <- GetAssayData(seurat[j, ], "counts")
          rownames(new_row) <- paste0(rownames(new_row), "-alt")
          if (rownames(new_row) %in% rownames(new_mat)) {
            rownames(new_row) <- paste0(rownames(new_row), "2")
          }
          new_mat <- rbind(new_mat, new_row)
        }
      }
    }
  } else {
    seurat <- seurat[rownames(seurat) %in% gene_anno$external_gene_name, ]
    new_mat <- GetAssayData(seurat, "counts")
  }
  gene_anno <- gene_anno[gene_anno$external_gene_name %in% rownames(new_mat), ]
  gene_anno <- gene_anno[order(match(gene_anno$external_gene_name, rownames(new_mat))), ]
  rownames(new_mat) <- gene_anno$ensembl_gene_id

  seurat <- CreateSeuratObject(new_mat, meta.data = seurat[[]])
  rm(new_mat)
  seurat$idents <- factor(
    c(
      rep(names(idents)[1], dim(condition1_seurat)[2]),
      rep(names(idents)[2], dim(condition2_seurat)[2])
    ),
    levels = names(idents)
  )
  rm(condition1_seurat, condition2_seurat)
  seurat@active.ident <- seurat$idents
  seurat$groups <- paste0(seurat$idents, "_", seurat$subject)
  # ----------------------------------------------------------------------------------------------
  
  saveRDS(seurat, rds)
  saveRDS(gene_anno, rds2)
}
# ------------------------------------------------------------------------------

sub_name <- "sctransform"
seurat <- cluster_pipeline(
  seurat, cache_dir, sub_name, protocol, vars_to_regress, parallel_override, cc = FALSE
)
seurat

# Create MDS plots to identify potential confounders and latent factors.
# ----------------------------------------------------------------------
rds <- file.path(cache_dir, paste0(gene_set_name, "_d.rds"))
if (file.exists(rds)) {
  d <- readRDS(rds)
} else {
  d <- dist(t(GetAssayData(seurat, "scale.data")))
  saveRDS(d, rds)
}
mds <- cmdscale(d, 2)
colnames(mds) <- paste0("mds_", 1:2)
seurat[["mds"]] <- CreateDimReducObject(mds, key = "mds_", assay = DefaultAssay(seurat))
rm(mds)
add_df <- data.frame(Embeddings(seurat, "mds"))
names(add_df) <- paste0("mds", seq(ncol(add_df)))
seurat$mds1 <- add_df$mds1
seurat$mds2 <- add_df$mds2

red_dim_plot(seurat, "umap1", "umap2", "idents", "cat")
red_dim_plot(seurat, "mds1", "mds2", "idents", "cat")
for (i in 1:length(metadata_to_plot)) {
  print(names(metadata_to_plot)[i])
  if ("no_legend" %in% metadata_to_plot[[i]] && "no_label" %in% metadata_to_plot[[i]]) {
    print(red_dim_plot(seurat, "mds1", "mds2", names(metadata_to_plot)[i]) + NoLegend())
  } else if ("no_legend" %in% metadata_to_plot[[i]]) {
    print(red_dim_plot(seurat, "mds1", "mds2", names(metadata_to_plot)[i], "cat") + NoLegend())
  } else if ("no_label" %in% metadata_to_plot[[i]]) {
    print(red_dim_plot(seurat, "mds1", "mds2", names(metadata_to_plot)[i]))
  } else {
    print(red_dim_plot(seurat, "mds1", "mds2", names(metadata_to_plot)[i], "cat"))
  }
}
# ----------------------------------------------------------------------

hist(as.matrix(GetAssayData(seurat)))

# Aggregate features to into "pseudo-bulk" counts.
# ------------------------------------------------
rds <- file.path(cache_dir, paste0(gene_set_name, "_pseudobulk.rds"))
if (file.exists(rds)) {
  sce <- readRDS(rds)
} else {
  sce <- as.SingleCellExperiment(seurat)
  counts(sce) <- logcounts(sce)
  sce <- suppressWarnings(
    aggregateAcrossCells(
      sce, colData(sce)[ , c("idents", "subject")],
      use_exprs_values = "counts", BPPARAM = MulticoreParam()
    )
  )
  # keep <- filterByExpr(counts(sce), group = sce$idents)
  # sce <- sce[keep, ]
  dge <- DGEList(counts(sce))
  dge <- calcNormFactors(dge)
  # dge <- voom(dge, design, plot = TRUE)
  logcounts(sce) <- cpm(dge, log = TRUE, prior.count = 3)
  rm(dge)
  # logcounts(sce) <- log2(counts(sce) + 1)
  # sce <- logNormCounts(sce)
  # logcounts(sce) <- counts(sce)
  saveRDS(sce, rds)
}
print("Pseudo-bulk counts distribution")
hist(logcounts(sce))
# ------------------------------------------------

# Subset other objects.
# ---------------------
gene_anno <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(sce), ]
seurat <- seurat[rownames(seurat) %in% rownames(sce), ]

tmp <- lapply(
  geneIds(gene_sets), function(x, y) na.omit(fastmatch::fmatch(x, y)), gene_anno$ensembl_gene_id
)
tmp <- filterGeneSets(tmp, 5, 500)
gene_sets <- gene_sets[names(gene_sets) %in% names(tmp)]
#----------------------

# Run GSVA on all cells.
# ----------------------
# rds <- file.path(cache_dir, paste0(gene_set_name, "_gsva_allcells.rds"))
# if (file.exists(rds)) {
#   gsva <- readRDS(rds)
# } else {
#   gsva <- gsva(as.matrix(GetAssayData(seurat)), gene_sets, min.sz = 5, max.sz = 500)
#   saveRDS(gsva, rds)
# }
# seurat[["GSVA"]] <- CreateAssayObject(gsva)
# DefaultAssay(seurat) <- "GSVA"
# print("GSVA distribution (all cells)")
# hist(as.matrix(GetAssayData(seurat, "counts")))
# # ----------------------

# Run GSVA on pseudo-bulk counts.
# -------------------------------
rds <- file.path(cache_dir, paste0(gene_set_name, "_gsva_pseudobulk.rds"))
if (file.exists(rds)) {
  gsva <- readRDS(rds)
} else {
  gsva <- gsva(logcounts(sce), gene_sets, method = "ssgsea")
  colnames(gsva) <- sce$groups
  saveRDS(gsva, rds)
}
# -------------------------------

# Add previous UMAP to Seurat object.
# -----------------------------------
umap <- as.matrix(data.frame(seurat$umap1, seurat$umap2))
colnames(umap) <- paste0("umap_", 1:2)
seurat[["umap"]] <- CreateDimReducObject(umap, key = "umap_", assay = DefaultAssay(seurat))
rm(umap)
# -----------------------------------
```

## DE `r gene_set_name` Ontology Tables

```{r}
# Fit model on pseudo-bulk GSVA results.
# --------------------------------------
design <- model.matrix(~ 0 + sce$idents + sce$library_date)
colnames(design)[1:2] <- names(idents)
colnames(design)[3] <- "library_date"
fit <- lmFit(gsva, design)
contrast_mat <- makeContrasts(CA1_Braak56-CA1_control, levels = design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat), robust = TRUE)
tests <- decideTests(cont_fit, p.value = 0.05)
tests_up <- tests[tests[ , 1] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, ]))
if (length(discard) > 0) {
  tests_up <- tests_up[-discard, ]
}
tests_down <- tests[tests[ , 1] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, ]))
if (length(discard) > 0) {
  tests_down <- tests_down[-discard, ]
}
results <- topTable(cont_fit, 1, Inf, p.value = 0.05)
# --------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
gsva_list_pseudobulk <- list(all, up, down)
names(gsva_list_pseudobulk) <- c(
  "All DE", paste0("Upregulated in ", names(idents)[1]),
  paste0("Downregulated in ", names(idents)[2])
)

# Fit a strict model on pseudo-bulk GSVA results.
# -----------------------------------------------
design <- model.matrix(~ 0 + sce$idents + sce$library_date)
colnames(design)[1:2] <- names(idents)
colnames(design)[3] <- "library_date"
fit <- lmFit(gsva, design)
contrast_mat <- makeContrasts(CA1_Braak56-CA1_control, levels = design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat), robust = TRUE)
tests <- decideTests(cont_fit, p.value = 0.01)
tests_up <- tests[tests[ , 1] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, ]))
if (length(discard) > 0) {
  tests_up <- tests_up[-discard, ]
}
tests_down <- tests[tests[ , 1] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, ]))
if (length(discard) > 0) {
  tests_down <- tests_down[-discard, ]
}
results <- topTable(cont_fit, 1, Inf, p.value = 0.01)
# -----------------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
gsva_list_pseudobulk_strict <- list(all, up, down)
names(gsva_list_pseudobulk_strict) <- c(
  "All DE", paste0("Upregulated in ", names(idents)[1]),
  paste0("Downregulated in ", names(idents)[2])
)

# Fit a stricter model on pseudo-bulk GSVA results.
# -------------------------------------------------
design <- model.matrix(~ 0 + sce$idents + sce$library_date)
colnames(design)[1:2] <- names(idents)
colnames(design)[3] <- "library_date"
fit <- lmFit(gsva, design)
contrast_mat <- makeContrasts(CA1_Braak56-CA1_control, levels = design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat), robust = TRUE)
tests <- decideTests(cont_fit, p.value = 0.005)
tests_up <- tests[tests[ , 1] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, ]))
if (length(discard) > 0) {
  tests_up <- tests_up[-discard, ]
}
tests_down <- tests[tests[ , 1] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, ]))
if (length(discard) > 0) {
  tests_down <- tests_down[-discard, ]
}
results <- topTable(cont_fit, 1, Inf, p.value = 0.005)
# -------------------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
gsva_list_pseudobulk_stricter <- list(all, up, down)
names(gsva_list_pseudobulk_stricter) <- c(
  "All DE", paste0("Upregulated in ", names(idents)[1]),
  paste0("Downregulated in ", names(idents)[2])
)

# Fit the strictest model on pseudo-bulk GSVA results.
# ----------------------------------------------------
design <- model.matrix(~ 0 + sce$idents + sce$library_date)
colnames(design)[1:2] <- names(idents)
colnames(design)[3] <- "library_date"
fit <- lmFit(gsva, design)
contrast_mat <- makeContrasts(CA1_Braak56-CA1_control, levels = design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat), robust = TRUE)
tests <- decideTests(cont_fit, p.value = 0.001)
tests_up <- tests[tests[ , 1] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, ]))
if (length(discard) > 0) {
  tests_up <- tests_up[-discard, ]
}
tests_down <- tests[tests[ , 1] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, ]))
if (length(discard) > 0) {
  tests_down <- tests_down[-discard, ]
}
results <- topTable(cont_fit, 1, Inf, p.value = 0.001)
# ----------------------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
gsva_list_pseudobulk_strictest <- list(all, up, down)
names(gsva_list_pseudobulk_strictest) <- c(
  "All DE", paste0("Upregulated in ", names(idents)[1]),
  paste0("Downregulated in ", names(idents)[2])
)
```

**`r names(idents)[1]` Upregulated `r gene_set_name` Pseudo-bulk**

```{r}
datatable_download_exp(gsva_list_pseudobulk[[2]])
```

**`r names(idents)[1]` Upregulated `r gene_set_name` Pseudo-bulk Strict Model**

```{r}
datatable_download_exp(gsva_list_pseudobulk_strict[[2]])
```

**`r names(idents)[1]` Upregulated `r gene_set_name` Pseudo-bulk Stricter Model**

```{r}
datatable_download_exp(gsva_list_pseudobulk_stricter[[2]])
```

**`r names(idents)[1]` Upregulated `r gene_set_name` Pseudo-bulk Strictest Model**

```{r}
datatable_download_exp(gsva_list_pseudobulk_strictest[[2]])
```

**`r names(idents)[1]` Downregulated `r gene_set_name` Pseudo-bulk**

```{r}
datatable_download_exp(gsva_list_pseudobulk[[3]])
```

**`r names(idents)[1]` Downregulated `r gene_set_name` Pseudo-bulk Strict**

```{r}
datatable_download_exp(gsva_list_pseudobulk_strict[[3]])
```

**`r names(idents)[1]` Downregulated `r gene_set_name` Pseudo-bulk Stricter**

```{r}
datatable_download_exp(gsva_list_pseudobulk_stricter[[3]])
```

**`r names(idents)[1]` Downregulated `r gene_set_name` Pseudo-bulk Strict**

```{r}
datatable_download_exp(gsva_list_pseudobulk_strictest[[3]])
```

## DE Genes Tables

```{r}
# Fit model on pseudo-bulk counts.
# --------------------------------
design <- model.matrix(~ 0 + sce$idents + sce$library_date)
colnames(design)[1:2] <- names(idents)
colnames(design)[3] <- "library_date"
fit <- lmFit(logcounts(sce), design)
contrast_mat <- makeContrasts(CA1_Braak56-CA1_control, levels = design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat), robust = TRUE, trend = TRUE)
tests <- decideTests(cont_fit, p.value = 0.05)
tests_up <- tests[tests[ , 1] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, ]))
if (length(discard) > 0) {
  tests_up <- tests_up[-discard, ]
}
tests_down <- tests[tests[ , 1] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, ]))
if (length(discard) > 0) {
  tests_down <- tests_down[-discard, ]
}
results <- topTable(cont_fit, 1, Inf, p.value = 0.05)
# --------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
markers_list_pseudobulk <- list(all, up, down)

# Fit a strict model on pseudo-bulk counts.
# -----------------------------------------
design <- model.matrix(~ 0 + sce$idents + sce$library_date)
colnames(design)[1:2] <- names(idents)
colnames(design)[3] <- "library_date"
fit <- lmFit(logcounts(sce), design)
contrast_mat <- makeContrasts(CA1_Braak56-CA1_control, levels = design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat), robust = TRUE, trend = TRUE)
tests <- decideTests(cont_fit, p.value = 0.01)
tests_up <- tests[tests[ , 1] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, ]))
if (length(discard) > 0) {
  tests_up <- tests_up[-discard, ]
}
tests_down <- tests[tests[ , 1] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, ]))
if (length(discard) > 0) {
  tests_down <- tests_down[-discard, ]
}
results <- topTable(cont_fit, 1, Inf, p.value = 0.01)
# -----------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
markers_list_pseudobulk_strict <- list(all, up, down)

# Fit a stricter model on pseudo-bulk counts.
# -------------------------------------------
design <- model.matrix(~ 0 + sce$idents + sce$library_date)
colnames(design)[1:2] <- names(idents)
colnames(design)[3] <- "library_date"
fit <- lmFit(logcounts(sce), design)
contrast_mat <- makeContrasts(CA1_Braak56-CA1_control, levels = design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat), robust = TRUE, trend = TRUE)
tests <- decideTests(cont_fit, p.value = 0.005)
tests_up <- tests[tests[ , 1] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, ]))
if (length(discard) > 0) {
  tests_up <- tests_up[-discard, ]
}
tests_down <- tests[tests[ , 1] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, ]))
if (length(discard) > 0) {
  tests_down <- tests_down[-discard, ]
}
results <- topTable(cont_fit, 1, Inf, p.value = 0.005)
# -------------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
markers_list_pseudobulk_stricter <- list(all, up, down)

# Fit the strictest model on pseudo-bulk counts.
# ----------------------------------------------
design <- model.matrix(~ 0 + sce$idents + sce$library_date)
colnames(design)[1:2] <- names(idents)
colnames(design)[3] <- "library_date"
fit <- lmFit(logcounts(sce), design)
contrast_mat <- makeContrasts(CA1_Braak56-CA1_control, levels = design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat), robust = TRUE, trend = TRUE)
tests <- decideTests(cont_fit, p.value = 0.001)
tests_up <- tests[tests[ , 1] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, ]))
if (length(discard) > 0) {
  tests_up <- tests_up[-discard, ]
}
tests_down <- tests[tests[ , 1] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, ]))
if (length(discard) > 0) {
  tests_down <- tests_down[-discard, ]
}
results <- topTable(cont_fit, 1, Inf, p.value = 0.001)
# ----------------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
markers_list_pseudobulk_strictest <- list(all, up, down)

# Convert ENSEMBL IDs back to gene symbols.
# -----------------------------------------
genes <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(markers_list_pseudobulk[[2]]), ]
genes <- genes[order(match(genes$ensembl_gene_id, rownames(markers_list_pseudobulk[[2]]))), ]
up <- markers_list_pseudobulk[[2]]
rownames(up) <- genes$external_gene_name

genes <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(markers_list_pseudobulk[[3]]), ]
genes <- genes[order(match(genes$ensembl_gene_id, rownames(markers_list_pseudobulk[[3]]))), ]
down <- markers_list_pseudobulk[[3]]
rownames(down) <- genes$external_gene_name

all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
markers_list_symbols_pseudobulk <- list(all, up, down)

genes <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(markers_list_pseudobulk_strict[[2]]), ]
genes <- genes[order(match(genes$ensembl_gene_id, rownames(markers_list_pseudobulk_strict[[2]]))), ]
up <- markers_list_pseudobulk_strict[[2]]
rownames(up) <- genes$external_gene_name

genes <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(markers_list_pseudobulk_strict[[3]]), ]
genes <- genes[order(match(genes$ensembl_gene_id, rownames(markers_list_pseudobulk_strict[[3]]))), ]
down <- markers_list_pseudobulk_strict[[3]]
rownames(down) <- genes$external_gene_name

all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
markers_list_symbols_pseudobulk_strict <- list(all, up, down)

genes <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(markers_list_pseudobulk_stricter[[2]]), ]
genes <- genes[
  order(match(genes$ensembl_gene_id, rownames(markers_list_pseudobulk_stricter[[2]]))),
]
up <- markers_list_pseudobulk_stricter[[2]]
rownames(up) <- genes$external_gene_name

genes <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(markers_list_pseudobulk_stricter[[3]]), ]
genes <- genes[
  order(match(genes$ensembl_gene_id, rownames(markers_list_pseudobulk_stricter[[3]]))),
]
down <- markers_list_pseudobulk_stricter[[3]]
rownames(down) <- genes$external_gene_name

all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
markers_list_symbols_pseudobulk_stricter <- list(all, up, down)

genes <- gene_anno[
  gene_anno$ensembl_gene_id %in% rownames(markers_list_pseudobulk_strictest[[2]]),
]
genes <- genes[
  order(match(genes$ensembl_gene_id, rownames(markers_list_pseudobulk_strictest[[2]]))),
]
up <- markers_list_pseudobulk_strictest[[2]]
rownames(up) <- genes$external_gene_name

genes <- gene_anno[
  gene_anno$ensembl_gene_id %in% rownames(markers_list_pseudobulk_strictest[[3]]),
]
genes <- genes[
  order(match(genes$ensembl_gene_id, rownames(markers_list_pseudobulk_strictest[[3]]))),
]
down <- markers_list_pseudobulk_strictest[[3]]
rownames(down) <- genes$external_gene_name

all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
markers_list_symbols_pseudobulk_strictest <- list(all, up, down)
# -----------------------------------------

# Create new Seurat object with gene symbols.
# -------------------------------------------
new_mat <- GetAssayData(seurat)
rownames(new_mat) <- gene_anno$external_gene_name
seurat_symbols <- CreateSeuratObject(new_mat, meta.data = seurat[[]])
Idents(seurat_symbols) <- "idents"
seurat_symbols <- SetAssayData(
  seurat_symbols, new.data = log2(as.matrix(GetAssayData(seurat_symbols, "counts")) + 1)
)
seurat_symbols <- SetAssayData(
  seurat_symbols, "scale.data", t(scale(t(GetAssayData(seurat_symbols)), FALSE))
)
# DefaultAssay(seurat) <- "GSVA"
# -------------------------------------------

# color2 <- colorRampPalette(brewer.pal(9, "YlOrBr"))(round(max(GetAssayData(seurat_symbols))))
# color2[1] <- "#DDDDDD"
```

**`r names(idents)[1]` Upregulated Pseudo-bulk**

```{r}
datatable_download_exp(markers_list_symbols_pseudobulk[[2]])
```

**`r names(idents)[1]` Upregulated Pseudo-bulk Strict Model**

```{r}
datatable_download_exp(markers_list_symbols_pseudobulk_strict[[2]])
```

**`r names(idents)[1]` Upregulated Pseudo-bulk Stricter Model**

```{r}
datatable_download_exp(markers_list_symbols_pseudobulk_stricter[[2]])
```

**`r names(idents)[1]` Upregulated Pseudo-bulk Strictest Model**

```{r}
datatable_download_exp(markers_list_symbols_pseudobulk_strictest[[2]])
```

**`r names(idents)[1]` Downregulated Pseudo-bulk**

```{r}
datatable_download_exp(markers_list_symbols_pseudobulk[[3]])
```

**`r names(idents)[1]` Downregulated Pseudo-bulk Strict Model**

```{r}
datatable_download_exp(markers_list_symbols_pseudobulk_strict[[3]])
```

**`r names(idents)[1]` Downregulated Pseudo-bulk Stricter Model**

```{r}
datatable_download_exp(markers_list_symbols_pseudobulk_stricter[[3]])
```

**`r names(idents)[1]` Downregulated Pseudo-bulk Strictest Model**

```{r}
datatable_download_exp(markers_list_symbols_pseudobulk_strictest[[3]])
```

## DE GO Ontology Heatmaps

```{r}
anno_color <- hue_pal()(length(unique(sce$subject)))
anno <- HeatmapAnnotation(
  donor = seurat_symbols$subject,
  cell_type = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = c(names(idents)[[1]], names(idents)[[2]]),
    labels_gp = gpar(col = "black", fontsize = 20)
  ),
  col = list(
    donor = c(
      "S11/020" = anno_color[1], "S09/230" = anno_color[2], "S09/324" = anno_color[3],
      "S09/066" = anno_color[4], "S10/103" = anno_color[5], "S12/001" = anno_color[6]
    )
  )
)
```

### `r gene_set_name` Ontologies

```{r, fig.width = 20, fig.height = 5}
genes <- rownames(markers_list_symbols_pseudobulk[[1]])
draw(
  Heatmap(
    t(
      apply(
        as.matrix(GetAssayData(seurat_symbols)[genes, , drop = FALSE]), 1,
        function (x) (x - min(x)) / (max(x) - min(x))
      )
    ),
    color3,
    cluster_columns = FALSE,
    show_column_names = FALSE,
    column_split = seurat_symbols$idents,
    show_heatmap_legend = FALSE,
    top_annotation = anno,
    column_title_gp = gpar(fontsize = 20),
    column_gap = unit(2, "mm")
  )
)
```

# References

This is the concluding section of the document, where we output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
