---
title: "Investigating Selective Vulnerability in Alzheimer's Disease"
subtitle: "Demonstration of the GeneFunnel Gene Set Enrichment Tool"
author:
  - name: "Emir Turkes (emir.turkes.19@ucl.ac.uk) and Prof. Karen E. Duff PhD (k.duff@ucl.ac.uk)"
  - name: "UK Dementia Research Institute at UCL"
date: '`r strftime(Sys.time(), "%B %d, %Y")`'
bibliography: '../../../../`r unlist(strsplit(getwd(), "/"))[4]`.bib'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: false
    theme: lumen
    highlight: haddock
    toc: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_file = file.path(
    "..", "..", "..",
    "..", "results", unlist(strsplit(getwd(), "/"))[6],
    unlist(strsplit(getwd(), "/"))[7], unlist(strsplit(getwd(), "/"))[8], "02-simulation.html"
  ))})

---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
.tocify-subheader .tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 35px; text-indent: 0;
}
div.main-container {max-width: none; width: 100%;}
</style>

*Source code and Docker/Singularity container for rerunning this analysis can be found at https://github.com/duff-lab-team/GeneFunnel-shiny (currently private).*

# About {.tabset}

This document demonstrates an analysis using the *GeneFunnel* gene set enrichment tool.
*GeneFunnel* is an R package providing an optimized pipeline for integrated gene set enrichment, differential expression, and hub gene analysis for gene expression data.
Here we apply *GeneFunnel* on wild-type mouse single-cell RNAseq (scRNAseq) data from the Allen Institute [@leinGenomewideAtlasGene2007] to investigate selective vulnerability in Alzheimer's Disease (AD), where certain cell types are preferentially affected by pathology over others.
We hypothesize that physiological expression patterns correlating with the pathological stage (i.e. Braak stage) of a cell type drive mechanisms associated with vulnerability.  

Please select a tab below to explore the data.
Start from the left to understand the whole pipeline, or the default selected tab which starts the main results.

```{r}
# Some standard boilerplate.
# --------------------------
#    This file is part of GeneFunnel-shiny.
#    Copyright (C) 2019-2021  Emir Turkes, The Duff Lab, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    The Duff Lab can be contacted at dufflab1@gmail.com

knitr::opts_chunk$set(fig.width = 10, fig.height = 7)
options(stringsAsFactors = FALSE)
packages <- c(
  "conflicted", "scales", "RColorBrewer", "Seurat", "dplyr", "ggplot2", "ggrepel", "GSEABase",
  "biomaRt", "Biostrings", "SingleCellExperiment", "SCnorm", "viridis", "scater", "GSVA", "limma",
  "DT", "dendextend", "ComplexHeatmap", "GO.db", "tm", "reshape2", "igraph", "qgraph", "pals",
  "cluster", "future", "edgeR", "circlize", "factoextra"
)
invisible(suppressPackageStartupMessages(lapply(packages, library, character.only = TRUE)))
conflict_prefer("strsplit", "Biostrings", quiet = TRUE)
conflict_prefer("intersect", "GSEABase", quiet = TRUE)
conflict_prefer("select", "AnnotationDbi", quiet = TRUE)
conflict_prefer("degree", "igraph", quiet = TRUE)
conflict_prefer("strsplit", "Biostrings", quiet = TRUE)
conflict_prefer("cpm", "edgeR", quiet = TRUE)
conflict_prefer("groups", "igraph", quiet = TRUE)

# Can be found at: https://github.com/eturkes/utils/blob/main/GeneFunnel-shiny/utils.R
# -------------------------------------------------------------------------------------
source(file.path("..", "..", "..", "utils.R"))
# -------------------------------------------------------------------------------------

protocol <- c("human", "droplet", "single-cell", "umis") # See `cluster_pipeline` in `utils.R`.
vars_to_regress <- NULL # See `cluster_pipeline` in `utils.R`.
parallel_override <- NULL # See `parallel_plan` in `utils.R`.
gene_set_name <- "GO" # What gene set are we using.
subset_names <- "neuronal" # From what subset of the data are we looking at.
gene_set_list <- "GO"

if (protocol[1] == "human") {
  organism <- "hsapiens"
} else if (protocol[1] == "mouse") {
  organism <- "mmusculus"
}

# Clusters of interest.
# ---------------------
idents <- vector("list", 2)
idents[[1]] <- c("seurat1_3", "seurat1_8")
idents[[2]] <- c("seurat2_7", "seurat2_9")
names(idents) <- c("BA9_L45_EX_RORB_NFT", "BA9_L45_EX_RORB_normal")
# ---------------------

# Metadata to plot after dimensionality reduction and clustering.
# Values in list can include "no_legend and/or "no_label" to exclude those.
# -------------------------------------------------------------------------
metadata_to_plot <- vector("list", 3)
names(metadata_to_plot) <- c("seurat_clusters", "donor", "Phase")
metadata_to_plot$donor <- "no_label"
metadata_to_plot$Phase <- "no_label"
# -------------------------------------------------------------------------

# Clustering function for gene sets.
# ----------------------------------
nbclust_fun <- function(x, k) {
  list(cluster = cutree(hclust(get_dist((x), "spearman"), "ward.D2"), k))
}
# ----------------------------------

color <- colorRampPalette(rev(brewer.pal(9, "RdBu")))(100) # For main gene set heatmap.
color2 <- colorRampPalette(rev(brewer.pal(11, "RdYlBu")))(100) # For gene heatmaps.
`%notin%` <- Negate(`%in%`)

data_name <- unlist(strsplit(getwd(), "/"))[6] # Name of dataset.
assets_dir <- file.path("..", "..", "..", "..", "cache", data_name, "comb", "05-gsva", "02")
results_dir <- file.path("..", "..", "..", "..", "results")
# --------------------------
```

## Prep

```{r}
```

The highlighted clusters were used as input to the main *GeneFunnel* pipeline, as shown below.
Note that gray boxes indicate optional input and green boxes indicate the main outputs.


Applying the pipeline, after subsetting and downsampling we arrived at the following reduced dimensions:

```{r, fig.width = 10, fig.height = 10}
section_no <- 1
gene_set_name <- gene_set_list[section_no]

# Load and filter gene sets.
# --------------------------
gene_sets_orig <- getGmt(
  file.path(
      "..", "..", "..", "..",
      "assets", "gene-sets", paste0(organism, ".", gene_set_name, ".comb.name.gmt")
  )
)
rds <- file.path(assets_dir, paste0(gene_set_name, "_gene_sets.rds"))
if (file.exists(rds)) {
  gene_sets <- readRDS(rds)
} else {
  gene_sets <- gene_sets_orig
  gene_sets <- filterGeneSets(gene_sets, 10, 50)
  for (i in seq(length(gene_sets@.Data))) {
    go_id <- gene_sets[[i]]@setName
    suppressWarnings(gene_sets[[i]]@setName <- gene_sets[[i]]@shortDescription)
    suppressWarnings(gene_sets[[i]]@shortDescription <- go_id)
  }
  saveRDS(gene_sets, rds)
}
# --------------------------

gs_names <- c(names(gene_sets)[grep("GABA | gamma-aminobutyric", names(gene_sets))], "positive regulation of glutamate secretion", "positive regulation of synaptic transmission, glutamatergic", "positive regulation of excitatory postsynaptic potential")
for (i in seq_along(gs_names)) {
  print(gene_sets[[gs_names[i]]]@setName)
  print(gene_sets[[gs_names[i]]]@geneIds)
}

gs1 <- c(gene_sets[[gs_names[1]]]@geneIds,
gene_sets[[gs_names[2]]]@geneIds,
gene_sets[[gs_names[3]]]@geneIds)
gs2 <- c(
gene_sets[[gs_names[4]]]@geneIds,
gene_sets[[gs_names[5]]]@geneIds,
gene_sets[[gs_names[6]]]@geneIds)

universe <- unique(unlist(geneIds(gene_sets)))

Nsamples <- 100
Mgenes <- length(unique(unlist(geneIds(gene_sets))))
orig_mat <- matrix(rnorm(Nsamples * Mgenes, 5, 10), Mgenes, Nsamples)
colnames(orig_mat) <- paste0('sample', 1:Nsamples)
rownames(orig_mat) <- unique(unlist(geneIds(gene_sets)))
# let half the samples be in one biological state and the other half a different one
# simulate differential expression for genes in the geneset
orig_mat[gs1, 1:round(Nsamples/2)] <- rnorm(length(gs1)*round(Nsamples/2), 10, 10)
orig_mat[gs2, 1:round(Nsamples/2)] <- rnorm(length(gs2)*round(Nsamples/2), 1, 10)
# we can visualize this weak differential expression visually in a heatmap
# visualize weakly differentially expressed genes and another 50 genes
vi <- c(c(gs1, gs2), universe[1:50])
col <- c(rep(1, times = length(gs1)), rep(3, times = length(gs2)), rep(2, times = 50))
# label supposedly differentially expressed genes
heatmap(orig_mat[vi,], Rowv=NA, Colv=NA, scale="none",
col=colorRampPalette(c("blue", "white", "red"))(100),
RowSideColors = rainbow(3)[col])

# run differential expression analysis
vals <- sapply(1:nrow(orig_mat), function(i) {
pv <- t.test(orig_mat[i, 1:round(Nsamples/2)], orig_mat[i, round(Nsamples/2+1):Nsamples])$p.val
pv
})
names(vals) <- rownames(orig_mat)
vals <- -log10(vals)
# look at -log10(p values) for genes that are supposedly differentially expressed
barplot(sort(vals[c(gs1, gs2)], decreasing=TRUE), ylim=c(0, 10))
# multiple testing correction line
bonf <- function(a, n) { 1 - (1-a) ** (1/n) }
abline(h = -log10(bonf(0.05, nrow(orig_mat))), col="red")

rds <- file.path(assets_dir, paste0(gene_set_name, "_gsva_pseudobulk.rds"))
if (file.exists(rds)) {
  gsva <- readRDS(rds)
} else {
  gsva <- gsva(orig_mat, gene_sets, method = "ssgsea", ssgsea.norm = FALSE)
  saveRDS(gsva, rds)
}

vi <- c(gs_names, names(gene_sets)[1:10])
col <- c(rep(1, times = 3), rep(3, times = 3), rep(2, times = 10))
# label supposedly differentially expressed genes
heatmap(gsva[vi,], Rowv=NA, Colv=NA, scale="none",
col=colorRampPalette(c("blue", "white", "red"))(100),
RowSideColors = rainbow(3)[col], margins = c(8, 25))

idents <- factor(c(rep("DE", times = 50), rep("not_DE", times = 50)))
```

```{r}
# Fit model on pseudo-bulk GSVA results.
# --------------------------------------
design <- model.matrix(~ 0 + idents)
colnames(design) <- levels(idents)
fit <- lmFit(gsva, design)
contrast_mat <- makeContrasts(DE-not_DE, levels = design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat))
tests <- decideTests(cont_fit, p.value = 5e-2)
tests_up <- tests[tests[ , 1] == 1, ]
tests_down <- tests[tests[ , 1] == -1, ]
results <- topTable(cont_fit, number = Inf)

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
gsva_list_pseudobulk <- list(all, up, down)
names(gsva_list_pseudobulk) <- c(
  "All DE", paste0("Upregulated in ", names(idents)[1]),
  paste0("Downregulated in ", names(idents)[2])
)
# --------------------------------------

# Fit a strict model on pseudo-bulk GSVA results.
# -----------------------------------------------
design <- model.matrix(~ 0 + idents)
colnames(design) <- levels(idents)
fit <- lmFit(gsva, design)
contrast_mat <- makeContrasts(DE-not_DE, levels = design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat))
tests <- decideTests(cont_fit, p.value = 1e-2)
tests_up <- tests[tests[ , 1] == 1, ]
tests_down <- tests[tests[ , 1] == -1, ]
results <- topTable(cont_fit, number = Inf)

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
gsva_list_pseudobulk_strict <- list(all, up, down)
names(gsva_list_pseudobulk_strict) <- c(
  "All DE", paste0("Upregulated in ", names(idents)[1]),
  paste0("Downregulated in ", names(idents)[2])
)
# -----------------------------------------------

# Fit a stricter model on pseudo-bulk GSVA results.
# -------------------------------------------------
design <- model.matrix(~ 0 + idents)
colnames(design) <- levels(idents)
fit <- lmFit(gsva, design)
contrast_mat <- makeContrasts(DE-not_DE, levels = design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat))
tests <- decideTests(cont_fit, p.value = 5e-3)
tests_up <- tests[tests[ , 1] == 1, ]
tests_down <- tests[tests[ , 1] == -1, ]
results <- topTable(cont_fit, number = Inf)

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
gsva_list_pseudobulk_stricter <- list(all, up, down)
names(gsva_list_pseudobulk_stricter) <- c(
  "All DE", paste0("Upregulated in ", names(idents)[1]),
  paste0("Downregulated in ", names(idents)[2])
)
# -------------------------------------------------

# Fit the strictest model on pseudo-bulk GSVA results.
# ----------------------------------------------------
design <- model.matrix(~ 0 + idents)
colnames(design) <- levels(idents)
fit <- lmFit(gsva, design)
contrast_mat <- makeContrasts(DE-not_DE, levels = design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat))
tests <- decideTests(cont_fit, p.value = 1e-3)
tests_up <- tests[tests[ , 1] == 1, ]
tests_down <- tests[tests[ , 1] == -1, ]
results <- topTable(cont_fit, number = Inf)

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
gsva_list_pseudobulk_strictest <- list(all, up, down)
names(gsva_list_pseudobulk_strictest) <- c(
  "All DE", paste0("Upregulated in ", names(idents)[1]),
  paste0("Downregulated in ", names(idents)[2])
)
# ----------------------------------------------------
```

**Upregulated in Direction of `r names(idents)[1]` `r gene_set_name` Pseudobulk Initial Model**

```{r}
ncol <- ncol(gsva_list_pseudobulk[[2]])
datatable_download_exp(gsva_list_pseudobulk[[2]][(ncol-3):ncol])
```

**Upregulated in Direction of `r names(idents)[1]` `r gene_set_name` Pseudobulk Strict Model**

```{r}
ncol <- ncol(gsva_list_pseudobulk_strict[[2]])
datatable_download_exp(gsva_list_pseudobulk_strict[[2]][(ncol-3):ncol])
```

**Upregulated in Direction of `r names(idents)[1]` `r gene_set_name` Pseudobulk Stricter Model**

```{r}
ncol <- ncol(gsva_list_pseudobulk_stricter[[2]])
datatable_download_exp(gsva_list_pseudobulk_stricter[[2]][(ncol-3):ncol])
```

**Upregulated in Direction of `r names(idents)[1]` `r gene_set_name` Pseudobulk Strictest Model**

```{r}
ncol <- ncol(gsva_list_pseudobulk_strictest[[2]])
datatable_download_exp(gsva_list_pseudobulk_strictest[[2]][(ncol-3):ncol])
```

**Downregulated in Direction of `r names(idents)[1]` `r gene_set_name` Pseudobulk Initial Model**

```{r}
ncol <- ncol(gsva_list_pseudobulk[[3]])
datatable_download_exp(gsva_list_pseudobulk[[3]][(ncol-3):ncol])
```

**Downregulated in Direction of `r names(idents)[1]` `r gene_set_name` Pseudobulk Strict Model**

```{r}
ncol <- ncol(gsva_list_pseudobulk_strict[[3]])
datatable_download_exp(gsva_list_pseudobulk_strict[[3]][(ncol-3):ncol])
```

**Downregulated in Direction of `r names(idents)[1]` `r gene_set_name` Pseudobulk Stricter Model**

```{r}
ncol <- ncol(gsva_list_pseudobulk_stricter[[3]])
datatable_download_exp(gsva_list_pseudobulk_stricter[[3]][(ncol-3):ncol])
```

**Downregulated in Direction of `r names(idents)[1]` `r gene_set_name` Pseudobulk Strictest Model**

```{r}
ncol <- ncol(gsva_list_pseudobulk_strictest[[3]])
datatable_download_exp(gsva_list_pseudobulk_strictest[[3]][(ncol-3):ncol])
```

**Number of DE Gene Sets Per Model**

```{r}
data <- data.frame(
  x = c(
    dim(gsva_list_pseudobulk[[1]])[1], dim(gsva_list_pseudobulk_strict[[1]])[1],
    dim(gsva_list_pseudobulk_stricter[[1]])[1], dim(gsva_list_pseudobulk_strictest[[1]])[1]
  ),
  y = c(
    dim(gsva_list_pseudobulk[[1]])[1], dim(gsva_list_pseudobulk_strict[[1]])[1],
    dim(gsva_list_pseudobulk_stricter[[1]])[1], dim(gsva_list_pseudobulk_strictest[[1]])[1]
  )
)
ggplot(data, aes(x, y)) +
  geom_point() +
  geom_smooth(method = lm) +
  geom_label_repel(
    aes(label = c("initial_model", "strict_model", "stricter_model", "strictest_model"))
  ) +
  labs(x = "Number of DE gene sets", y = "Number of DE gene sets")
```

## DE Genes

We fit the same models as before, this time on the genes themselves.
Please collapse the code block for details.

```{r}
# Fit model on pseudo-bulk counts.
# --------------------------------
design <- model.matrix(~ 0 + idents)
colnames(design) <- levels(idents)
fit <- lmFit(orig_mat, design)
contrast_mat <- makeContrasts(DE-not_DE, levels = design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat), trend = TRUE)
tests <- decideTests(cont_fit, p.value = 5e-2, adjust.method = "none")
tests_up <- tests[tests[ , 1] == 1, ]
tests_down <- tests[tests[ , 1] == -1, ]
results <- topTable(cont_fit, number = Inf)
# --------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
markers_list_pseudobulk <- list(all, up, down)

# Fit a strict model on pseudo-bulk counts.
# -----------------------------------------
design <- model.matrix(~ 0 + idents)
colnames(design) <- levels(idents)
fit <- lmFit(orig_mat, design)
contrast_mat <- makeContrasts(DE-not_DE, levels = design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat), trend = TRUE)
tests <- decideTests(cont_fit, p.value = 1e-2, adjust.method = "none")
tests_up <- tests[tests[ , 1] == 1, ]
tests_down <- tests[tests[ , 1] == -1, ]
results <- topTable(cont_fit, number = Inf)
# -----------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
markers_list_pseudobulk_strict <- list(all, up, down)

# Fit a stricter model on pseudo-bulk counts.
# -------------------------------------------
design <- model.matrix(~ 0 + idents)
colnames(design) <- levels(idents)
fit <- lmFit(orig_mat, design)
contrast_mat <- makeContrasts(DE-not_DE, levels = design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat), trend = TRUE)
tests <- decideTests(cont_fit, p.value = 5e-3, adjust.method = "none")
tests_up <- tests[tests[ , 1] == 1, ]
tests_down <- tests[tests[ , 1] == -1, ]
results <- topTable(cont_fit, number = Inf)
# -------------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
markers_list_pseudobulk_stricter <- list(all, up, down)

# Fit the strictest model on pseudo-bulk counts.
# ----------------------------------------------
design <- model.matrix(~ 0 + idents)
colnames(design) <- levels(idents)
fit <- lmFit(orig_mat, design)
contrast_mat <- makeContrasts(DE-not_DE, levels = design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat), trend = TRUE)
tests <- decideTests(cont_fit, p.value = 1e-3, adjust.method = "none")
tests_up <- tests[tests[ , 1] == 1, ]
tests_down <- tests[tests[ , 1] == -1, ]
results <- topTable(cont_fit, number = Inf)
# ----------------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
markers_list_pseudobulk_strictest <- list(all, up, down)
```

**Upregulated in Direction of `r names(idents)[1]` `r gene_set_name` Pseudobulk Initial Model**

```{r}
ncol <- ncol(markers_list_pseudobulk[[2]])
datatable_download_exp(markers_list_pseudobulk[[2]][(ncol-3):ncol])
```

**Upregulated in Direction of `r names(idents)[1]` `r gene_set_name` Pseudobulk Strict Model**

```{r}
ncol <- ncol(markers_list_pseudobulk_strict[[2]])
datatable_download_exp(markers_list_pseudobulk_strict[[2]][(ncol-3):ncol])
```

**Upregulated in Direction of `r names(idents)[1]` `r gene_set_name` Pseudobulk Stricter Model**

```{r}
ncol <- ncol(markers_list_pseudobulk_stricter[[2]])
datatable_download_exp(markers_list_pseudobulk_stricter[[2]][(ncol-3):ncol])
```

**Upregulated in Direction of `r names(idents)[1]` `r gene_set_name` Pseudobulk Strictest Model**

```{r}
ncol <- ncol(markers_list_pseudobulk_strictest[[2]])
datatable_download_exp(markers_list_pseudobulk_strictest[[2]][(ncol-3):ncol])
```

**Downregulated in Direction of `r names(idents)[1]` `r gene_set_name` Pseudobulk Initial Model**

```{r}
ncol <- ncol(markers_list_pseudobulk[[3]])
datatable_download_exp(markers_list_pseudobulk[[3]][(ncol-3):ncol])
```

**Downregulated in Direction of `r names(idents)[1]` `r gene_set_name` Pseudobulk Strict Model**

```{r}
ncol <- ncol(markers_list_pseudobulk_strict[[3]])
datatable_download_exp(markers_list_pseudobulk_strict[[3]][(ncol-3):ncol])
```

**Downregulated in Direction of `r names(idents)[1]` `r gene_set_name` Pseudobulk Stricter Model**

```{r}
ncol <- ncol(markers_list_pseudobulk_stricter[[3]])
datatable_download_exp(markers_list_pseudobulk_stricter[[3]][(ncol-3):ncol])
```

**Downregulated in Direction of `r names(idents)[1]` `r gene_set_name` Pseudobulk Strictest Model**

```{r}
ncol <- ncol(markers_list_pseudobulk_strictest[[3]])
datatable_download_exp(markers_list_pseudobulk_strictest[[3]][(ncol-3):ncol])
```

**Number of DE Genes Per Model**

```{r}
data <- data.frame(
  x = c(
    dim(markers_list_pseudobulk[[1]])[1], dim(markers_list_pseudobulk_strict[[1]])[1],
    dim(markers_list_pseudobulk_stricter[[1]])[1], dim(markers_list_pseudobulk_strictest[[1]])[1]
  ),
  y = c(
    dim(markers_list_pseudobulk[[1]])[1], dim(markers_list_pseudobulk_strict[[1]])[1],
    dim(markers_list_pseudobulk_stricter[[1]])[1], dim(markers_list_pseudobulk_strictest[[1]])[1]
  )
)
ggplot(data, aes(x, y)) +
  geom_point() +
  geom_smooth(method = lm) +
  geom_label_repel(
    aes(label = c("initial_model", "strict_model", "stricter_model", "strictest_model"))
  ) +
  labs(x = "Number of DE genes", y = "Number of DE genes")
```

## Gene Set Heatmaps {.active}

In this section, DE gene sets and genes are aggregated, filtered, and optimally clustered following the pipeline described in the Flowchart 2 of the section `Prep`.
The main output is a large heatmap with word cloud summarizations of the gene set cluster modules.
We also show heatmaps for each module individually.
The label color indicates the most stringent model that was fit, where black is the most lenient and from yellow to orange to red is more stringent.
The most stringent models intend to test for gene sets that are incrementally increased or decreased across the four cell types in the order of their columns, while less stringent models have relaxed conditions.
Below are the exact contrasts used for each model:

```
# Black
(EC_Reln-IN_Sst, CA1_Wfs1-IN_Sst, CA3_Bok-IN_Sst) # All must be DE.
(EC_Reln-CA1_Wfs1, CA1_Wfs1-CA3_Bok) # Must NOT be DE in opposite direction.

# Yellow
(EC_Reln-IN_Sst, EC_Reln-CA3_Bok, EC_Reln-CA1_Wfs1) # All must be DE.
(CA1_Wfs1-CA3_Bok, CA3_Bok-IN_Sst) # Must NOT be DE in opposite direction.

# Orange
(EC_Reln-IN_Sst, EC_Reln-CA3_Bok, CA1_Wfs1-IN_Sst, CA1_Wfs1-CA3_Bok) # All must be DE.
(EC_Reln-CA1_Wfs1, CA3_Bok-IN_Sst) # Must NOT be DE in opposite direction.

# Red
(EC_Reln-CA1_Wfs1, CA1_Wfs1-CA3_Bok, CA3_Bok-IN_Sst) # All must be DE.
```

```{r, fig.width = 20, fig.height = 10}
# TODO: Comments and refactoring.
# -------------------------------

gene_anno <- data.frame(ensembl_gene_id = rownames(orig_mat))

color4 <- colorRampPalette(c("blue", "white", "red"))(100)

color3 <- colorRamp2(
  c(min(gsva), (max(gsva) + min(gsva)) / 2, max(gsva)),
  c(brewer.pal(3, "RdBu")[3], brewer.pal(3, "RdBu")[2], brewer.pal(3, "RdBu")[1])
)

color2 <- colorRamp2(
  c(min(gsva), (max(gsva) + min(gsva)) / 2, max(gsva)),
  c(brewer.pal(3, "RdBu")[3], brewer.pal(3, "RdBu")[2], brewer.pal(3, "RdBu")[1])
)

gsva_list_pseudobulk_all <- list(
  c(
    rownames(gsva_list_pseudobulk[[1]]), rownames(gsva_list_pseudobulk_strict[[1]]),
    rownames(gsva_list_pseudobulk_stricter[[1]]), rownames(gsva_list_pseudobulk_strictest[[1]])
  ),
  c(
    rownames(gsva_list_pseudobulk[[2]]), rownames(gsva_list_pseudobulk_strict[[2]]),
    rownames(gsva_list_pseudobulk_stricter[[2]]), rownames(gsva_list_pseudobulk_strictest[[2]])
  ),
  c(
    rownames(gsva_list_pseudobulk[[3]]), rownames(gsva_list_pseudobulk_strict[[3]]),
    rownames(gsva_list_pseudobulk_stricter[[3]]), rownames(gsva_list_pseudobulk_strictest[[3]])
  )
)

markers_list_pseudobulk_all <- list(
  c(
    rownames(markers_list_pseudobulk[[1]]),
    rownames(markers_list_pseudobulk_strict[[1]]),
    rownames(markers_list_pseudobulk_stricter[[1]]),
    rownames(markers_list_pseudobulk_strictest[[1]])
  ),
  c(
    rownames(markers_list_pseudobulk[[2]]),
    rownames(markers_list_pseudobulk_strict[[2]]),
    rownames(markers_list_pseudobulk_stricter[[2]]),
    rownames(markers_list_pseudobulk_strictest[[2]])
  ),
  c(
    rownames(markers_list_pseudobulk[[3]]),
    rownames(markers_list_pseudobulk_strict[[3]]),
    rownames(markers_list_pseudobulk_stricter[[3]]),
    rownames(markers_list_pseudobulk_strictest[[3]])
  )
)

mat <- gsva
mat <- mat[rownames(mat) %in% gsva_list_pseudobulk_all[[1]], ]
mat_up <- mat[rownames(mat) %in% gsva_list_pseudobulk_all[[2]], ]
mat_down <- mat[rownames(mat) %in% gsva_list_pseudobulk_all[[3]], ]
gene_sets_sub_up <- gene_sets[names(gene_sets) %in% rownames(mat_up)]
up_vector <- vector()
for (j in seq(length(gene_sets_sub_up@.Data))) {
  if (any(gene_sets_sub_up[[j]]@geneIds %in% rownames(markers_list_pseudobulk[[2]]))) {
    up_vector <- append(up_vector, names(gene_sets_sub_up)[j])
  }
}
gene_sets_sub_down <- gene_sets[names(gene_sets) %in% rownames(mat_down)]
down_vector <- c()
for (j in seq(length(gene_sets_sub_down@.Data))) {
  if (any(gene_sets_sub_down[[j]]@geneIds %in% rownames(markers_list_pseudobulk[[3]]))) {
    down_vector <- append(down_vector, names(gene_sets_sub_down)[j])
  }
}
all_vector <- append(up_vector, down_vector)
mat <- mat[rownames(mat) %in% all_vector, ]
gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(mat)]
gene_sets_sub_ids <- gene_sets_sub
for (j in seq(length(gene_sets_sub_ids@.Data))) {
  suppressWarnings(gene_sets_sub_ids[[j]]@setName <- gene_sets_sub_ids[[j]]@shortDescription)
}

discard <- computeGeneSetsOverlapMax(
  gene_sets_sub_ids, unique(unlist(geneIds(gene_sets_sub_ids)))
)
discard[!upper.tri(discard)] <- 0
discard <- colnames(discard)[colSums(discard) == 0]
discard <- which(names(gene_sets_sub_ids) %in% discard)
mat <- mat[-discard, ]
gene_sets_sub <- gene_sets_sub[-discard, ]
gene_sets_sub_ids <- gene_sets_sub_ids[-discard, ]

gene_set_overlap <- computeGeneSetsOverlapMax(
  gene_sets_sub_ids, unique(unlist(geneIds(gene_sets_sub_ids)))
)
clustering <- hclust(get_dist(gene_set_overlap, "spearman"), "ward.D2")

anno <- HeatmapAnnotation(
  cell_type = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = c(levels(idents)[[1]], levels(idents)[[2]]),
    labels_gp = gpar(col = "black", fontsize = 20)
  )
)

modules_up <- rownames(mat)[rownames(mat) %in% gsva_list_pseudobulk_all[[2]]]
modules_down <- rownames(mat)[rownames(mat) %in% gsva_list_pseudobulk_all[[3]]]

module_gene_sets_sub <- gene_sets[names(gene_sets) %in% modules_up]
frequent_genes_tmp <- as.data.frame(table(unlist(geneIds(module_gene_sets_sub))))
frequent_genes_tmp <- frequent_genes_tmp[
  frequent_genes_tmp$Var1 %in% gene_anno$ensembl_gene_id,
]
rownames(frequent_genes_tmp) <- NULL
frequent_genes_up <- frequent_genes_tmp

module_gene_sets_sub <- gene_sets[names(gene_sets) %in% modules_down]
frequent_genes_tmp <- as.data.frame(table(unlist(geneIds(module_gene_sets_sub))))
frequent_genes_tmp <- frequent_genes_tmp[
  frequent_genes_tmp$Var1 %in% gene_anno$ensembl_gene_id,
]
rownames(frequent_genes_tmp) <- NULL
frequent_genes_down <- frequent_genes_tmp
frequent_genes <- rbind(frequent_genes_up, frequent_genes_down)

node_data <- melt(geneIds(gene_sets_sub))
node_data <- node_data[node_data$value %in% frequent_genes$Var1, ]
names(node_data)[1] <- "ensembl_gene_id"
names(node_data)[2] <- "gene_set"
node_data$direction <- ifelse(
  node_data$gene_set %in% gsva_list_pseudobulk_all[[2]], "up", "down"
)

unique_direction <- as.data.frame.matrix(table(node_data[ , c(1, 3)]), )
unique_direction$down <- unique_direction$down * -1
module_membership_norm <- unique_direction
module_membership_norm <- module_membership_norm[
  abs(rowSums(module_membership_norm)) >= 2,
]

hub_gene_all_GO <- table(unlist(geneIds(gene_sets)))
hub_gene_all_GO <- hub_gene_all_GO[
  names(hub_gene_all_GO) %in% rownames(module_membership_norm)
]

module_membership_thresh <- median(
  (abs(rowSums(module_membership_norm)) / hub_gene_all_GO) *
    abs(rowSums(module_membership_norm))
)
# module_membership_thresh <- summary(
#   as.numeric(
#     (abs(rowSums(module_membership_norm)) / hub_gene_all_GO) *
#       abs(rowSums(module_membership_norm))
#   )
# )[2][[1]]

modules_with_hubs <- vector("list", 26)
modules_with_hubs[[1]] <- 0
hub_gene_list <- list(gene_anno)
for (l in 2:26) {
  set.seed(1)
  heatmap <- Heatmap(
    mat, cluster_rows = clustering, cluster_columns = FALSE, show_column_names = FALSE,
    show_row_names = FALSE, show_heatmap_legend = FALSE, row_split = l,
  )
  modules <- suppressWarnings(row_order(heatmap))
  hub_gene_list <- vector("list", length(modules))
  for (m in seq_along(modules)) {
    module_sets <- rownames(mat)[modules[[m]]]
    modules_up <- module_sets[module_sets %in% gsva_list_pseudobulk_all[[2]]]
    modules_down <- module_sets[module_sets %in% gsva_list_pseudobulk_all[[3]]]
    module_hub_genes <- vector()
    
    if (length(modules_up) == 0) {
    } else {
      module_gene_sets_sub <- gene_sets[names(gene_sets) %in% modules_up]
      frequent_genes <- as.data.frame(table(unlist(geneIds(module_gene_sets_sub))))
      hub_gene_all_GO_sub <- hub_gene_all_GO[
        names(hub_gene_all_GO) %in% frequent_genes$Var1
      ]
      frequent_genes <- frequent_genes[frequent_genes$Var1 %in% names(hub_gene_all_GO), ]
      frequent_genes$Freq <-
        (frequent_genes$Freq / hub_gene_all_GO_sub) * frequent_genes$Freq
      frequent_genes <- frequent_genes[
        frequent_genes$Freq >= module_membership_thresh,
      ]
      module_hub_genes_tmp <- gene_anno[
        gene_anno$ensembl_gene_id %in% frequent_genes$Var1,
      ]
      module_hub_genes_tmp <- module_hub_genes_tmp[
        module_hub_genes_tmp %in% markers_list_pseudobulk_all[[2]]
      ]
      module_hub_genes <- module_hub_genes_tmp
    }
    
    if (length(modules_down) == 0) {
    } else {
      module_gene_sets_sub <- gene_sets[names(gene_sets) %in% modules_down]
      frequent_genes <- as.data.frame(table(unlist(geneIds(module_gene_sets_sub))))
      hub_gene_all_GO_sub <- hub_gene_all_GO[
        names(hub_gene_all_GO) %in% frequent_genes$Var1
      ]
      frequent_genes <- frequent_genes[frequent_genes$Var1 %in% names(hub_gene_all_GO), ]
      frequent_genes$Freq <-
        (frequent_genes$Freq / hub_gene_all_GO_sub) * frequent_genes$Freq
      frequent_genes <- frequent_genes[
        frequent_genes$Freq >= module_membership_thresh,
      ]
      module_hub_genes_tmp <- gene_anno[
        gene_anno$ensembl_gene_id %in% frequent_genes$Var1,
      ]
      module_hub_genes_tmp <- module_hub_genes_tmp[
        module_hub_genes_tmp %in% markers_list_pseudobulk_all[[3]]
      ]
      module_hub_genes <- append(module_hub_genes, module_hub_genes_tmp)
    }
    
    hub_gene_list[[m]] <- module_hub_genes
  }
  modules_with_hubs[[l]] <- length(which(lengths(hub_gene_list) != 0)) / l
}
optimal_k <- 27 - which.max(rev(modules_with_hubs))
module_colors <- cols25(optimal_k)
clustering <- color_branches(clustering, optimal_k, col = module_colors)
set.seed(1)
heatmap <- Heatmap(
  mat, cluster_rows = clustering, cluster_columns = FALSE, show_column_names = FALSE,
  show_row_names = FALSE, show_heatmap_legend = FALSE, row_split = optimal_k
)

modules <- suppressWarnings(row_order(heatmap))
module_names <- modules
go_anno <- select(GO.db, names(gene_sets_sub_ids), c("GOID", "TERM", "DEFINITION"), "GOID")
for (l in seq_along(module_names)) {
  module_names[[l]] <- paste0(
    rownames(mat)[module_names[[l]]], " ", go_anno$DEFINITION[module_names[[l]]]
  )
}
# for (l in seq_along(module_names)) {
#   module_names[[l]] <- paste0(rownames(mat)[module_names[[l]]])
# }
top_words_tmp <- data.frame(
  gene_sets = unlist(module_names),
  module = rep(seq_along(module_names), lapply(module_names, length))
)
top_words_all <- suppressWarnings(
  as.character(
    tapply(top_words_tmp$gene_sets, top_words_tmp$module, word_cloud)
  )
)
top_words <- suppressWarnings(
  as.character(
    tapply(top_words_tmp$gene_sets, top_words_tmp$module, word_cloud, width = 5)
  )
)
row_anno <-  HeatmapAnnotation(
  top_words = anno_block(
    gp = gpar(col = module_colors), labels = seq(optimal_k),
    labels_rot = 0, labels_gp = gpar(fontsize = 20)
  ),
  which = "row"
)

row_colors <- rownames(mat)
row_colors <- data.frame(
  row_colors = row_colors,
  DE_strict = row_colors %in% rownames(gsva_list_pseudobulk_strict[[1]]),
  DE_stricter = row_colors %in% rownames(gsva_list_pseudobulk_stricter[[1]]),
  DE_strictest = row_colors %in% rownames(gsva_list_pseudobulk_strictest[[1]])
) %>%
  mutate(
    type = case_when(
      DE_strictest == TRUE ~ "firebrick",
      DE_stricter == TRUE ~ "darkorange",
      DE_strict == TRUE ~ "darkgoldenrod",
      TRUE ~ "black"
    )
  )
row_colors <- row_colors$type

# idx <- which.max(nchar(rownames(mat)))
# rownames(mat)[idx] <- "oxidoreductase activity, acting on paired donors..."

draw(
  Heatmap(
    mat,
    color4,
    cluster_rows = clustering,
    cluster_columns = FALSE,
    row_names_max_width = max_text_width(rownames(mat), gpar(fontsize = 10)),
    show_column_names = FALSE,
    column_split = idents,
    top_annotation = anno,
    rect_gp = gpar(col = "lightgray"),
    column_title = " ",
    row_split = optimal_k,
    row_title = top_words,
    row_title_rot = 0,
    row_title_gp = gpar(fontsize = 12),
    row_names_gp = gpar(fontsize = 10, col = unlist(row_colors)),
    left_annotation = row_anno,
    row_gap = unit(2, "mm"),
    column_gap = unit(2, "mm"),
    heatmap_legend_param = list(
      title = "scaled ssGSEA score", direction = "horizontal", legend_width = unit(7.5, "cm")
    )
  ),
  heatmap_legend_side = "top",
  annotation_legend_side = "top"
)

hub_gene_data <- vector()
for (l in seq_along(modules)) {
  module_sets <- rownames(mat)[modules[[l]]]
  modules_up <- module_sets[module_sets %in% gsva_list_pseudobulk_all[[2]]]
  modules_down <- module_sets[module_sets %in% gsva_list_pseudobulk_all[[3]]]
  frequent_genes <- vector()
  frequent_genes_up <- vector()
  frequent_genes_down <- vector()
  
  if (length(modules_up) == 0) {
  } else {
    module_gene_sets_sub <- gene_sets[names(gene_sets) %in% modules_up]
    frequent_genes_tmp <- as.data.frame(table(unlist(geneIds(module_gene_sets_sub))))
    frequent_genes_tmp <- frequent_genes_tmp[
      frequent_genes_tmp$Var1 %in% gene_anno$ensembl_gene_id,
    ]
    rownames(frequent_genes_tmp) <- NULL
    if (any(frequent_genes_tmp$Var1 %in% markers_list_pseudobulk_all[[1]])) {
      discard <- which(
        frequent_genes_tmp$Var1 %in% markers_list_pseudobulk_all[[1]] &
          frequent_genes_tmp$Var1 %in% markers_list_pseudobulk_all[[3]]
      )
      if (length(discard) > 0) {
        frequent_genes_tmp <- frequent_genes_tmp[-discard, ]
      }
    }
    frequent_genes_up <- frequent_genes_tmp
  }
  
  if (length(modules_down) == 0) {
  } else {
    module_gene_sets_sub <- gene_sets[names(gene_sets) %in% modules_down]
    frequent_genes_tmp <- as.data.frame(table(unlist(geneIds(module_gene_sets_sub))))
    frequent_genes_tmp <- frequent_genes_tmp[
      frequent_genes_tmp$Var1 %in% gene_anno$ensembl_gene_id,
    ]
    rownames(frequent_genes_tmp) <- NULL
    if (any(frequent_genes_tmp$Var1 %in% markers_list_pseudobulk_all[[1]])) {
      discard <- which(
        frequent_genes_tmp$Var1 %in% markers_list_pseudobulk_all[[1]] &
          frequent_genes_tmp$Var1 %in% markers_list_pseudobulk_all[[2]]
      )
      if (length(discard) > 0) {
        frequent_genes_tmp <- frequent_genes_tmp[-discard, ]
      }
    }
    frequent_genes_down <- frequent_genes_tmp
  }
  
  frequent_genes <- rbind(frequent_genes_up, frequent_genes_down)
  if (dim(frequent_genes)[1] != 0) {
    module_gene_sets_sub <- gene_sets[names(gene_sets) %in% module_sets]
    node_data <- melt(geneIds(module_gene_sets_sub))
    node_data <- node_data[node_data$value %in% frequent_genes$Var1, ]
    names(node_data)[1] <- "ensembl_gene_id"
    names(node_data)[2] <- "gene_set"
    gene_anno_sub <- gene_anno[
      gene_anno$ensembl_gene_id %in% node_data$ensembl_gene_id, -3
    ]
    node_data <- node_data[node_data$ensembl_gene_id %in% gene_anno_sub, ]
    node_data$direction <- ifelse(
      node_data$gene_set %in% gsva_list_pseudobulk_all[[2]], "up", "down"
    )
    node_data$DE <- ifelse(
      node_data$ensembl_gene_id %in% markers_list_pseudobulk_all[[1]], "yes", "no"
    )
    node_data$module <- l
    hub_gene_data <- rbind(hub_gene_data, node_data)
  }
}

add <- data.frame(external_gene_name = hub_gene_data$ensembl_gene_id)
hub_gene_data1 <- cbind(hub_gene_data[ , 1:2], add)
hub_gene_data <- cbind(hub_gene_data1, hub_gene_data[ , 3:5])

unique_direction <- as.data.frame.matrix(table(hub_gene_data[ , c(3, 4)]), )
unique_direction$down <- unique_direction$down * -1
discard <- which(rowSums(unique_direction) == 0)
if (length(discard) > 0) {
  unique_direction <- unique_direction[-discard, ]
}
module_membership_norm <- unique_direction
unique_direction <- ifelse(rowSums(unique_direction) > 0, "up", "down")
hub_gene_data <- hub_gene_data[
  hub_gene_data$external_gene_name %in% names(unique_direction),
]

rownames(hub_gene_data) <- NULL
discard <- vector()
for (i_gene in seq_along(hub_gene_data$external_gene_name)) {
  sub_gene <- hub_gene_data$external_gene_name[i_gene]
  sub_unique_direction <- unique_direction[which(names(unique_direction) == sub_gene)]
  if (hub_gene_data$direction[i_gene] != sub_unique_direction[[1]]) {
    discard <- append(discard, i_gene)
  }
}
hub_gene_data <- hub_gene_data[-discard, ]

frequent_genes_tmp <- as.data.frame(table(hub_gene_data$external_gene_name))
frequent_genes <- frequent_genes_tmp
hub_gene_data <- hub_gene_data[
  hub_gene_data$external_gene_name %in% frequent_genes$Var1,
]

hub_gene_data_up <- hub_gene_data[which(hub_gene_data$direction == "up"), ]
hub_gene_data_down <- hub_gene_data[which(hub_gene_data$direction == "down"), ]

# hub_mat <- as.data.frame.matrix(table(hub_gene_data_up[ , c(3, 2)]), )
# hub_mat <- hub_mat == 1
# co_mat <- hub_mat %*% t(hub_mat)
# diag(co_mat) <- 0
# graph_up <- graph_from_adjacency_matrix(co_mat, "upper", TRUE)
# 
# edges_up <- get.edgelist(graph_up, FALSE)
# set.seed(1)
# layout_up <- qgraph.layout.fruchtermanreingold(
#   edges_up, vcount = vcount(graph_up), E(graph_up)$weight,
#   area = vcount(graph_up) ^ 2, repulse.rad = vcount(graph_up) ^ 2
# )
# 
# hub_gene_all_GO <- table(unlist(geneIds(gene_sets)))
# hub_gene_all_GO <- hub_gene_all_GO[
#   names(hub_gene_all_GO) %in% unique(hub_gene_data_up$ensembl_gene_id)
# ]
# sub_gene_anno <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(hub_gene_all_GO), ]
# rownames(sub_gene_anno) <- NULL
# sub_gene_anno <- sub_gene_anno[
#   match(rownames(hub_gene_all_GO), sub_gene_anno)
# ]
# rownames(hub_gene_all_GO) <- sub_gene_anno
# hub_gene_all_GO_up <- hub_gene_all_GO[order(names(hub_gene_all_GO))]
# 
# module_membership_up <- as.data.frame.matrix(table(hub_gene_data_up[ , c(3, 6)]), )
# module_membership_binary_up <- module_membership_up
# module_membership_binary_up[module_membership_binary_up > 0] <- 1
# module_membership_list_up <- lapply(
#   split(module_membership_up, row.names(module_membership_up)), unlist
# )
# module_network_colors_up <- module_colors[unique(hub_gene_data_up$module)]
# V(graph_up)$pie.color = list(module_network_colors_up)
# 
# module_membership_norm_up <- module_membership_norm[
#   rownames(module_membership_norm) %in% rownames(module_membership_up),
# ]
# 
# V(graph_up)$vertex.pie = list(module_network_colors_up)
# V(graph_up)$vertex.shape <- "pie"
# V(graph_up)$vertex.color <- module_membership_list_up
# V(graph_up)$vertex.label.family = "sans"
# 
# DE_up <- vector()
# for (l in seq(length(V(graph_up)))) {
#   if (V(graph_up)[l]$name %in% rownames(markers_list_pseudobulk[[2]])) {
#     DE_up <- append(DE_up, "yes")
#   } else {
#     DE_up <- append(DE_up, "no")
#   }
# }
# 
# DE_opp_up <- vector()
# for (l in seq(length(V(graph_up)))) {
#   if (V(graph_up)[l]$name %in% markers_list_pseudobulk_all[[3]]) {
#     DE_opp_up <- append(DE_opp_up, "yes")
#   } else {
#     DE_opp_up <- append(DE_opp_up, "no")
#   }
# }
# 
# DE_strict_up <- vector()
# for (l in seq(length(V(graph_up)))) {
#   if (V(graph_up)[l]$name %in% rownames(markers_list_pseudobulk_strict[[2]])) {
#     DE_strict_up <- append(DE_strict_up, "yes")
#   } else {
#     DE_strict_up <- append(DE_strict_up, "no")
#   }
# }
# 
# DE_stricter_up <- vector()
# for (l in seq(length(V(graph_up)))) {
#   if (V(graph_up)[l]$name %in% rownames(markers_list_pseudobulk_stricter[[2]])) {
#     DE_stricter_up <- append(DE_stricter_up, "yes")
#   } else {
#     DE_stricter_up <- append(DE_stricter_up, "no")
#   }
# }
# 
# DE_strictest_up <- vector()
# for (l in seq(length(V(graph_up)))) {
#   if (V(graph_up)[l]$name %in% rownames(markers_list_pseudobulk_strictest[[2]])) {
#     DE_strictest_up <- append(DE_strictest_up, "yes")
#   } else {
#     DE_strictest_up <- append(DE_strictest_up, "no")
#   }
# }
# 
# direction_up <- vector()
# for (l in seq(length(E(graph_up)))) {
#   hub_gene_sub_up <- hub_gene_data_up[
#     hub_gene_data_up$external_gene_name %in% head_of(graph_up, l)[[1]]$name,
#     ]
#   direction_up <- append(direction_up, unique(hub_gene_sub_up$direction))
# }
# E(graph_up)$edge.color <- ifelse(direction_up == "up", "firebrick1", "dodgerblue")
# 
# hub_mat <- as.data.frame.matrix(table(hub_gene_data_down[ , c(3, 2)]), )
# hub_mat <- hub_mat == 1
# co_mat <- hub_mat %*% t(hub_mat)
# diag(co_mat) <- 0
# graph_down <- graph_from_adjacency_matrix(co_mat, "upper", TRUE)
# 
# edges_down <- get.edgelist(graph_down, FALSE)
# set.seed(1)
# layout_down <- qgraph.layout.fruchtermanreingold(
#   edges_down, vcount = vcount(graph_down), E(graph_down)$weight,
#   area = vcount(graph_down) ^ 2, repulse.rad = vcount(graph_down) ^ 2
# )
# 
# hub_gene_all_GO <- table(unlist(geneIds(gene_sets)))
# hub_gene_all_GO <- hub_gene_all_GO[
#   names(hub_gene_all_GO) %in% unique(hub_gene_data_down$ensembl_gene_id)
# ]
# sub_gene_anno <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(hub_gene_all_GO), ]
# rownames(sub_gene_anno) <- NULL
# sub_gene_anno <- sub_gene_anno[
#   match(rownames(hub_gene_all_GO), sub_gene_anno$ensembl_gene_id),
# ]
# rownames(hub_gene_all_GO) <- sub_gene_anno$external_gene_name
# hub_gene_all_GO_down <- hub_gene_all_GO[order(names(hub_gene_all_GO))]
# 
# module_membership_down <- as.data.frame.matrix(table(hub_gene_data_down[ , c(3, 6)]), )
# module_membership_binary_down <- module_membership_down
# module_membership_binary_down[module_membership_binary_down > 0] <- 1
# module_membership_list_down <- lapply(
#   split(module_membership_down, row.names(module_membership_down)), unlist
# )
# module_network_colors_down <- module_colors[unique(hub_gene_data_down$module)]
# V(graph_down)$pie.color = list(module_network_colors_down)
# 
# module_membership_norm_down <- module_membership_norm[
#   rownames(module_membership_norm) %in% rownames(module_membership_down),
# ]
# 
# V(graph_down)$pie.color = list(module_network_colors_down)
# V(graph_down)$vertex.shape <- "pie"
# V(graph_down)$vertex.color <- module_membership_list_down
# V(graph_down)$vertex.label.family = "sans"
# 
# DE_down <- vector()
# for (l in seq(length(V(graph_down)))) {
#   if (V(graph_down)[l]$name %in% rownames(markers_list_pseudobulk[[3]])) {
#     DE_down <- append(DE_down, "yes")
#   } else {
#     DE_down <- append(DE_down, "no")
#   }
# }
# 
# DE_opp_down <- vector()
# for (l in seq(length(V(graph_down)))) {
#   if (V(graph_down)[l]$name %in% markers_list_pseudobulk_all[[2]]) {
#     DE_opp_down <- append(DE_opp_down, "yes")
#   } else {
#     DE_opp_down <- append(DE_opp_down, "no")
#   }
# }
# 
# DE_strict_down <- vector()
# for (l in seq(length(V(graph_down)))) {
#   if (V(graph_down)[l]$name %in% rownames(markers_list_pseudobulk_strict[[3]])) {
#     DE_strict_down <- append(DE_strict_down, "yes")
#   } else {
#     DE_strict_down <- append(DE_strict_down, "no")
#   }
# }
# 
# DE_stricter_down <- vector()
# for (l in seq(length(V(graph_down)))) {
#   if (
#     V(graph_down)[l]$name %in% rownames(markers_list_pseudobulk_stricter[[3]])
#   ) {
#     DE_stricter_down <- append(DE_stricter_down, "yes")
#   } else {
#     DE_stricter_down <- append(DE_stricter_down, "no")
#   }
# }
# 
# DE_strictest_down <- vector()
# for (l in seq(length(V(graph_down)))) {
#   if (
#     V(graph_down)[l]$name %in% rownames(markers_list_pseudobulk_strictest[[3]])
#   ) {
#     DE_strictest_down <- append(DE_strictest_down, "yes")
#   } else {
#     DE_strictest_down <- append(DE_strictest_down, "no")
#   }
# }
# 
# direction_down <- vector()
# for (l in seq(length(E(graph_down)))) {
#   hub_gene_sub_down <- hub_gene_data_down[
#     hub_gene_data_down$external_gene_name %in% head_of(graph_down, l)[[1]]$name,
#     ]
#   direction_down <- append(direction_down, unique(hub_gene_sub_down$direction))
# }
# E(graph_down)$edge.color <- ifelse(direction_down == "up", "firebrick1", "dodgerblue")

hub_mat <- as.data.frame.matrix(table(hub_gene_data[ , c(3, 2)]), )
hub_mat <- hub_mat == 1
co_mat <- hub_mat %*% t(hub_mat)
diag(co_mat) <- 0
graph <- graph_from_adjacency_matrix(co_mat, "upper", TRUE)

edges <- get.edgelist(graph, FALSE)
set.seed(1)
layout <- qgraph.layout.fruchtermanreingold(
  edges, vcount = vcount(graph), E(graph)$weight,
  area = vcount(graph) ^ 2, repulse.rad = vcount(graph) ^ 2
)

hub_gene_all_GO <- table(unlist(geneIds(gene_sets)))
hub_gene_all_GO <- hub_gene_all_GO[
  names(hub_gene_all_GO) %in% unique(hub_gene_data$ensembl_gene_id)
]
# sub_gene_anno <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(hub_gene_all_GO), ]
# rownames(sub_gene_anno) <- NULL
# sub_gene_anno <- sub_gene_anno[
#   match(rownames(hub_gene_all_GO), sub_gene_anno$ensembl_gene_id),
# ]
# rownames(hub_gene_all_GO) <- sub_gene_anno$external_gene_name
hub_gene_all_GO <- hub_gene_all_GO[order(names(hub_gene_all_GO))]

module_membership <- as.data.frame.matrix(table(hub_gene_data[ , c(3, 6)]), )
module_membership_binary <- module_membership
module_membership_binary[module_membership_binary > 0] <- 1
module_membership_list <- lapply(
  split(module_membership, row.names(module_membership)), unlist
)
module_network_colors <- module_colors
module_membership_norm <- module_membership_norm[
  rownames(module_membership_norm) %in% rownames(module_membership),
]

V(graph)$pie.color = list(module_network_colors)
V(graph)$vertex.shape <- "pie"
V(graph)$vertex.color <- module_membership_list
V(graph)$vertex.label.family = "sans"

module_max_genes <- apply(module_membership_norm, 2, function (x) which(x == max(x)))
module_max_genes <- rownames(module_membership_norm)[unlist(module_max_genes)]
module_max <- vector()
for (l in seq(length(V(graph)))) {
  if (V(graph)[l]$name %in% module_max_genes) {
    module_max <- append(module_max, "yes")
  } else {
    module_max <- append(module_max, "no")
  }
}

unique_markers <- rownames(markers_list_pseudobulk[[2]])[
  rownames(markers_list_pseudobulk[[2]]) %notin%
    hub_gene_data_down$external_gene_name
]
unique_markers <- append(
  unique_markers,
  rownames(markers_list_pseudobulk[[3]])[
    rownames(markers_list_pseudobulk[[3]]) %notin%
      hub_gene_data_up$external_gene_name
  ]
)
DE <- vector()
for (l in seq(length(V(graph)))) {
  if (V(graph)[l]$name %in% unique_markers) {
    DE <- append(DE, "yes")
  } else {
    DE <- append(DE, "no")
  }
}

direction <- vector()
for (l in seq(length(E(graph)))) {
  hub_gene_sub <- hub_gene_data[
    hub_gene_data$external_gene_name %in% head_of(graph, l)[[1]]$name,
  ]
  direction <- append(direction, unique(hub_gene_sub$direction))
}
E(graph)$edge.color <- ifelse(direction == "up", "firebrick1", "dodgerblue")
# -------------------------------
```

```{r, fig.height = 20, fig.width = 20}
# module_names <- modules
# for (l in seq_along(module_names)) {
#   module_names[[l]] <- paste0(rownames(mat)[module_names[[l]]])
# }
# row_colors <- module_names
# for (l in seq_along(row_colors)) {
#   tmp <- data.frame(
#     row_colors = row_colors[[l]],
#     DE_strict = row_colors[[l]] %in% rownames(gsva_list_pseudobulk_strict[[1]]),
#     DE_stricter = row_colors[[l]] %in% rownames(gsva_list_pseudobulk_stricter[[1]]),
#     DE_strictest = row_colors[[l]] %in% rownames(gsva_list_pseudobulk_strictest[[1]])
#   ) %>%
#     mutate(
#       type = case_when(
#         DE_strictest == TRUE ~ "firebrick",
#         DE_stricter == TRUE ~ "darkorange",
#         DE_strict == TRUE ~ "darkgoldenrod",
#         TRUE ~ "black"
#       )
#     )
#   tmp <- tmp$type
#   row_colors[[l]] <- tmp
# }
# 
# for (l in seq_along(module_names)) {
#   exp_mat <- mat[rownames(mat) %in% module_names[[l]], ]
#   sort_mat <- mat[unlist(suppressWarnings(row_order(heatmap))), ]
#   sort_mat <- sort_mat[rownames(sort_mat) %in% rownames(exp_mat), ]
#   exp_mat <- exp_mat[match(rownames(sort_mat), rownames(exp_mat)), ]
#   draw(
#     Heatmap(
#       t(
#         apply(
#           exp_mat, 1,
#           function (x) (x - min(x)) / (max(x) - min(x))
#         )
#       ),
#       color,
#       cluster_columns = FALSE,
#       cluster_rows = FALSE,
#       row_names_max_width = max_text_width(rownames(exp_mat), gpar(fontsize = 20)),
#       show_column_names = FALSE,
#       column_split = sce$idents,
#       top_annotation = anno,
#       row_names_gp = gpar(fontsize = 20, col = row_colors[[l]]),
#       rect_gp = gpar(col = "lightgray"),
#       column_title = paste0("Module ", l, ":\n", top_words[l]),
#       column_title_gp = gpar(fontsize = 20),
#       column_gap = unit(2, "mm"),
#       heatmap_legend_param = list(
#         title = "scaled ssGSEA score", direction = "horizontal", legend_width = unit(7.5, "cm")
#       )
#     ),
#     heatmap_legend_side = "top",
#     annotation_legend_side = "top"
#   )
# }
```

## Hub Gene Network

In order to pinpoint genes that may have the most functional contribution to a model, we create a graph network following the criteria in Flowchart 2 of the section `Prep`.
Note that due to plotting errors in R Markdown, we simply load pre-computed images in and leave the code commented out for now.

```{r, fig.width = 20, fig.height = 40}
# TODO: Comments, refactoring, and fix R Markdown plots.
# ------------------------------------------------------
DE_alt <- c(
  rownames(markers_list_pseudobulk[[1]]),
  rownames(markers_list_pseudobulk_strict[[1]]),
  rownames(markers_list_pseudobulk_stricter[[1]]),
  rownames(markers_list_pseudobulk_strictest[[1]])
)

all_DE_match <- as.data.frame(
  table(V(graph)$name[match(DE_alt, V(graph)$name)]), stringsAsFactors = FALSE
)
all_DE_match_tmp <- data.frame(Var1 = V(graph)$name[V(graph)$name %notin% DE_alt], Freq = 0)
all_DE_match <- rbind(all_DE_match, all_DE_match_tmp)
all_DE_match <- all_DE_match[order(all_DE_match$Var1), ]

module_membership_median <- median(
  (abs(rowSums(module_membership_norm)) / hub_gene_all_GO) *
  abs(rowSums(module_membership_norm))
)
membership <-
  (abs(rowSums(module_membership_norm)) / hub_gene_all_GO) * abs(rowSums(module_membership_norm))

vertex_frame_color <- data.frame(DE = DE) %>%
  mutate(
    type = case_when(
      DE == "yes" ~ "white",
      TRUE ~ "darkgray"
    )
  )
vertex_frame_color <- vertex_frame_color$type

vertex_label_color <- data.frame(DE = DE) %>%
  mutate(
    type = case_when(
      DE == "yes" ~ "white",
      TRUE ~ "darkgray"
    )
  )
vertex_label_color <- vertex_label_color$type

vertex_label_font <- data.frame(DE = DE) %>%
  mutate(
    type = case_when(
      TRUE ~ 4
    )
  )
vertex_label_font <- vertex_label_font$type

vertex_label <- data.frame(DE = DE) %>%
  mutate(
    type = case_when(
      TRUE ~ V(graph)$name
    )
  )
vertex_label <- vertex_label$type
vertex_label[vertex_label == "NA"] <- NA

png(
  file.path(results_dir, "otero-garcia-2020-ssRNAseq", "comb", "05-gsva", "simulation-network.png"), 150,
  150, "in",
  res = 72
)
par(bg = "black", mar = c(5.1, 4.1, 4.1, 450.1), xpd = TRUE)
plot(
    graph,
    rescale = FALSE,
    xlim = c(-1, 1),
    ylim = c(-1, 1),
    layout = norm_coords(layout),
    vertex.size = sqrt(
      (abs(rowSums(module_membership_norm)) / hub_gene_all_GO) *
      abs(rowSums(module_membership_norm))
    ) * 4,
    vertex.label.color = vertex_label_color,
    edge.width = E(graph)$weight ^ 2 / 4,
    vertex.shape = "pie",
    vertex.pie = module_membership_list,
    edge.color = ifelse(direction == "up", "firebrick1", "dodgerblue"),
    vertex.label.family = "sans",
    vertex.label.font = vertex_label_font,
    vertex.label.cex = ifelse(
      DE == "yes", (all_DE_match$Freq + 1), 1
    ),
    vertex.frame.color = vertex_frame_color,
    vertex.label = vertex_label
)
legend(
    "right",
    legend = paste0(
      "Module ", unique(hub_gene_data$module),
      ": ", gsub("\n", " ", top_words[unique(hub_gene_data$module)])
    ),
    pch = 20,
    col = module_network_colors,
    pt.cex = 15,
    cex = 10,
    bty = "n",
    text.col = "white",
    inset = c(-0.9, 0)
)

# DE_alt <- c(
#   rownames(markers_list_pseudobulk[[1]]),
#   rownames(markers_list_pseudobulk_strict[[1]]),
#   rownames(markers_list_pseudobulk_stricter[[1]]),
#   rownames(markers_list_pseudobulk_strictest[[1]])
# )
# 
# all_DE_match <- as.data.frame(
#   table(V(graph)$name[match(DE_alt, V(graph)$name)]), stringsAsFactors = FALSE
# )
# all_DE_match_tmp <- data.frame(Var1 = V(graph)$name[V(graph)$name %notin% DE_alt], Freq = 0)
# all_DE_match <- rbind(all_DE_match, all_DE_match_tmp)
# all_DE_match <- all_DE_match[order(all_DE_match$Var1), ]
# 
# module_membership_median <- median(
#   (abs(rowSums(module_membership_norm)) / hub_gene_all_GO) *
#   abs(rowSums(module_membership_norm))
# )
# # module_membership_thresh <- summary(
# #   as.numeric(
# #     (abs(rowSums(module_membership_norm)) / hub_gene_all_GO) *
# #       abs(rowSums(module_membership_norm))
# #   )
# # )[5][[1]]
# membership <-
#   (abs(rowSums(module_membership_norm)) / hub_gene_all_GO) * abs(rowSums(module_membership_norm))
# 
# vertex_frame_color <- data.frame(DE = DE) %>%
#   mutate(
#     type = case_when(
#       DE == "yes" & membership >= module_membership_thresh ~ "white",
#       TRUE ~ "darkgray"
#     )
#   )
# vertex_frame_color <- vertex_frame_color$type
# 
# vertex_label_color <- data.frame(DE = DE) %>%
#   mutate(
#     type = case_when(
#       DE == "yes" & membership >= module_membership_thresh ~ "white",
#       TRUE ~ "darkgray"
#     )
#   )
# vertex_label_color <- vertex_label_color$type
# 
# vertex_label_font <- data.frame(DE = DE) %>%
#   mutate(
#     type = case_when(
#       TRUE ~ 4
#     )
#   )
# vertex_label_font <- vertex_label_font$type
# 
# vertex_label <- data.frame(DE = DE) %>%
#   mutate(
#     type = case_when(
#       DE == "yes" & membership >= module_membership_thresh ~ V(graph)$name
#     )
#   )
# vertex_label <- vertex_label$type
# vertex_label[vertex_label == "NA"] <- NA
# 
# png(
#   file.path(results_dir, "otero-garcia-2020-ssRNAseq", "comb", "05-gsva", "network.png"), 150,
#   150, "in",
#   res = 72
# )
# par(bg = "black", mar = c(5.1, 4.1, 4.1, 450.1), xpd = TRUE)
# plot(
#     graph,
#     rescale = FALSE,
#     xlim = c(-1, 1),
#     ylim = c(-1, 1),
#     layout = norm_coords(layout),
#     vertex.size = sqrt(
#       (abs(rowSums(module_membership_norm)) / hub_gene_all_GO) *
#       abs(rowSums(module_membership_norm))
#     ) * 5,
#     vertex.label.color = vertex_label_color,
#     edge.width = E(graph)$weight ^ 2 / 4,
#     vertex.shape = "pie",
#     vertex.pie = module_membership_list,
#     edge.color = ifelse(direction == "up", "firebrick1", "dodgerblue"),
#     vertex.label.family = "sans",
#     vertex.label.font = vertex_label_font,
#     vertex.label.cex = ifelse(
#       DE == "yes" & membership >= module_membership_thresh, all_DE_match$Freq / 2, 1
#     ),
#     vertex.frame.color = vertex_frame_color,
#     vertex.label = vertex_label
# )
# legend(
#     "right",
#     legend = paste0(
#       "Module ", unique(hub_gene_data$module),
#       ": ", gsub("\n", " ", top_words[unique(hub_gene_data$module)])
#     ),
#     pch = 20,
#     col = module_network_colors,
#     pt.cex = 15,
#     cex = 10,
#     bty = "n",
#     text.col = "white",
#     inset = c(-1, 0)
# )
# 
# all_DE_match <- as.data.frame(
#   table(V(graph_up)$name[match(DE_alt, V(graph_up)$name)]), stringsAsFactors = FALSE
# )
# all_DE_match_tmp <- data.frame(Var1 = V(graph_up)$name[V(graph_up)$name %notin% DE_alt], Freq = 0)
# all_DE_match <- rbind(all_DE_match, all_DE_match_tmp)
# all_DE_match <- all_DE_match[order(all_DE_match$Var1), ]
# 
# module_membership_median_up <- median(
#   (abs(rowSums(module_membership_norm_up)) / hub_gene_all_GO_up) *
#   abs(rowSums(module_membership_norm_up))
# )
# 
# vertex_frame_color <- data.frame(DE_up = DE_up) %>%
#   mutate(
#     type = case_when(
#       (
#         DE_up == "yes" | DE_strict_up == "yes" | DE_stricter_up == "yes" | DE_strictest_up == "yes"
#       ) & DE_opp_up == "no" ~ "white",
#       TRUE ~ "darkgray"
#     )
#   )
# vertex_frame_color <- vertex_frame_color$type
# 
# vertex_label_color <- data.frame(DE_up = DE_up) %>%
#   mutate(
#     type = case_when(
#       (
#         DE_up == "yes" | DE_strict_up == "yes" | DE_stricter_up == "yes" | DE_strictest_up == "yes"
#       ) & DE_opp_up == "no" ~ "white",
#       TRUE ~ "darkgray"
#     )
#   )
# vertex_label_color <- vertex_label_color$type
# 
# vertex_label_font <- data.frame(DE_up = DE_up) %>%
#   mutate(
#     type = case_when(
#       TRUE ~ 2
#     )
#   )
# vertex_label_font <- vertex_label_font$type
# 
# vertex_label <- data.frame(DE_up = DE_up) %>%
#   mutate(
#     type = case_when(
#       TRUE ~ V(graph_up)$name
#     )
#   )
# vertex_label <- vertex_label$type
# vertex_label[vertex_label == "NA"] <- NA
# 
# png(
#   file.path(results_dir, "kampmann-2020-snRNAseq", "05-gsva", "03-Reln-Wfs1-Bok-network-up.png"), 150,
#   150, "in",
#   res = 72
# )
# par(bg = "black", mar = c(5.1, 4.1, 4.1, 250.1), xpd = TRUE)
# plot(
#     graph_up,
#     rescale = FALSE,
#     xlim = c(-1, 1),
#     ylim = c(-1, 1),
#     layout = norm_coords(layout_up),
#     vertex.size = sqrt(
#       (abs(rowSums(module_membership_norm_up)) / hub_gene_all_GO_up) *
#       abs(rowSums(module_membership_norm_up))
#     ) * 5,
#     vertex.label.color = vertex_label_color,
#     edge.width = E(graph_up)$weight ^ 2 / 4,
#     vertex.shape = "pie",
#     vertex.pie = module_membership_list_up,
#     edge.color = ifelse(direction_up == "up", "firebrick1", "dodgerblue"),
#     vertex.label.family = "sans",
#     vertex.label.font = vertex_label_font,
#     vertex.label.cex = ifelse(
#       (
#         DE_up == "yes" | DE_strict_up == "yes" | DE_stricter_up == "yes" | DE_strictest_up == "yes"
#       ) &
#         DE_opp_up == "no",
#       (all_DE_match$Freq + 3) * 2, 3
#     ),
#     vertex.frame.color = vertex_frame_color,
#     vertex.label = vertex_label
# )
# legend(
#     "right",
#     legend = paste0(
#       "Module ", unique(hub_gene_data_up$module),
#       ": ", gsub("\n", " ", top_words[unique(hub_gene_data_up$module)])
#     ),
#     pch = 20,
#     col = module_network_colors_up,
#     pt.cex = 12,
#     cex = 8,
#     bty = "n",
#     text.col = "white",
#     inset = c(-0.5, 0)
# )
# 
# all_DE_match <- as.data.frame(
#   table(V(graph_down)$name[match(DE_alt, V(graph_down)$name)]), stringsAsFactors = FALSE
# )
# all_DE_match_tmp <- data.frame(
#   Var1 = V(graph_down)$name[V(graph_down)$name %notin% DE_alt], Freq = 0
# )
# all_DE_match <- rbind(all_DE_match, all_DE_match_tmp)
# all_DE_match <- all_DE_match[order(all_DE_match$Var1), ]
# 
# module_membership_median_down <- median(
#   (abs(rowSums(module_membership_norm_down)) / hub_gene_all_GO_down) *
#   abs(rowSums(module_membership_norm_down))
# )
# 
# vertex_frame_color <- data.frame(DE_down = DE_down) %>%
#   mutate(
#     type = case_when(
#       (
#         DE_down == "yes" | DE_strict_down == "yes" |
#           DE_stricter_down == "yes" | DE_strictest_down == "yes"
#       ) &
#         DE_opp_down == "no" ~ "white",
#       TRUE ~ "darkgray"
#     )
#   )
# vertex_frame_color <- vertex_frame_color$type
# 
# vertex_label_color <- data.frame(DE_down = DE_down) %>%
#   mutate(
#     type = case_when(
#       (
#         DE_down == "yes" | DE_strict_down == "yes" |
#           DE_stricter_down == "yes" | DE_strictest_down == "yes"
#       ) &
#         DE_opp_down == "no" ~ "white",
#       TRUE ~ "darkgray"
#     )
#   )
# vertex_label_color <- vertex_label_color$type
# 
# vertex_label_font <- data.frame(DE_down = DE_down) %>%
#   mutate(
#     type = case_when(
#       TRUE ~ 2
#     )
#   )
# vertex_label_font <- vertex_label_font$type
# 
# vertex_label <- data.frame(DE_down = DE_down) %>%
#   mutate(
#     type = case_when(
#       TRUE ~ V(graph_down)$name
#     )
#   )
# vertex_label <- vertex_label$type
# vertex_label[vertex_label == "NA"] <- NA
# 
# png(
#   file.path(results_dir, "kampmann-2020-snRNAseq", "05-gsva", "03-Reln-Wfs1-Bok-network-down.png"), 150,
#   150, "in",
#   res = 72
# )
# par(bg = "black", mar = c(5.1, 4.1, 4.1, 250.1), xpd = TRUE)
# plot(
#     graph_down,
#     rescale = FALSE,
#     xlim = c(-1, 1),
#     ylim = c(-1, 1),
#     layout = norm_coords(layout_down),
#     vertex.size = sqrt(
#       (abs(rowSums(module_membership_norm_down)) / hub_gene_all_GO_down) *
#       abs(rowSums(module_membership_norm_down))
#     ) * 7,
#     vertex.label.color = vertex_label_color,
#     edge.width = E(graph_down)$weight ^ 2 / 2,
#     vertex.shape = "pie",
#     vertex.pie = module_membership_list_down,
#     edge.color = ifelse(direction_down == "up", "firebrick1", "dodgerblue"),
#     vertex.label.family = "sans",
#     vertex.label.font = vertex_label_font,
#     vertex.label.cex = ifelse(
#       (
#         DE_down == "yes" | DE_strict_down == "yes" |
#           DE_stricter_down == "yes" | DE_strictest_down == "yes"
#       ) &
#         DE_opp_down == "no",
#       (all_DE_match$Freq + 3) * 2, 3
#     ),
#     vertex.frame.color = vertex_frame_color,
#     vertex.label = vertex_label
# )
# legend(
#     "right",
#     legend = paste0(
#       "Module ", unique(hub_gene_data_down$module),
#       ": ", gsub("\n", " ", top_words[unique(hub_gene_data_down$module)])
#     ),
#     pch = 20,
#     col = module_network_colors_down,
#     pt.cex = 12,
#     cex = 8,
#     bty = "n",
#     text.col = "white",
#     inset = c(-0.5, 0)
# )
# ------------------------------------------------------
```

## References

This is the concluding section of the document, where we output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
