---
title: '02 Subsets - `r unlist(strsplit(getwd(), "/"))[6]`'
author:
  - name: "Emir Turkes [emir.turkes@eturkes.com]"
  - name: "UK Dementia Research Institute at UCL"
date: '`r strftime(Sys.time(), "%B %d, %Y")`'
bibliography: '../../../`r unlist(strsplit(getwd(), "/"))[4]`.bib'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_file = file.path(
    "..", "..", "..", "results",
    unlist(strsplit(getwd(), "/"))[6], unlist(strsplit(getwd(), "/"))[7], "02-subsets.html"
  ))})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
.tocify-subheader .tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 35px; text-indent: 0;
}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of the [Tau Vulnerability Project](https://github.com/eturkes/tau-vulnerability).*

In this document we create several subsets of the data and rerun the clustering pipeline in `01_prep.Rmd`.
The data here is derived from @`r unlist(strsplit(getwd(), "/"))[6]` and will be referenced using the name ``r unlist(strsplit(getwd(), "/"))[6]``.

```{r}
#    This file is part of tau-vulnerability.
#    Copyright (C) 2019-2021  Emir Turkes, UK DRI at UCL, Columbia University Medical Center
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

options(stringsAsFactors = FALSE)
packages <- c(
  "conflicted", "Seurat", "future", "parallel", "biomaRt", "dplyr",
  "ggplot2", "ggrepel", "viridis", "SingleCellExperiment", "Biostrings", "zinbwave"
)
invisible(suppressPackageStartupMessages(lapply(packages, library, character.only = TRUE)))
conflict_prefer("strsplit", "Biostrings", quiet = TRUE)
conflict_prefer("setdiff", "dplyr", quiet = TRUE)
source(file.path("..", "..", "utils.R"))

# Use UMI protocol due to poor performance of reads protocol.
# -----------------------------------------------------------
protocol <- c("human", "droplet", "single-cell", "umis") # See `cluster_pipeline` in `utils.R`.
# -----------------------------------------------------------

group <- "normal" # Extract group.
vars_to_regress <- NULL # See `cluster_pipeline` in `utils.R`.
parallel_override <- NULL # See `parallel_plan` in `utils.R`.

# Metadata to plot after dimensionality reduction and clustering.
# Values in list can include "no_legend and/or "no_label" to exclude those.
# -------------------------------------------------------------------------
metadata_to_plot <- vector("list", 3)
names(metadata_to_plot) <- c("seurat_clusters", "donor", "Phase")
metadata_to_plot$donor <- "no_label"
metadata_to_plot$Phase <- "no_label"
# -------------------------------------------------------------------------

# List subsets and their associated labels/idents.
# ------------------------------------------------
subsets <- vector("list", 1)
names(subsets) <- "neuronal"
subsets[[1]] <- c(26, 27, 28)
# ------------------------------------------------

analysis_no <- 2
data_name <- unlist(strsplit(getwd(), "/"))[6] # Name of dataset.
assets_dir <- file.path(getwd(), "..", "..", "..", "assets") # Backed up data.
results_dir <- file.path(getwd(), "..", "..", "..", "results")

# Unique cache directory for each analysis number.
# ------------------------------------------------
cache_dir <- file.path(
  "..", "..", "..", "cache", data_name, unlist(strsplit(getwd(), "/"))[7], paste0("0", analysis_no)
)
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}
# ------------------------------------------------

knitr::opts_chunk$set(fig.width = 12, fig.height = 12)
```

We start by reading in the original data.

```{r}
subset_counter <- 1

seurat_orig <- readRDS(file.path(cache_dir, "..", paste0("0", analysis_no - 1), "all_seurat.rds"))
seurat_orig

if ("subsets_mito_percent" %in% names(seurat_orig[[]])) {
  mito <- 1
} else {
  mito <- NULL
}

if (is.null(subsets)) {
  print("No further subsets necessary.")
  rds <- file.path(cache_dir, "all_seurat.rds")
  if (file.exists(rds)) {
    seurat_orig <- readRDS(rds)
  } else {
    saveRDS(seurat_orig, rds)
  }

  for (i in 1:length(metadata_to_plot)) {
    print(names(metadata_to_plot)[i])
    if ("no_legend" %in% metadata_to_plot[[i]] && "no_label" %in% metadata_to_plot[[i]]) {
      print(red_dim_plot(seurat_orig, "umap1", "umap2", names(metadata_to_plot)[i]) + NoLegend())
    } else if ("no_legend" %in% metadata_to_plot[[i]]) {
      print(
        red_dim_plot(seurat_orig, "umap1", "umap2", names(metadata_to_plot)[i], "cat") + NoLegend()
      )
    } else if ("no_label" %in% metadata_to_plot[[i]]) {
      print(red_dim_plot(seurat_orig, "umap1", "umap2", names(metadata_to_plot)[i]))
    } else {
      print(red_dim_plot(seurat_orig, "umap1", "umap2", names(metadata_to_plot)[i], "cat"))
    }
  }
  
  if (protocol[4] == "umis") {
    print("nFeature_SCT")
    red_dim_plot(seurat_orig, "umap1", "umap2", "nFeature_SCT", "cont")
    print("nCount_SCT")
    red_dim_plot(seurat_orig, "umap1", "umap2", "nCount_SCT", "cont")
  } else if (protocol[4] == "reads") {
    print("nFeature_RNA")
    red_dim_plot(seurat_orig, "umap1", "umap2", "nFeature_RNA", "cont")
    print("nCount_RNA")
    red_dim_plot(seurat_orig, "umap1", "umap2", "nCount_RNA", "cont")
  }
  
  if (length(mito) != 0) {
    print("subsets_mito_percent")
    red_dim_plot(seurat_orig, "umap1", "umap2", "subsets_mito_percent", "cont")
  }
}
```

# `r names(subsets[subset_counter])`

We rerun a pipeline on the subset that applies normalization and scaling, dimensionality reduction, and clustering.

```{r, dpi = 300, fig.width = 10, fig.height = 10}
sub_name <- names(subsets[subset_counter])
subset_counter <- subset_counter + 1
seurat <- seurat_orig

if ("neuronal" %in% unlist(strsplit(sub_name, "_"))) {
  if (protocol[1] == "human") {
    genes <- "SNAP25"
  } else if (protocol[1] == "mouse") {
    genes <- "Snap25"
  }
  FeaturePlot(seurat, genes, order = TRUE, cols = c("lightgrey", "red"), max.cutoff = "q10")
}
markers <- FindAllMarkers(seurat, only.pos = TRUE)
saveRDS(markers, file.path(cache_dir, "markers.rds"))

seurat <- subset(
  seurat, idents = setdiff(1:length(unique(seurat$seurat_clusters)) - 1, subsets[[1]])
)
seurat

if ("neuronal" %in% unlist(strsplit(sub_name, "_"))) {
  if (protocol[1] == "human") {
    genes <- "SNAP25"
  } else if (protocol[1] == "mouse") {
    genes <- "Snap25"
  }
  FeaturePlot(seurat, genes, order = TRUE, cols = c("lightgrey", "red"), max.cutoff = "q10")
}

print("Remove genes not expressed in any cell.")
DefaultAssay(seurat) <- "RNA"
dim(seurat)
seurat <- seurat[rowSums(as.matrix(GetAssayData(seurat, slot = "counts")) > 0) > 0, ]
dim(seurat)

seurat <- cluster_pipeline(
  seurat, cache_dir, sub_name, protocol, vars_to_regress, parallel_override
)
seurat

for (i in 1:length(metadata_to_plot)) {
  print(names(metadata_to_plot)[i])
  if ("no_legend" %in% metadata_to_plot[[i]] && "no_label" %in% metadata_to_plot[[i]]) {
    print(red_dim_plot(seurat, "umap1", "umap2", names(metadata_to_plot)[i]) + NoLegend())
  } else if ("no_legend" %in% metadata_to_plot[[i]]) {
    print(red_dim_plot(seurat, "umap1", "umap2", names(metadata_to_plot)[i], "cat") + NoLegend())
  } else if ("no_label" %in% metadata_to_plot[[i]]) {
    print(red_dim_plot(seurat, "umap1", "umap2", names(metadata_to_plot)[i]))
  } else {
    print(red_dim_plot(seurat, "umap1", "umap2", names(metadata_to_plot)[i], "cat"))
  }
}

if (protocol[4] == "umis") {
  print("nFeature_SCT")
  red_dim_plot(seurat, "umap1", "umap2", "nFeature_SCT", "cont")
  print("nCount_SCT")
  red_dim_plot(seurat, "umap1", "umap2", "nCount_SCT", "cont")
} else if (protocol[4] == "reads") {
  print("nFeature_RNA")
  red_dim_plot(seurat, "umap1", "umap2", "nFeature_RNA", "cont")
  print("nCount_RNA")
  red_dim_plot(seurat, "umap1", "umap2", "nCount_RNA", "cont")
}

if (length(mito) != 0) {
  print("subsets_mito_percent")
  red_dim_plot(seurat, "umap1", "umap2", "subsets_mito_percent", "cont")
}
```

# References

This is the concluding section of the document. Here we output the `sessionInfo` and create a bibliography for works cited.

```{r}
sessionInfo()
```
