---
title: '03 Vulnerable Types - `r unlist(strsplit(getwd(), "/"))[6]`'
author:
  - name: "Emir Turkes [emir.turkes@eturkes.com]"
  - name: "UK Dementia Research Institute at UCL"
date: '`r strftime(Sys.time(), "%B %d, %Y")`'
bibliography: '../../../`r unlist(strsplit(getwd(), "/"))[4]`.bib'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_file = file.path(
    "..", "..", "..", "results",
    unlist(strsplit(getwd(), "/"))[6], unlist(strsplit(getwd(), "/"))[7], "03-vulnerable-types.html"
  ))})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
.tocify-subheader .tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 35px; text-indent: 0;
}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of the [Tau Vulnerability Project](https://github.com/eturkes/tau-vulnerability).*

In this document we probe for vulnerable cell types.
The data here is derived from @`r unlist(strsplit(getwd(), "/"))[6]` and will be referenced using the name ``r unlist(strsplit(getwd(), "/"))[6]``.

```{r}
#    This file is part of tau-vulnerability.
#    Copyright (C) 2019-2021  Emir Turkes, UK DRI at UCL, Columbia University Medical Center
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

options(stringsAsFactors = FALSE)
packages <- c(
  "conflicted", "Seurat", "future", "parallel", "biomaRt", "dplyr", "ggplot2", "ggrepel", "viridis",
  "limma"
)
invisible(suppressPackageStartupMessages(lapply(packages, library, character.only = TRUE)))
source(file.path("..", "..", "utils.R"))

# Use UMI protocol due to poor performance of reads protocol.
# -----------------------------------------------------------
protocol <- c("human", "droplet", "single-nuc", "umis") # See `cluster_pipeline` in `utils.R`.
# -----------------------------------------------------------

vars_to_regress <- NULL # See `cluster_pipeline` in `utils.R`.
parallel_override <- NULL # See `parallel_plan` in `utils.R`.

# Metadata to plot after dimensionality reduction and clustering.
# Values in list can include "no_legend and/or "no_label" to exclude those.
# -------------------------------------------------------------------------
metadata_to_plot <- vector("list", 6)
names(metadata_to_plot) <- c(
  "seurat_clusters", "clusterAssignment", "subclusterAssignment",
  "PatientID", "SampleBatch", "Phase"
)
metadata_to_plot$PatientID <- "no_label"
metadata_to_plot$SampleBatch <- "no_label"
metadata_to_plot$Phase <- "no_label"
# -------------------------------------------------------------------------

subset_names <- "all" # Subsetted Seurats to loop over.

analysis_no <- 3
data_name <- unlist(strsplit(getwd(), "/"))[6] # Name of dataset.
assets_dir <- file.path(getwd(), "..", "..", "..", "assets") # Backed up data.
results_dir <- file.path(getwd(), "..", "..", "..", "results")

# Unique cache directory for each analysis number.
# ------------------------------------------------
cache_dir <- file.path(
  "..", "..", "..", "cache", data_name, unlist(strsplit(getwd(), "/"))[7], paste0("0", analysis_no)
)
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}
# ------------------------------------------------

if (protocol[1] == "human") {
  organism <- "hsapiens"
} else if (protocol[1] == "mouse") {
  organism <- "mmusculus"
}

knitr::opts_chunk$set(fig.width = 10, fig.height = 7)
```

We use data saved from previous analyses.

```{r}
subset_list <- c(readRDS(file.path(cache_dir, "..", "02", paste0(subset_names[1], "_seurat.rds"))))

for (i in 1:length(subset_list)) {
  print(subset_names[i])
  print(red_dim_plot(subset_list[[i]], "umap1", "umap2", "seurat_clusters", "cat"))

  if (protocol[1] == "human") {
    plots <- FeaturePlot(
      subset_list[[i]], c("SLC17A7", "GAD1"), order = TRUE, blend = TRUE,
      combine = FALSE, blend.threshold = 0, max.cutoff = "q10"
    )
  } else if (protocol[1] == "mouse") {
    plots <- FeaturePlot(
      subset_list[[i]], c("Slc17a7", "Gad1"), order = TRUE, blend = TRUE,
      combine = FALSE, blend.threshold = 0, max.cutoff = "q10"
    )
  }

  print(CombinePlots(plots[3:4], legend = "none"))
}
```

# NFT Detection

```{r, dpi = 300, fig.width = 15, fig.height = 10}
NFT_sig <- readRDS(
  file.path(
    cache_dir, "..", "..", "..", "otero-garcia-2020-ssRNAseq",
    "comb", "05-gsva", "01", "sctransform_seurat.rds"
  )
)

NFT_sig$batch <- sub("^[^_]*_", "", NFT_sig$donor)
discard <- which(NFT_sig$batch == "2")
NFT_sig <- NFT_sig[ , -discard]
DefaultAssay(NFT_sig) <- "RNA"
NFT_sig[["SCT"]] <- NULL
NFT_sig <- NFT_sig[rowSums(as.matrix(GetAssayData(NFT_sig, slot = "counts")) > 0) > 0, ]

seurat <- subset_list[[1]]
mart <- useEnsembl("ensembl", paste0(organism, "_gene_ensembl"))
attributes <- c("external_gene_name", "ensembl_gene_id")
gene_anno <- getBM(attributes, "external_gene_name", rownames(seurat), mart)
gene_anno <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(NFT_sig), ]
dup <- gene_anno$external_gene_name[which(duplicated(gene_anno$external_gene_name))]
gene_anno <- gene_anno[-which(gene_anno$external_gene_name == dup), ]
seurat <- seurat[rownames(seurat) %in% gene_anno$external_gene_name, ]
new_mat <- GetAssayData(seurat, "counts")
gene_anno <- gene_anno[gene_anno$external_gene_name %in% rownames(new_mat), ]
gene_anno <- gene_anno[order(match(gene_anno$external_gene_name, rownames(new_mat))), ]
rownames(new_mat) <- gene_anno$ensembl_gene_id
seurat <- CreateSeuratObject(new_mat, meta.data = seurat[[]])
NFT_sig <- NFT_sig[rownames(NFT_sig) %in% rownames(seurat), ]
rm(new_mat)

parallel_plan(NFT_sig, parallel_override)
NFT_sig <- suppressWarnings(SCTransform(NFT_sig, vars.to.regress = vars_to_regress, verbose = FALSE))
batch_corrected_mat <- removeBatchEffect(GetAssayData(NFT_sig), NFT_sig$batch)
NFT_sig <- SetAssayData(NFT_sig, "data", batch_corrected_mat)

set.seed(1)
NFT_sig <- RunPCA(NFT_sig, verbose = FALSE)
add_df <- data.frame(Embeddings(NFT_sig)[ , 1:2])
names(add_df) <- paste0("pca", seq(ncol(add_df)))
NFT_sig$pca1 <- add_df$pca1
NFT_sig$pca2 <- add_df$pca2
reduction <- "pca"
dims <- 1:30 # Dimensions for downstream computations.

NFT_sig <- RunUMAP(NFT_sig, dims, reduction, min.dist = 0.75, return.model = TRUE, verbose = FALSE)
add_df <- data.frame(Embeddings(NFT_sig, "umap"))
names(add_df) <- paste0("umap", seq(ncol(add_df)))
NFT_sig$umap1 <- add_df$umap1
NFT_sig$umap2 <- add_df$umap2

# resolution <- 0.15
# NFT_sig <- FindNeighbors(NFT_sig, reduction, 1:30, verbose = FALSE)
# NFT_sig <- FindClusters(NFT_sig, resolution = resolution, verbose = FALSE)

set.seed(3)
kmeans <- kmeans(data.frame(Embeddings(NFT_sig)[ , c(2, 4)]), 2)
NFT_sig$seurat_clusters <- factor(kmeans$cluster)

NFT_sig@active.ident <- NFT_sig$seurat_clusters
NFT_sig <- RenameIdents(NFT_sig, "1" = "NFT", "2" = "normal")
NFT_sig <- UpdateSeuratObject(NFT_sig)
red_dim_plot(NFT_sig, "pca1", "pca2", "idents", "cat")
NFT_sig$idents <- NFT_sig@active.ident
red_dim_plot(NFT_sig, "pca1", "pca2", "idents", "cat")
red_dim_plot(NFT_sig, "pca1", "pca2", "donor", "cat")

NFT_sig <- readRDS(
  file.path(
    cache_dir, "..", "..", "..", "otero-garcia-2020-ssRNAseq",
    "comb", "05-gsva", "01", "NFT_sig.rds"
  )
)

NFT_sig@active.ident <- NFT_sig$seurat_clusters
NFT_sig <- RenameIdents(NFT_sig, "1" = "NFT", "2" = "normal")
NFT_sig <- UpdateSeuratObject(NFT_sig)
red_dim_plot(NFT_sig, "pca1", "pca2", "idents", "cat")
NFT_sig$idents <- NFT_sig@active.ident
red_dim_plot(NFT_sig, "pca1", "pca2", "idents", "cat")
red_dim_plot(NFT_sig, "pca1", "pca2", "donor", "cat")

sub_name <- "sctransform"
seurat <- cluster_pipeline(
  seurat, cache_dir, sub_name, protocol, vars_to_regress, parallel_override, cc = FALSE
)
# seurat <- suppressWarnings(SCTransform(seurat, verbose = FALSE))
# NFT_sig <- suppressWarnings(SCTransform(NFT_sig, verbose = FALSE))

seurat@active.ident <- seurat$seurat_clusters
red_dim_plot(seurat, "umap1", "umap2", "seurat_clusters", "cat")
for (i in 1:length(metadata_to_plot)) {
  print(names(metadata_to_plot)[i])
  if ("no_legend" %in% metadata_to_plot[[i]] && "no_label" %in% metadata_to_plot[[i]]) {
    print(red_dim_plot(seurat, "umap1", "umap2", names(metadata_to_plot)[i]) + NoLegend())
  } else if ("no_legend" %in% metadata_to_plot[[i]]) {
    print(red_dim_plot(seurat, "umap1", "umap2", names(metadata_to_plot)[i], "cat") + NoLegend())
  } else if ("no_label" %in% metadata_to_plot[[i]]) {
    print(red_dim_plot(seurat, "umap1", "umap2", names(metadata_to_plot)[i]))
  } else {
    print(red_dim_plot(seurat, "umap1", "umap2", names(metadata_to_plot)[i], "cat"))
  }
}

seurat <- cluster_pipeline(
  seurat, cache_dir, sub_name, protocol, vars_to_regress, parallel_override, cc = FALSE
)
seurat@active.ident <- seurat$seurat_clusters
seurat <- UpdateSeuratObject(seurat)

# anchors <- FindTransferAnchors(NFT_sig, seurat, "SCT", reduction = "cca", l2.norm = FALSE)
anchors <- FindTransferAnchors(NFT_sig, seurat, "SCT", npcs = 4, dims = c(2, 4))
predictions <- TransferData(anchors, NFT_sig@active.ident, weight.reduction = seurat[["pca"]], dims = 1:30, prediction.assay = TRUE)
# predictions <- TransferData(anchors, NFT_sig@active.ident, weight.reduction = "cca", prediction.assay = TRUE)
seurat[["NFT_sig"]] <- predictions
DefaultAssay(seurat) <- "NFT_sig"

new_mat <- GetAssayData(seurat)
rownames(new_mat) <- c("normal", "NFT", "max")
new_seurat <- CreateSeuratObject(new_mat)
add_df <- Embeddings(seurat, "umap")
new_seurat[["umap"]] <- CreateDimReducObject(embeddings = add_df, key = "UMAP_", assay = DefaultAssay(new_seurat))

plots <- FeaturePlot(
  new_seurat, c("NFT", "normal"), order = TRUE, blend = TRUE,
  combine = FALSE, blend.threshold = 0, max.cutoff = 0.5
)
print(CombinePlots(plots[3:4], legend = "none") + NoLegend())

# seurat <- MapQuery(anchors, seurat, NFT_sig, "idents", reference.reduction = "pca", reduction.model = "umap")

# NFT_sig$id <- "reference"
# seurat$id <- "query"
# refquery <- merge(NFT_sig, seurat)
# refquery[["pca"]] <- merge(NFT_sig[["pca"]], seurat[["ref.pca"]])
# refquery <- RunUMAP(refquery, reduction = "pca", dims = 1:50)
# DimPlot(refquery, group.by = "id", shuffle = TRUE)
# 
# DimPlot(seurat, reduction = "ref.umap", group.by = "predicted.id", label = TRUE, label.size = 3, repel = TRUE)
# red_dim_plot(seurat, "umap1", "umap2", "predicted.id", "cat")
# 
# seurat[["NFT_sig"]] <- predictions
# DefaultAssay(seurat) <- "NFT_sig"
# plots <- FeaturePlot(
#   seurat, c("NFT", "normal"), order = TRUE, blend = TRUE,
#   combine = FALSE
# )
# print(CombinePlots(plots[3:4], legend = "none") + NoLegend())
```

# EC Layer II

## Set 1

```{r}
if (protocol[1] == "human") {
  genes <- c("RASGRP2", "SH3BGRL2")
} else if (protocol[1] == "mouse") {
  genes <- c("Rasgrp2", "Sh3bgrl2")
}

gene_plot(subset_list, subset_names, genes)
```

## Set 2

```{r, dpi = 300, fig.width = 15, fig.height = 10}
if (protocol[1] == "human") {
  genes <- c("RELN", "RORB")
} else if (protocol[1] == "mouse") {
  genes <- c("Reln", "Nefh")
}

gene_plot(subset_list, subset_names, genes)
```

## Set 3

```{r}
if (protocol[1] == "human") {
  genes <- c("RORB", "IL1RAPL2")
} else if (protocol[1] == "mouse") {
  genes <- c("Rorb", "Il1rapl2")
}

gene_plot(subset_list, subset_names, genes)
```

## Set 4

```{r}
if (protocol[1] == "human") {
  genes <- c("CALB1", "CUX2")
} else if (protocol[1] == "mouse") {
  genes <- c("Calb1", "Cux2")
}

gene_plot(subset_list, subset_names, genes)
```

## Set 5

```{r}
if (protocol[1] == "human") {
  genes <- c("BCL11B", "BMPR1B")
} else if (protocol[1] == "mouse") {
  genes <- c("Bcl11b", "Bmpr1b")
}

gene_plot(subset_list, subset_names, genes)
```

## Set 6

```{r}
if (protocol[1] == "human") {
  genes <- c("PDGFD", "LAMA3")
} else if (protocol[1] == "mouse") {
  genes <- c("Pdgfd", "Lama3")
}

gene_plot(subset_list, subset_names, genes)
```

## Combined

```{r, dpi = 300, fig.width = 15, fig.height = 10}
if (protocol[1] == "human") {
  genes <- c(
    "RASGRP2", "SH3BGRL2", "RELN", "NEFH", "RORB", "IL1RAPL2",
    "CALB1", "CUX2", "BCL11B", "BMPR1B", "PDGFD", "LAMA3"
  )
} else if (protocol[1] == "mouse") {
  genes <- c(
    "Rasgrp2", "Sh3bgrl2", "Reln", "Nefh", "Rorb", "Il1rapl2",
    "Calb1", "Cux2", "Bcl11b", "Bmpr1b", "Pdgfd", "Lama3"
  )
}

gene_plot(subset_list, subset_names, genes)
```

# Subiculum

## Set 1

```{r}
if (protocol[1] == "human") {
  genes <- c("NEFH", "CARTPT")
} else if (protocol[1] == "mouse") {
  genes <- c("Nefh", "Cartpt")
}

gene_plot(subset_list, subset_names, genes)
```

## Set 2

```{r}
if (protocol[1] == "human") {
  genes <- c("RELN", "CARTPT")
} else if (protocol[1] == "mouse") {
  genes <- c("Reln", "Cartpt")
}

gene_plot(subset_list, subset_names, genes)
```

## Combined

```{r}
if (protocol[1] == "human") {
  genes <- c("CARTPT", "RELN", "NEFH")
} else if (protocol[1] == "mouse") {
  genes <- c("Cartpt", "Reln", "Nefh")
}

gene_plot(subset_list, subset_names, genes)
```

# HIP CA 1

## Set 1

```{r}
if (protocol[1] == "human") {
  genes <- c("SSTR4", "NOV")
} else if (protocol[1] == "mouse") {
  genes <- c("Sstr4", "Nov")
}

gene_plot(subset_list, subset_names, genes)
```

## Set 2

```{r}
if (protocol[1] == "human") {
  genes <- c("RELN", "NEFH")
} else if (protocol[1] == "mouse") {
  genes <- c("Reln", "Nefh")
}

gene_plot(subset_list, subset_names, genes)
```

## Combined

```{r}
if (protocol[1] == "human") {
  genes <- c("SSTR4", "NOV", "RELN", "NEFH")
} else if (protocol[1] == "mouse") {
  genes <- c("Sstr4", "Nov", "Reln", "Nefh")
}

gene_plot(subset_list, subset_names, genes)
```

# BA9 L2-3 EX

## Superficial

```{r}
if (protocol[1] == "human") {
  genes <- c("TBR1", "SATB2", "WFS1", "CUX2", "LAMP5", "SERPINE2")
} else if (protocol[1] == "mouse") {
  genes <- c("Tbr1", "Satb2", "Wfs1", "Cux2", "Lamp5", "Serpine2")
}

gene_plot(subset_list, subset_names, genes)
```

## Deep

```{r}
if (protocol[1] == "human") {
  genes <- c("CUX2", "COL5A2")
} else if (protocol[1] == "mouse") {
  genes <- c("Cux2", "Col5a2")
}

gene_plot(subset_list, subset_names, genes)
```

# BA9 L4-5 EX

## Set 1

```{r}
if (protocol[1] == "human") {
  genes <- c("RORB", "GAL1")
} else if (protocol[1] == "mouse") {
  genes <- c("Rorb", "Gal1")
}

gene_plot(subset_list, subset_names, genes)
```

## Set 2

```{r}
if (protocol[1] == "human") {
  genes <- c("RORB", "PCP4")
} else if (protocol[1] == "mouse") {
  genes <- c("Rorb", "Pcp4")
}

gene_plot(subset_list, subset_names, genes)
```

# BA9 L5-6 EX

```{r}
if (protocol[1] == "human") {
  genes <- c("NFIA", "THEMIS")
} else if (protocol[1] == "mouse") {
  genes <- c("Nfia", "Themis")
}

gene_plot(subset_list, subset_names, genes)
```

# BA10 L4 EX

```{r}
if (protocol[1] == "human") {
  genes <- c("LINGO1", "RASGEF1B", "SLC26A3")
} else if (protocol[1] == "mouse") {
  genes <- c("Lingo1", "Rasgef1b", "Slc26a3")
}

gene_plot(subset_list, subset_names, genes)
```

# BA9 Chandelier

GAD2 expresssion should be low!

```{r}
if (protocol[1] == "human") {
  genes <- c("PVALB", "GAD1", "GAD2", "SCUBE3")
} else if (protocol[1] == "mouse") {
  genes <- c("Pvalb", "Gad1", "Gad2", "Scube3")
}

gene_plot(subset_list, subset_names, genes)
```

# References

This is the concluding section of the document, where we output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
