---
title:
  '01 Analysis - `r unlist(strsplit(getwd(), "/"))[length(unlist(strsplit(getwd(), "/"))) - 2]`'
author:
  - name: "Emir Turkes [emir.turkes@eturkes.com]"
  - name: "UK Dementia Research Institute at UCL"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(
    inputFile,
    encoding = encoding,
    output_file = file.path(
      "..", "..", "results",
      unlist(strsplit(getwd(), "/"))[length(unlist(strsplit(getwd(), "/")))], "01-analysis.html"
    )
  )})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of the [Tau Vulnerability Project](https://github.com/eturkes/tau-vulnerability).*

The data here is derived from @`r unlist(strsplit(getwd(), "/"))[length(unlist(strsplit(getwd(), "/")))]` and will be referenced using the name ``r unlist(strsplit(getwd(), "/"))[length(unlist(strsplit(getwd(), "/")))]``.

```{r, boilerplate}
# Load in necessary boilerplate and libraries.
# --------------------------------------------

#    This file is part of tau-vulnerability.
#    Copyright (C) 2019-2020  Emir Turkes, UK DRI at UCL, Columbia University Medical Center
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

# These should be checked per document.
# -------------------------------------
options(stringsAsFactors = FALSE)
packages <- c(
  "conflicted", "ComplexHeatmap", "UpSetR", "biomaRt"
)
invisible(suppressPackageStartupMessages(lapply(packages, library, character.only = TRUE)))
source(file.path("..", "utils.R"))

`%notin%` <- Negate(`%in%`)
# -------------------------------------

# Everything else in this chunk remains generally unchanged.
# ----------------------------------------------------------
analysis_no <- 1
data_name <- unlist(strsplit(getwd(), "/"))[6] # Name of dataset.
assets_dir <- file.path("..", "..", "assets") # Backed up data.

# Unique cache directory for each analysis number.
# ------------------------------------------------
cache_dir <- file.path("..", "..", "cache", data_name, paste0("0", analysis_no))
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}
# ------------------------------------------------
# ----------------------------------------------------------
```

# With Leng 2020 snRNAseq

```{r, fig.height = 10, fig.width = 12}
chen_hub_gene_data <- read.delim(file.path(assets_dir, "overlap", "chen-hub-gene-data.tsv"))
allen_mouse_03_hub_gene_data <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-03-hub-gene-data.tsv")
)
roussarie_hub_gene_data <- read.delim(
  file.path(assets_dir, "overlap", "roussarie-hub-gene-data.tsv")
)
leng_CTRL_hub_gene_data <- read.delim(
  file.path(assets_dir, "overlap", "leng-CTRL-hub-gene-data.tsv")
)

pathways <- vector("list", 4)
pathways[[1]] <- unique(chen_hub_gene_data$gene_set)
pathways[[2]] <- unique(allen_mouse_03_hub_gene_data$gene_set)
pathways[[3]] <- unique(roussarie_hub_gene_data$gene_set)
pathways[[4]] <- unique(leng_CTRL_hub_gene_data$gene_set)
names(pathways) <- c(
  "Chen Spatial Transcriptomics", "Allen Mouse scRNAseq",
  "Roussarie bacTRAP RNAseq", "Leng Human snRNAseq"
)

comb_mat <- as.data.frame(list_to_matrix(pathways))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "Pathways"
rownames(comb_mat) <- NULL

set.seed(1)
upset(comb_mat, text.scale = 2)

rds <- file.path(cache_dir, "leng_mouse_ortho.rds")
if (file.exists(rds)) {
  leng_mouse_ortho <- readRDS(rds)
} else {
  leng_mouse_ortho <- human_to_mouse_genes(unique(leng_CTRL_hub_gene_data$external_gene_name))
  saveRDS(leng_mouse_ortho, rds)
}

genes <- vector("list", 4)
genes[[1]] <- unique(chen_hub_gene_data$external_gene_name)
genes[[2]] <- unique(allen_mouse_03_hub_gene_data$external_gene_name)
genes[[3]] <- unique(roussarie_hub_gene_data$external_gene_name)
genes[[4]] <- unique(leng_mouse_ortho)
names(genes) <- c(
  "Chen Spatial Transcriptomics", "Allen Mouse scRNAseq",
  "Roussarie bacTRAP RNAseq", "Leng Human snRNAseq"
)

comb_mat <- as.data.frame(list_to_matrix(genes))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "Genes"
rownames(comb_mat) <- NULL

set.seed(1)
upset(comb_mat, text.scale = 2)

leng_DEGs <- leng_CTRL_hub_gene_data[which(leng_CTRL_hub_gene_data$DE == "yes"), ]
rds <- file.path(cache_dir, "leng_DEGs_mouse_ortho.rds")
if (file.exists(rds)) {
  leng_DEGs_mouse_ortho <- readRDS(rds)
} else {
  leng_DEGs_mouse_ortho <- human_to_mouse_genes(unique(leng_DEGs$external_gene_name))
  saveRDS(leng_DEGs_mouse_ortho, rds)
}

chen_DEGs <- chen_hub_gene_data[which(chen_hub_gene_data$DE == "yes"), ]
allen_mouse_03_DEGs <- allen_mouse_03_hub_gene_data[
  which(allen_mouse_03_hub_gene_data$DE == "yes"),
]
roussarie_DEGs <- roussarie_hub_gene_data[which(roussarie_hub_gene_data$DE == "yes"), ]

DEGs <- vector("list", 4)
DEGs[[1]] <- unique(chen_DEGs$external_gene_name)
DEGs[[2]] <- unique(allen_mouse_03_DEGs$external_gene_name)
DEGs[[3]] <- unique(roussarie_DEGs$external_gene_name)
DEGs[[4]] <- unique(leng_mouse_ortho)
names(DEGs) <- c(
  "Chen Spatial Transcriptomics", "Allen Mouse scRNAseq",
  "Roussarie bacTRAP RNAseq", "Leng Human snRNAseq"
)

comb_mat <- as.data.frame(list_to_matrix(DEGs))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "DEGs"
rownames(comb_mat) <- NULL

set.seed(1)
upset(comb_mat, text.scale = 2)
```

# Without Leng 2020 snRNAseq

```{r, fig.height = 10, fig.width = 12}
chen_hub_gene_data <- read.delim(file.path(assets_dir, "overlap", "chen-hub-gene-data.tsv"))
allen_mouse_03_hub_gene_data <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-03-hub-gene-data.tsv")
)
roussarie_hub_gene_data <- read.delim(
  file.path(assets_dir, "overlap", "roussarie-hub-gene-data.tsv")
)

pathways <- vector("list", 3)
pathways[[1]] <- unique(chen_hub_gene_data$gene_set)
pathways[[2]] <- unique(allen_mouse_03_hub_gene_data$gene_set)
pathways[[3]] <- unique(roussarie_hub_gene_data$gene_set)
names(pathways) <- c(
  "Chen Spatial Transcriptomics", "Allen Mouse scRNAseq", "Roussarie bacTRAP RNAseq"
)

comb_mat <- as.data.frame(list_to_matrix(pathways))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "Pathways"
rownames(comb_mat) <- NULL

set.seed(1)
upset(comb_mat, text.scale = 2)

genes <- vector("list", 3)
genes[[1]] <- unique(chen_hub_gene_data$external_gene_name)
genes[[2]] <- unique(allen_mouse_03_hub_gene_data$external_gene_name)
genes[[3]] <- unique(roussarie_hub_gene_data$external_gene_name)
names(genes) <- c(
  "Chen Spatial Transcriptomics", "Allen Mouse scRNAseq", "Roussarie bacTRAP RNAseq"
)

comb_mat <- as.data.frame(list_to_matrix(genes))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "Genes"
rownames(comb_mat) <- NULL

set.seed(1)
upset(comb_mat, text.scale = 2)

chen_DEGs <- chen_hub_gene_data[which(chen_hub_gene_data$DE == "yes"), ]
allen_mouse_03_DEGs <- allen_mouse_03_hub_gene_data[
  which(allen_mouse_03_hub_gene_data$DE == "yes"),
]
roussarie_DEGs <- roussarie_hub_gene_data[which(roussarie_hub_gene_data$DE == "yes"), ]

DEGs <- vector("list", 3)
DEGs[[1]] <- unique(chen_DEGs$external_gene_name)
DEGs[[2]] <- unique(allen_mouse_03_DEGs$external_gene_name)
DEGs[[3]] <- unique(roussarie_DEGs$external_gene_name)
names(DEGs) <- c(
  "Chen Spatial Transcriptomics", "Allen Mouse scRNAseq", "Roussarie bacTRAP RNAseq"
)

comb_mat <- as.data.frame(list_to_matrix(DEGs))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "DEGs"
rownames(comb_mat) <- NULL

set.seed(1)
upset(comb_mat, text.scale = 2)
```

# References

This is the concluding section of the document, where we output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
