---
title:
  'Overlap Analysis - `r unlist(strsplit(getwd(), "/"))[length(unlist(strsplit(getwd(), "/"))) - 2]`'
author:
  - name: "Emir Turkes [emir.turkes@eturkes.com]"
  - name: "UK Dementia Research Institute at UCL"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(
    inputFile,
    encoding = encoding,
    output_file = file.path(
      "..", "..", "results",
      unlist(strsplit(getwd(), "/"))[length(unlist(strsplit(getwd(), "/")))], "overlap.html"
    )
  )})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of the [Tau Vulnerability Project](https://github.com/eturkes/tau-vulnerability).*

The data here is derived from @`r unlist(strsplit(getwd(), "/"))[length(unlist(strsplit(getwd(), "/")))]` and will be referenced using the name ``r unlist(strsplit(getwd(), "/"))[length(unlist(strsplit(getwd(), "/")))]``.

```{r, boilerplate}
# Load in necessary boilerplate and libraries.
# --------------------------------------------

#    This file is part of tau-vulnerability.
#    Copyright (C) 2019-2020  Emir Turkes, UK DRI at UCL, Columbia University Medical Center
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

# These should be checked per document.
# -------------------------------------
options(stringsAsFactors = FALSE)
packages <- c(
  "conflicted", "ComplexHeatmap", "ComplexUpset", "biomaRt", "DT", "ggplot2"
)
invisible(suppressPackageStartupMessages(lapply(packages, library, character.only = TRUE)))
source(file.path("..", "utils.R"))

`%notin%` <- Negate(`%in%`)
# -------------------------------------

# Everything else in this chunk remains generally unchanged.
# ----------------------------------------------------------
analysis_no <- 1
data_name <- unlist(strsplit(getwd(), "/"))[6] # Name of dataset.
assets_dir <- file.path("..", "..", "assets") # Backed up data.

# Unique cache directory for each analysis number.
# ------------------------------------------------
cache_dir <- file.path("..", "..", "cache", data_name, paste0("0", analysis_no))
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}
# ------------------------------------------------
# ----------------------------------------------------------
```

# Analysis

```{r, fig.height = 10, fig.width = 12}
chen_hub_gene_data <- read.delim(file.path(assets_dir, "overlap", "chen-hub-gene-data.tsv"))
allen_mouse_03_hub_gene_data <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-03-hub-gene-data.tsv")
)
allen_mouse_04_hub_gene_data <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-04-hub-gene-data.tsv")
)
roussarie_hub_gene_data <- read.delim(
  file.path(assets_dir, "overlap", "roussarie-hub-gene-data.tsv")
)

pruned_gene_sets <- vector("list", 4)
pruned_gene_sets[[1]] <- unique(chen_hub_gene_data$gene_set)
pruned_gene_sets[[2]] <- unique(roussarie_hub_gene_data$gene_set)
pruned_gene_sets[[3]] <- unique(allen_mouse_03_hub_gene_data$gene_set)
pruned_gene_sets[[4]] <- unique(allen_mouse_04_hub_gene_data$gene_set)
names(pruned_gene_sets) <- c(
  "Chen Spatial Transcriptomics", "Roussarie bacTRAP RNAseq",
  "Allen Mouse scRNAseq", "Allen Mouse scRNAseq (with IN)"
)

comb_mat <- as.data.frame(list_to_matrix(pruned_gene_sets))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "pruned_gene_sets"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:5]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

set.seed(1)
UpSetR::upset(comb_mat, sets = rev(colnames(comb_mat)[2:5]), keep.order = TRUE, text.scale = 2)
grid.text("Pruned Gene Set Overlap", 0.65, 0.95, gp = gpar(fontsize = 20))
datatable_download(comb_mat)

pruned_gene_set_genes <- vector("list", 4)
pruned_gene_set_genes[[1]] <- unique(chen_hub_gene_data$external_gene_name)
pruned_gene_set_genes[[2]] <- unique(roussarie_hub_gene_data$external_gene_name)
pruned_gene_set_genes[[3]] <- unique(allen_mouse_03_hub_gene_data$external_gene_name)
pruned_gene_set_genes[[4]] <- unique(allen_mouse_04_hub_gene_data$external_gene_name)
names(pruned_gene_set_genes) <- c(
  "Chen Spatial Transcriptomics", "Roussarie bacTRAP RNAseq",
  "Allen Mouse scRNAseq", "Allen Mouse scRNAseq (with IN)"
)

comb_mat <- as.data.frame(list_to_matrix(pruned_gene_set_genes))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "pruned_gene_set_genes"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:5]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

set.seed(1)
UpSetR::upset(comb_mat, sets = rev(colnames(comb_mat)[2:5]), keep.order = TRUE, text.scale = 2)
grid.text("Pruned Gene Set Genes Overlap", 0.65, 0.95, gp = gpar(fontsize = 20))
datatable_download(comb_mat)

chen_hub_genes <- chen_hub_gene_data[which(chen_hub_gene_data$DE == "yes"), ]
allen_mouse_03_hub_genes <- allen_mouse_03_hub_gene_data[
  which(allen_mouse_03_hub_gene_data$DE == "yes"),
]
allen_mouse_04_hub_genes <- allen_mouse_04_hub_gene_data[
  which(allen_mouse_04_hub_gene_data$DE == "yes"),
]
roussarie_hub_genes <- roussarie_hub_gene_data[which(roussarie_hub_gene_data$DE == "yes"), ]

hub_genes <- vector("list", 4)
hub_genes[[1]] <- unique(chen_hub_genes$external_gene_name)
hub_genes[[2]] <- unique(roussarie_hub_genes$external_gene_name)
hub_genes[[3]] <- unique(allen_mouse_03_hub_genes$external_gene_name)
hub_genes[[4]] <- unique(allen_mouse_04_hub_genes$external_gene_name)
names(hub_genes) <- c(
  "Chen Spatial Transcriptomics", "Roussarie bacTRAP RNAseq",
  "Allen Mouse scRNAseq", "Allen Mouse scRNAseq (with IN)"
)

comb_mat <- as.data.frame(list_to_matrix(hub_genes))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "hub_genes"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:5]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

set.seed(1)
UpSetR::upset(comb_mat, sets = rev(colnames(comb_mat)[2:5]), keep.order = TRUE, text.scale = 2)
grid.text("Hub Gene Overlap", 0.65, 0.95, gp = gpar(fontsize = 20))
datatable_download(comb_mat)

chen_gene_sets <- read.delim(
  file.path(assets_dir, "overlap", "chen-gene-sets-all.tsv"), header = FALSE
)
allen_mouse_03_gene_sets <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-03-gene-sets-all.tsv"), header = FALSE
)
allen_mouse_04_gene_sets <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-04-gene-sets-all.tsv"), header = FALSE
)
roussarie_gene_sets <- read.delim(
  file.path(assets_dir, "overlap", "roussarie-gene-sets-all.tsv"), header = FALSE
)

gene_sets <- vector("list", 4)
gene_sets[[1]] <- unique(chen_gene_sets$V1)
gene_sets[[2]] <- unique(roussarie_gene_sets$V1)
gene_sets[[3]] <- unique(allen_mouse_03_gene_sets$V1)
gene_sets[[4]] <- unique(allen_mouse_04_gene_sets$V1)
names(gene_sets) <- c(
  "Chen Spatial Transcriptomics", "Roussarie bacTRAP RNAseq",
  "Allen Mouse scRNAseq", "Allen Mouse scRNAseq (with IN)"
)

comb_mat <- as.data.frame(list_to_matrix(gene_sets))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "gene_sets"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:5]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

set.seed(1)
UpSetR::upset(comb_mat, sets = rev(colnames(comb_mat)[2:5]), keep.order = TRUE, text.scale = 2)
grid.text("Gene Set Overlap", 0.65, 0.95, gp = gpar(fontsize = 20))
datatable_download(comb_mat)

chen_DEGs <- read.delim(
  file.path(assets_dir, "overlap", "chen-genes-all.tsv"), header = FALSE
)
allen_mouse_03_DEGs <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-03-genes-all.tsv"), header = FALSE
)
allen_mouse_04_DEGs <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-04-genes-all.tsv"), header = FALSE
)
roussarie_DEGs <- read.delim(
  file.path(assets_dir, "overlap", "roussarie-genes-all.tsv"), header = FALSE
)

DEGs <- vector("list", 4)
DEGs[[1]] <- unique(chen_DEGs$V1)
DEGs[[2]] <- unique(roussarie_DEGs$V1)
DEGs[[3]] <- unique(allen_mouse_03_DEGs$V1)
DEGs[[4]] <- unique(allen_mouse_04_DEGs$V1)
names(DEGs) <- c(
  "Chen Spatial Transcriptomics", "Roussarie bacTRAP RNAseq",
  "Allen Mouse scRNAseq", "Allen Mouse scRNAseq (with IN)"
)

comb_mat <- as.data.frame(list_to_matrix(DEGs))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "DEGs"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:5]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

set.seed(1)
UpSetR::upset(comb_mat, sets = rev(colnames(comb_mat)[2:5]), keep.order = TRUE, text.scale = 2)
grid.text("DEG Overlap", 0.65, 0.95, gp = gpar(fontsize = 20))
datatable_download(comb_mat)

chen_word_cloud <- read.delim(
  file.path(assets_dir, "overlap", "chen-top-words.tsv"), header = FALSE
)
allen_mouse_03_word_cloud <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-03-top-words.tsv"), header = FALSE
)
allen_mouse_04_word_cloud <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-04-top-words.tsv"), header = FALSE
)
roussarie_word_cloud <- read.delim(
  file.path(assets_dir, "overlap", "roussarie-top-words.tsv"), header = FALSE
)

word_cloud <- vector("list", 4)
word_cloud[[1]] <- unique(chen_word_cloud$V1)
word_cloud[[2]] <- unique(roussarie_word_cloud$V1)
word_cloud[[3]] <- unique(allen_mouse_03_word_cloud$V1)
word_cloud[[4]] <- unique(allen_mouse_04_word_cloud$V1)
names(word_cloud) <- c(
  "Chen Spatial Transcriptomics", "Roussarie bacTRAP RNAseq",
  "Allen Mouse scRNAseq", "Allen Mouse scRNAseq (with IN)"
)

comb_mat <- as.data.frame(list_to_matrix(word_cloud))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "word_cloud_from_description"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:5]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

set.seed(1)
UpSetR::upset(comb_mat, sets = rev(colnames(comb_mat)[2:5]), keep.order = TRUE, text.scale = 2)
grid.text("Word Cloud (From Description)", 0.65, 0.95, gp = gpar(fontsize = 20))
datatable_download(comb_mat)

chen_word_cloud_minimal <- read.delim(
  file.path(assets_dir, "overlap", "chen-top-words-minimal.tsv"), header = FALSE
)
allen_mouse_03_word_cloud_minimal <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-03-top-words-minimal.tsv"), header = FALSE
)
allen_mouse_04_word_cloud_minimal <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-04-top-words-minimal.tsv"), header = FALSE
)
roussarie_word_cloud_minimal <- read.delim(
  file.path(assets_dir, "overlap", "roussarie-top-words-minimal.tsv"), header = FALSE
)

word_cloud_minimal <- vector("list", 4)
word_cloud_minimal[[1]] <- unique(chen_word_cloud_minimal$V1)
word_cloud_minimal[[2]] <- unique(roussarie_word_cloud_minimal$V1)
word_cloud_minimal[[3]] <- unique(allen_mouse_03_word_cloud_minimal$V1)
word_cloud_minimal[[4]] <- unique(allen_mouse_04_word_cloud_minimal$V1)
names(word_cloud_minimal) <- c(
  "Chen Spatial Transcriptomics", "Roussarie bacTRAP RNAseq",
  "Allen Mouse scRNAseq", "Allen Mouse scRNAseq (with IN)"
)

comb_mat <- as.data.frame(list_to_matrix(word_cloud_minimal))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "word_cloud"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:5]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

set.seed(1)
UpSetR::upset(comb_mat, sets = rev(colnames(comb_mat)[2:5]), keep.order = TRUE, text.scale = 2)
grid.text("Word Cloud Overlap", 0.65, 0.95, gp = gpar(fontsize = 20))
datatable_download(comb_mat)
```

# Analysis (No IN)

```{r, fig.height = 10, fig.width = 12}
chen_hub_gene_data <- read.delim(file.path(assets_dir, "overlap", "chen-hub-gene-data.tsv"))
allen_mouse_03_hub_gene_data <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-03-hub-gene-data.tsv")
)
allen_mouse_04_hub_gene_data <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-04-hub-gene-data.tsv")
)
roussarie_hub_gene_data <- read.delim(
  file.path(assets_dir, "overlap", "roussarie-hub-gene-data.tsv")
)

pruned_gene_sets <- vector("list", 3)
pruned_gene_sets[[1]] <- unique(chen_hub_gene_data$gene_set)
pruned_gene_sets[[2]] <- unique(roussarie_hub_gene_data$gene_set)
pruned_gene_sets[[3]] <- unique(allen_mouse_03_hub_gene_data$gene_set)
names(pruned_gene_sets) <- c(
  "Chen Spatial Transcriptomics", "Roussarie bacTRAP RNAseq", "Allen Mouse scRNAseq"
)

comb_mat <- as.data.frame(list_to_matrix(pruned_gene_sets))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "pruned_gene_sets"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:4]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

set.seed(1)
UpSetR::upset(comb_mat, sets = rev(colnames(comb_mat)[2:4]), keep.order = TRUE, text.scale = 2)
grid.text("Pruned Gene Set Overlap", 0.65, 0.95, gp = gpar(fontsize = 20))
datatable_download(comb_mat)

pruned_gene_set_genes <- vector("list", 3)
pruned_gene_set_genes[[1]] <- unique(chen_hub_gene_data$external_gene_name)
pruned_gene_set_genes[[2]] <- unique(roussarie_hub_gene_data$external_gene_name)
pruned_gene_set_genes[[3]] <- unique(allen_mouse_03_hub_gene_data$external_gene_name)
names(pruned_gene_set_genes) <- c(
  "Chen Spatial Transcriptomics", "Roussarie bacTRAP RNAseq", "Allen Mouse scRNAseq"
)

comb_mat <- as.data.frame(list_to_matrix(pruned_gene_set_genes))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "pruned_gene_set_genes"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:4]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

set.seed(1)
UpSetR::upset(comb_mat, sets = rev(colnames(comb_mat)[2:4]), keep.order = TRUE, text.scale = 2)
grid.text("Pruned Gene Set Genes Overlap", 0.65, 0.95, gp = gpar(fontsize = 20))
datatable_download(comb_mat)

chen_hub_genes <- chen_hub_gene_data[which(chen_hub_gene_data$DE == "yes"), ]
allen_mouse_03_hub_genes <- allen_mouse_03_hub_gene_data[
  which(allen_mouse_03_hub_gene_data$DE == "yes"),
]
allen_mouse_04_hub_genes <- allen_mouse_04_hub_gene_data[
  which(allen_mouse_04_hub_gene_data$DE == "yes"),
]
roussarie_hub_genes <- roussarie_hub_gene_data[which(roussarie_hub_gene_data$DE == "yes"), ]

hub_genes <- vector("list", 3)
hub_genes[[1]] <- unique(chen_hub_genes$external_gene_name)
hub_genes[[2]] <- unique(roussarie_hub_genes$external_gene_name)
hub_genes[[3]] <- unique(allen_mouse_03_hub_genes$external_gene_name)
names(hub_genes) <- c(
  "Chen Spatial Transcriptomics", "Roussarie bacTRAP RNAseq", "Allen Mouse scRNAseq"
)

comb_mat <- as.data.frame(list_to_matrix(hub_genes))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "hub_genes"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:4]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

set.seed(1)
UpSetR::upset(comb_mat, sets = rev(colnames(comb_mat)[2:4]), keep.order = TRUE, text.scale = 2)
grid.text("Hub Gene Overlap", 0.65, 0.95, gp = gpar(fontsize = 20))
datatable_download(comb_mat)

chen_gene_sets <- read.delim(
  file.path(assets_dir, "overlap", "chen-gene-sets-all.tsv"), header = FALSE
)
allen_mouse_03_gene_sets <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-03-gene-sets-all.tsv"), header = FALSE
)
allen_mouse_04_gene_sets <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-04-gene-sets-all.tsv"), header = FALSE
)
roussarie_gene_sets <- read.delim(
  file.path(assets_dir, "overlap", "roussarie-gene-sets-all.tsv"), header = FALSE
)

gene_sets <- vector("list", 3)
gene_sets[[1]] <- unique(chen_gene_sets$V1)
gene_sets[[2]] <- unique(roussarie_gene_sets$V1)
gene_sets[[3]] <- unique(allen_mouse_03_gene_sets$V1)
names(gene_sets) <- c(
  "Chen Spatial Transcriptomics", "Roussarie bacTRAP RNAseq", "Allen Mouse scRNAseq"
)

comb_mat <- as.data.frame(list_to_matrix(gene_sets))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "gene_sets"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:4]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

set.seed(1)
UpSetR::upset(comb_mat, sets = rev(colnames(comb_mat)[2:4]), keep.order = TRUE, text.scale = 2)
grid.text("Gene Set Overlap", 0.65, 0.95, gp = gpar(fontsize = 20))
datatable_download(comb_mat)

chen_DEGs <- read.delim(
  file.path(assets_dir, "overlap", "chen-genes-all.tsv"), header = FALSE
)
allen_mouse_03_DEGs <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-03-genes-all.tsv"), header = FALSE
)
allen_mouse_04_DEGs <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-04-genes-all.tsv"), header = FALSE
)
roussarie_DEGs <- read.delim(
  file.path(assets_dir, "overlap", "roussarie-genes-all.tsv"), header = FALSE
)

DEGs <- vector("list", 3)
DEGs[[1]] <- unique(chen_DEGs$V1)
DEGs[[2]] <- unique(roussarie_DEGs$V1)
DEGs[[3]] <- unique(allen_mouse_03_DEGs$V1)
names(DEGs) <- c(
  "Chen Spatial Transcriptomics", "Roussarie bacTRAP RNAseq", "Allen Mouse scRNAseq"
)

comb_mat <- as.data.frame(list_to_matrix(DEGs))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "DEGs"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:4]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

set.seed(1)
UpSetR::upset(comb_mat, sets = rev(colnames(comb_mat)[2:4]), keep.order = TRUE, text.scale = 2)
grid.text("DEG Overlap", 0.65, 0.95, gp = gpar(fontsize = 20))
datatable_download(comb_mat)

chen_word_cloud <- read.delim(
  file.path(assets_dir, "overlap", "chen-top-words.tsv"), header = FALSE
)
allen_mouse_03_word_cloud <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-03-top-words.tsv"), header = FALSE
)
allen_mouse_04_word_cloud <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-04-top-words.tsv"), header = FALSE
)
roussarie_word_cloud <- read.delim(
  file.path(assets_dir, "overlap", "roussarie-top-words.tsv"), header = FALSE
)

word_cloud <- vector("list", 3)
word_cloud[[1]] <- unique(chen_word_cloud$V1)
word_cloud[[2]] <- unique(roussarie_word_cloud$V1)
word_cloud[[3]] <- unique(allen_mouse_03_word_cloud$V1)
names(word_cloud) <- c(
  "Chen Spatial Transcriptomics", "Roussarie bacTRAP RNAseq", "Allen Mouse scRNAseq"
)

comb_mat <- as.data.frame(list_to_matrix(word_cloud))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "word_cloud_from_description"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:4]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

set.seed(1)
UpSetR::upset(comb_mat, sets = rev(colnames(comb_mat)[2:4]), keep.order = TRUE, text.scale = 2)
grid.text("Word Cloud (From Description)", 0.65, 0.95, gp = gpar(fontsize = 20))
datatable_download(comb_mat)

chen_word_cloud_minimal <- read.delim(
  file.path(assets_dir, "overlap", "chen-top-words-minimal.tsv"), header = FALSE
)
allen_mouse_03_word_cloud_minimal <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-03-top-words-minimal.tsv"), header = FALSE
)
allen_mouse_04_word_cloud_minimal <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-04-top-words-minimal.tsv"), header = FALSE
)
roussarie_word_cloud_minimal <- read.delim(
  file.path(assets_dir, "overlap", "roussarie-top-words-minimal.tsv"), header = FALSE
)

word_cloud_minimal <- vector("list", 3)
word_cloud_minimal[[1]] <- unique(chen_word_cloud_minimal$V1)
word_cloud_minimal[[2]] <- unique(roussarie_word_cloud_minimal$V1)
word_cloud_minimal[[3]] <- unique(allen_mouse_03_word_cloud_minimal$V1)
names(word_cloud_minimal) <- c(
  "Chen Spatial Transcriptomics", "Roussarie bacTRAP RNAseq", "Allen Mouse scRNAseq"
)

comb_mat <- as.data.frame(list_to_matrix(word_cloud_minimal))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "word_cloud"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:4]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

set.seed(1)
UpSetR::upset(comb_mat, sets = rev(colnames(comb_mat)[2:4]), keep.order = TRUE, text.scale = 2)
grid.text("Word Cloud Overlap", 0.65, 0.95, gp = gpar(fontsize = 20))
datatable_download(comb_mat)
```

# Analysis (With Kampmann)

```{r, fig.height = 10, fig.width = 15}
chen_hub_gene_data <- read.delim(file.path(assets_dir, "overlap", "chen-hub-gene-data.tsv"))
allen_mouse_03_hub_gene_data <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-03-hub-gene-data.tsv")
)
allen_mouse_04_hub_gene_data <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-04-hub-gene-data.tsv")
)
roussarie_hub_gene_data <- read.delim(
  file.path(assets_dir, "overlap", "roussarie-hub-gene-data.tsv")
)
leng_CTRL_hub_gene_data <- read.delim(
  file.path(assets_dir, "overlap", "kampmann-CTRL-hub-gene-data.tsv")
)
leng_AD_CTRL_hub_gene_data <- read.delim(
  file.path(assets_dir, "overlap", "kampmann-Braak6-hub-gene-data.tsv")
)

pruned_gene_sets <- vector("list", 6)
pruned_gene_sets[[1]] <- unique(chen_hub_gene_data$gene_set)
pruned_gene_sets[[2]] <- unique(roussarie_hub_gene_data$gene_set)
pruned_gene_sets[[3]] <- unique(allen_mouse_03_hub_gene_data$gene_set)
pruned_gene_sets[[4]] <- unique(allen_mouse_04_hub_gene_data$gene_set)
pruned_gene_sets[[5]] <- unique(leng_CTRL_hub_gene_data$gene_set)
pruned_gene_sets[[6]] <- unique(leng_AD_CTRL_hub_gene_data$gene_set)
names(pruned_gene_sets) <- c(
  "Chen WT Mouse ST - EC CA1 CA3", "Roussarie WT Mouse bacTRAP - EC CA1 CA3",
  "Allen WT Mouse scRNAseq - EC CA1 CA3", "Allen WT Mouse scRNAseq - EC CA1 CA3 IN",
  "Leng CTRL Human snRNAseq - ECvuln ECstable", "Leng AD/CTRL Human snRNAseq - ECvuln Case/Control"
)

comb_mat <- as.data.frame(list_to_matrix(pruned_gene_sets))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "pruned_gene_sets"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:7]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

upset(
  comb_mat,
  rev(colnames(comb_mat)[2:7]),
  set_sizes = FALSE,
  sort_sets = FALSE,
  themes = upset_default_themes(text = element_text(size = 20))
) + ggtitle("Pruned Gene Set Overlap")
datatable_download(comb_mat)

rds <- file.path(cache_dir, "leng_CTRL_mouse_ortho.rds")
if (file.exists(rds)) {
  leng_CTRL_mouse_ortho <- readRDS(rds)
} else {
  leng_CTRL_mouse_ortho <- human_to_mouse_genes(unique(leng_CTRL_hub_gene_data$external_gene_name))
  saveRDS(leng_CTRL_mouse_ortho, rds)
}
rds <- file.path(cache_dir, "leng_AD_CTRL_mouse_ortho.rds")
if (file.exists(rds)) {
  leng_AD_CTRL_mouse_ortho <- readRDS(rds)
} else {
  leng_AD_CTRL_mouse_ortho <- human_to_mouse_genes(
    unique(leng_AD_CTRL_hub_gene_data$external_gene_name)
  )
  saveRDS(leng_AD_CTRL_mouse_ortho, rds)
}

pruned_gene_set_genes <- vector("list", 4)
pruned_gene_set_genes[[1]] <- unique(chen_hub_gene_data$external_gene_name)
pruned_gene_set_genes[[2]] <- unique(roussarie_hub_gene_data$external_gene_name)
pruned_gene_set_genes[[3]] <- unique(allen_mouse_03_hub_gene_data$external_gene_name)
pruned_gene_set_genes[[4]] <- unique(allen_mouse_04_hub_gene_data$external_gene_name)
pruned_gene_set_genes[[5]] <- unique(leng_CTRL_mouse_ortho)
pruned_gene_set_genes[[6]] <- unique(leng_AD_CTRL_mouse_ortho)
names(pruned_gene_set_genes) <- c(
  "Chen WT Mouse ST - EC CA1 CA3", "Roussarie WT Mouse bacTRAP - EC CA1 CA3",
  "Allen WT Mouse scRNAseq - EC CA1 CA3", "Allen WT Mouse scRNAseq - EC CA1 CA3 IN",
  "Leng CTRL Human snRNAseq - ECvuln ECstable", "Leng AD/CTRL Human snRNAseq - ECvuln Case/Control"
)

comb_mat <- as.data.frame(list_to_matrix(pruned_gene_set_genes))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "pruned_gene_set_genes"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:7]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

upset(
  comb_mat,
  rev(colnames(comb_mat)[2:7]),
  set_sizes = FALSE,
  sort_sets = FALSE,
  themes = upset_default_themes(text = element_text(size = 20))
) + ggtitle("Pruned Gene Set Genes Overlap")
datatable_download(comb_mat)

chen_hub_genes <- chen_hub_gene_data[which(chen_hub_gene_data$DE == "yes"), ]
allen_mouse_03_hub_genes <- allen_mouse_03_hub_gene_data[
  which(allen_mouse_03_hub_gene_data$DE == "yes"),
]
allen_mouse_04_hub_genes <- allen_mouse_04_hub_gene_data[
  which(allen_mouse_04_hub_gene_data$DE == "yes"),
]
roussarie_hub_genes <- roussarie_hub_gene_data[which(roussarie_hub_gene_data$DE == "yes"), ]

leng_CTRL_hub_genes <- leng_CTRL_hub_gene_data[which(leng_CTRL_hub_gene_data$DE == "yes"), ]
rds <- file.path(cache_dir, "leng_CTRL_hub_genes_mouse_ortho.rds")
if (file.exists(rds)) {
  leng_CTRL_hub_genes_mouse_ortho <- readRDS(rds)
} else {
  leng_CTRL_hub_genes_mouse_ortho <- human_to_mouse_genes(
    unique(leng_CTRL_hub_genes$external_gene_name)
  )
  saveRDS(leng_CTRL_hub_genes_mouse_ortho, rds)
}
leng_AD_CTRL_hub_genes <- leng_AD_CTRL_hub_gene_data[
  which(leng_AD_CTRL_hub_gene_data$DE == "yes"),
]
rds <- file.path(cache_dir, "leng_AD_CTRL_hub_genes_mouse_ortho.rds")
if (file.exists(rds)) {
  leng_AD_CTRL_hub_genes_mouse_ortho <- readRDS(rds)
} else {
  leng_AD_CTRL_hub_genes_mouse_ortho <- human_to_mouse_genes(
    unique(leng_AD_CTRL_hub_genes$external_gene_name)
  )
  saveRDS(leng_AD_CTRL_hub_genes_mouse_ortho, rds)
}

hub_genes <- vector("list", 4)
hub_genes[[1]] <- unique(chen_hub_genes$external_gene_name)
hub_genes[[2]] <- unique(roussarie_hub_genes$external_gene_name)
hub_genes[[3]] <- unique(allen_mouse_03_hub_genes$external_gene_name)
hub_genes[[4]] <- unique(allen_mouse_04_hub_genes$external_gene_name)
hub_genes[[5]] <- unique(leng_CTRL_hub_genes_mouse_ortho)
hub_genes[[6]] <- unique(leng_AD_CTRL_hub_genes_mouse_ortho)
names(hub_genes) <- c(
  "Chen WT Mouse ST - EC CA1 CA3", "Roussarie WT Mouse bacTRAP - EC CA1 CA3",
  "Allen WT Mouse scRNAseq - EC CA1 CA3", "Allen WT Mouse scRNAseq - EC CA1 CA3 IN",
  "Leng CTRL Human snRNAseq - ECvuln ECstable", "Leng AD/CTRL Human snRNAseq - ECvuln Case/Control"
)

comb_mat <- as.data.frame(list_to_matrix(hub_genes))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "hub_genes"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:7]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

upset(
  comb_mat,
  rev(colnames(comb_mat)[2:7]),
  set_sizes = FALSE,
  sort_sets = FALSE,
  themes = upset_default_themes(text = element_text(size = 20))
) + ggtitle("Hub Gene Overlap")
datatable_download(comb_mat)

chen_gene_sets <- read.delim(
  file.path(assets_dir, "overlap", "chen-gene-sets-all.tsv"), header = FALSE
)
allen_mouse_03_gene_sets <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-03-gene-sets-all.tsv"), header = FALSE
)
allen_mouse_04_gene_sets <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-04-gene-sets-all.tsv"), header = FALSE
)
roussarie_gene_sets <- read.delim(
  file.path(assets_dir, "overlap", "roussarie-gene-sets-all.tsv"), header = FALSE
)
leng_CTRL_gene_sets <- read.delim(
  file.path(assets_dir, "overlap", "kampmann-CTRL-gene-sets-all.tsv"), header = FALSE
)
leng_AD_CTRL_gene_sets <- read.delim(
  file.path(assets_dir, "overlap", "kampmann-Braak6-gene-sets-all.tsv"), header = FALSE
)

gene_sets <- vector("list", 6)
gene_sets[[1]] <- unique(chen_gene_sets$V1)
gene_sets[[2]] <- unique(roussarie_gene_sets$V1)
gene_sets[[3]] <- unique(allen_mouse_03_gene_sets$V1)
gene_sets[[4]] <- unique(allen_mouse_04_gene_sets$V1)
gene_sets[[5]] <- unique(leng_CTRL_gene_sets$V1)
gene_sets[[6]] <- unique(leng_AD_CTRL_gene_sets$V1)
names(gene_sets) <- c(
  "Chen WT Mouse ST - EC CA1 CA3", "Roussarie WT Mouse bacTRAP - EC CA1 CA3",
  "Allen WT Mouse scRNAseq - EC CA1 CA3", "Allen WT Mouse scRNAseq - EC CA1 CA3 IN",
  "Leng CTRL Human snRNAseq - ECvuln ECstable", "Leng AD/CTRL Human snRNAseq - ECvuln Case/Control"
)

comb_mat <- as.data.frame(list_to_matrix(gene_sets))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "gene_sets"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:7]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

upset(
  comb_mat,
  rev(colnames(comb_mat)[2:7]),
  set_sizes = FALSE,
  sort_sets = FALSE,
  themes = upset_default_themes(text = element_text(size = 20))
) + ggtitle("Gene Set Overlap")
datatable_download(comb_mat)

chen_DEGs <- read.delim(
  file.path(assets_dir, "overlap", "chen-genes-all.tsv"), header = FALSE
)
allen_mouse_03_DEGs <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-03-genes-all.tsv"), header = FALSE
)
allen_mouse_04_DEGs <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-04-genes-all.tsv"), header = FALSE
)
roussarie_DEGs <- read.delim(
  file.path(assets_dir, "overlap", "roussarie-genes-all.tsv"), header = FALSE
)
leng_CTRL_DEGs <- read.delim(
  file.path(assets_dir, "overlap", "kampmann-CTRL-genes-all.tsv")
)
leng_AD_CTRL_DEGs <- read.delim(
  file.path(assets_dir, "overlap", "kampmann-Braak6-genes-all.tsv")
)

rds <- file.path(cache_dir, "leng_CTRL_DEGs_mouse_ortho.rds")
if (file.exists(rds)) {
  leng_CTRL_DEGs_mouse_ortho <- readRDS(rds)
} else {
  leng_CTRL_DEGs_mouse_ortho <- human_to_mouse_genes(unique(leng_CTRL_DEGs))
  saveRDS(leng_CTRL_DEGs_mouse_ortho, rds)
}
rds <- file.path(cache_dir, "leng_AD_CTRL_DEGs_mouse_ortho.rds")
if (file.exists(rds)) {
  leng_AD_CTRL_DEGs_mouse_ortho <- readRDS(rds)
} else {
  leng_AD_CTRL_DEGs_mouse_ortho <- human_to_mouse_genes(unique(leng_AD_CTRL_DEGs))
  saveRDS(leng_AD_CTRL_DEGs_mouse_ortho, rds)
}

DEGs <- vector("list", 4)
DEGs[[1]] <- unique(chen_DEGs$V1)
DEGs[[2]] <- unique(roussarie_DEGs$V1)
DEGs[[3]] <- unique(allen_mouse_03_DEGs$V1)
DEGs[[4]] <- unique(allen_mouse_04_DEGs$V1)
DEGs[[5]] <- unique(leng_CTRL_DEGs_mouse_ortho)
DEGs[[6]] <- unique(leng_AD_CTRL_DEGs_mouse_ortho)
names(DEGs) <- c(
  "Chen WT Mouse ST - EC CA1 CA3", "Roussarie WT Mouse bacTRAP - EC CA1 CA3",
  "Allen WT Mouse scRNAseq - EC CA1 CA3", "Allen WT Mouse scRNAseq - EC CA1 CA3 IN",
  "Leng CTRL Human snRNAseq - ECvuln ECstable", "Leng AD/CTRL Human snRNAseq - ECvuln Case/Control"
)

comb_mat <- as.data.frame(list_to_matrix(DEGs))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "DEGs"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:7]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

upset(
  comb_mat,
  rev(colnames(comb_mat)[2:7]),
  set_sizes = FALSE,
  sort_sets = FALSE,
  themes = upset_default_themes(text = element_text(size = 20))
) + ggtitle("DEG Overlap")
datatable_download(comb_mat)

chen_word_cloud <- read.delim(
  file.path(assets_dir, "overlap", "chen-top-words.tsv"), header = FALSE
)
allen_mouse_03_word_cloud <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-03-top-words.tsv"), header = FALSE
)
allen_mouse_04_word_cloud <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-04-top-words.tsv"), header = FALSE
)
roussarie_word_cloud <- read.delim(
  file.path(assets_dir, "overlap", "roussarie-top-words.tsv"), header = FALSE
)
leng_CTRL_word_cloud <- read.delim(
  file.path(assets_dir, "overlap", "kampmann-CTRL-top-words.tsv"), header = FALSE
)
leng_CTRL_AD_word_cloud <- read.delim(
  file.path(assets_dir, "overlap", "kampmann-Braak6-top-words.tsv"), header = FALSE
)

word_cloud <- vector("list", 4)
word_cloud[[1]] <- unique(chen_word_cloud$V1)
word_cloud[[2]] <- unique(roussarie_word_cloud$V1)
word_cloud[[3]] <- unique(allen_mouse_03_word_cloud$V1)
word_cloud[[4]] <- unique(allen_mouse_04_word_cloud$V1)
word_cloud[[5]] <- unique(leng_CTRL_word_cloud$V1)
word_cloud[[6]] <- unique(leng_CTRL_AD_word_cloud$V1)
names(word_cloud) <- c(
  "Chen WT Mouse ST - EC CA1 CA3", "Roussarie WT Mouse bacTRAP - EC CA1 CA3",
  "Allen WT Mouse scRNAseq - EC CA1 CA3", "Allen WT Mouse scRNAseq - EC CA1 CA3 IN",
  "Leng CTRL Human snRNAseq - ECvuln ECstable", "Leng AD/CTRL Human snRNAseq - ECvuln Case/Control"
)

comb_mat <- as.data.frame(list_to_matrix(word_cloud))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "word_cloud_from_description"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:7]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

upset(
  comb_mat,
  rev(colnames(comb_mat)[2:7]),
  set_sizes = FALSE,
  sort_sets = FALSE,
  themes = upset_default_themes(text = element_text(size = 20))
) + ggtitle("Word Cloud (From Description)")
datatable_download(comb_mat)

chen_word_cloud_minimal <- read.delim(
  file.path(assets_dir, "overlap", "chen-top-words-minimal.tsv"), header = FALSE
)
allen_mouse_03_word_cloud_minimal <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-03-top-words-minimal.tsv"), header = FALSE
)
allen_mouse_04_word_cloud_minimal <- read.delim(
  file.path(assets_dir, "overlap", "allen-mouse-04-top-words-minimal.tsv"), header = FALSE
)
roussarie_word_cloud_minimal <- read.delim(
  file.path(assets_dir, "overlap", "roussarie-top-words-minimal.tsv"), header = FALSE
)
leng_CTRL_word_cloud_minimal <- read.delim(
  file.path(assets_dir, "overlap", "kampmann-CTRL-top-words-minimal.tsv"), header = FALSE
)
leng_CTRL_AD_word_cloud_minimal <- read.delim(
  file.path(assets_dir, "overlap", "kampmann-Braak6-top-words-minimal.tsv"), header = FALSE
)

word_cloud_minimal <- vector("list", 4)
word_cloud_minimal[[1]] <- unique(chen_word_cloud_minimal$V1)
word_cloud_minimal[[2]] <- unique(roussarie_word_cloud_minimal$V1)
word_cloud_minimal[[3]] <- unique(allen_mouse_03_word_cloud_minimal$V1)
word_cloud_minimal[[4]] <- unique(allen_mouse_04_word_cloud_minimal$V1)
word_cloud_minimal[[5]] <- unique(leng_CTRL_word_cloud_minimal$V1)
word_cloud_minimal[[6]] <- unique(leng_CTRL_AD_word_cloud_minimal$V1)
names(word_cloud_minimal) <- c(
  "Chen WT Mouse ST - EC CA1 CA3", "Roussarie WT Mouse bacTRAP - EC CA1 CA3",
  "Allen WT Mouse scRNAseq - EC CA1 CA3", "Allen WT Mouse scRNAseq - EC CA1 CA3 IN",
  "Leng CTRL Human snRNAseq - ECvuln ECstable", "Leng AD/CTRL Human snRNAseq - ECvuln Case/Control"
)

comb_mat <- as.data.frame(list_to_matrix(word_cloud_minimal))
comb_mat <- cbind(rownames(comb_mat), comb_mat)
colnames(comb_mat)[1] <- "word_cloud"
comb_mat <- comb_mat[order(rowSums(comb_mat[ , 2:7]), decreasing = TRUE), ]
rownames(comb_mat) <- NULL

upset(
  comb_mat,
  rev(colnames(comb_mat)[2:7]),
  set_sizes = FALSE,
  sort_sets = FALSE,
  themes = upset_default_themes(text = element_text(size = 20))
) + ggtitle("Word Cloud Overlap")
datatable_download(comb_mat)
```

# References

This is the concluding section of the document, where we output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
