---
title: '03 EC L2 Lateral/Medial CA1 CA3 - `r unlist(strsplit(getwd(), "/"))[6]`'
author:
  - name: "Emir Turkes [emir.turkes@eturkes.com]"
  - name: "UK Dementia Research Institute at UCL"
date: '`r strftime(Sys.time(), "%B %d, %Y")`'
bibliography: '../../../`r unlist(strsplit(getwd(), "/"))[4]`.bib'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_file = file.path(
    "..", "..", "..", "results",
    unlist(strsplit(getwd(), "/"))[6], unlist(strsplit(getwd(), "/"))[7], "03-EC-LM-CA.html"
  ))})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
.tocify-subheader .tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 35px; text-indent: 0;
}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of the [Tau Vulnerability Project](https://github.com/eturkes/tau-vulnerability).*

In this document we perform GSVA analysis.
The data here is derived from @`r unlist(strsplit(getwd(), "/"))[6]` and will be referenced using the name `r unlist(strsplit(getwd(), "/"))[6]`.

```{r}
#    Copyright (C) 2019-2020  Emir Turkes
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

# These should be checked per document.
# -------------------------------------
options(stringsAsFactors = FALSE)
packages <- c(
  "conflicted", "scales", "RColorBrewer", "Seurat", "dplyr", "ggplot2", "ggrepel",
  "GSEABase", "biomaRt", "Biostrings", "SingleCellExperiment", "SCnorm", "viridis", "scater",
  "GSVA", "limma", "DT", "GOSemSim", "factoextra", "ComplexHeatmap"
)
invisible(suppressPackageStartupMessages(lapply(packages, library, character.only = TRUE)))
conflict_prefer("strsplit", "Biostrings", quiet = TRUE)
conflict_prefer("intersect", "GSEABase", quiet = TRUE)
source(file.path(getwd(), "..", "..", "utils.R"))

# Use UMI protocol due to poor performance of reads protocol.
# -----------------------------------------------------------
protocol <- c("mouse", "smart-seq", "single-cell", "umis") # See `cluster_pipeline` in `utils.R`.
# -----------------------------------------------------------

# Metadata to plot after dimensionality reduction and clustering.
# Values in list can include "no_legend and/or "no_label" to exclude those.
# -------------------------------------------------------------------------
metadata_to_plot <- vector("list", 5)
names(metadata_to_plot) <- c("idents", "cluster_stem", "sex_label", "age_label", "donor_id")
# -------------------------------------------------------------------------
# --------------------------------------------

gene_set_list <- "GO.comb"
iss_codebook_name <- "codebook_comb.txt" # Filename of in-situ sequencing codebook to reference.
subset_names <- "EC_HIP_neuronal"
downsample <- 87 # TODO: Automate this.

# Clusters of interest.
# ---------------------
idents <- vector("list", 4)
idents[[1]] <- "L2 IT ENTl"
idents[[2]] <- "L2 IT ENTm"
idents[[3]] <- 1
idents[[4]] <- 14
names(idents) <- c("EC_L2_Lat", "EC_L2_Med", "CA1", "CA3")
# ---------------------

# Everything else in this chunk remains generally unchanged.
# ----------------------------------------------------------
analysis_no <- 03
data_name <- unlist(strsplit(getwd(), "/"))[6] # Name of dataset.
assets_dir <- file.path(getwd(), "..", "..", "..", "assets") # Backed up data.

# Unique cache directory for each analysis number.
# ------------------------------------------------
cache_dir <- file.path(
  getwd(), "..", "..", "..", "cache", data_name, "05-gsva", paste0("0", analysis_no)
)
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}
# ------------------------------------------------

if (protocol[1] == "human") {
  organism <- "hsapiens"
} else if (protocol[1] == "mouse") {
  organism <- "mmusculus"
}

knitr::opts_chunk$set(fig.width = 12, fig.height = 12)
# ----------------------------------------------------------
```

A statistical model is built that finds gene sets that follow a trajectory in our clusters of interest.
We use `lmFit` from limma to fit the models to both GSVA results and the normalized count data, extract significant results, and then see look at results overlapping between the two in the following subsections.
Besides statistical thresholds, an important parameter is the number of significant results required in the `decideTests` output.

```{r}
seurat <- readRDS(file.path(cache_dir, "..", "..", "02", paste0(subset_names[1], "_seurat.rds")))
seurat
red_dim_plot(seurat, "umap1", "umap2", "seurat_clusters", "cat")
red_dim_plot(seurat, "umap1", "umap2", "cluster_stem", "cat")
red_dim_plot(seurat, "umap1", "umap2", "donor_id", "cat")
```

# GO BP/CC/MF

```{r, fig.height = 5, fig.width = 7.5}
section_no <- 1
gene_set_name <- gene_set_list[section_no]
seurat <- readRDS(file.path(cache_dir, "..", "..", "02", paste0(subset_names[1], "_seurat.rds")))

rds <- file.path(cache_dir, paste0(gene_set_name, "_gene_sets.rds"))
rds2 <- file.path(cache_dir, paste0(gene_set_name, "_gene_sets_filtered.rds"))
if (file.exists(rds) & file.exists(rds2)) {
  gene_sets <- readRDS(rds)
  gene_sets_filtered <- readRDS(rds2)
} else {
  gene_sets <- getGmt(
    file.path(assets_dir, "gene-sets", paste0(organism, ".", gene_set_name, ".ENSG.gmt"))
  )
  gene_sets <- filterGeneSets(gene_sets, 5, 500)
  gene_sets_filtered <- gene_sets # Put aside for further filtering later.
  for (i in seq(length(gene_sets@.Data))) {
    go_id <- gene_sets[[i]]@setName
    suppressWarnings(gene_sets[[i]]@setName <- gene_sets[[i]]@shortDescription)
    suppressWarnings(gene_sets[[i]]@shortDescription <- go_id)
  }

  filter_sets <- read.delim(file.path(assets_dir, "gene-sets", "genes-09-00593-s001-s5.tsv"))
  filter_sets <- filter_sets$GO.TERM
  gene_sets_filtered <- gene_sets_filtered[names(gene_sets_filtered) %in% filter_sets]
  for (i in seq(length(gene_sets_filtered@.Data))) {
    go_id <- gene_sets_filtered[[i]]@setName
    suppressWarnings(gene_sets_filtered[[i]]@setName <- gene_sets_filtered[[i]]@shortDescription)
    suppressWarnings(gene_sets_filtered[[i]]@shortDescription <- go_id)
  }

  saveRDS(gene_sets, rds)
  saveRDS(gene_sets_filtered, rds2)
}
gene_sets

iss_codebook <- read.csv(file.path(assets_dir, "gene-sets", iss_codebook_name), sep = "\t")
print("In-situ sequencing (ISS) codebook")
head(iss_codebook)

# Create a Seurat object containing only clusters of interest with downsampling.
# Also convert gene symbols to ENSEMBL IDs.
# ------------------------------------------------------------------------------
rds <- file.path(cache_dir, paste0(gene_set_name, "_seurat_cleaned.rds"))
rds2 <- file.path(cache_dir, paste0(gene_set_name, "_gene_anno.rds"))
if (file.exists(rds) & file.exists(rds2)) {
  seurat <- readRDS(rds)
  gene_anno <- readRDS(rds2)
} else {

  # Remove obsolete data.
  # ---------------------
  DefaultAssay(seurat) <- "RNA"
  seurat[["SCT"]] <- NULL
  # ---------------------

  # Create subsets.
  # ---------------
  seurat1 <- subset(seurat, subset = donor_id == 453 | donor_id == 454 | donor_id == 448)
  expr <- FetchData(seurat1, "cluster_stem")

  condition1_seurat <- seurat1[ , which(expr == idents[[1]])]
  set.seed(1)
  condition1_seurat <- subset(
    condition1_seurat, cells = sample(Cells(condition1_seurat), downsample)
  )
  sample1_seurat <- subset(condition1_seurat, subset = donor_id == 453)
  sample2_seurat <- subset(condition1_seurat, subset = donor_id == 454)
  sample3_seurat <- subset(condition1_seurat, subset = donor_id == 448)
  condition1_seurat <- merge(sample1_seurat, c(sample2_seurat, sample3_seurat))

  condition2_seurat <- seurat1[ , which(expr == idents[[2]])]
  set.seed(1)
  condition2_seurat <- subset(
    condition2_seurat, cells = sample(Cells(condition2_seurat), downsample)
  )
  sample1_seurat <- subset(condition2_seurat, subset = donor_id == 453)
  sample2_seurat <- subset(condition2_seurat, subset = donor_id == 454)
  sample3_seurat <- subset(condition2_seurat, subset = donor_id == 448)
  condition2_seurat <- merge(sample1_seurat, c(sample2_seurat, sample3_seurat))

  seurat2 <- subset(seurat, subset = donor_id == 490 | donor_id == 492 | donor_id == 493)
  expr <- FetchData(seurat2, "seurat_clusters")

  condition3_seurat <- seurat2[ , which(expr == idents[[3]])]
  set.seed(1)
  condition3_seurat <- subset(
    condition3_seurat, cells = sample(Cells(condition3_seurat), downsample)
  )
  sample1_seurat <- subset(condition3_seurat, subset = donor_id == 490)
  sample2_seurat <- subset(condition3_seurat, subset = donor_id == 492)
  sample3_seurat <- subset(condition3_seurat, subset = donor_id == 493)
  condition3_seurat <- merge(sample1_seurat, c(sample2_seurat, sample3_seurat))

  condition4_seurat <- seurat2[ , which(expr == idents[[4]])]
  set.seed(1)
  condition4_seurat <- subset(
    condition4_seurat, cells = sample(Cells(condition4_seurat), downsample)
  )
  sample1_seurat <- subset(condition4_seurat, subset = donor_id == 490)
  sample2_seurat <- subset(condition4_seurat, subset = donor_id == 492)
  sample3_seurat <- subset(condition4_seurat, subset = donor_id == 493)
  condition4_seurat <- merge(sample1_seurat, c(sample2_seurat, sample3_seurat))
  
  seurat <- merge(condition1_seurat, c(condition2_seurat, condition3_seurat, condition4_seurat))
  rm(
    condition1_seurat, condition2_seurat, condition3_seurat, condition4_seurat, sample1_seurat,
    sample2_seurat, sample3_seurat, seurat1, seurat2
  )
  seurat <- seurat[rowSums(as.matrix(GetAssayData(seurat, slot = "counts")) > 0) > 0, ]
  # ---------------

  # Get gene annotations.
  # ---------------------
  mart <- useEnsembl("ensembl", paste0(organism, "_gene_ensembl"))
  attributes <- c("external_gene_name", "ensembl_gene_id")
  gene_anno <- getBM(attributes, "external_gene_name", rownames(seurat), mart)
  gene_anno <- gene_anno[gene_anno$ensembl_gene_id %in% unique(unlist(geneIds(gene_sets))), ]
  # ---------------------

  # For gene symbols with multiple ENSEMBL IDs, duplicate the gene symbol to have an identical row
  # for each ENSEMBL ID.
  # ----------------------------------------------------------------------------------------------
  dup <- gene_anno[duplicated(gene_anno$external_gene_name), ]
  if (nrow(dup) > 0) {
    for (i in 1:dim(dup)[1]) {
      for (j in 1:dim(gene_anno)[1]) {
        if (dup$ensembl_gene_id[i] == gene_anno$ensembl_gene_id[j]) {
          gene_anno$external_gene_name[j] <- paste0(gene_anno$external_gene_name[j], "-alternative")
        }
      }
    }
    gene_anno$external_gene_name[which(duplicated(gene_anno$external_gene_name))] <-
      "Ccl27a-alternative2" # TODO: Remove hardcoding.
    if (any(duplicated(gene_anno$external_gene_name))) {
      stop("Duplicates in gene_anno.")
    }
    seurat <- seurat[rownames(seurat) %in% gene_anno$external_gene_name, ]
    new_mat <- GetAssayData(seurat, "counts")
    for (i in 1:dim(dup)[1]) {
      for (j in 1:dim(seurat)[1]) {
        if (dup$external_gene_name[i] == rownames(seurat)[j]) {
          new_row <- GetAssayData(seurat[j, ], "counts")
          rownames(new_row) <- paste0(rownames(new_row), "-alternative")
          if (rownames(new_row) %in% rownames(new_mat)) {
            rownames(new_row) <- paste0(rownames(new_row), "2")
          }
          new_mat <- rbind(new_mat, new_row)
        }
      }
    }
  } else {
    seurat <- seurat[rownames(seurat) %in% gene_anno$external_gene_name, ]
    new_mat <- GetAssayData(seurat, "counts")
  }
  gene_anno <- gene_anno[gene_anno$external_gene_name %in% rownames(new_mat), ]
  gene_anno <- gene_anno[order(match(gene_anno$external_gene_name, rownames(new_mat))), ]
  rownames(new_mat) <- gene_anno$ensembl_gene_id

  seurat <- CreateSeuratObject(new_mat, meta.data = seurat[[]])
  rm(new_mat)
  seurat$idents <- factor(
    c(
      rep(names(idents)[1], downsample), rep(names(idents)[2], downsample),
      rep(names(idents)[3], downsample), rep(names(idents)[4], downsample)
    ),
    levels = names(idents)
  )
  seurat@active.ident <- seurat$idents
  seurat$groups <- paste0(seurat$idents, "_", seurat$donor_id)
  # ----------------------------------------------------------------------------------------------
  
  saveRDS(seurat, rds)
  saveRDS(gene_anno, rds2)
}
# ------------------------------------------------------------------------------

# Renormalize using SCnorm using sample-wise condition information and gene-wise gene length and GC
# content information.
# -------------------------------------------------------------------------------------------------
rds <- file.path(cache_dir, paste0(gene_set_name, "_scnorm.rds"))
if (file.exists(rds)) {
  sce <- readRDS(rds)
} else {
  sce <- SingleCellExperiment(list(counts = as.matrix(GetAssayData(seurat))))
  groups <- factor(rep(c(0, 1), c(downsample, downsample)))
  row_data <- data.frame(
    getGeneLengthAndGCContent(rownames(sce), "mmusculus_gene_ensembl", "biomart")
  )
  sce <- SCnorm(
    sce, groups, withinSample = c(row_data$gc, row_data$length),
    useZerosToScale = TRUE, NCores = detectCores()
  )
  discard <- unique(c(metadata(sce)$GenesFilteredOut[[1]], metadata(sce)$GenesFilteredOut[[2]]))
  sce <- sce[-which(rownames(sce) %in% discard), ]
  saveRDS(sce, rds)
}
# -------------------------------------------------------------------------------------------------

# Subset other objects.
# ---------------------
gene_anno <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(sce), ]
seurat <- seurat[rownames(seurat) %in% rownames(sce), ]
#----------------------

# Add normalized data to Seurat object.
# -------------------------------------
seurat <- SetAssayData(seurat, new.data = normcounts(sce))
seurat <- SetAssayData(seurat, "scale.data", log2(normcounts(sce) + 1))
# -------------------------------------

# Create MDS plots to identify potential confounders and latent factors.
# ----------------------------------------------------------------------
rds <- file.path(cache_dir, paste0(gene_set_name, "_d.rds"))
if (file.exists(rds)) {
  d <- readRDS(rds)
} else {
  d <- dist(t(scale(GetAssayData(seurat, "scale.data"))))
  saveRDS(d, rds)
}
mds <- cmdscale(d, 2)
colnames(mds) <- paste0("mds_", 1:2)
seurat[["mds"]] <- CreateDimReducObject(mds, key = "mds_", assay = DefaultAssay(seurat))
rm(mds)
add_df <- data.frame(Embeddings(seurat, "mds"))
names(add_df) <- paste0("mds", seq(ncol(add_df)))
seurat$mds1 <- add_df$mds1
seurat$mds2 <- add_df$mds2

red_dim_plot(seurat, "umap1", "umap2", "idents", "cat")
for (i in 1:length(metadata_to_plot)) {
  print(names(metadata_to_plot)[i])
  print(red_dim_plot(seurat, "mds1", "mds2", names(metadata_to_plot)[i], "cat"))
}
# ----------------------------------------------------------------------

print("SCnorm normalized counts distribution")
hist(as.matrix(GetAssayData(seurat, "scale.data")))

# Aggregate features to into "pseudo-bulk" counts.
# ------------------------------------------------
rds <- file.path(cache_dir, paste0(gene_set_name, "_pseudobulk.rds"))
if (file.exists(rds)) {
  sce <- readRDS(rds)
} else {
  sce <- as.SingleCellExperiment(seurat)
  counts(sce) <- logcounts(sce)
  sce <- suppressWarnings(
    aggregateAcrossCells(
      sce, colData(sce)[ , c("idents", "donor_id")], average = TRUE,
      use_exprs_values = "counts", BPPARAM = MulticoreParam()
    )
  )
  logcounts(sce) <- log2(counts(sce) + 1)
  saveRDS(sce, rds)
}
print("Pseudo-bulk counts distribution")
hist(logcounts(sce))
# ------------------------------------------------

# Run GSVA on all cells.
# ----------------------
rds <- file.path(cache_dir, paste0(gene_set_name, "_gsva_allcells.rds"))
if (file.exists(rds)) {
  gsva <- readRDS(rds)
} else {
  gsva <- gsva(as.matrix(GetAssayData(seurat, "scale.data")), gene_sets)
  saveRDS(gsva, rds)
}
seurat[["GSVA"]] <- CreateAssayObject(gsva)
DefaultAssay(seurat) <- "GSVA"
print("GSVA distribution (all cells)")
hist(as.matrix(GetAssayData(seurat, "counts")))
# ----------------------

# Run GSVA on pseudo-bulk counts.
# -------------------------------
rds <- file.path(cache_dir, paste0(gene_set_name, "_gsva_pseudobulk.rds"))
if (file.exists(rds)) {
  gsva <- readRDS(rds)
} else {
  gsva <- gsva(logcounts(sce), gene_sets)
  colnames(gsva) <- sce$groups
  saveRDS(gsva, rds)
}
# -------------------------------

# Add previous UMAP to Seurat object.
# -----------------------------------
umap <- as.matrix(data.frame(seurat$umap1, seurat$umap2))
colnames(umap) <- paste0("umap_", 1:2)
seurat[["umap"]] <- CreateDimReducObject(umap, key = "umap_", assay = DefaultAssay(seurat))
rm(umap)
# -----------------------------------
```

## All Results

```{r}
# Fit model on pseudo-bulk GSVA results.
# --------------------------------------
design <- model.matrix(~ 0 + sce$idents, as.data.frame(gsva))
colnames(design) <- names(idents)
fit <- lmFit(gsva, design)
contrast_mat <- makeContrasts(
  EC_L2_Lat-CA3, EC_L2_Lat-CA1, EC_L2_Lat-EC_L2_Med, EC_L2_Med-CA3,
  EC_L2_Med-CA1, CA1-CA3, levels = design
)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat))
tests <- decideTests(cont_fit, "nestedF")
tests_up <- tests[tests[ , 1] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, -1]))
tests_up <- tests_up[-discard, ]
tests_down <- tests[tests[ , 1] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, -1]))
tests_down <- tests_down[-discard, ]
results <- topTable(cont_fit, 1, Inf, p.value = 0.05)
# --------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
gsva_list_pseudobulk <- list(all, up, down)
names(gsva_list_pseudobulk) <- c(
  "All DE", paste0("Upregulated in ", names(idents)[1]),
  paste0("Downregulated in ", names(idents)[2])
)

# Fit model on all cells GSVA results.
# ------------------------------------
DefaultAssay(seurat) <- "GSVA"
design <- model.matrix(~ 0 + seurat$idents, as.data.frame(GetAssayData(seurat, "counts")))
colnames(design) <- names(idents)
fit <- lmFit(as.data.frame(GetAssayData(seurat, "counts")), design)
contrast_mat <- makeContrasts(
  EC_L2_Lat-CA3, EC_L2_Lat-CA1, EC_L2_Lat-EC_L2_Med, EC_L2_Med-CA3,
  EC_L2_Med-CA1, CA1-CA3, levels = design
)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat))
tests <- decideTests(cont_fit, "nestedF")
tests_up <- tests[tests[ , 1] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, -1]))
tests_up <- tests_up[-discard, ]
tests_down <- tests[tests[ , 1] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, -1]))
tests_down <- tests_down[-discard, ]
results <- topTable(cont_fit, 1, Inf, p.value = 0.05)
# --------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
gsva_list_allcells <- list(all, up, down)
names(gsva_list_allcells) <- c(
  "All DE", paste0("Upregulated in ", names(idents)[1]),
  paste0("Downregulated in ", names(idents)[2])
)

color <- colorRampPalette(rev(brewer.pal(9, "RdBu")))(100)
```

**`r names(idents)[1]` Upregulated DE Gene Sets Pseudo-bulk**

```{r}
datatable_download_exp(gsva_list_pseudobulk[[2]])
```

**`r names(idents)[1]` Upregulated DE Gene Sets All Cells**

```{r}
datatable_download_exp(gsva_list_allcells[[2]])
```

**`r names(idents)[1]` Downregulated DE Gene Sets Pseudo-bulk**

```{r}
datatable_download_exp(gsva_list_pseudobulk[[3]])
```

**`r names(idents)[1]` Downregulated DE Gene Sets All Cells**

```{r}
datatable_download_exp(gsva_list_allcells[[3]])
```

```{r, fig.width = 15}
anno_color <- hue_pal()(length(unique(sce$donor_id)))
gsva_heatmap <- vector("list", 3)
nbclust_fun <- function(x, k) {
  list(cluster = cutree(hclust(as.dist(1 - x), "ward.D2"), k))
}

for (i in seq_along(gsva_list_pseudobulk)) {
  if (nrow(gsva_list_pseudobulk[[i]]) > 0) {
    mat <- gsva
    mat <- mat[rownames(mat) %in% rownames(gsva_list_pseudobulk[[i]]), ]
    gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(mat)]
    gene_sets_sub <- gene_sets_sub[names(gene_sets_sub) %in% names(gene_sets_filtered)]
    mat <- mat[rownames(mat) %in% names(gene_sets_sub), ]
    if (is.null(nrow(mat)) || dim(mat)[1] == 0) {
    } else {
      gene_set_overlap <- computeGeneSetsOverlap(
        gene_sets_sub, unique(unlist(geneIds(gene_sets_sub)))
      )
      clustering <- hclust(as.dist(1 - gene_set_overlap), "ward.D2")
      ha <- HeatmapAnnotation(
        donor = sce$donor_id,
        cell_type = anno_block(
          gp = gpar(fill = "lightgray", col = "lightgray"),
          labels = c(
            names(idents)[[1]], names(idents)[[2]], names(idents)[[3]], names(idents)[[4]]
          ),
          labels_gp = gpar(col = "black")
        ),
        col = list(
          donor = c(
            "453" = anno_color[1], "454" = anno_color[2], "448" = anno_color[3],
            "490" = anno_color[4], "492" = anno_color[5], "493" = anno_color[6]
          )
        )
      )
      if (nrow(mat) >= 4) {
        nbclust <- fviz_nbclust(
          gene_set_overlap, nbclust_fun, k.max = round(nrow(mat) / 2),
          print.summary = FALSE, verbose = FALSE
        )
        optimal_k <- nbclust$data
        optimal_k <- as.numeric(optimal_k$clusters[which.max(optimal_k$y)])
        set.seed(1)
        gsva_heatmap[[i]] <- Heatmap(
          mat,
          color,
          cluster_rows = clustering,
          cluster_columns = FALSE,
          row_names_max_width = max_text_width(rownames(mat)),
          show_column_names = FALSE,
          column_split = sce$idents,
          show_heatmap_legend = FALSE,
          top_annotation = ha,
          rect_gp = gpar(col = "lightgray"),
          column_title = paste0(names(gsva_list_pseudobulk)[i], " AD Relevent GO Terms"),
          row_split = optimal_k,
          row_gap = unit(2, "mm"),
          column_gap = unit(2, "mm")
        )
      } else {
        set.seed(1)
        gsva_heatmap[[i]] <- Heatmap(
          mat,
          color,
          cluster_rows = clustering,
          cluster_columns = FALSE,
          row_names_max_width = max_text_width(rownames(mat)),
          show_column_names = FALSE,
          column_split = sce$idents,
          show_heatmap_legend = FALSE,
          top_annotation = ha,
          rect_gp = gpar(col = "lightgray"),
          column_title = paste0(names(gsva_list_pseudobulk)[i], " AD Relevent GO Terms"),
          row_gap = unit(2, "mm"),
          column_gap = unit(2, "mm")
        )
      }
      draw(gsva_heatmap[[i]])

      mat <- as.matrix(GetAssayData(seurat, "counts"))
      mat <- mat[rownames(mat) %in% rownames(gsva_list_pseudobulk[[i]]), ]
      mat <- mat[rownames(mat) %in% names(gene_sets_sub), ]
      ha <- HeatmapAnnotation(
        donor = seurat$donor_id,
        cell_type = anno_block(
          gp = gpar(fill = "lightgray", col = "lightgray"),
          labels = c(
            names(idents)[[1]], names(idents)[[2]], names(idents)[[3]], names(idents)[[4]]
          ),
          labels_gp = gpar(col = "black")
        ),
        col = list(
          donor = c(
            "453" = anno_color[1], "454" = anno_color[2], "448" = anno_color[3],
            "490" = anno_color[4], "492" = anno_color[5], "493" = anno_color[6]
          )
        )
      )
      if (nrow(mat) >= 4) {
        set.seed(1)
        draw(
          Heatmap(
            mat,
            color,
            cluster_rows = clustering,
            cluster_columns = FALSE,
            row_names_max_width = max_text_width(rownames(mat)),
            show_column_names = FALSE,
            column_split = seurat$idents,
            show_heatmap_legend = FALSE,
            top_annotation = ha,
            column_title = paste0(names(gsva_list_allcells)[i], " AD Relevent GO Terms"),
            row_split = optimal_k,
            row_gap = unit(2, "mm"),
            column_gap = unit(2, "mm")
          )
        )
      } else {
        set.seed(1)
        draw(
          Heatmap(
            mat,
            color,
            cluster_rows = clustering,
            cluster_columns = FALSE,
            row_names_max_width = max_text_width(rownames(mat)),
            show_column_names = FALSE,
            column_split = seurat$idents,
            show_heatmap_legend = FALSE,
            top_annotation = ha,
            column_title = paste0(names(gsva_list_allcells)[i], " AD Relevent GO Terms"),
            row_gap = unit(2, "mm"),
            column_gap = unit(2, "mm")
          )
        )
      }
      print(
        FeaturePlot(
          seurat, rownames(mat)[unlist(row_order(gsva_heatmap[[i]]))[1:4]], order = TRUE,
          cols = c("lawngreen", "firebrick1"), ncol = 2
        )
      )
      print(VlnPlot(seurat, rownames(mat)[unlist(row_order(gsva_heatmap[[i]]))[1:6]], ncol = 2))
    }
  }
}

ontology <- c("BP", "CC", "MF")
for (k in seq_along(ontology)) {
  if (organism == "hsapiens") {
    go_data <- godata("org.Hs.eg.db", "ENSEMBL", ontology[k])
  } else if (organism == "mmusculus") {
    go_data <- godata("org.Mm.eg.db", "ENSEMBL", ontology[k])
  }
  for (i in seq_along(gsva_list_pseudobulk)) {
    if (nrow(gsva_list_pseudobulk[[i]]) > 0) {
      mat <- gsva
      mat <- mat[rownames(mat) %in% rownames(gsva_list_pseudobulk[[i]]), ]
      gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(mat)]
      gene_sets_sub <- gene_sets_sub[names(gene_sets_sub) %in% names(gene_sets_filtered)]
      mat <- mat[rownames(mat) %in% names(gene_sets_sub), ]
      gene_sets_sub_ids <- gene_sets_sub
      for (j in seq(length(gene_sets_sub_ids@.Data))) {
        suppressWarnings(gene_sets_sub_ids[[j]]@setName <- gene_sets_sub_ids[[j]]@shortDescription)
      }
      keep <- which(names(gene_sets_sub_ids) %in% go_data@geneAnno$GO)
      gene_sets_sub_ids <- gene_sets_sub_ids[keep, ]
      gene_sets_sub <- gene_sets_sub[keep, ]
      mat <- mat[keep, ]
      if (is.null(nrow(mat)) || dim(mat)[1] == 0) {
      } else {
        gene_set_overlap <- mgoSim(
          names(gene_sets_sub_ids), names(gene_sets_sub_ids), go_data, "Jiang", combine = NULL
        )
        gene_set_overlap <- gene_set_overlap[rowSums(is.na(gene_set_overlap)) != ncol(gene_set_overlap), ]
        gene_set_overlap <- gene_set_overlap[ , colSums(is.na(gene_set_overlap)) != nrow(gene_set_overlap)]
        keep <- which(names(gene_sets_sub_ids) %in% rownames(gene_set_overlap))
        gene_sets_sub_ids <- gene_sets_sub_ids[keep, ]
        gene_sets_sub <- gene_sets_sub[keep, ]
        mat <- mat[keep, ]
        clustering <- hclust(as.dist(1 - gene_set_overlap), "ward.D2")
        ha <- HeatmapAnnotation(
          donor = sce$donor_id,
          cell_type = anno_block(
            gp = gpar(fill = "lightgray", col = "lightgray"),
            labels = c(
              names(idents)[[1]], names(idents)[[2]], names(idents)[[3]], names(idents)[[4]]
            ),
            labels_gp = gpar(col = "black")
          ),
          col = list(
            donor = c(
              "453" = anno_color[1], "454" = anno_color[2], "448" = anno_color[3],
              "490" = anno_color[4], "492" = anno_color[5], "493" = anno_color[6]
            )
          )
        )
        if (nrow(mat) >= 4) {
          nbclust <- fviz_nbclust(
            gene_set_overlap, nbclust_fun, k.max = round(nrow(mat) / 2),
            print.summary = FALSE, verbose = FALSE
          )
          optimal_k <- nbclust$data
          optimal_k <- as.numeric(optimal_k$clusters[which.max(optimal_k$y)])
          set.seed(1)
          gsva_heatmap[[i]] <- Heatmap(
            mat,
            color,
            cluster_rows = clustering,
            cluster_columns = FALSE,
            row_names_max_width = max_text_width(rownames(mat)),
            show_column_names = FALSE,
            column_split = sce$idents,
            show_heatmap_legend = FALSE,
            top_annotation = ha,
            rect_gp = gpar(col = "lightgray"),
            column_title = paste0(
              names(gsva_list_pseudobulk)[i], " AD Relevent GO ", ontology[k], " Terms"
            ),
            row_split = optimal_k,
            row_gap = unit(2, "mm"),
            column_gap = unit(2, "mm")
          )
        } else {
          set.seed(1)
          gsva_heatmap[[i]] <- Heatmap(
            mat,
            color,
            cluster_rows = clustering,
            cluster_columns = FALSE,
            row_names_max_width = max_text_width(rownames(mat)),
            show_column_names = FALSE,
            column_split = sce$idents,
            show_heatmap_legend = FALSE,
            top_annotation = ha,
            rect_gp = gpar(col = "lightgray"),
            column_title = paste0(
              names(gsva_list_pseudobulk)[i], " AD Relevent GO ", ontology[k], " Terms"
            ),
            row_gap = unit(2, "mm"),
            column_gap = unit(2, "mm")
          )
        }
        draw(gsva_heatmap[[i]])

        mat <- as.matrix(GetAssayData(seurat, "counts"))
        mat <- mat[rownames(mat) %in% rownames(gsva_list_pseudobulk[[i]]), ]
        mat <- mat[rownames(mat) %in% names(gene_sets_sub), ]
        ha <- HeatmapAnnotation(
          donor = seurat$donor_id,
          cell_type = anno_block(
            gp = gpar(fill = "lightgray", col = "lightgray"),
            labels = c(
              names(idents)[[1]], names(idents)[[2]], names(idents)[[3]], names(idents)[[4]]
            ),
            labels_gp = gpar(col = "black")
          ),
          col = list(
            donor = c(
              "453" = anno_color[1], "454" = anno_color[2], "448" = anno_color[3],
              "490" = anno_color[4], "492" = anno_color[5], "493" = anno_color[6]
            )
          )
        )
        if (nrow(mat) >= 4) {
          set.seed(1)
          draw(
            Heatmap(
              mat,
              color,
              cluster_rows = clustering,
              cluster_columns = FALSE,
              row_names_max_width = max_text_width(rownames(mat)),
              show_column_names = FALSE,
              column_split = seurat$idents,
              show_heatmap_legend = FALSE,
              top_annotation = ha,
              column_title = paste0(
                names(gsva_list_allcells)[i], " AD Relevent GO ", ontology[k], " Terms"
              ),
              row_split = optimal_k,
              row_gap = unit(2, "mm"),
              column_gap = unit(2, "mm")
            )
          )
        } else {
          set.seed(1)
          draw(
            Heatmap(
              mat,
              color,
              cluster_rows = clustering,
              cluster_columns = FALSE,
              row_names_max_width = max_text_width(rownames(mat)),
              show_column_names = FALSE,
              column_split = seurat$idents,
              show_heatmap_legend = FALSE,
              top_annotation = ha,
              column_title = paste0(
                names(gsva_list_allcells)[i], " AD Relevent GO ", ontology[k], " Terms"
              ),
              row_gap = unit(2, "mm"),
              column_gap = unit(2, "mm")
            )
          )
        }
        print(
          FeaturePlot(
            seurat, rownames(mat)[unlist(row_order(gsva_heatmap[[i]]))[1:4]], order = TRUE,
            cols = c("lawngreen", "firebrick1"), ncol = 2
          )
        )
        print(VlnPlot(seurat, rownames(mat)[unlist(row_order(gsva_heatmap[[i]]))[1:6]], ncol = 2))
      }
    }
  }
}

# Fit model on pseudo-bulk counts.
# --------------------------------
design <- model.matrix(~ 0 + sce$idents, as.data.frame(logcounts(sce)))
colnames(design) <- names(idents)
fit <- lmFit(logcounts(sce), design)
contrast_mat <- makeContrasts(
  EC_L2_Lat-CA3, EC_L2_Lat-CA1, EC_L2_Lat-EC_L2_Med, EC_L2_Med-CA3,
  EC_L2_Med-CA1, CA1-CA3, levels = design
)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat))
tests <- decideTests(cont_fit, "nestedF")
tests_up <- tests[tests[ , 1] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, -1]))
tests_up <- tests_up[-discard, ]
tests_down <- tests[tests[ , 1] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, -1]))
tests_down <- tests_down[-discard, ]
results <- topTable(cont_fit, 1, Inf, p.value = 0.05)
# --------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
markers_list_pseudobulk <- list(all, up, down)

# Fit model on all cells counts.
# ------------------------------
DefaultAssay(seurat) <- "RNA"
design <- model.matrix(~ 0 + seurat$idents, as.data.frame(GetAssayData(seurat, "scale.data")))
colnames(design) <- names(idents)
fit <- lmFit(as.data.frame(GetAssayData(seurat, "scale.data")), design)
contrast_mat <- makeContrasts(
  EC_L2_Lat-CA3, EC_L2_Lat-CA1, EC_L2_Lat-EC_L2_Med, EC_L2_Med-CA3,
  EC_L2_Med-CA1, CA1-CA3, levels = design
)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat))
tests <- decideTests(cont_fit, "nestedF")
tests_up <- tests[tests[ , 1] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, -1]))
tests_up <- tests_up[-discard, ]
tests_down <- tests[tests[ , 1] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, -1]))
tests_down <- tests_down[-discard, ]
results <- topTable(cont_fit, 1, Inf, p.value = 0.05)
# --------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
markers_list_allcells <- list(all, up, down)

# Convert ENSEMBL IDs back to gene symbols.
# -----------------------------------------
genes <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(markers_list_pseudobulk[[2]]), ]
genes <- genes[order(match(genes$ensembl_gene_id, rownames(markers_list_pseudobulk[[2]]))), ]
markers_list_symbols_up_pseudobulk <- markers_list_pseudobulk[[2]]
rownames(markers_list_symbols_up_pseudobulk) <- genes$external_gene_name

genes <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(markers_list_pseudobulk[[3]]), ]
genes <- genes[order(match(genes$ensembl_gene_id, rownames(markers_list_pseudobulk[[3]]))), ]
markers_list_symbols_down_pseudobulk <- markers_list_pseudobulk[[3]]
rownames(markers_list_symbols_down_pseudobulk) <- genes$external_gene_name

genes <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(markers_list_allcells[[2]]), ]
genes <- genes[order(match(genes$ensembl_gene_id, rownames(markers_list_allcells[[2]]))), ]
markers_list_symbols_up_allcells <- markers_list_allcells[[2]]
rownames(markers_list_symbols_up_allcells) <- genes$external_gene_name

genes <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(markers_list_allcells[[3]]), ]
genes <- genes[order(match(genes$ensembl_gene_id, rownames(markers_list_allcells[[3]]))), ]
markers_list_symbols_down_allcells <- markers_list_allcells[[3]]
rownames(markers_list_symbols_down_allcells) <- genes$external_gene_name
# -----------------------------------------
```

**`r names(idents)[1]` Upregulated DE Genes Pseudo-bulk**

```{r}
datatable_download_exp(markers_list_symbols_up_pseudobulk)
```

**`r names(idents)[1]` Upregulated DE Genes All Cells**

```{r}
datatable_download_exp(markers_list_symbols_up_allcells)
```

**`r names(idents)[1]` Downregulated DE Genes Pseudo-bulk**

```{r}
datatable_download_exp(markers_list_symbols_down_pseudobulk)
```

**`r names(idents)[1]` Downregulated DE Genes All Cells**

```{r}
datatable_download_exp(markers_list_symbols_down_allcells)
```

```{r}
# Check if any markers are in ISS codebook.
# -----------------------------------------
iss_genes <- rownames(markers_list_symbols_up_pseudobulk)[
  rownames(markers_list_symbols_up_pseudobulk) %in% iss_codebook[ , 1]
]
paste0("Upregulated pseudo-bulk genes in ISS codebook: ", paste(iss_genes, collapse = ", "))

iss_genes <- rownames(markers_list_symbols_down_pseudobulk)[
  rownames(markers_list_symbols_down_pseudobulk) %in% iss_codebook[ , 1]
]
paste0("Downregulated pseudo-bulk genes in ISS codebook: ", paste(iss_genes, collapse = ", "))
# -----------------------------------------

# Create new Seurat object with gene symbols.
# -------------------------------------------
new_mat <- GetAssayData(seurat)
rownames(new_mat) <- gene_anno$external_gene_name
seurat_symbols <- CreateSeuratObject(new_mat, meta.data = seurat[[]])
Idents(seurat_symbols) <- "idents"
seurat_symbols <- SetAssayData(
  seurat_symbols, new.data = log2(as.matrix(GetAssayData(seurat_symbols, "counts")) + 1)
)
seurat_symbols <- SetAssayData(
  seurat_symbols, "scale.data", t(scale(t(GetAssayData(seurat_symbols))))
)
# -------------------------------------------

color2 <- colorRampPalette(brewer.pal(9, "YlOrBr"))(round(max(GetAssayData(seurat_symbols))))
color2[1] <- "#DDDDDD"
```

### `r names(idents)[1]` Upregulation

We display DE genes within gene sets that are both upregulated in the direction of the trajectory start point.

```{r, fig.height = 5, fig.width = 7.5}
# TODO: Fix problem of contents appearing out of order.
# -----------------------------------------------------
set <- 2 # Up or downregulation?
if (nrow(gsva_list_pseudobulk[[set]]) > 0) {
  mat <- gsva_list_pseudobulk[[set]]
  gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(mat)]
  gene_sets_sub <- gene_sets_sub[names(gene_sets_sub) %in% names(gene_sets_filtered)]
  mat <- mat[rownames(mat) %in% names(gene_sets_sub), ]
  for (i in 1:nrow(mat)) {
    genes <- which(
      gene_sets[[rownames(mat)[i]]]@geneIds %in% rownames(markers_list_pseudobulk[[set]])
    )
    genes <- gene_sets[[rownames(mat)[i]]]@geneIds[genes]
    genes <- gene_anno[gene_anno$ensembl_gene_id %in% genes, ]
    genes <- genes$external_gene_name
    title <- paste0(
      gene_sets[[rownames(mat)[i]]]@shortDescription, ": ", gene_sets[[rownames(mat)[i]]]@setName
    )
    print(title)

    if (length(genes) > 0) {
      print(
        paste0(
          gene_sets[[rownames(mat)[i]]]@shortDescription, " DE genes: ",
          paste(genes, collapse = ", ")
        )
      )
      iss_genes <- genes[genes %in% iss_codebook[ , 1]]
      print(
        paste0(
          gene_sets[[rownames(mat)[i]]]@shortDescription, " DE genes in ISS codebook: ",
          paste(iss_genes, collapse = ", ")
        )
      )
      ha <- HeatmapAnnotation(
        donor = seurat_symbols$donor_id,
        cell_type = anno_block(
          gp = gpar(fill = "lightgray", col = "lightgray"),
          labels = c(
            names(idents)[[1]], names(idents)[[2]], names(idents)[[3]], names(idents)[[4]]
          ),
          labels_gp = gpar(col = "black")
        ),
        col = list(
          donor = c(
            "453" = anno_color[1], "454" = anno_color[2], "448" = anno_color[3],
            "490" = anno_color[4], "492" = anno_color[5], "493" = anno_color[6]
          )
        )
      )
      draw(
        Heatmap(
          as.matrix(GetAssayData(seurat_symbols)[genes, , drop = FALSE]), color2,
          cluster_rows = FALSE, cluster_columns = FALSE,
          show_column_names = FALSE, column_split = seurat_symbols$idents,
          show_heatmap_legend = FALSE, top_annotation = ha,
          column_title = title, column_gap = unit(2, "mm")
        )
      )
    }

    genes <- gene_sets[[rownames(mat)[i]]]@geneIds
    genes <- gene_anno[gene_anno$ensembl_gene_id %in% genes, ]
    genes <- genes$external_gene_name
    if (length(genes) > 0) {
      print(
        paste0(
          gene_sets[[rownames(mat)[i]]]@shortDescription,
          " all genes: ",
          paste(genes, collapse = ", ")
        )
      )
      iss_genes <- genes[genes %in% iss_codebook[ , 1]]
      print(
        paste0(
          gene_sets[[rownames(mat)[i]]]@shortDescription,
          " all genes in ISS codebook: ",
          paste(iss_genes, collapse = ", ")
        )
      )
      ha <- HeatmapAnnotation(
        donor = seurat_symbols$donor_id,
        cell_type = anno_block(
          gp = gpar(fill = "lightgray", col = "lightgray"),
          labels = c(
            names(idents)[[1]], names(idents)[[2]], names(idents)[[3]], names(idents)[[4]]
          ),
          labels_gp = gpar(col = "black")
        ),
        col = list(
          donor = c(
            "453" = anno_color[1], "454" = anno_color[2], "448" = anno_color[3],
            "490" = anno_color[4], "492" = anno_color[5], "493" = anno_color[6]
          )
        )
      )
      draw(
        Heatmap(
          as.matrix(GetAssayData(seurat_symbols)[genes, , drop = FALSE]), color2,
          cluster_rows = FALSE, cluster_columns = FALSE,
          show_column_names = FALSE, column_split = seurat_symbols$idents,
          show_heatmap_legend = FALSE, top_annotation = ha,
          column_title = title, column_gap = unit(2, "mm")
        )
      )
    }
  }
}
# -----------------------------------------------------
```

### `r names(idents)[1]` Downregulation

We display DE genes within gene sets that are both upregulated in the direction of the trajectory start point.

```{r, fig.height = 5, fig.width = 7.5}
# TODO: Fix problem of contents appearing out of order.
# -----------------------------------------------------
set <- 3 # Up or downregulation?
if (nrow(gsva_list_pseudobulk[[set]]) > 0) {
  mat <- gsva_list_pseudobulk[[set]]
  gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(mat)]
  gene_sets_sub <- gene_sets_sub[names(gene_sets_sub) %in% names(gene_sets_filtered)]
  mat <- mat[rownames(mat) %in% names(gene_sets_sub), ]
  for (i in 1:nrow(mat)) {
    genes <- which(
      gene_sets[[rownames(mat)[i]]]@geneIds %in% rownames(markers_list_pseudobulk[[set]])
    )
    genes <- gene_sets[[rownames(mat)[i]]]@geneIds[genes]
    genes <- gene_anno[gene_anno$ensembl_gene_id %in% genes, ]
    genes <- genes$external_gene_name
    title <- paste0(
      gene_sets[[rownames(mat)[i]]]@shortDescription, ": ", gene_sets[[rownames(mat)[i]]]@setName
    )
    print(title)

    if (length(genes) > 0) {
      print(
        paste0(
          gene_sets[[rownames(mat)[i]]]@shortDescription, " DE genes: ",
          paste(genes, collapse = ", ")
        )
      )
      iss_genes <- genes[genes %in% iss_codebook[ , 1]]
      print(
        paste0(
          gene_sets[[rownames(mat)[i]]]@shortDescription, " DE genes in ISS codebook: ",
          paste(iss_genes, collapse = ", ")
        )
      )
      ha <- HeatmapAnnotation(
        donor = seurat_symbols$donor_id,
        cell_type = anno_block(
          gp = gpar(fill = "lightgray", col = "lightgray"),
          labels = c(
            names(idents)[[1]], names(idents)[[2]], names(idents)[[3]], names(idents)[[4]]
          ),
          labels_gp = gpar(col = "black")
        ),
        col = list(
          donor = c(
            "453" = anno_color[1], "454" = anno_color[2], "448" = anno_color[3],
            "490" = anno_color[4], "492" = anno_color[5], "493" = anno_color[6]
          )
        )
      )
      draw(
        Heatmap(
          as.matrix(GetAssayData(seurat_symbols)[genes, , drop = FALSE]), color2,
          cluster_rows = FALSE, cluster_columns = FALSE,
          show_column_names = FALSE, column_split = seurat_symbols$idents,
          show_heatmap_legend = FALSE, top_annotation = ha,
          column_title = title, column_gap = unit(2, "mm")
        )
      )
    }

    genes <- gene_sets[[rownames(mat)[i]]]@geneIds
    genes <- gene_anno[gene_anno$ensembl_gene_id %in% genes, ]
    genes <- genes$external_gene_name
    if (length(genes) > 0) {
      print(
        paste0(
          gene_sets[[rownames(mat)[i]]]@shortDescription,
          " all genes: ",
          paste(genes, collapse = ", ")
        )
      )
      iss_genes <- genes[genes %in% iss_codebook[ , 1]]
      print(
        paste0(
          gene_sets[[rownames(mat)[i]]]@shortDescription,
          " all genes in ISS codebook: ",
          paste(iss_genes, collapse = ", ")
        )
      )
      ha <- HeatmapAnnotation(
        donor = seurat_symbols$donor_id,
        cell_type = anno_block(
          gp = gpar(fill = "lightgray", col = "lightgray"),
          labels = c(
            names(idents)[[1]], names(idents)[[2]], names(idents)[[3]], names(idents)[[4]]
          ),
          labels_gp = gpar(col = "black")
        ),
        col = list(
          donor = c(
            "453" = anno_color[1], "454" = anno_color[2], "448" = anno_color[3],
            "490" = anno_color[4], "492" = anno_color[5], "493" = anno_color[6]
          )
        )
      )
      draw(
        Heatmap(
          as.matrix(GetAssayData(seurat_symbols)[genes, , drop = FALSE]), color2,
          cluster_rows = FALSE, cluster_columns = FALSE,
          show_column_names = FALSE, column_split = seurat_symbols$idents,
          show_heatmap_legend = FALSE, top_annotation = ha,
          column_title = title, column_gap = unit(2, "mm")
        )
      )
    }
  }
}
# -----------------------------------------------------
```

# References

This is the concluding section of the document, where we output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
