---
title:
  '01 Analysis - `r unlist(strsplit(getwd(), "/"))[length(unlist(strsplit(getwd(), "/"))) - 2]`'
author:
  - name: "Emir Turkes [emir.turkes@eturkes.com]"
  - name: "UK Dementia Research Institute at UCL"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(
    inputFile,
    encoding = encoding,
    output_file = file.path(
      "..", "..", "results",
      unlist(strsplit(getwd(), "/"))[length(unlist(strsplit(getwd(), "/")))], "01-analysis.html"
    )
  )})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of the [Tau Vulnerability Project](https://github.com/eturkes/tau-vulnerability).*

The data here is derived from @`r unlist(strsplit(getwd(), "/"))[length(unlist(strsplit(getwd(), "/")))]` and will be referenced using the name ``r unlist(strsplit(getwd(), "/"))[length(unlist(strsplit(getwd(), "/")))]``.

```{r, boilerplate}
# Load in necessary boilerplate and libraries.
# --------------------------------------------

#    This file is part of tau-vulnerability.
#    Copyright (C) 2019-2021  Emir Turkes, UK DRI at UCL, Columbia University Medical Center
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

# These should be checked per document.
# -------------------------------------
options(stringsAsFactors = FALSE)
packages <- c(
  "conflicted", "maEndToEnd", "biomaRt", "DT", "GSEABase", "GSVA", "scales", "circlize",
  "ComplexHeatmap"
)
invisible(suppressPackageStartupMessages(lapply(packages, library, character.only = TRUE)))
source(file.path(getwd(), "..", "utils.R"))

conflict_prefer("strsplit", "Biostrings", quiet = TRUE)
conflict_prefer("rowMedians", "Biobase", quiet = TRUE)
conflict_prefer("summarize", "dplyr", quiet = TRUE)
conflict_prefer("filter", "dplyr", quiet = TRUE)
conflict_prefer("intersect", "dplyr", quiet = TRUE)

analysis_no <- 1
color <- colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100)
color_legend <- c("lightgray", "darkgray")
gene_set_list <- "GO.BP"
organism <- "hsapiens"
# -------------------------------------

# Everything else in this chunk remains generally unchanged.
# ----------------------------------------------------------
analysis_no <- 1
data_name <- unlist(strsplit(getwd(), "/"))[6] # Name of dataset.
assets_dir <- file.path(getwd(), "..", "..", "assets") # Backed up data.

# Unique cache directory for each analysis number.
# ------------------------------------------------
cache_dir <- file.path(getwd(), "..", "..", "cache", data_name, paste0("0", analysis_no))
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}
# ------------------------------------------------

knitr::opts_chunk$set(fig.width = 12, fig.height = 12)
# ----------------------------------------------------------
```

# All Samples

First, we run a standard pipeline on the raw data for QC and normalization.

```{r}
sdrf <- read.delim(normalizePath(file.path(assets_dir, data_name, "E-GEOD-4757.sdrf.txt")))
rownames(sdrf) <- sdrf$Array.Data.File
sdrf <- AnnotatedDataFrame(sdrf)

data <- read.celfiles(
  filenames = file.path(assets_dir, data_name, sdrf$Array.Data.File), phenoData = sdrf,
  verbose = FALSE
)
pData(data) <- pData(data)[, c(
  "Source.Name", "Characteristics.individual.", "Characteristics..clinical.information."
)]

boxplot(data, main = "Boxplot of log2-intensities for Raw Data")
data <- rma(data)
boxplot(data, main = "Boxplot of log2-intensities for Normalized Data")

intensity_threshold <- 5
sample_threshold <- min(table(data$Characteristics..clinical.information.))
hist_res <- hist(
  rowMedians(exprs(data)), 100, col = "cornsilk",
  freq = FALSE, main = "Histogram of Median Intensities", border = "antiquewhite4",
  xlab = "Median intensities"
)
hist_res <- hist(
  rowMedians(exprs(data)), 100, col = "cornsilk",
  freq = FALSE, main = "Histogram of Median Intensities", border = "antiquewhite4",
  xlab = "Median intensities"
)
abline(v = intensity_threshold, col = "coral4", lwd = 2)
meets_cutoff <- apply(
  exprs(data), 1, function(x) { sum(x > intensity_threshold) >= sample_threshold }
)
table(meets_cutoff)
data <- subset(data, meets_cutoff)
hist_res <- hist(
  rowMedians(exprs(data)), 100, col = "cornsilk",
  freq = FALSE, main = "Histogram of Median Intensities", border = "antiquewhite4",
  xlab = "Median intensities"
)

PCA <- prcomp(t(exprs(data)), scale = FALSE)
percentVar <- round(100 * PCA$sdev ^ 2 / sum(PCA$sdev ^ 2), 1)
gg_df <- data.frame(
  PC1 = PCA$x[ , 1],
  PC2 = PCA$x[ , 2],
  Disease = data$Characteristics..clinical.information.,
  Individual = data$Characteristics.individual.
)
gg_df$Individual <- factor(gg_df$Individual)
ggplot(gg_df, aes(PC1, PC2)) +
  geom_point(aes(shape = Individual, colour = Disease)) +
  ggtitle("PCA") +
  xlab(paste0("PC1, VarExp: ", percentVar[1], "%")) +
  ylab(paste0("PC2, VarExp: ", percentVar[2], "%")) +
  scale_shape_manual(values = 1:nlevels(gg_df$Individual)) +
  theme(plot.title = element_text(hjust = 0.5))

disease_names <- ifelse(
  str_detect(data$Characteristics..clinical.information., "histopathologically"),
  "Normal",
  "NFT"
)
heatmap_anno <- data.frame(Disease = disease_names)
rownames(heatmap_anno) <- rownames(pData(data))
dists <- as.matrix(dist(t(exprs(data))))
rownames(dists) <- rownames(pData(data))
col <- rev(colorRampPalette(brewer.pal(9, "YlOrRd"))(255))
colnames(dists) <- NULL
diag(dists) <- NA
ann_colors <- list(Disease = c(Normal = "blue4", NFT = "cadetblue2"))
pheatmap(
  dists, col = col,
  annotation_row = heatmap_anno, annotation_colors = ann_colors,
  legend = TRUE, treeheight_row = 0,
  legend_breaks = c(min(dists, na.rm = TRUE), max(dists, na.rm = TRUE)),
  legend_labels = (c("small distance", "large distance")), main = "Heatmap for Normalized Data"
)

section_no <- 1
gene_set_name <- gene_set_list[section_no]

gene_sets_orig <- getGmt(
  file.path(
      "..", "..", "assets", "gene-sets", paste0("hsapiens", ".", gene_set_name, ".ENSG.gmt")
  )
)
rds <- file.path(cache_dir, paste0(gene_set_name, "_gene_sets.rds"))
if (file.exists(rds)) {
  gene_sets <- readRDS(rds)
} else {
  gene_sets <- gene_sets_orig
  gene_sets <- filterGeneSets(gene_sets, 3, 500)
  for (i in seq(length(gene_sets@.Data))) {
    go_id <- gene_sets[[i]]@setName
    suppressWarnings(gene_sets[[i]]@setName <- gene_sets[[i]]@shortDescription)
    suppressWarnings(gene_sets[[i]]@shortDescription <- go_id)
  }
  saveRDS(gene_sets, rds)
}

rds <- file.path(cache_dir, "gene_anno.rds")
if (file.exists(rds)) {
  gene_anno <- readRDS(rds)
} else {
  mart <- useEnsembl("ensembl", "hsapiens_gene_ensembl", "http://uswest.ensembl.org/")
  attributes <- c("affy_hg_u133_plus_2", "external_gene_name", "ensembl_gene_id", "chromosome_name")
  gene_anno <- getBM(
    attributes = attributes, filters = "affy_hg_u133_plus_2",
    values = featureNames(data), mart = mart
  )
  saveRDS(gene_anno, rds)
}
gene_anno <- gene_anno[gene_anno$chromosome_name %in% c(1:22), ]
gene_anno <- gene_anno[gene_anno$ensembl_gene_id %in% unique(unlist(geneIds(gene_sets))), ]

keep <- lapply(
  geneIds(gene_sets), function(x, y) na.omit(fastmatch::fmatch(x, y)), gene_anno$ensembl_gene_id
)
keep <- filterGeneSets(keep, 3, 500)
gene_sets <- gene_sets[names(gene_sets) %in% names(keep)]

# We use a modified overlap coefficient on the gene sets which divides the number of genes in the
# larger set by the number of shared genes with the smaller set.
# We perform this twice, once to sort the sets so that more general terms are towards the top, and
# again to remove sets with identical genes, keeping the more general term.
# ------------------------------------------------------------------------------------------------
rds <- file.path(cache_dir, paste0(gene_set_name, "_gene_set_overlap.rds"))
if (file.exists(rds)) {
  overlap <- readRDS(rds)
} else {
  overlap <- computeGeneSetsOverlapMax(gene_sets, unique(unlist(geneIds(gene_sets))))
  saveRDS(overlap, rds)
}
keep <- rowSums(overlap)
keep <- keep[order(keep, decreasing = TRUE)]
gene_sets_sorted <- gene_sets[match(names(keep), names(gene_sets))]
rds <- file.path(cache_dir, paste0(gene_set_name, "_gene_set_overlap_sorted.rds"))
if (file.exists(rds)) {
  overlap <- readRDS(rds)
} else {
  overlap <- computeGeneSetsOverlapMax(gene_sets_sorted, unique(unlist(geneIds(gene_sets_sorted))))
  saveRDS(overlap, rds)
}

overlap[upper.tri(overlap)] <- 0
diag(overlap) <- 0
keep <- apply(overlap, 1, max)
keep <- keep[keep < 1]
gene_sets <- gene_sets[names(gene_sets) %in% names(keep)]

gene_anno <- gene_anno[gene_anno$ensembl_gene_id %in% unique(unlist(geneIds(gene_sets))), ]
gene_anno <- gene_anno[!duplicated(gene_anno$affy_hg_u133_plus_2), ]
gene_anno <- gene_anno[!duplicated(gene_anno$external_gene_name), ]

keep <- lapply(
  geneIds(gene_sets), function(x, y) na.omit(fastmatch::fmatch(x, y)), gene_anno$ensembl_gene_id
)
keep <- filterGeneSets(keep, 3, 500)
gene_sets <- gene_sets[names(gene_sets) %in% names(keep)]

gene_anno <- gene_anno[gene_anno$ensembl_gene_id %in% unique(unlist(geneIds(gene_sets))), ]

data <- data[which(rownames(data) %in% gene_anno$affy_hg_u133_plus_2), ]
fData(data) <- gene_anno
rownames(data) <- fData(data)$ensembl_gene_id

print("Original gene sets")
gene_sets_orig
print("Filtered gene sets")
gene_sets

NFT <- data[ , grep("neurofibrillary", data$Characteristics..clinical.information.)]
normal <- data[ , grep("histopathologically", data$Characteristics..clinical.information.)]
conditions <- c("NFT", "Normal")
comb_mat <- cbind(exprs(NFT), exprs(normal))
rownames(comb_mat) <- fData(data)$ensembl_gene_id
groups <- factor(rep(conditions, c(dim(NFT)[2], dim(normal)[2])))
levels(groups) <- conditions
batch <- factor(
  gsub(
    rep(data$Characteristics.individual.[!duplicated(data$Characteristics.individual.)], 2),
    pattern = " |-",
    replacement = "_"
  )
)

rds <- file.path(cache_dir, paste0(gene_set_name, "_gsva.rds"))
if (file.exists(rds)) {
  gsva <- readRDS(rds)
} else {
  gsva <- gsva(comb_mat, gene_sets, method = "ssgsea", ssgsea.norm = FALSE)
  colnames(gsva) <- groups
  saveRDS(gsva, rds)
}
```

## DE `r gene_set_name` Ontology Tables

```{r}
# Fit model on pseudo-bulk GSVA results.
# --------------------------------------
design <- model.matrix(~ 0 + groups + batch)
colnames(design)[1:2] <- levels(groups)
colnames(design)[3] <- "individual"
fit <- lmFit(gsva, design)
contrast_mat <- makeContrasts(NFT-Normal, levels = design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat))
tests <- decideTests(cont_fit, p.value = 0.1)
tests_up <- tests[tests[ , 1] == 1, ]
discard <- which(rownames(tests_up) %in% rownames(tests_up[rowSums(tests_up == -1) > 0, ]))
if (length(discard) > 0) {
  tests_up <- tests_up[-discard, ]
}
tests_down <- tests[tests[ , 1] == -1, ]
discard <- which(rownames(tests_down) %in% rownames(tests_down[rowSums(tests_down == 1) > 0, ]))
if (length(discard) > 0) {
  tests_down <- tests_down[-discard, ]
}
results <- topTable(cont_fit, number = Inf)
# --------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
gsva_list_pseudobulk <- list(all, up, down)
names(gsva_list_pseudobulk) <- c(
  "All DE", paste0("Upregulated in ", levels(groups)[1]),
  paste0("Downregulated in ", levels(groups)[2])
)
```

**`r levels(groups)[1]` Upregulated `r gene_set_name`**

```{r}
datatable_download_exp(gsva_list_pseudobulk[[2]])
```

**`r levels(groups)[1]` Downregulated `r gene_set_name`**

```{r}
datatable_download_exp(gsva_list_pseudobulk[[3]])
```

# DEGs

We calculate DEGs (differentially expressed genes).

```{r}
# Fit model on pseudo-bulk counts.
# --------------------------------
design <- model.matrix(~ 0 + groups + batch)
colnames(design)[1:2] <- levels(groups)
colnames(design)[3] <- "individual"
fit <- lmFit(comb_mat, design)
contrast_mat <- makeContrasts(NFT-Normal, levels = design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat), robust = TRUE)
tests <- decideTests(cont_fit, p.value = 0.1)
tests_up <- tests[tests[ , 1] == 1, ]
tests_down <- tests[tests[ , 1] == -1, ]
results <- topTable(cont_fit, 1, Inf)
# --------------------------------------

up <- results[rownames(results) %in% rownames(tests_up), ]
down <- results[rownames(results) %in% rownames(tests_down), ]
all <- rbind(up, down)
if (length(all) > 0) {
  all <- all[order(-all$B), ]
}
markers_list_pseudobulk <- list(all, up, down)

genes <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(markers_list_pseudobulk[[2]]), ]
genes <- genes[order(match(genes$ensembl_gene_id, rownames(markers_list_pseudobulk[[2]]))), ]
up <- markers_list_pseudobulk[[2]]
rownames(up) <- genes$external_gene_name

genes <- gene_anno[gene_anno$ensembl_gene_id %in% rownames(markers_list_pseudobulk[[3]]), ]
genes <- genes[order(match(genes$ensembl_gene_id, rownames(markers_list_pseudobulk[[3]]))), ]
down <- markers_list_pseudobulk[[3]]
rownames(down) <- genes$external_gene_name

all <- rbind(up, down)
all <- all[order(-all$B), ]
markers_list_symbols_pseudobulk <- list(all, up, down)
```

**`r levels(groups)[1]` Upregulated Pseudo-bulk**

```{r}
datatable_download_exp(markers_list_symbols_pseudobulk[[2]])
```

**`r levels(groups)[1]` Downregulated Pseudo-bulk**

```{r}
datatable_download_exp(markers_list_symbols_pseudobulk[[3]])
```

## DE GO Ontology Heatmaps

### `r gene_set_name` Ontologies

```{r, fig.width = 20, fig.height = 5}
anno_color <- hue_pal()(length(unique(batch)))
gsva_heatmap <- vector("list", 3)

color2 <- colorRamp2(
  c(min(gsva), (max(gsva) + min(gsva)) / 2, max(gsva)),
  c(brewer.pal(3, "RdBu")[3], brewer.pal(3, "RdBu")[2], brewer.pal(3, "RdBu")[1])
)

i <- 1
mat <- gsva
mat <- mat[rownames(mat) %in% rownames(gsva_list_pseudobulk[[i]]), ]
gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(mat)]
# gene_sets_sub <- gene_sets_sub[names(gene_sets_sub) %in% names(gene_sets_filtered)]
# mat <- mat[rownames(mat) %in% names(gene_sets_sub), ]
gene_sets_sub_ids <- gene_sets_sub
for (j in seq(length(gene_sets_sub_ids@.Data))) {
  suppressWarnings(gene_sets_sub_ids[[j]]@setName <- gene_sets_sub_ids[[j]]@shortDescription)
}
# keep <- which(names(gene_sets_sub_ids) %in% go_data@geneAnno$GO)
# gene_sets_sub_ids <- gene_sets_sub_ids[keep, ]
# gene_sets_sub <- gene_sets_sub[keep, ]
# mat <- mat[keep, ]
anno <- HeatmapAnnotation(
  donor = batch,
  cell_type = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = c(levels(groups)[1], levels(groups)[2]),
    labels_gp = gpar(col = "black", fontsize = 20)
  ),
  col = list(
    donor = c(
      "individual_00_28" = anno_color[1], "individual_01_22" = anno_color[2],
      "individual_95_08" = anno_color[3], "individual_96_18" = anno_color[4],
      "individual_96_31" = anno_color[5], "individual_97_01" = anno_color[6],
      "individual_98_18" = anno_color[6], "individual_99_21" = anno_color[8],
      "individual_99_28" = anno_color[9], "individual_99_33" = anno_color[10]
    )
  )
)

draw(
  Heatmap(
    t(apply(mat, 1, function (x) (x - min(x)) / (max(x) - min(x)))), color,
    cluster_columns = FALSE,
    cluster_rows = FALSE,
    row_names_max_width = max_text_width(rownames(mat), gpar(fontsize = 20)),
    show_column_names = FALSE,
    column_split = groups,
    top_annotation = anno,
    row_names_gp = gpar(fontsize = 20),
    rect_gp = gpar(col = "lightgray"),
    column_title = paste0(
      "All Significant GO ", unlist(strsplit(gene_set_name, "\\."))[2], " Terms"
    ),
    column_title_gp = gpar(fontsize = 20),
    heatmap_legend_param = list(title = "ssGSEA score"),
    column_gap = unit(2, "mm")
  )
)

heat_data <- data[rownames(data) %in% rownames(markers_list_pseudobulk[[1]]), ]
rownames(heat_data) <- fData(heat_data)$external_gene_name
draw(
  Heatmap(
    t(apply(exprs(heat_data), 1, function (x) (x - min(x)) / (max(x) - min(x)))), color,
    cluster_columns = FALSE,
    cluster_rows = FALSE,
    row_names_max_width = max_text_width(rownames(mat), gpar(fontsize = 20)),
    show_column_names = FALSE,
    column_split = groups,
    top_annotation = anno,
    row_names_gp = gpar(fontsize = 20),
    rect_gp = gpar(col = "lightgray"),
    column_title = paste0("All Significant Genes"),
    column_title_gp = gpar(fontsize = 20),
    heatmap_legend_param = list(title = "ssGSEA score"),
    column_gap = unit(2, "mm")
  )
)

rownames(data) <- fData(data)$external_gene_name
for (i in seq_along(rownames(mat))) {
  genes <- gene_sets[[rownames(mat)[i]]]@geneIds
  genes <- gene_anno[gene_anno$ensembl_gene_id %in% genes, ]
  genes <- genes$external_gene_name
  if (length(genes) > 0) {
    print(
      paste0(
        gene_sets[[rownames(mat)[i]]]@shortDescription, " all genes: ",
        paste(genes, collapse = ", ")
      )
    )
    heat_data <- data[rownames(data) %in% genes, ]
    draw(
      Heatmap(
        t(apply(exprs(heat_data), 1, function (x) (x - min(x)) / (max(x) - min(x)))), color,
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        row_names_max_width = max_text_width(rownames(mat), gpar(fontsize = 20)),
        show_column_names = FALSE,
        column_split = groups,
        top_annotation = anno,
        row_names_gp = gpar(fontsize = 20),
        rect_gp = gpar(col = "lightgray"),
        column_title = rownames(mat)[i],
        column_title_gp = gpar(fontsize = 20),
        heatmap_legend_param = list(title = "ssGSEA score"),
        column_gap = unit(2, "mm")
      )
    )
  }
}
```

# References

This is the concluding section of the document. Here we output the `sessionInfo` and create a bibliography for works cited.

```{r}
sessionInfo()
```
