<!---
    This file is part of tau-vulnerability.
    Copyright (C) 2019  Emir Turkes

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

    Emir Turkes can be contacted at eturkes@bu.edu
-->

# Miscellaneous Notes

## Our objective

1. #### Identify differentially expressed genes that cluster with markers of excitatory cells vs. inhibitory cells in human RNAseq datasets, looking at pathways other than tau homeostasis.
1. #### Correlate results with morphology, ontology, physiology, etc. of different cell populations in different regions to better understand selective vulnerability of excitatory cells to tauopathy.

## Datasets to be used

Human normal post-mortem single-nucleus RNAseq annotated datasets from [Fu, 2019](https://www.nature.com/articles/s41593-018-0298-7) tau homeostasis paper:
* GTEx ([phs000424](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000424.v7.p2), Regev)
* SCAP-T ([phs000833](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000833.v7.p1), Lake)

As a reference, mouse RNAseq data from [Tasic, 2016](https://www.nature.com/articles/nn.4216):
* celltax ([GSE71585](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE71585), Menon & Harris)

Allen Institute should have fresh RNAseq data from the EC, but we need to follow up on it:
* There is human data for the MTG, but not sure if contains EC ([link1](http://celltypes.brain-map.org/), [link2](http://celltypes.brain-map.org/rnaseq))

## High-level pipeline

##### Generic roadmap provided by [Conesa, 2016](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0881-8):

![RNAseq-pipeline](./notes/Screenshot_20190110_152039.png)

## GTEx in detail

### Some incongruities:
* Fu, 2019 used GTEx data from the [Single Cell Portal](https://portals.broadinstitute.org/single_cell/study/dronc-seq-single-nucleus-rna-seq-on-human-archived-brain).
    * This data was not initially available for download on 2019/01/04, but now [it is](/home/eturkes/Documents/projects/Human_DroNc-seq).
    * Now also available at the [GTEx Portal](https://gtexportal.org/home/datasets).
        * Single Cell Portal seems to have [more data](./notes/Screenshot_from_2019-02-06_12-05-38.png).
    * The data was published in [Habib, 2017](https://www.nature.com/nmeth/journal/v14/n10/full/nmeth.4407.html).
    * The accession code is phs000424.v8.p1, which is [not yet on dbGaP](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000424.v8.p1).
        * "GTEx data will be released through dbGaP on a periodic basis as it is generated, and there is no formal publication embargo. However, the GTEx Consortium requests that users of GTEx data respect the data producer's interest to publish a full analysis of the data first." [source](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000424.v4.p1)
    * We applied for access to dbGaP and were approved for an older version, [phs000424.v7.p2](https://dbgap.ncbi.nlm.nih.gov/aa/wga.cgi?page=request_details_pi&arid=288901&filter=arid).
        * This version does not seem to contain DroNC-seq set.

### Characteristics of DroNC-seq

#### Summary:
* ##### Samples:
    * 3T3 mouse cell line
        * 5,636 nuclei
    * Adult frozen mouse brain tissue
        * 19,561 nuclei
    * Archived frozen adult human post-mortem tissue
        * 19,550 nuclei

#### 3T3 mouse cell line
* 3,295 genes (4,643 transcripts)

#### Adult frozen mouse brain tissue
* 2,731 genes (3,653 transcripts)

#### Archived frozen adult human post-mortem tissue
* 1,683 genes (2,187 transcripts)

### Characteristics of v7.p2

#### Summary:
* ##### Sample size:

![sample-size](./notes/Screenshot_from_2019-02-06_15-52-03.png)

* ##### Sample variables:
    * ASE_SNP_CNV_Expression: Allele-specific expression for SNPs and INDELs
    * ASE_SNP_Expression: Allele-specific expression data
    * Array_SNP: SNP genotypes obtained using standard or custom microarrays
    * Array_totRNA_Expression: Expression data for total RNA samples (array data)
    * Imputation_SNP: Imputed SNP genotypes
    * Seq_DNA_CNV: CNV genotypes derived from sequence data
    * Seq_DNA_SNP_CNV: SNP and CNV genotypes derived from sequence data
    * Seq_DNA_WholeExome: Whole exome sequencing
    * Seq_DNA_WholeGenome: Whole genome sequencing
    * Seq_RNA: Whole transcriptome sequencing
    * Seq_RNA_Expression: Expression levels derived from RNA sequences
    * Seq_RNA_WTSS: Whole transcriptome shotgun sequencing (RNA-seq)
![molecular-data-stats](./notes/Screenshot_from_2019-02-06_15-26-30.png)

* ##### Brain regions

![brain-regions](./notes/Screenshot_from_2019-02-06_15-42-05.png)

* ##### Cause of death

![cause-of-death](./notes/Screenshot_from_2019-02-06_14-57-40.png)

* ##### Bioinformatics pipeline
    * [RNAseq alignment, expression quantification, and quality control](https://github.com/broadinstitute/gtex-pipeline/tree/master/rnaseq):
![RNAseq](./notes/Screenshot_from_2019-02-07_07-38-21.png)
    * [Generation of gene models and annotations for RNAseq and eQTL](https://github.com/broadinstitute/gtex-pipeline/tree/master/gene_model)
![gene-models](./notes/Screenshot_from_2019-02-07_07-53-53.png)

#### Resources:
* [Study metadata](/home/eturkes/ncbi/dbGaP-20886/files/66697/PhenoGenotypeFiles/RootStudyConsentSet_phs000424.GTEx.v7.p2.c1.GRU/StudyMetaFiles) from dbGaP
* [Phenotype data](/home/eturkes/ncbi/dbGaP-20886/files/66706/PhenoGenotypeFiles/RootStudyConsentSet_phs000424.GTEx.v7.p2.c1.GRU/PhenotypeFiles) from dbGaP
* [Methods](https://gtexportal.org/home/documentationPage) from GTEx Portal
* [Dataset summary](https://gtexportal.org/home/tissueSummaryPage) from GTEx Portal
