---
title: "09 Habib Neuronal"
author:
  - name: "Emir Turkes [et2628@cumc.columbia.edu]"
  - name: "Columbia University"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
bibliography: "../tau-vulnerability.bib"
biblio-style: apalike
link-citations: true
output:
  html_document:
    code_folding: show
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
knit:
  (function(inputFile, encoding) {
    rmarkdown::render(
      inputFile, encoding = encoding, output_file = "../results/09-habib-neuronal.html")})
---

```{r, include = FALSE}
#    This file is part of tau-vulnerability.
#    Copyright (C) 2019  Emir Turkes
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

knitr::opts_chunk$set(fig.width = 15, fig.height = 13)
```

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of the [Tau Vulnerability Project](https://github.com/eturkes/tau-vulnerability).*

This document follows up its preparation file to provide and in-depth analysis of neuronal populations in the dataset from @habib_massively_2017.
We start by setting some global variables and loading in any required packages.

```{r}
assets_dir <- file.path(getwd(), "..", "assets")
results_dir <- file.path(getwd(), "..", "results")

packages <- c(
  "conflicted", "SingleCellExperiment", "magrittr", "dplyr", "ggplot2", "ggrepel", "S4Vectors",
  "SummarizedExperiment", "DropletUtils", "scran", "BiocSingular", "scater", "Rtsne", "svd", "SC3",
  "DT", "data.table", "Seurat", "uwot", "viridis", "WGCNA", "pheatmap", "RColorBrewer")
invisible(suppressPackageStartupMessages(lapply(packages, library, character.only = TRUE)))

conflict_prefer("which", "BiocGenerics")
options(stringsAsFactors = FALSE)

# Create a unique cache and results data directory for each iterated section.
if (!dir.exists(file.path(assets_dir, "cache", "09"))) {
  dir.create(file.path(assets_dir, "cache", "09"), recursive = TRUE)}
if (!dir.exists(file.path(results_dir, "data", "09"))) {
  dir.create(file.path(results_dir, "data", "09"), recursive = TRUE)}

# ggplot2 function providing custom aesthetics and automatic placement of categorical labels.
# For continuous data, a colorbar is implemented.
dim_red_plot <- function(data, x, y, col, type) {
  gg <- ggplot(data, aes_string(x = x, y = y, color = col)) +
    geom_point(alpha = 0.35, stroke = 0.05, shape = 21, aes_string(fill = col)) +
    theme_classic() +
    theme(
      legend.position = "right", plot.title = element_text(hjust = 0.5),
      legend.title = element_blank()) +
    guides(color = guide_legend(override.aes = list(alpha = 1)))
    if (type == "cat") {
      gg <- gg + geom_label_repel(data = label_df2, aes(label = label), show.legend = FALSE, size = 3)
    } else if (type == "cont") {
      gg <- ggplot(data, aes_string(x = x, y = y)) +
        geom_point(alpha = 0.35, stroke = 0.05, aes_string(color = col)) +
        theme_classic() +
        theme(
          legend.position = "right", plot.title = element_text(hjust = 0.5),
          legend.title = element_blank()) +
        scale_colour_viridis()}
  gg}

# Adds download buttons.
datatable_custom <- function(dt) {
  datatable(
    dt,
    extensions = "Buttons", options = list(dom = "Blfrtip", buttons = list(
      "copy", "print",
      list(extend = "collection", buttons = c("csv", "excel", "pdf"), text = "Download"))))}
```

We load in the data from the previous analysis.

```{r}
sce_hvg <- readRDS(file.path(results_dir, "data", "08", "sce_hvg.rds"))
sce_hvg
```

# Seurat DEG

In this section, we calculate differentially expressed genes (DEGs) and identify marker genes for each cluster using multiple resolutions in order to build a hierarchy of marker genes.

## 0.01 Resolution

```{r}
seurat_hvg <- readRDS(file.path(assets_dir, "cache", "08", "seurat_hvg_res0.01.rds"))
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache", "09", "hvg_seurat_markers_res0.01")
if (file.exists(rds)) {
  hvg_seurat_markers <- readRDS(rds)
} else {
  hvg_seurat_markers <- FindAllMarkers(seurat_hvg, only.pos = TRUE)
  saveRDS(hvg_seurat_markers, rds)}
```

```{r}
top <- hvg_seurat_markers %>% group_by(cluster) %>% top_n(n = 30, wt = avg_logFC)
DoHeatmap(seurat_hvg, features = top$gene, size = 3) +
  NoLegend() + scale_fill_gradientn(colors = c("blue", "white", "red"))

top <- hvg_seurat_markers %>% group_by(cluster) %>% top_n(n = 1, wt = avg_logFC)
datatable_custom(top)
VlnPlot(seurat_hvg, features = top$gene, ncol = 2)
FeaturePlot(
  seurat_hvg, reduction = "UMAP", features = top$gene, order = TRUE,
  cols = c("lightgrey", "red"), max.cutoff = 20, ncol = 2)
DotPlot(seurat_hvg, features = top$gene, cols = c("blue", "red")) + RotatedAxis()

datatable_custom(hvg_seurat_markers %>% group_by(cluster) %>% top_n(n = 250, wt = avg_logFC))
```

## 0.2 Resolution

```{r}
seurat_hvg <- readRDS(file.path(assets_dir, "cache", "08", "seurat_hvg_res0.2.rds"))
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache", "09", "hvg_seurat_markers_res0.2")
if (file.exists(rds)) {
  hvg_seurat_markers <- readRDS(rds)
} else {
  hvg_seurat_markers <- FindAllMarkers(seurat_hvg, only.pos = TRUE)
  saveRDS(hvg_seurat_markers, rds)}
```

```{r}
top <- hvg_seurat_markers %>% group_by(cluster) %>% top_n(n = 15, wt = avg_logFC)
DoHeatmap(seurat_hvg, features = top$gene, size = 3) +
  NoLegend() + scale_fill_gradientn(colors = c("blue", "white", "red"))

top <- hvg_seurat_markers %>% group_by(cluster) %>% top_n(n = 1, wt = avg_logFC)
datatable_custom(top)
VlnPlot(seurat_hvg, features = top$gene, ncol = 3)
FeaturePlot(
  seurat_hvg, reduction = "UMAP", features = top$gene, order = TRUE,
  cols = c("lightgrey", "red"), max.cutoff = 20, ncol = 3)
DotPlot(seurat_hvg, features = top$gene, cols = c("blue", "red")) + RotatedAxis()

datatable_custom(hvg_seurat_markers %>% group_by(cluster) %>% top_n(n = 100, wt = avg_logFC))
```

## ~1.3 Resolution

As this is the final level of clustering, we also use the markers to create new annotations.

```{r}
seurat_hvg <- readRDS(file.path(assets_dir, "cache", "08", "seurat_hvg_res.rds"))
seurat_clusters_id <- c(
"SLC17A7.TSHZ2.RORB",
"SLC17A7.TESPA1.COL5A2",
"TAC1.MEPE",
"SLC17A7.TESPA1.CARTPT",
"CXCL14",
"TAC1.SST",
"SLC17A7.TESPA1.ENC1",
"SLC17A7.TRABD2A",
"SLC17A7.TESPA1.PDE4C",
"CXCL14.KIT",
"SLC17A7.TSHZ2.SEMA3E",
"SLC17A7.TESPA1.NECAB1",
"SLC17A7.PTGDS",
"CXCL14.CNR1")
names(seurat_clusters_id) <- levels(seurat_hvg)
seurat_hvg <- RenameIdents(seurat_hvg, seurat_clusters_id)
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache", "09", "hvg_seurat_markers_res")
if (file.exists(rds)) {
  hvg_seurat_markers <- readRDS(rds)
} else {
  hvg_seurat_markers <- FindAllMarkers(seurat_hvg, only.pos = TRUE)
  saveRDS(hvg_seurat_markers, rds)}
```

```{r}
top <- hvg_seurat_markers %>% group_by(cluster) %>% top_n(n = 5, wt = avg_logFC)
DoHeatmap(seurat_hvg, features = top$gene, size = 3) +
  NoLegend() + scale_fill_gradientn(colors = c("blue", "white", "red"))

top <- hvg_seurat_markers %>% group_by(cluster) %>% top_n(n = 1, wt = avg_logFC)
datatable_custom(top)
VlnPlot(seurat_hvg, features = top$gene[1:7], ncol = 3)
VlnPlot(seurat_hvg, features = top$gene[8:14], ncol = 3)
FeaturePlot(
  seurat_hvg, reduction = "UMAP", features = top$gene[1:7], order = TRUE,
  cols = c("lightgrey", "red"), max.cutoff = 20, ncol = 3)
FeaturePlot(
  seurat_hvg, reduction = "UMAP", features = top$gene[8:14], order = TRUE,
  cols = c("lightgrey", "red"), max.cutoff = 20, ncol = 3)
DotPlot(seurat_hvg, features = top$gene, cols = c("blue", "red")) + RotatedAxis()

datatable_custom(hvg_seurat_markers %>% group_by(cluster) %>% top_n(n = 30, wt = avg_logFC))
```

We visualize the UMAP using the new annotations.

```{r, fig.width = 8.5, fig.height = 7}
# Set up data frame for ggplot2.
gg_df <- data.frame(
  colData(sce_hvg)[ , c(
    "cell_id_stem", paste0("habib_tsne", 1:2), "log10_total_features_by_counts",
    "habib_cluster_name", paste0("sub_hvg_ppk_umap", 1:2))])
rownames(gg_df) <- NULL
gg_df$habib_cluster_name <- factor(gg_df$habib_cluster_name)

gg_df$hvg_seurat_clusters_res <- seurat_hvg$seurat_clusters
gg_df$hvg_seurat_clusters_res_name <- NA
hvg_seurat_clusters_res <- as.integer(gg_df$hvg_seurat_clusters_res)
for (i in 1:length(hvg_seurat_clusters_res)) {
  if (hvg_seurat_clusters_res[i] == 1) {
    gg_df[i, ncol(gg_df)] <- "SLC17A7.TSHZ2.RORB"
  } else if (hvg_seurat_clusters_res[i] == 2) {
    gg_df[i, ncol(gg_df)] <- "SLC17A7.TESPA1.COL5A2"
  } else if (hvg_seurat_clusters_res[i] == 3) {
    gg_df[i, ncol(gg_df)] <- "TAC1.MEPE"
  } else if (hvg_seurat_clusters_res[i] == 4) {
    gg_df[i, ncol(gg_df)] <- "SLC17A7.TESPA1.CARTPT"
  } else if (hvg_seurat_clusters_res[i] == 5) {
    gg_df[i, ncol(gg_df)] <- "CXCL14"
  } else if (hvg_seurat_clusters_res[i] == 6) {
    gg_df[i, ncol(gg_df)] <- "TAC1.SST"
  } else if (hvg_seurat_clusters_res[i] == 7) {
    gg_df[i, ncol(gg_df)] <- "SLC17A7.TESPA1.ENC1"
  } else if (hvg_seurat_clusters_res[i] == 8) {
    gg_df[i, ncol(gg_df)] <- "SLC17A7.TRABD2A"
  } else if (hvg_seurat_clusters_res[i] == 9) {
    gg_df[i, ncol(gg_df)] <- "SLC17A7.TESPA1.PDE4C"
  } else if (hvg_seurat_clusters_res[i] == 10) {
    gg_df[i, ncol(gg_df)] <- "CXCL14.KIT"
  } else if (hvg_seurat_clusters_res[i] == 11) {
    gg_df[i, ncol(gg_df)] <- "SLC17A7.TSHZ2.SEMA3E"
  } else if (hvg_seurat_clusters_res[i] == 12) {
    gg_df[i, ncol(gg_df)] <- "SLC17A7.TESPA1.NECAB1"
  } else if (hvg_seurat_clusters_res[i] == 13) {
    gg_df[i, ncol(gg_df)] <- "SLC17A7.PTGDS"
  } else if (hvg_seurat_clusters_res[i] == 14) {
    gg_df[i, ncol(gg_df)] <- "CXCL14.CNR1"}}

gg_df$hvg_seurat_clusters_res_name <- factor(gg_df$hvg_seurat_clusters_res_name)
colData(sce_hvg) <- cbind(
  colData(sce_hvg),
  hvg_seurat_clusters_res = gg_df$hvg_seurat_clusters_res,
  hvg_seurat_clusters_res_name = gg_df$hvg_seurat_clusters_res_name)

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters_res_name = levels(gg_df$hvg_seurat_clusters_res_name),
  label = levels(gg_df$hvg_seurat_clusters_res_name))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters_res_name) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "hvg_seurat_clusters_res_name", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat ~1.3 Resolution Clusters Subset HVG PPK UMAP")
```

# IHC Genes

Next, we examine the expression of a set genes for which we have antibodies to conduct immunohistochemistry.
Due to the removal of some genes from QC and the HVG section, we create a new Seurat object with minimal processing but with HVG metadata.

```{r}
sce_orig <- readRDS(file.path(results_dir, "data", "08", "sce_orig.rds"))
keep <- readRDS(file.path(results_dir, "data", "08", "keep1.rds"))
sce_orig <- sce_orig[ , keep]
keep <- readRDS(file.path(results_dir, "data", "08", "keep2.rds"))
sce_orig <- sce_orig[ , keep]
sce_orig <- calculateQCMetrics(sce_orig)
detected_genes <- rowSums(counts(sce_orig) >= 1)
sce_orig <- sce_orig[which(detected_genes >= 1), ]

seurat_orig <- as.Seurat(sce_orig, counts = "counts", data = NULL)
umap <- as.matrix(data.frame(sce_hvg$sub_hvg_ppk_umap1, sce_hvg$sub_hvg_ppk_umap2))
colnames(umap) <- paste0("UMAP_", 1:2)
rownames(umap) <- colnames(sce_orig)
seurat_orig[["UMAP"]] <- CreateDimReducObject(
  embeddings = umap, key = "UMAP_", assay = DefaultAssay(seurat_orig))
Idents(object = seurat_orig) <- Idents(object = seurat_hvg)
seurat_orig <- NormalizeData(seurat_orig)
seurat_orig <- ScaleData(seurat_orig, features = rownames(seurat_orig))

genes <- c(
  "GRIA1", "GRIA2", "SATB1", "SATB2", "CALB1", "CALB2",
  "GAD1", "GAD2", "WFS1", "PVALB", "SST", "TBR1")
rm(umap)
DoHeatmap(seurat_orig, slot = "data", features = genes, size = 3) +
  NoLegend() +
  scale_fill_gradient2(
    low = rev(c('#D1E5F0','#67A9CF','#2166AC')), mid = "white",
    high = rev(c('#B2182B','#EF8A62','#FDDBC7')), midpoint = 0,
    guide = "colourbar", aesthetics = "fill")
VlnPlot(seurat_orig, features = genes, ncol = 3)
FeaturePlot(
  seurat_orig, reduction = "UMAP", features = genes, order = TRUE,
  cols = c("lightgrey", "red"), max.cutoff = 20, ncol = 3)
DotPlot(seurat_orig, features = genes, cols = c("blue", "red")) + RotatedAxis()
```

For closer inspection of the relationships between IHC genes, we only examine the most specific genes (those present in less than 1/8 of nuclei) and only nuclei with expression in two or more IHC genes.

```{r}
# Subset by most specific IHC genes.
keep <- which((rownames(x = seurat_orig)) %in% genes)
heatmap_mat <- seurat_orig[keep, ]
heatmap_mat <- as.matrix(GetAssayData(object = heatmap_mat, slot = "data"))

cells_per_gene <- rowSums(x = heatmap_mat > 0)
cells_per_gene

specificity_cutoff <- dim(seurat_orig)[2] / 8
specificity_cutoff

keep <- cells_per_gene < specificity_cutoff
genes <- names(cells_per_gene)
genes <- genes[keep]
genes

# Subset by nuclei with IHC coexpression.
keep <- which((rownames(x = seurat_orig)) %in% genes)
sub_seurat <- seurat_orig[keep, ]
heatmap_mat <- as.matrix(GetAssayData(object = sub_seurat, slot = "data"))
nonzero_cols <- (colSums(heatmap_mat != 0))
sub_seurat <- sub_seurat[ , nonzero_cols > 1]

heatmap_mat <- heatmap_mat[ , nonzero_cols > 1]
color <- colorRampPalette(brewer.pal(n = 7, name = "Reds"))(100)
color[1] <- "#DDDDDD"
pheatmap(
  heatmap_mat, angle_col = 0, fontsize = 9, color = color, cluster_cols = FALSE)
DoHeatmap(sub_seurat, slot = "data", features = genes, size = 3) +
  NoLegend() +
  scale_fill_gradient2(
    low = rev(c('#D1E5F0','#67A9CF','#2166AC')), mid = "white",
    high = rev(c('#B2182B','#EF8A62','#FDDBC7')), midpoint = 0,
    guide = "colourbar", aesthetics = "fill")
VlnPlot(sub_seurat, features = genes, ncol = 2)
FeaturePlot(
  sub_seurat, reduction = "UMAP", features = genes, order = TRUE,
  cols = c("lightgrey", "red"), max.cutoff = 20, ncol = 2)
DotPlot(sub_seurat, features = genes, cols = c("blue", "red")) + RotatedAxis()
```

# AD Genes

These are genes identified as significant risk factors for sporadic AD.

```{r}
genes <- c("MAPT", "APP", "APOE", "BIN1", "PICALM", "CLU")
DoHeatmap(seurat_orig, slot = "data", features = genes, size = 3) +
  NoLegend() +
  scale_fill_gradient2(
    low = rev(c('#D1E5F0','#67A9CF','#2166AC')), mid = "white",
    high = rev(c('#B2182B','#EF8A62','#FDDBC7')), midpoint = 0,
    guide = "colourbar", aesthetics = "fill")
VlnPlot(seurat_orig, features = genes, ncol = 2)
FeaturePlot(
  seurat_orig, reduction = "UMAP", features = genes, order = TRUE,
  cols = c("lightgrey", "red"), max.cutoff = 20, ncol = 2)
DotPlot(seurat_orig, features = genes, cols = c("blue", "red")) + RotatedAxis()
```

# Proteostasis Genes

These are genes identified in @fu_tau_2019.

```{r}
genes <- c("UCHL1", "NDRG4", "SYN1", "KCNC1")
DoHeatmap(seurat_orig, slot = "data", features = genes, size = 3) +
  NoLegend() +
  scale_fill_gradient2(
    low = rev(c('#D1E5F0','#67A9CF','#2166AC')), mid = "white",
    high = rev(c('#B2182B','#EF8A62','#FDDBC7')), midpoint = 0,
    guide = "colourbar", aesthetics = "fill")
VlnPlot(seurat_orig, features = genes, ncol = 2)
FeaturePlot(
  seurat_orig, reduction = "UMAP", features = genes, order = TRUE,
  cols = c("lightgrey", "red"), max.cutoff = 20, ncol = 2)
DotPlot(seurat_orig, features = genes, cols = c("blue", "red")) + RotatedAxis()
```

# Ken Harris Genes

These are genes that were found to frequently associate with proteostasis genes.

```{r}
genes <- c("CCK", "NOS1", "CNR1", "CKB")
DoHeatmap(seurat_orig, slot = "data", features = genes, size = 3) +
  NoLegend() +
  scale_fill_gradient2(
    low = rev(c('#D1E5F0','#67A9CF','#2166AC')), mid = "white",
    high = rev(c('#B2182B','#EF8A62','#FDDBC7')), midpoint = 0,
    guide = "colourbar", aesthetics = "fill")
VlnPlot(seurat_orig, features = genes, ncol = 2)
FeaturePlot(
  seurat_orig, reduction = "UMAP", features = genes, order = TRUE,
  cols = c("lightgrey", "red"), max.cutoff = 20, ncol = 2)
DotPlot(seurat_orig, features = genes, cols = c("blue", "red")) + RotatedAxis()
```

# Vulnerable Types

The below marker genes sets describe cell-types believed to be selectively vulnerable to tau pathology.

## Ex1

```{r}
genes <- c("CUX2", "LAMP5", "SERPINE2")
DoHeatmap(seurat_orig, slot = "data", features = genes, size = 3) +
  NoLegend() +
  scale_fill_gradient2(
    low = rev(c('#D1E5F0','#67A9CF','#2166AC')), mid = "white",
    high = rev(c('#B2182B','#EF8A62','#FDDBC7')), midpoint = 0,
    guide = "colourbar", aesthetics = "fill")
VlnPlot(seurat_orig, features = genes, ncol = 2)
FeaturePlot(
  seurat_orig, reduction = "UMAP", features = genes, order = TRUE,
  cols = c("lightgrey", "red"), max.cutoff = 20, ncol = 2)
DotPlot(seurat_orig, features = genes, cols = c("blue", "red")) + RotatedAxis()
```

## Ex2

```{r}
genes <- c("CUX2", "COL5A2")
DoHeatmap(seurat_orig, slot = "data", features = genes, size = 3) +
  NoLegend() +
  scale_fill_gradient2(
    low = rev(c('#D1E5F0','#67A9CF','#2166AC')), mid = "white",
    high = rev(c('#B2182B','#EF8A62','#FDDBC7')), midpoint = 0,
    guide = "colourbar", aesthetics = "fill")
VlnPlot(seurat_orig, features = genes, ncol = 1)
plots <- FeaturePlot(
  seurat_orig, reduction = "UMAP", features = genes,
  order = TRUE, blend = TRUE, combine = FALSE)
CombinePlots(plots[3:4], legend = "none")
DotPlot(seurat_orig, features = genes, cols = c("blue", "red")) + RotatedAxis()
```

## Ex3

No GAL available!

```{r}
genes <- c("RORB")
DoHeatmap(seurat_orig, slot = "data", features = genes, size = 3) +
  NoLegend() +
  scale_fill_gradient2(
    low = rev(c('#D1E5F0','#67A9CF','#2166AC')), mid = "white",
    high = rev(c('#B2182B','#EF8A62','#FDDBC7')), midpoint = 0,
    guide = "colourbar", aesthetics = "fill")
VlnPlot(seurat_orig, features = genes, ncol = 1)
FeaturePlot(
  seurat_orig, reduction = "UMAP", features = genes, order = TRUE,
  cols = c("lightgrey", "red"), max.cutoff = 20, ncol = 1)
DotPlot(seurat_orig, features = genes, cols = c("blue", "red")) + RotatedAxis()
```

## Ex6

```{r}
genes <- c("RORB", "PCP4")
DoHeatmap(seurat_orig, slot = "data", features = genes, size = 3) +
  NoLegend() +
  scale_fill_gradient2(
    low = rev(c('#D1E5F0','#67A9CF','#2166AC')), mid = "white",
    high = rev(c('#B2182B','#EF8A62','#FDDBC7')), midpoint = 0,
    guide = "colourbar", aesthetics = "fill")
VlnPlot(seurat_orig, features = genes, ncol = 1)
plots <- FeaturePlot(
  seurat_orig, reduction = "UMAP", features = genes,
  order = TRUE, blend = TRUE, combine = FALSE)
CombinePlots(plots[3:4], legend = "none")
DotPlot(seurat_orig, features = genes, cols = c("blue", "red")) + RotatedAxis()
```

## Ex8

No THEMIS available!

```{r}
genes <- c("NFIA")
DoHeatmap(seurat_orig, slot = "data", features = genes, size = 3) +
  NoLegend() +
  scale_fill_gradient2(
    low = rev(c('#D1E5F0','#67A9CF','#2166AC')), mid = "white",
    high = rev(c('#B2182B','#EF8A62','#FDDBC7')), midpoint = 0,
    guide = "colourbar", aesthetics = "fill")
VlnPlot(seurat_orig, features = genes, ncol = 1)
FeaturePlot(
  seurat_orig, reduction = "UMAP", features = genes, order = TRUE,
  cols = c("lightgrey", "red"), max.cutoff = 20, ncol = 1)
DotPlot(seurat_orig, features = genes, cols = c("blue", "red")) + RotatedAxis()
```

## Chandelier

GAD2 expression should be low!
SCUBE3 not available!

```{r}
genes <- c("PVALB", "GAD1", "GAD2")
DoHeatmap(seurat_orig, slot = "data", features = genes, size = 3) +
  NoLegend() +
  scale_fill_gradient2(
    low = rev(c('#D1E5F0','#67A9CF','#2166AC')), mid = "white",
    high = rev(c('#B2182B','#EF8A62','#FDDBC7')), midpoint = 0,
    guide = "colourbar", aesthetics = "fill")
VlnPlot(seurat_orig, features = genes, ncol = 2)
FeaturePlot(
  seurat_orig, reduction = "UMAP", features = genes, order = TRUE,
  cols = c("lightgrey", "red"), max.cutoff = 20, ncol = 2)
DotPlot(seurat_orig, features = genes, cols = c("blue", "red")) + RotatedAxis()
```

## From Mathys et al., 2019

### Excitatory

```{r}
genes <- c("RASGEF1B", "MTRNR2L8", "SLC26A3", "DHFR", "LINGO1")
DoHeatmap(seurat_orig, slot = "data", features = genes, size = 3) +
  NoLegend() +
  scale_fill_gradient2(
    low = rev(c('#D1E5F0','#67A9CF','#2166AC')), mid = "white",
    high = rev(c('#B2182B','#EF8A62','#FDDBC7')), midpoint = 0,
    guide = "colourbar", aesthetics = "fill")
VlnPlot(seurat_orig, features = genes, ncol = 2)
FeaturePlot(
  seurat_orig, reduction = "UMAP", features = genes, order = TRUE,
  cols = c("lightgrey", "red"), max.cutoff = 20, ncol = 2)
DotPlot(seurat_orig, features = genes, cols = c("blue", "red")) + RotatedAxis()
```

### Inhibitory

```{r}
genes <- c("DHFR", "MTRNR2L12", "GPS2")
DoHeatmap(seurat_orig, slot = "data", features = genes, size = 3) +
  NoLegend() +
  scale_fill_gradient2(
    low = rev(c('#D1E5F0','#67A9CF','#2166AC')), mid = "white",
    high = rev(c('#B2182B','#EF8A62','#FDDBC7')), midpoint = 0,
    guide = "colourbar", aesthetics = "fill")
VlnPlot(seurat_orig, features = genes, ncol = 2)
FeaturePlot(
  seurat_orig, reduction = "UMAP", features = genes, order = TRUE,
  cols = c("lightgrey", "red"), max.cutoff = 20, ncol = 2)
DotPlot(seurat_orig, features = genes, cols = c("blue", "red")) + RotatedAxis()
```

# Spared Types

In contrast, the following cell-types are believed to be spared.

## Ex4

```{r}
genes <- c("RORB", "SERPINE2")
DoHeatmap(seurat_orig, slot = "data", features = genes, size = 3) +
  NoLegend() +
  scale_fill_gradient2(
    low = rev(c('#D1E5F0','#67A9CF','#2166AC')), mid = "white",
    high = rev(c('#B2182B','#EF8A62','#FDDBC7')), midpoint = 0,
    guide = "colourbar", aesthetics = "fill")
VlnPlot(seurat_orig, features = genes, ncol = 1)
plots <- FeaturePlot(
  seurat_orig, reduction = "UMAP", features = genes,
  order = TRUE, blend = TRUE, combine = FALSE)
CombinePlots(plots[3:4], legend = "none")
DotPlot(seurat_orig, features = genes, cols = c("blue", "red")) + RotatedAxis()
```

## Ex9

```{r}
genes <- c("NR4A2", "NTNG2")
DoHeatmap(seurat_orig, slot = "data", features = genes, size = 3) +
  NoLegend() +
  scale_fill_gradient2(
    low = rev(c('#D1E5F0','#67A9CF','#2166AC')), mid = "white",
    high = rev(c('#B2182B','#EF8A62','#FDDBC7')), midpoint = 0,
    guide = "colourbar", aesthetics = "fill")
VlnPlot(seurat_orig, features = genes, ncol = 1)
plots <- FeaturePlot(
  seurat_orig, reduction = "UMAP", features = genes,
  order = TRUE, blend = TRUE, combine = FALSE)
CombinePlots(plots[3:4], legend = "none")
DotPlot(seurat_orig, features = genes, cols = c("blue", "red")) + RotatedAxis()
```

# Pathway Analysis

![](../assets/cytoscape/09/habib-2017-snRNAseq-pathways-nolabels-wlegend.svg)

# WGCNA

```{r}
# Prepare data.
n_sets <- length(levels(x = seurat_orig))
wgcna_input <- vector(mode = "list", length = n_sets)
for (i in 1:n_sets) {
  cluster_data <- subset(x = seurat_orig, idents = levels(x = seurat_orig)[i])
  cluster_data <- as.data.frame(as.matrix(GetAssayData(object = cluster_data, slot = "data")))
  wgcna_input[[i]] <- list(data = as.data.frame(t(cluster_data)))
}
wgcna_size <- checkSets(wgcna_input) # Check if satisfies conditions for WGCNA.
```

# References

This is the concluding section of the document. Here we write relevant results to disk, output the `sessionInfo`, and create a bibliography for works cited.

```{r}
saveRDS(sce_hvg, file.path(results_dir, "data", "09", "sce_hvg.rds"))

sessionInfo()
```
