---
title: "Tau Vulnerability Report"
author:
  - name: "Emir Turkes [et2628@cumc.columbia.edu]"
  - name: "Columbia University"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
bibliography: "../tau-vulnerability.bib"
biblio-style: apalike
link-citations: true
output:
  html_document:
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: false
knit:
  (function(inputFile, encoding) {
    rmarkdown::render(
      inputFile, encoding = encoding, output_file = "../results/tau-vulnerability-report.html")})
---

```{r, include = FALSE}
#    This file is part of tau-vulnerability.
#    Copyright (C) 2019  Emir Turkes
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

knitr::opts_chunk$set(fig.width = 8.5, fig.height = 7)
```

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
.tocify-subheader .tocify-subheader .tocify-item {
  font-size: 0.90em; padding-left: 35px; text-indent: 0;}
div.main-container {max-width: none; width: 100%;}
</style>

*This is an in-depth analysis to characterize differential vulnerability to tauopathy.*

The background for this data is as follows:

- snRNAseq data from @habib_massively_2017 cleaned and analyzed in [github.com/eturkes/habib-2017-snRNAseq](https://github.com/eturkes/habib-2017-snRNAseq).

This analysis was performed in R except where noted.
The source code and instructions for rerunning the analysis can be found at [github.com/eturkes/tau-vulnerability](https://github.com/eturkes/tau-vulnerability).

# Final Results

**Read just this section for the final results of the analysis and a summary of the methods.**

# ~~~ Breakdown of Methods ~~~ {-}

**Sections from here to the end break down the methods used and are optional to read.**

We start by loading in any required packages and setting some global variables.

```{r}
packages <- c(
  "conflicted", "SingleCellExperiment", "magrittr", "dplyr", "ggplot2", "ggrepel", "S4Vectors",
  "SummarizedExperiment", "DropletUtils", "scran", "BiocSingular", "scater", "Rtsne", "svd",
  "SC3", "DT", "data.table", "Seurat", "uwot", "viridis")
invisible(suppressPackageStartupMessages(lapply(packages, library, character.only = TRUE)))

options(stringsAsFactors = FALSE)

assets_dir <- file.path(getwd(), "..", "assets")
results_dir <- file.path(getwd(), "..", "results")

# ggplot2 function providing custom aesthetics and automatic placement of categorical labels.
# For continuous data, a colorbar is implemented.
dim_red_plot <- function(data, x, y, col, type) {
  gg <- ggplot(data, aes_string(x = x, y = y, color = col)) +
    geom_point(alpha = 0.35, stroke = 0.05, shape = 21, aes_string(fill = col)) +
    theme_classic() +
    theme(
      legend.position = "right", plot.title = element_text(hjust = 0.5),
      legend.title = element_blank()) +
    guides(color = guide_legend(override.aes = list(alpha = 1)))
    if (type == "cat") {
      gg <- gg + geom_label_repel(data = label_df2, aes(label = label), show.legend = FALSE)
    } else if (type == "cont") {
      gg <- ggplot(data, aes_string(x = x, y = y)) +
        geom_point(alpha = 0.35, stroke = 0.05, aes_string(color = col)) +
        theme_classic() +
        theme(
          legend.position = "right", plot.title = element_text(hjust = 0.5),
          legend.title = element_blank()) +
        scale_colour_viridis()}
  gg}

# From SC3 source, needed to look at more marker genes than the top 10.
organise_marker_genes <- function(object, k, p_val, auroc) {
  dat <- rowData(object)[ , c(
    paste0("sc3_", k, "_markers_clusts"), paste0("sc3_", k, "_markers_auroc"),
    paste0("sc3_", k, "_markers_padj"), "feature_symbol")]
  dat <- dat[dat[ , paste0("sc3_", k, "_markers_padj")] < p_val & !is.na(
    dat[, paste0("sc3_", k, "_markers_padj")]), ]
  dat <- dat[dat[ , paste0("sc3_", k, "_markers_auroc")] > auroc, ]
  d <- NULL
  for (i in sort(unique(dat[, paste0("sc3_", k, "_markers_clusts")]))) {
    tmp <- dat[dat[, paste0("sc3_", k, "_markers_clusts")] == i, ]
    tmp <- tmp[order(tmp[, paste0("sc3_", k, "_markers_auroc")], decreasing = TRUE), ]
    d <- rbind(d, tmp)}
  if (nrow(dat) > 0) {
    return(d)
  } else {
    return(NULL)}}

# Adds download buttons.
datatable_custom <- function(dt) {
  datatable(
    dt,
    extensions = "Buttons", options = list(dom = "Blfrtip", buttons = list(
      "copy", "print",
      list(extend = "collection", buttons = c("csv", "excel", "pdf"), text = "Download"))))}
```

# Original Data

We start by reading in data from previous broad analyses and displaying their final results.

## Habib 2017 snRNAseq {.tabset}

The cleaned dataset here is derived from the analysis in [github.com/eturkes/habib-2017-snRNAseq](https://github.com/eturkes/habib-2017-snRNAseq), which uses droplet-based UMI data from @habib_massively_2017.
As cluster labels and tSNE coordinates were released by the authors, the main figure was recreated in the analysis, along with a successful attempt at replicating that data.
Please see the analysis for details of the processing, which was performed in close accordance to what was done in @habib_massively_2017.

```{r}
sce <- readRDS(file.path(assets_dir, "habib-2017-snRNAseq", "sce.rds"))
sce

# Set up data frame for ggplot2.
gg_df <- data.frame(
  colData(sce)[ , c(
    paste0("habib_tsne", 1:2), paste0("hvg_tsne", 1:2),
    "habib_cluster_name", "seurat_clusters_name")])
rownames(gg_df) <- NULL
gg_df$habib_cluster_name <- factor(gg_df$habib_cluster_name)
gg_df$seurat_clusters_name <- factor(gg_df$seurat_clusters_name)
rm(sce)
```

### Habib Clusters Habib tSNE

```{r}
# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  habib_cluster_name = levels(gg_df$habib_cluster_name), label = levels(gg_df$habib_cluster_name))
label_df2 <- gg_df %>%
  group_by(habib_cluster_name) %>%
  summarize(habib_tsne1 = median(habib_tsne1), habib_tsne2 = median(habib_tsne2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "habib_tsne1", y = "habib_tsne2",col = "habib_cluster_name", type = "cat") +
  xlab("tSNE 1") + ylab("tSNE 2") + ggtitle("Habib Clusters Habib tSNE")
```

### Seurat Clusters Habib tSNE

```{r}
# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters_name = levels(gg_df$seurat_clusters_name),
  label = levels(gg_df$seurat_clusters_name))
label_df2 <- gg_df %>%
  group_by(seurat_clusters_name) %>%
  summarize(habib_tsne1 = median(habib_tsne1), habib_tsne2 = median(habib_tsne2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "habib_tsne1", y = "habib_tsne2", col = "seurat_clusters_name", type = "cat") +
  xlab("tSNE 1") + ylab("tSNE 2") + ggtitle("Seurat Clusters Habib tSNE")
```

### Seurat Clusters HVG tSNE

```{r}
# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters_name = levels(gg_df$seurat_clusters_name),
  label = levels(gg_df$seurat_clusters_name))
label_df2 <- gg_df %>%
  group_by(seurat_clusters_name) %>%
  summarize(hvg_tsne1 = median(hvg_tsne1), hvg_tsne2 = median(hvg_tsne2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "hvg_tsne1", y = "hvg_tsne2", col = "seurat_clusters_name", type = "cat") +
  xlab("tSNE 1") + ylab("tSNE 2") + ggtitle("Seurat Clusters HVG tSNE")
```

### Habib Clusters HVG tSNE

```{r}
# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  habib_cluster_name = levels(gg_df$habib_cluster_name), label = levels(gg_df$habib_cluster_name))
label_df2 <- gg_df %>%
  group_by(habib_cluster_name) %>%
  summarize(hvg_tsne1 = median(hvg_tsne1), hvg_tsne2 = median(hvg_tsne2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "hvg_tsne1", y = "hvg_tsne2", col = "habib_cluster_name", type = "cat") +
  xlab("tSNE 1") + ylab("tSNE 2") + ggtitle("Habib Clusters HVG tSNE")
```

# Habib 2017 snRNAseq

The following subsections work with data derived from the analysis in [github.com/eturkes/habib-2017-snRNAseq](https://github.com/eturkes/habib-2017-snRNAseq).

```{r}
# Create a unique cache for each iterated section.
if (!dir.exists(file.path(assets_dir, "cache1"))) {
  dir.create(file.path(assets_dir, "cache1"))}

sce <- readRDS(file.path(assets_dir, "habib-2017-snRNAseq", "sce_orig.rds"))
sce
```

## Additional Processing

Further modifications to the data before subsetting to points of interest.

### QC

We remove nuclei overabundant in mitochondrial/ribosomal gene expression, a common indicator of low quality nuclei.

```{r}
mito_drop <- isOutlier(sce$pct_counts_mito, nmads = 3, type = "higher")
ribo_drop <- isOutlier(sce$pct_counts_ribo, nmads = 3, type = "higher")
keep <- !(mito_drop | ribo_drop)
datatable(data.table(
  "By Mitochondrial" = sum(mito_drop), "By Ribosomal" = sum(ribo_drop),
  Remaining = sum(keep)))
sce <- sce[ , keep]
dim(sce)[2]
```

We consider the expression profile of genes each time nuclei are removed.
A distinctly negative skew can be seen due to lowly expressed genes in our dataset.

```{r, fig.height = 5}
sce <- calculateQCMetrics(sce)
par(mfrow = c(1, 3), mar = c(5, 4, 1, 1))
hist(
  log10(rowData(sce)$mean_counts + 1e-6), col = "grey80",  main = "",
  breaks = 40, xlab = "log10(Average Number of UMI + 1e-6)")
hist(
  log10(rowData(sce)$n_cells_by_counts + 1), col = "grey80", main = "",
  breaks = 40, xlab = "log10(Number of Expressed Nuclei + 1)")
plot(
  log10(rowData(sce)$mean_counts + 1e-6), pch = 16,
  col = rgb(0, 0, 0, 0.4), log10(rowData(sce)$n_cells_by_counts + 1),
  xlab = "log10(Average Number of UMI + 1e-6)", ylab = "log10(Number of Expressed Nuclei + 1)")
```

Therefore, we remove genes in the manner specified in @habib_massively_2017.

> A gene is considered detected in a cell if it has at least two unique UMIs (transcripts) associated with it.
For each analysis, genes were removed that were detected in less than 10 nuclei.

```{r}
detected_genes <- rowSums(counts(sce) >= 2)
sce <- sce[which(detected_genes >= 10), ]
dim(sce)[1]
```

We see a robust improvement in normality.

```{r, fig.height = 5}
sce <- calculateQCMetrics(sce)
par(mfrow = c(1, 3), mar = c(5, 4, 1, 1))
hist(
  log10(rowData(sce)$mean_counts + 1e-6), col = "grey80",  main = "",
  breaks = 40, xlab = "log10(Average Number of UMI + 1e-6)")
hist(
  log10(rowData(sce)$n_cells_by_counts + 1), col = "grey80", main = "",
  breaks = 40, xlab = "log10(Number of Expressed Nuclei + 1)")
plot(
  log10(rowData(sce)$mean_counts + 1e-6), pch = 16,
  col = rgb(0, 0, 0, 0.4), log10(rowData(sce)$n_cells_by_counts + 1),
  xlab = "log10(Average Number of UMI + 1e-6)", ylab = "log10(Number of Expressed Nuclei + 1)")
```

We then display some additional summary statistics.
Apparent is the disproportionately high expression of encoding lncRNAs MALAT1 and MEG3, both of which are known to have higher expression in nuclei, as noted in @habib_massively_2017.

```{r}
plotHighestExprs(sce, exprs_values = "counts")
plotExprsFreqVsMean(sce)
```

We use `computeSumFactors` from the `scran` package to perform normalization.
This function uses a linear deconvolution system to account for expected variation across different cell types/sizes, producing "size factor" values that indicate the extent to which nuclei should be scaled.
We also perform a rough clustering of our nuclei, which improves accuracy on highly heterogenous data such as that from the brain.

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache1", "quickCluster.rds")
if (file.exists(rds)) {
  quickCluster <- readRDS(rds)
} else {
  set.seed(1)
  quickCluster <- quickCluster(sce, use.ranks = FALSE, BSPARAM = IrlbaParam())
  saveRDS(quickCluster, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache1", "size_factors.rds")
if (file.exists(rds)) {
  size_factors <- readRDS(rds)
} else {
  size_factors <- computeSumFactors(sce, cluster = quickCluster, min.mean = 0.1)
  saveRDS(size_factors, rds)}
```

In an experiment where most systematic differences between nuclei are driven by capture efficiency and sequencing depth, we should see a correlation between size factors and library size.

```{r}
sce <- size_factors
par(mfrow = c(1, 2), mar = c(5, 4, 2, 1), bty = "n")
smoothScatter(
  sce$total_counts, sizeFactors(sce), log = "xy", xlab = "Total Counts", ylab = "Size Factors")
plot(
  sce$total_counts, sizeFactors(sce), log = "xy", xlab = "Total Counts",
  ylab = "Size Factors", cex = 0.3, pch = 20, col = rgb(0.1, 0.2, 0.7, 0.3))
abline(h = 0.05)
```

The plots are well correlated, confirming the source of bias.
We therefore add normalized expression data to our SCE object using the calculated size factors.

```{r}
sce <- normalize(sce)
sce
```

### Dimensionality Reduction

#### Technical Summation

We start by applying `denoisePCA` from the `scran` package, which can automatically select the number of PCs to retain by modeling technical noise.

```{r}
new_trend <- makeTechTrend(x = sce)
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache1", "denoise_pca.rds")
if (file.exists(rds)) {
  pca <- readRDS(rds)
} else {
  set.seed(1)
  pca <- denoisePCA(sce, technical = new_trend, BSPARAM = IrlbaParam())
  saveRDS(pca, rds)}
```

The results suggest retaining the following number of PCs:

```{r}
sce <- pca
ncol(reducedDim(sce, "PCA"))
par(mfrow = c(1, 1))
plot(
  log10(attr(reducedDim(sce), "percentVar")), xlab = "PC",
  ylab = "log10(Proportion of Variance Explained)", pch = 20,
  cex = 0.6, col = rgb(0.8, 0.2, 0.2, 0.5))
abline(v = ncol(reducedDim(sce, "PCA")), lty = 2, col = "red")
```

We assess the effectiveness of these PCs by using them to compute UMAP coordinates.

```{r}
rds <- file.path(assets_dir, "cache1", "denoise_umap.rds")
if (file.exists(rds)) {
  umap <- readRDS(rds)
} else {
  set.seed(1)
  umap <- runUMAP(sce, min_dist = 0.5)
  saveRDS(umap, rds)}
```

```{r}
sce <- umap
add_df <- data.frame(reducedDim(sce, "UMAP"))
names(add_df) <- paste0("denoise_umap", seq(ncol(add_df)))
colData(sce) <- cbind(
  colData(sce), denoise_umap1 = add_df$denoise_umap1, denoise_umap2 = add_df$denoise_umap2)

# Set up data frame for ggplot2.
gg_df <- data.frame(
  colData(sce)[ , c(
    "cell_id_stem", paste0("habib_tsne", 1:2),
    "log10_total_features_by_counts", "habib_cluster_name")],
  add_df)
rownames(gg_df) <- NULL
gg_df$habib_cluster_name <- factor(gg_df$habib_cluster_name)

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  habib_cluster_name = levels(gg_df$habib_cluster_name),
  label = levels(gg_df$habib_cluster_name))
label_df2 <- gg_df %>%
  group_by(habib_cluster_name) %>%
  summarize(denoise_umap1 = median(denoise_umap1), denoise_umap2 = median(denoise_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "denoise_umap1", y = "denoise_umap2", col = "habib_cluster_name", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Habib Clusters denoisePCA UMAP")
dim_red_plot(
  data = gg_df, x = "denoise_umap1", y = "denoise_umap2", col = "cell_id_stem", type = "other") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Sample Distribution denoisePCA UMAP")
dim_red_plot(
  data = gg_df, x = "denoise_umap1", y = "denoise_umap2",
  col = "log10_total_features_by_counts", type = "cont") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("log10(Total Features) denoisePCA UMAP")
```

#### PROPACK SVD

As an alternative, we can also calculate a truncated SVD (singular value decomposition) from which PCs can be extracted from it.
In our testing, this approach, particularly the PROPACK implementation, was found to have high power in distinguishing subpopulations of common cell types.

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache1", "ppk.rds")
if (file.exists(rds)) {
  ppk <- readRDS(rds)
} else {
  set.seed(1)
  ppk <- propack.svd(scale(t(as.matrix(logcounts(sce)))), neig = 50)
  saveRDS(ppk, rds)}
```

```{r}
pca <- t(ppk$d * t(ppk$u))
retain_pcs <- 20
retain_pcs
par(mfrow = c(1, 1))
plot(
  log10(ppk$d), xlab = "PC", ylab = "log10(Proportion of Variance Explained)",
  pch = 20, cex = 0.6, col = rgb(0.8, 0.2, 0.2, 0.5))
abline(v = retain_pcs, lty = 2, col = "red")
```

```{r}
rds <- file.path(assets_dir, "cache1", "ppk_umap.rds")
if (file.exists(rds)) {
  umap <- readRDS(rds)
} else {
  set.seed(1)
  umap <- umap(pca[ , 1:retain_pcs], min_dist = 0.5)
  saveRDS(umap, rds)}
```

```{r}
reducedDims(sce) <- SimpleList(PCA = pca, UMAP = umap)
add_df <- data.frame(reducedDim(sce, "UMAP"))
names(add_df) <- paste0("ppk_umap", seq(ncol(add_df)))
colData(sce) <- cbind(
  colData(sce), ppk_umap1 = add_df$ppk_umap1, ppk_umap2 = add_df$ppk_umap2)
gg_df <- data.frame(gg_df, add_df)

# Setup for automatic placement of cluster labels.
label_df2 <- gg_df %>%
  group_by(habib_cluster_name) %>%
  summarize(ppk_umap1 = median(ppk_umap1), ppk_umap2 = median(ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "ppk_umap1", y = "ppk_umap2", col = "habib_cluster_name", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Habib Clusters PPK UMAP")
dim_red_plot(
  data = gg_df, x = "ppk_umap1", y = "ppk_umap2", col = "cell_id_stem", type = "other") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Sample Distribution PPK UMAP")
dim_red_plot(
  data = gg_df, x = "ppk_umap1", y = "ppk_umap2",
  col = "log10_total_features_by_counts", type = "cont") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("log10(Total Features) PPK UMAP")
```

#### HVG

We subset to the most highly variable genes (HVGs) to reduce noise in our dataset.
As we do not have spike-in transcripts, we start by modeling the technical noise as Poisson and creating a fitted trend on that basis.

```{r}
new_trend <- makeTechTrend(x = sce)
fit <- trendVar(sce, use.spikes = FALSE, loess.args = list(span = 0.05))
par(mfrow = c(1, 1), mar = c(5, 4, 2, 1), bty = "n")
plot(
  fit$mean, fit$var, pch = 20,
  col = rgb(0.1, 0.2, 0.7, 0.6), xlab = "Mean Log-expression", ylab = "Variance")
curve(fit$trend(x), col = "orange", lwd = 2, add = TRUE)
curve(new_trend(x), col = "red", lwd = 2, add = TRUE)
legend(
  "top", legend = c("Poisson Noise", "Observed Trend"), lty = 1,
  lwd = 2, col = c("red", "orange"), bty = "n")
```

The plot shows a large discrepancy between the two trends, which we assume is contributed by each gene's biological component.
We extract the genes with the largest biological components and plot them here.

```{r}
fit$trend <- new_trend
dec <- decomposeVar(fit = fit)
top_dec <- rownames(dec[order(dec$bio, decreasing = TRUE), ])[1:10]
rm(fit)
plotExpression(sce, features = top_dec) +
  stat_summary(
    fun.y = median, fun.ymin = median, fun.ymax = median, geom = "crossbar", width = 0.3, alpha = 0.8)
```

We inspect the distribution of our FDR and biological component values to see how see should proceed.

```{r, fig.height = 6}
fdr_median <- median(dec$FDR, na.rm = TRUE)
fdr_median
bio_median <- median(dec$bio, na.rm = TRUE)
bio_median
par(mfrow = c(1, 2))
hist(log10(dec$FDR), breaks = 100, main = "")
abline(v = c(log10(fdr_median), log10(min(dec$FDR[dec$FDR > 0]))), lty = 2, col = "red")
hist(log10(dec$bio), breaks = 100, main = "")
abline(v = c(log10(bio_median), log10(max(dec$bio))), lty = 2, col = "red")
```

We see that thresholding at the median is able to capture most of the desired distribution of both values.
Let's see what number of genes we would subset to with these thresholds.

```{r}
keep <- which(dec$FDR < fdr_median & dec$bio >= bio_median)
sce_hvg <- sce[keep, ]
rm(dec)
dim(sce_hvg)[1]
```

This approach captures identifies about 5,000 HVGs, a suitable number for downstream analysis.
Due to the change in dimensions, we renormalize and calculate new QC metrics.

```{r}
sce_hvg <- calculateQCMetrics(sce_hvg)
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache1", "hvg_quickCluster.rds")
if (file.exists(rds)) {
  quickCluster <- readRDS(rds)
} else {
  set.seed(1)
  quickCluster <- quickCluster(sce_hvg, use.ranks = FALSE, BSPARAM = IrlbaParam())
  saveRDS(quickCluster, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache1", "hvg_size_factors.rds")
if (file.exists(rds)) {
  size_factors <- readRDS(rds)
} else {
  size_factors <- computeSumFactors(sce_hvg, cluster = quickCluster, min.mean = 0.1)
  saveRDS(size_factors, rds)}
```

```{r}
sce_hvg <- size_factors
sce_hvg <- normalize(sce_hvg)
rm(size_factors)
sce_hvg
```

We now test the effectiveness of our HVGs through the SVD+UMAP method described earlier.

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache1", "hvg_ppk.rds")
if (file.exists(rds)) {
  ppk <- readRDS(rds)
} else {
  set.seed(1)
  ppk <- propack.svd(scale(t(as.matrix(logcounts(sce_hvg)))), neig = 50)
  saveRDS(ppk, rds)}
```

```{r}
pca <- t(ppk$d * t(ppk$u))
hvg_retain_pcs <- 23
hvg_retain_pcs
par(mfrow = c(1, 1))
plot(
  log10(ppk$d), xlab = "PC", ylab = "log10(Proportion of Variance Explained)",
  pch = 20, cex = 0.6, col = rgb(0.8, 0.2, 0.2, 0.5))
abline(v = hvg_retain_pcs, lty = 2, col = "red")
```

```{r}
rds <- file.path(assets_dir, "cache1", "hvg_ppk_umap.rds")
if (file.exists(rds)) {
  umap <- readRDS(rds)
} else {
  set.seed(1)
  umap <- umap(pca[ , 1:hvg_retain_pcs], min_dist = 0.5)
  saveRDS(umap, rds)}
```

```{r}
reducedDims(sce) <- SimpleList(PCA = pca, UMAP = umap)
add_df <- data.frame(reducedDim(sce, "UMAP"))
names(add_df) <- paste0("hvg_ppk_umap", seq(ncol(add_df)))
colData(sce_hvg) <- cbind(
  colData(sce_hvg), hvg_ppk_umap1 = add_df$hvg_ppk_umap1, hvg_ppk_umap2 = add_df$hvg_ppk_umap2)
colData(sce) <- cbind(
  colData(sce), hvg_ppk_umap1 = add_df$hvg_ppk_umap1, hvg_ppk_umap2 = add_df$hvg_ppk_umap2)
gg_df <- data.frame(gg_df, add_df)
rm(pca, umap, add_df)

# Setup for automatic placement of cluster labels.
label_df2 <- gg_df %>%
  group_by(habib_cluster_name) %>%
  summarize(hvg_ppk_umap1 = median(hvg_ppk_umap1), hvg_ppk_umap2 = median(hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "hvg_ppk_umap1", y = "hvg_ppk_umap2", col = "habib_cluster_name", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Habib Clusters HVG PPK UMAP")
dim_red_plot(
  data = gg_df, x = "hvg_ppk_umap1", y = "hvg_ppk_umap2", col = "cell_id_stem", type = "other") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Sample Distribution HVG PPK UMAP")
dim_red_plot(
  data = gg_df, x = "hvg_ppk_umap1", y = "hvg_ppk_umap2",
  col = "log10_total_features_by_counts", type = "cont") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("log10(Total Features) HVG PPK UMAP")
```

### Clustering

#### SC3

Due computational demands of SC3, we do not run it on the full dataset.
However, we do use the `sc3_estimate_k` to predict an optimal $k$ for other clustering algorithms.

```{r}
rowData(sce)$feature_symbol = rowData(sce)$external_gene_name
rowData(sce_hvg)$feature_symbol = rowData(sce_hvg)$external_gene_name
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache1", "sc3_estimate_k.rds")
if (file.exists(rds)) {
  sc3_estimate_k <- readRDS(rds)
} else {
  set.seed(1)
  sc3_estimate_k <- sc3_estimate_k(sce)
  saveRDS(sc3_estimate_k, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache1", "hvg_sc3_estimate_k.rds")
if (file.exists(rds)) {
  hvg_sc3_estimate_k <- readRDS(rds)
} else {
  set.seed(1)
  hvg_sc3_estimate_k <- sc3_estimate_k(sce_hvg)
  saveRDS(hvg_sc3_estimate_k, rds)}
```

```{r}
sce <- sc3_estimate_k
rm(sc3_estimate_k)
metadata(sce)$sc3$k_estimation

sce_hvg <- hvg_sc3_estimate_k
rm(hvg_sc3_estimate_k)
metadata(sce_hvg)$sc3$k_estimation
```

Unfortunately, we see that SC3 was not able to provide a reasonable estimation of $k$.
Therefore, we rely on other means to decide the optimal number of clusters to obtain.

#### Seurat

Seurat is a nearest neighbor graph clustering method with a `resolution` parameter that indirectly influences the number of clusters generated.
Concerning `resolution`, the authors write:

> setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells.
Optimal resolution often increases for larger datasets.

As a compromise, we try simply scaling the median of these recommended settings (0.8, which is also the default) to the size of our dataset.

```{r}
resolution <- (dim(sce)[2] / 3000) * 0.8
resolution
```

We experiment with feature and PCA input derived from both all genes and HVGs.

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache1", "seurat.rds")
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {
  seurat <- as.Seurat(sce, counts = "counts", data = NULL)
  seurat <- NormalizeData(seurat)
  seurat <- FindVariableFeatures(seurat, nfeatures = dim(sce)[1])
  seurat <- ScaleData(seurat, features = rownames(seurat))
  seurat <- FindNeighbors(seurat, reduction = "PCA", dims = 1:retain_pcs)
  seurat <- FindClusters(seurat, resolution = resolution)
  saveRDS(seurat, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache1", "seurat_hvg.rds")
if (file.exists(rds)) {
  seurat_hvg <- readRDS(rds)
} else {
  seurat_hvg <- as.Seurat(sce_hvg, counts = "counts", data = NULL)
  seurat_hvg <- NormalizeData(seurat_hvg)
  seurat_hvg <- FindVariableFeatures(seurat_hvg, nfeatures = dim(sce_hvg)[1])
  seurat_hvg <- ScaleData(seurat_hvg, features = rownames(seurat_hvg))
  seurat_hvg <- FindNeighbors(seurat_hvg, reduction = "PCA", dims = 1:hvg_retain_pcs)
  seurat_hvg <- FindClusters(seurat_hvg, resolution = resolution)
  saveRDS(seurat_hvg, rds)}
```

```{r}
gg_df$seurat_clusters <- seurat$seurat_clusters
gg_df$hvg_seurat_clusters <- seurat_hvg$seurat_clusters
rm(seurat, seurat_hvg)

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters = levels(gg_df$seurat_clusters), label = levels(gg_df$seurat_clusters))
label_df2 <- gg_df %>%
  group_by(seurat_clusters) %>%
  summarize(ppk_umap1 = median(ppk_umap1), ppk_umap2 = median(ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "ppk_umap1", y = "ppk_umap2", col = "seurat_clusters", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Seurat Clusters PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters = levels(gg_df$hvg_seurat_clusters), label = levels(gg_df$hvg_seurat_clusters))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters) %>%
  summarize(ppk_umap1 = median(ppk_umap1), ppk_umap2 = median(ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "ppk_umap1", y = "ppk_umap2", col = "hvg_seurat_clusters", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("HVG Seurat Clusters PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters = levels(gg_df$seurat_clusters), label = levels(gg_df$seurat_clusters))
label_df2 <- gg_df %>%
  group_by(seurat_clusters) %>%
  summarize(hvg_ppk_umap1 = median(hvg_ppk_umap1), hvg_ppk_umap2 = median(hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "hvg_ppk_umap1", y = "hvg_ppk_umap2", col = "seurat_clusters", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Seurat Clusters HVG PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters = levels(gg_df$hvg_seurat_clusters), label = levels(gg_df$hvg_seurat_clusters))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters) %>%
  summarize(hvg_ppk_umap1 = median(hvg_ppk_umap1), hvg_ppk_umap2 = median(hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "hvg_ppk_umap1", y = "hvg_ppk_umap2", col = "hvg_seurat_clusters", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("HVG Seurat Clusters HVG PPK UMAP")
```

We see that Seurat produced 35 clusters with both datasets.
While they are similar, differences can be seen particularly regarding cells that lie between major clusters.

#### k-means

To validate the Seurat clusters, we also try k-means on both datasets with $k$ set to 35.

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache1", "km_35.rds")
if (file.exists(rds)) {
  km_35 <- readRDS(rds)
} else {
  km_35 <- kmeans(
    reducedDim(sce, "PCA")[ , 1:retain_pcs], centers = 35, iter.max = 1e8,
    nstart = 2500, algorithm = "MacQueen")
  saveRDS(km_35, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache1", "hvg_km_35.rds")
if (file.exists(rds)) {
  hvg_km_35 <- readRDS(rds)
} else {
  hvg_km_35 <- kmeans(
    reducedDim(sce_hvg, "PCA")[ , 1:hvg_retain_pcs], centers = 35, iter.max = 1e8,
    nstart = 2500, algorithm = "MacQueen")
  saveRDS(hvg_km_35, rds)}
```

```{r}
gg_df$km_35_clusters <- factor(km_35$cluster)
gg_df$hvg_km_35_clusters <- factor(hvg_km_35$cluster)
rm(km_35, hvg_km_35)

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  km_35_clusters = levels(gg_df$km_35_clusters), label = levels(gg_df$km_35_clusters))
label_df2 <- gg_df %>%
  group_by(km_35_clusters) %>%
  summarize(ppk_umap1 = median(ppk_umap1), ppk_umap2 = median(ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "ppk_umap1", y = "ppk_umap2", col = "km_35_clusters", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("KM 35 Clusters PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_km_35_clusters = levels(gg_df$hvg_km_35_clusters), label = levels(gg_df$hvg_km_35_clusters))
label_df2 <- gg_df %>%
  group_by(hvg_km_35_clusters) %>%
  summarize(ppk_umap1 = median(ppk_umap1), ppk_umap2 = median(ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "ppk_umap1", y = "ppk_umap2", col = "hvg_km_35_clusters", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("HVG KM 35 Clusters PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  km_35_clusters = levels(gg_df$km_35_clusters), label = levels(gg_df$km_35_clusters))
label_df2 <- gg_df %>%
  group_by(km_35_clusters) %>%
  summarize(hvg_ppk_umap1 = median(hvg_ppk_umap1), hvg_ppk_umap2 = median(hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "hvg_ppk_umap1", y = "hvg_ppk_umap2",
  col = "km_35_clusters", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("KM 35 Clusters HVG PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_km_35_clusters = levels(gg_df$hvg_km_35_clusters), label = levels(gg_df$hvg_km_35_clusters))
label_df2 <- gg_df %>%
  group_by(hvg_km_35_clusters) %>%
  summarize(hvg_ppk_umap1 = median(hvg_ppk_umap1), hvg_ppk_umap2 = median(hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "hvg_ppk_umap1", y = "hvg_ppk_umap2", col = "hvg_km_35_clusters", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("HVG KM 35 Clusters HVG PPK UMAP")
```

#### Labeling

After visually inspecting all of the cluster/UMAP combinations, we select the Seurat clusters using all genes overlaid onto the HVG PPK UMAP as being most suitable for our needs.
Besides the clusters partitioning cleanly, this combination provided the clearest separation of putative excitatory/inhibitory neurons from the rest of the cell population.
The two neuron populations are also well separated from each other with reasonable partitioning of clusters intermediate to them.
We therefore go ahead and label these clusters based on their proximity to clusters from @habib_massively_2017 and their UMAP coordinates.

```{r}
gg_df$seurat_clusters_name <- NA
seurat_clusters <- as.integer(gg_df$seurat_clusters)
for (i in 1:length(seurat_clusters)) {
  if (seurat_clusters[i] == 1) {
    gg_df[i, ncol(gg_df)] <- "exPFC1.1"
  } else if (seurat_clusters[i] == 2) {
    gg_df[i, ncol(gg_df)] <- "ODC2.1"
  } else if (seurat_clusters[i] == 3) {
    gg_df[i, ncol(gg_df)] <- "ODC1.2"
  } else if (seurat_clusters[i] == 4) {
    gg_df[i, ncol(gg_df)] <- "exPFC1.4"
  } else if (seurat_clusters[i] == 5) {
    gg_df[i, ncol(gg_df)] <- "GABA1.1"
  } else if (seurat_clusters[i] == 6) {
    gg_df[i, ncol(gg_df)] <- "ODC1.1"
  } else if (seurat_clusters[i] == 7) {
    gg_df[i, ncol(gg_df)] <- "exDG1.1"
  } else if (seurat_clusters[i] == 8) {
    gg_df[i, ncol(gg_df)] <- "exCA1"
  } else if (seurat_clusters[i] == 9) {
    gg_df[i, ncol(gg_df)] <- "exPFC1.3"
  } else if (seurat_clusters[i] == 10) {
    gg_df[i, ncol(gg_df)] <- "exDG1.2"
  } else if (seurat_clusters[i] == 11) {
    gg_df[i, ncol(gg_df)] <- "exCA3"
  } else if (seurat_clusters[i] == 12) {
    gg_df[i, ncol(gg_df)] <- "exPFC2"
  } else if (seurat_clusters[i] == 13) {
    gg_df[i, ncol(gg_df)] <- "ASC1.1"
  } else if (seurat_clusters[i] == 14) {
    gg_df[i, ncol(gg_df)] <- "GABA2.1"
  } else if (seurat_clusters[i] == 15) {
    gg_df[i, ncol(gg_df)] <- "exDG1.3"
  } else if (seurat_clusters[i] == 16) {
    gg_df[i, ncol(gg_df)] <- "ASC2.2"
  } else if (seurat_clusters[i] == 17) {
    gg_df[i, ncol(gg_df)] <- "ASC2.1"
  } else if (seurat_clusters[i] == 18) {
    gg_df[i, ncol(gg_df)] <- "ODC2.2"
  } else if (seurat_clusters[i] == 19) {
    gg_df[i, ncol(gg_df)] <- "GABA1.2"
  } else if (seurat_clusters[i] == 20) {
    gg_df[i, ncol(gg_df)] <- "exPFC1.2"
  } else if (seurat_clusters[i] == 21) {
    gg_df[i, ncol(gg_df)] <- "OPC1.1"
  } else if (seurat_clusters[i] == 22) {
    gg_df[i, ncol(gg_df)] <- "OPC1.2"
  } else if (seurat_clusters[i] == 23) {
    gg_df[i, ncol(gg_df)] <- "ASC1.2"
  } else if (seurat_clusters[i] == 24) {
    gg_df[i, ncol(gg_df)] <- "MG"
  } else if (seurat_clusters[i] == 25) {
    gg_df[i, ncol(gg_df)] <- "ASC1.5"
  } else if (seurat_clusters[i] == 26) {
    gg_df[i, ncol(gg_df)] <- "NSC"
  } else if (seurat_clusters[i] == 27) {
    gg_df[i, ncol(gg_df)] <- "GABA2.2"
  } else if (seurat_clusters[i] == 28) {
    gg_df[i, ncol(gg_df)] <- "exPFC1.6"
  } else if (seurat_clusters[i] == 29) {
    gg_df[i, ncol(gg_df)] <- "exPFC1.5"
  } else if (seurat_clusters[i] == 30) {
    gg_df[i, ncol(gg_df)] <- "Unlabeled3"
  } else if (seurat_clusters[i] == 31) {
    gg_df[i, ncol(gg_df)] <- "END"
  } else if (seurat_clusters[i] == 32) {
    gg_df[i, ncol(gg_df)] <- "New1"
  } else if (seurat_clusters[i] == 33) {
    gg_df[i, ncol(gg_df)] <- "New2"
  } else if (seurat_clusters[i] == 34) {
    gg_df[i, ncol(gg_df)] <- "Unlabeled1"
  } else if (seurat_clusters[i] == 35) {
    gg_df[i, ncol(gg_df)] <- "New3"}}

colData(sce_hvg) <- cbind(
  colData(sce_hvg), seurat_clusters = gg_df$seurat_clusters,
  seurat_clusters_name = gg_df$seurat_clusters_name)
colData(sce) <- cbind(
  colData(sce), seurat_clusters = gg_df$seurat_clusters,
  seurat_clusters_name = gg_df$seurat_clusters_name)
gg_df$seurat_clusters_name <- factor(gg_df$seurat_clusters_name)

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters_name = levels(gg_df$seurat_clusters_name),
  label = levels(gg_df$seurat_clusters_name))
label_df2 <- gg_df %>%
  group_by(seurat_clusters_name) %>%
  summarize(hvg_ppk_umap1 = median(hvg_ppk_umap1), hvg_ppk_umap2 = median(hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "hvg_ppk_umap1", y = "hvg_ppk_umap2",
  col = "seurat_clusters_name", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Seurat Clusters HVG PPK UMAP")
```

The final figure appears satisfactory for our needs.
A single caveat remains regarding the "Unlabeled3" population.
Our label system was not able to label it in a sane location, indicating that its location on the UMAP is quite distributed.
For the future, we consider coercing it into separate clusters or removing it entirely as technical noise.
We now save our SCE object for use in the following subsections.

```{r}
saveRDS(sce, file.path(assets_dir, "cache1", "sce.rds"))
saveRDS(sce_hvg, file.path(assets_dir, "cache1", "sce_hvg.rds"))
rm(sce, sce_hvg, gg_df, label_df, label_df2)
```

## EX/IN PFC Neurons

Due to findings of increased vulnerability to tauopathy in excitatory (EX) neurons vs. inhibitory (IN) neurons [@fu_tau_2019], we start by subsetting our data to these cell types.
We also look specifically in the PFC as it is an early affected region of Alzheimer's Disease (AD) and will complement current experimental samples in our lab.

```{r}
# Create a unique cache for each iterated section.
if (!dir.exists(file.path(assets_dir, "cache2"))) {
  dir.create(file.path(assets_dir, "cache2"))}

sce <- readRDS(file.path(assets_dir, "cache1", "sce.rds"))
regions <- "PFC"
c_types <- "exPFC|GABA|New1"
keep <- grepl(regions, colData(sce)$cell_id_stem)
sce <- sce[ , keep]
keep <- grepl(c_types, colData(sce)$seurat_clusters_name)
sce <- sce[ , keep]
sce
```

### QC

#### Lowly Expressed Genes

We remove genes that are now lowly expressed after the subset.

```{r}
sce <- calculateQCMetrics(sce)
detected_genes <- rowSums(counts(sce) >= 2)
sce <- sce[which(detected_genes >= 10), ]
dim(sce)[1]
```

#### Normalization

Due to the change in dimensions from subsetting and QC, we renormalize.

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "quickCluster.rds")
if (file.exists(rds)) {
  quickCluster <- readRDS(rds)
} else {
  set.seed(1)
  quickCluster <- quickCluster(sce, use.ranks = FALSE, BSPARAM = IrlbaParam())
  saveRDS(quickCluster, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "size_factors.rds")
if (file.exists(rds)) {
  size_factors <- readRDS(rds)
} else {
  size_factors <- computeSumFactors(sce, cluster = quickCluster, min.mean = 0.1)
  saveRDS(size_factors, rds)}
```

```{r}
sce <- normalize(sce)
rm(size_factors)
```

### Dimensionality Reduction

#### PROPACK SVD

We use PROPACK SVD+UMAP as promising results were obtained with that combination in the Additional Processing section.

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "ppk.rds")
if (file.exists(rds)) {
  ppk <- readRDS(rds)
} else {
  set.seed(1)
  ppk <- propack.svd(scale(t(as.matrix(logcounts(sce)))), neig = 50)
  saveRDS(ppk, rds)}
```

```{r}
pca <- t(ppk$d * t(ppk$u))
retain_pcs <- 24
retain_pcs
par(mfrow = c(1, 1))
plot(
  log10(ppk$d), xlab = "PC", ylab = "log10(Proportion of Variance Explained)",
  pch = 20, cex = 0.6, col = rgb(0.8, 0.2, 0.2, 0.5))
abline(v = retain_pcs, lty = 2, col = "red")
```

Interestingly, there appear to be "bridging" cells that have qualities of both EX and IN neurons.
These cells were also apparent when examining the whole dataset and are robust across a range of PCA/UMAP parameters.

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "ppk_umap.rds")
if (file.exists(rds)) {
  umap <- readRDS(rds)
} else {
  set.seed(1)
  umap <- umap(pca[ , 1:retain_pcs], min_dist = 0.2)
  saveRDS(umap, rds)}
```

```{r}
reducedDims(sce) <- SimpleList(PCA = pca, UMAP = umap)
add_df <- data.frame(reducedDim(sce, "UMAP"))
names(add_df) <- paste0("sub_ppk_umap", seq(ncol(add_df)))
colData(sce) <- cbind(
  colData(sce), sub_ppk_umap1 = add_df$sub_ppk_umap1, sub_ppk_umap2 = add_df$sub_ppk_umap2)

# Set up data frame for ggplot2.
gg_df <- data.frame(
  colData(sce)[ , c(
    "cell_id_stem", "log10_total_features_by_counts", "seurat_clusters_name")], add_df)
rownames(gg_df) <- NULL
gg_df$seurat_clusters_name <- factor(gg_df$seurat_clusters_name)

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters_name = levels(gg_df$seurat_clusters_name),
  label = levels(gg_df$seurat_clusters_name))
label_df2 <- gg_df %>%
  group_by(seurat_clusters_name) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2",
  col = "seurat_clusters_name", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Whole Set Seurat Clusters Subset PPK UMAP")
dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2", col = "cell_id_stem", type = "other") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Sample Distribution Subset PPK UMAP")
dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2",
  col = "log10_total_features_by_counts", type = "cont") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("log10(Total Features) PPK UMAP")
```

#### HVG

Though we already have ~5,000 genes, we assess how descriptive this set is.

```{r}
new_trend <- makeTechTrend(x = sce)
fit <- trendVar(sce, use.spikes = FALSE, loess.args = list(span = 0.05))
fit$trend <- new_trend
dec <- decomposeVar(fit = fit)
top_dec <- rownames(dec[order(dec$bio, decreasing = TRUE), ])[1:10]
rm(fit)
plotExpression(sce, features = top_dec) +
  stat_summary(
    fun.y = median, fun.ymin = median, fun.ymax = median, geom = "crossbar", width = 0.3, alpha = 0.8)

fdr_median <- median(dec$FDR, na.rm = TRUE)
fdr_median
bio_median <- median(dec$bio, na.rm = TRUE)
bio_median
```

We decide that the FDR median is too large to be used as a cutoff, so we use half of the median, which is more reasonable.

```{r, fig.height = 6}
fdr_half_median <- median(dec$FDR, na.rm = TRUE) / 2
fdr_half_median
bio_half_median <- median(dec$bio, na.rm = TRUE) / 2
bio_half_median
par(mfrow = c(1, 2))
hist(log10(dec$FDR), breaks = 100, main = "")
abline(v = c(log10(fdr_half_median), log10(min(dec$FDR[dec$FDR > 0]))), lty = 2, col = "red")
hist(log10(dec$bio), breaks = 100, main = "")
abline(v = c(log10(bio_half_median), log10(max(dec$bio))), lty = 2, col = "red")
```

```{r}
keep <- which(dec$FDR < fdr_half_median & dec$bio >= bio_half_median)
sce_hvg <- sce[keep, ]
rm(dec)
dim(sce_hvg)[1]
```

We retain ~2,000 HVGs, a suitable number for our analysis.

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "hvg_ppk.rds")
if (file.exists(rds)) {
  ppk <- readRDS(rds)
} else {
  set.seed(1)
  ppk <- propack.svd(scale(t(as.matrix(logcounts(sce_hvg)))), neig = 50)
  saveRDS(ppk, rds)}
```

```{r}
pca <- t(ppk$d * t(ppk$u))
hvg_retain_pcs <- 25
hvg_retain_pcs
par(mfrow = c(1, 1))
plot(
  log10(ppk$d), xlab = "PC", ylab = "log10(Proportion of Variance Explained)",
  pch = 20, cex = 0.6, col = rgb(0.8, 0.2, 0.2, 0.5))
abline(v = hvg_retain_pcs, lty = 2, col = "red")
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "hvg_ppk_umap.rds")
if (file.exists(rds)) {
  umap <- readRDS(rds)
} else {
  set.seed(1)
  umap <- umap(pca[ , 1:hvg_retain_pcs], min_dist = 0.2)
  saveRDS(umap, rds)}
```

```{r}
reducedDims(sce) <- SimpleList(PCA = pca, UMAP = umap)
add_df <- data.frame(reducedDim(sce, "UMAP"))
names(add_df) <- paste0("sub_hvg_ppk_umap", seq(ncol(add_df)))
colData(sce_hvg) <- cbind(
  colData(sce_hvg), sub_hvg_ppk_umap1 = add_df$sub_hvg_ppk_umap1,
  sub_hvg_ppk_umap2 = add_df$sub_hvg_ppk_umap2)
colData(sce) <- cbind(
  colData(sce), sub_hvg_ppk_umap1 = add_df$sub_hvg_ppk_umap1,
  sub_hvg_ppk_umap2 = add_df$sub_hvg_ppk_umap2)
gg_df <- data.frame(gg_df, add_df)
rm(pca, umap, add_df)

# Setup for automatic placement of cluster labels.
label_df2 <- gg_df %>%
  group_by(seurat_clusters_name) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "seurat_clusters_name", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Seurat Clusters Subset HVG PPK UMAP")
dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "cell_id_stem", type = "other") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Sample Distribution Subset HVG PPK UMAP")
dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "log10_total_features_by_counts", type = "cont") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("log10(Total Features) Subset HVG PPK UMAP")
```

We see that subsetting to HVGs further highlights an intertwined relationship between the EX/IN neurons.

### Clustering

We compute new clusters using the new `sce_hvg`.

#### Optimal K

As before, we first try estimating the optimal $k$ for our subset.

```{r}
rowData(sce)$feature_symbol = rowData(sce)$external_gene_name
rowData(sce_hvg)$feature_symbol = rowData(sce_hvg)$external_gene_name
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "sc3_estimate_k.rds")
if (file.exists(rds)) {
  sc3_estimate_k <- readRDS(rds)
} else {
  set.seed(1)
  sc3_estimate_k <- sc3_estimate_k(sce)
  saveRDS(sc3_estimate_k, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "hvg_sc3_estimate_k.rds")
if (file.exists(rds)) {
  hvg_sc3_estimate_k <- readRDS(rds)
} else {
  set.seed(1)
  hvg_sc3_estimate_k <- sc3_estimate_k(sce_hvg)
  saveRDS(hvg_sc3_estimate_k, rds)}
```

```{r}
sce <- sc3_estimate_k
rm(sc3_estimate_k)
metadata(sce)$sc3$k_estimation

sce_hvg <- hvg_sc3_estimate_k
rm(hvg_sc3_estimate_k)
metadata(sce_hvg)$sc3$k_estimation
```

Although more reasonable than before, the values still seem large considering the UMAP results.
To get another estimate we go through the Seurat clustering pipeline.

#### Seurat

As our cell count is much closer to the ~3,000 cells that Seurat gives a `resolution` recommendation of 0.4-1.2 for, we run it over that range, incrementing by 0.1 each time.

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat0.4.rds")
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {
  seurat <- as.Seurat(sce, counts = "counts", data = NULL)
  seurat <- NormalizeData(seurat)
  seurat <- FindVariableFeatures(seurat, nfeatures = dim(sce)[1])
  seurat <- ScaleData(seurat, features = rownames(seurat))
  seurat <- FindNeighbors(seurat, reduction = "PCA", dims = 1:retain_pcs)
  seurat <- FindClusters(seurat, resolution = 0.4)
  saveRDS(seurat, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat_hvg0.4.rds")
if (file.exists(rds)) {
  seurat_hvg <- readRDS(rds)
} else {
  seurat_hvg <- as.Seurat(sce_hvg, counts = "counts", data = NULL)
  seurat_hvg <- NormalizeData(seurat_hvg)
  seurat_hvg <- FindVariableFeatures(seurat_hvg, nfeatures = dim(sce_hvg)[1])
  seurat_hvg <- ScaleData(seurat_hvg, features = rownames(seurat_hvg))
  seurat_hvg <- FindNeighbors(seurat_hvg, reduction = "PCA", dims = 1:hvg_retain_pcs)
  seurat_hvg <- FindClusters(seurat_hvg, resolution = 0.4)
  saveRDS(seurat_hvg, rds)}
```

```{r}
gg_df$seurat_clusters0.4 <- seurat$seurat_clusters
gg_df$hvg_seurat_clusters0.4 <- seurat_hvg$seurat_clusters

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters0.4 = levels(gg_df$seurat_clusters0.4), label = levels(gg_df$seurat_clusters0.4))
label_df2 <- gg_df %>%
  group_by(seurat_clusters0.4) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2", col = "seurat_clusters0.4", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Subset Seurat 0.4 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters0.4 = levels(gg_df$hvg_seurat_clusters0.4),
  label = levels(gg_df$hvg_seurat_clusters0.4))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters0.4) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2",
  col = "hvg_seurat_clusters0.4", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 0.4 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters0.4 = levels(gg_df$seurat_clusters0.4), label = levels(gg_df$seurat_clusters0.4))
label_df2 <- gg_df %>%
  group_by(seurat_clusters0.4) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "seurat_clusters0.4", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset Seurat 0.4 Resolution Clusters Subset HVG PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters0.4 = levels(gg_df$hvg_seurat_clusters0.4),
  label = levels(gg_df$hvg_seurat_clusters0.4))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters0.4) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "hvg_seurat_clusters0.4", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 0.4 Resolution Clusters Subset HVG PPK UMAP")
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat0.5.rds")
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {
  seurat <- as.Seurat(sce, counts = "counts", data = NULL)
  seurat <- NormalizeData(seurat)
  seurat <- FindVariableFeatures(seurat, nfeatures = dim(sce)[1])
  seurat <- ScaleData(seurat, features = rownames(seurat))
  seurat <- FindNeighbors(seurat, reduction = "PCA", dims = 1:retain_pcs)
  seurat <- FindClusters(seurat, resolution = 0.5)
  saveRDS(seurat, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat_hvg0.5.rds")
if (file.exists(rds)) {
  seurat_hvg <- readRDS(rds)
} else {
  seurat_hvg <- as.Seurat(sce_hvg, counts = "counts", data = NULL)
  seurat_hvg <- NormalizeData(seurat_hvg)
  seurat_hvg <- FindVariableFeatures(seurat_hvg, nfeatures = dim(sce_hvg)[1])
  seurat_hvg <- ScaleData(seurat_hvg, features = rownames(seurat_hvg))
  seurat_hvg <- FindNeighbors(seurat_hvg, reduction = "PCA", dims = 1:hvg_retain_pcs)
  seurat_hvg <- FindClusters(seurat_hvg, resolution = 0.5)
  saveRDS(seurat_hvg, rds)}
```

```{r}
gg_df$seurat_clusters0.5 <- seurat$seurat_clusters
gg_df$hvg_seurat_clusters0.5 <- seurat_hvg$seurat_clusters

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters0.5 = levels(gg_df$seurat_clusters0.5), label = levels(gg_df$seurat_clusters0.5))
label_df2 <- gg_df %>%
  group_by(seurat_clusters0.5) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2", col = "seurat_clusters0.5", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Subset Seurat 0.5 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters0.5 = levels(gg_df$hvg_seurat_clusters0.5),
  label = levels(gg_df$hvg_seurat_clusters0.5))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters0.5) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2",
  col = "hvg_seurat_clusters0.5", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 0.5 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters0.5 = levels(gg_df$seurat_clusters0.5), label = levels(gg_df$seurat_clusters0.5))
label_df2 <- gg_df %>%
  group_by(seurat_clusters0.5) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "seurat_clusters0.5", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset Seurat 0.5 Resolution Clusters Subset HVG PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters0.5 = levels(gg_df$hvg_seurat_clusters0.5),
  label = levels(gg_df$hvg_seurat_clusters0.5))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters0.5) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "hvg_seurat_clusters0.5", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 0.5 Resolution Clusters Subset HVG PPK UMAP")
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat0.6.rds")
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {
  seurat <- as.Seurat(sce, counts = "counts", data = NULL)
  seurat <- NormalizeData(seurat)
  seurat <- FindVariableFeatures(seurat, nfeatures = dim(sce)[1])
  seurat <- ScaleData(seurat, features = rownames(seurat))
  seurat <- FindNeighbors(seurat, reduction = "PCA", dims = 1:retain_pcs)
  seurat <- FindClusters(seurat, resolution = 0.6)
  saveRDS(seurat, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat_hvg0.6.rds")
if (file.exists(rds)) {
  seurat_hvg <- readRDS(rds)
} else {
  seurat_hvg <- as.Seurat(sce_hvg, counts = "counts", data = NULL)
  seurat_hvg <- NormalizeData(seurat_hvg)
  seurat_hvg <- FindVariableFeatures(seurat_hvg, nfeatures = dim(sce_hvg)[1])
  seurat_hvg <- ScaleData(seurat_hvg, features = rownames(seurat_hvg))
  seurat_hvg <- FindNeighbors(seurat_hvg, reduction = "PCA", dims = 1:hvg_retain_pcs)
  seurat_hvg <- FindClusters(seurat_hvg, resolution = 0.6)
  saveRDS(seurat_hvg, rds)}
```

```{r}
gg_df$seurat_clusters0.6 <- seurat$seurat_clusters
gg_df$hvg_seurat_clusters0.6 <- seurat_hvg$seurat_clusters

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters0.6 = levels(gg_df$seurat_clusters0.6), label = levels(gg_df$seurat_clusters0.6))
label_df2 <- gg_df %>%
  group_by(seurat_clusters0.6) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2", col = "seurat_clusters0.6", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Subset Seurat 0.6 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters0.6 = levels(gg_df$hvg_seurat_clusters0.6),
  label = levels(gg_df$hvg_seurat_clusters0.6))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters0.6) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2",
  col = "hvg_seurat_clusters0.6", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 0.6 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters0.6 = levels(gg_df$seurat_clusters0.6), label = levels(gg_df$seurat_clusters0.6))
label_df2 <- gg_df %>%
  group_by(seurat_clusters0.6) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "seurat_clusters0.6", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset Seurat 0.6 Resolution Clusters Subset HVG PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters0.6 = levels(gg_df$hvg_seurat_clusters0.6),
  label = levels(gg_df$hvg_seurat_clusters0.6))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters0.6) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "hvg_seurat_clusters0.6", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 0.6 Resolution Clusters Subset HVG PPK UMAP")
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat0.7.rds")
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {
  seurat <- as.Seurat(sce, counts = "counts", data = NULL)
  seurat <- NormalizeData(seurat)
  seurat <- FindVariableFeatures(seurat, nfeatures = dim(sce)[1])
  seurat <- ScaleData(seurat, features = rownames(seurat))
  seurat <- FindNeighbors(seurat, reduction = "PCA", dims = 1:retain_pcs)
  seurat <- FindClusters(seurat, resolution = 0.7)
  saveRDS(seurat, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat_hvg0.7.rds")
if (file.exists(rds)) {
  seurat_hvg <- readRDS(rds)
} else {
  seurat_hvg <- as.Seurat(sce_hvg, counts = "counts", data = NULL)
  seurat_hvg <- NormalizeData(seurat_hvg)
  seurat_hvg <- FindVariableFeatures(seurat_hvg, nfeatures = dim(sce_hvg)[1])
  seurat_hvg <- ScaleData(seurat_hvg, features = rownames(seurat_hvg))
  seurat_hvg <- FindNeighbors(seurat_hvg, reduction = "PCA", dims = 1:hvg_retain_pcs)
  seurat_hvg <- FindClusters(seurat_hvg, resolution = 0.7)
  saveRDS(seurat_hvg, rds)}
```

```{r}
gg_df$seurat_clusters0.7 <- seurat$seurat_clusters
gg_df$hvg_seurat_clusters0.7 <- seurat_hvg$seurat_clusters
rm(seurat, seurat_hvg)

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters0.7 = levels(gg_df$seurat_clusters0.7), label = levels(gg_df$seurat_clusters0.7))
label_df2 <- gg_df %>%
  group_by(seurat_clusters0.7) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2", col = "seurat_clusters0.7", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Subset Seurat 0.7 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters0.7 = levels(gg_df$hvg_seurat_clusters0.7),
  label = levels(gg_df$hvg_seurat_clusters0.7))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters0.7) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2",
  col = "hvg_seurat_clusters0.7", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 0.7 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters0.7 = levels(gg_df$seurat_clusters0.7), label = levels(gg_df$seurat_clusters0.7))
label_df2 <- gg_df %>%
  group_by(seurat_clusters0.7) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "seurat_clusters0.7", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset Seurat 0.7 Resolution Clusters Subset HVG PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters0.7 = levels(gg_df$hvg_seurat_clusters0.7),
  label = levels(gg_df$hvg_seurat_clusters0.7))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters0.7) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "hvg_seurat_clusters0.7", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 0.7 Resolution Clusters Subset HVG PPK UMAP")
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat0.8.rds")
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {
  seurat <- as.Seurat(sce, counts = "counts", data = NULL)
  seurat <- NormalizeData(seurat)
  seurat <- FindVariableFeatures(seurat, nfeatures = dim(sce)[1])
  seurat <- ScaleData(seurat, features = rownames(seurat))
  seurat <- FindNeighbors(seurat, reduction = "PCA", dims = 1:retain_pcs)
  seurat <- FindClusters(seurat, resolution = 0.8)
  saveRDS(seurat, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat_hvg0.8.rds")
if (file.exists(rds)) {
  seurat_hvg <- readRDS(rds)
} else {
  seurat_hvg <- as.Seurat(sce_hvg, counts = "counts", data = NULL)
  seurat_hvg <- NormalizeData(seurat_hvg)
  seurat_hvg <- FindVariableFeatures(seurat_hvg, nfeatures = dim(sce_hvg)[1])
  seurat_hvg <- ScaleData(seurat_hvg, features = rownames(seurat_hvg))
  seurat_hvg <- FindNeighbors(seurat_hvg, reduction = "PCA", dims = 1:hvg_retain_pcs)
  seurat_hvg <- FindClusters(seurat_hvg, resolution = 0.8)
  saveRDS(seurat_hvg, rds)}
```

```{r}
gg_df$seurat_clusters0.8 <- seurat$seurat_clusters
gg_df$hvg_seurat_clusters0.8 <- seurat_hvg$seurat_clusters
rm(seurat, seurat_hvg)

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters0.8 = levels(gg_df$seurat_clusters0.8), label = levels(gg_df$seurat_clusters0.8))
label_df2 <- gg_df %>%
  group_by(seurat_clusters0.8) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2", col = "seurat_clusters0.8", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Subset Seurat 0.8 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters0.8 = levels(gg_df$hvg_seurat_clusters0.8),
  label = levels(gg_df$hvg_seurat_clusters0.8))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters0.8) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2",
  col = "hvg_seurat_clusters0.8", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 0.8 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters0.8 = levels(gg_df$seurat_clusters0.8), label = levels(gg_df$seurat_clusters0.8))
label_df2 <- gg_df %>%
  group_by(seurat_clusters0.8) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "seurat_clusters0.8", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset Seurat 0.8 Resolution Clusters 0.8 Subset HVG PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters0.8 = levels(gg_df$hvg_seurat_clusters0.8),
  label = levels(gg_df$hvg_seurat_clusters0.8))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters0.8) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "hvg_seurat_clusters0.8", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 0.8 Resolution Clusters Subset HVG PPK UMAP")
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat0.9.rds")
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {
  seurat <- as.Seurat(sce, counts = "counts", data = NULL)
  seurat <- NormalizeData(seurat)
  seurat <- FindVariableFeatures(seurat, nfeatures = dim(sce)[1])
  seurat <- ScaleData(seurat, features = rownames(seurat))
  seurat <- FindNeighbors(seurat, reduction = "PCA", dims = 1:retain_pcs)
  seurat <- FindClusters(seurat, resolution = 0.9)
  saveRDS(seurat, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat_hvg0.9.rds")
if (file.exists(rds)) {
  seurat_hvg <- readRDS(rds)
} else {
  seurat_hvg <- as.Seurat(sce_hvg, counts = "counts", data = NULL)
  seurat_hvg <- NormalizeData(seurat_hvg)
  seurat_hvg <- FindVariableFeatures(seurat_hvg, nfeatures = dim(sce_hvg)[1])
  seurat_hvg <- ScaleData(seurat_hvg, features = rownames(seurat_hvg))
  seurat_hvg <- FindNeighbors(seurat_hvg, reduction = "PCA", dims = 1:hvg_retain_pcs)
  seurat_hvg <- FindClusters(seurat_hvg, resolution = 0.9)
  saveRDS(seurat_hvg, rds)}
```

```{r}
gg_df$seurat_clusters0.9 <- seurat$seurat_clusters
gg_df$hvg_seurat_clusters0.9 <- seurat_hvg$seurat_clusters
rm(seurat, seurat_hvg)

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters0.9 = levels(gg_df$seurat_clusters0.9), label = levels(gg_df$seurat_clusters0.9))
label_df2 <- gg_df %>%
  group_by(seurat_clusters0.9) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2", col = "seurat_clusters0.9", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Subset Seurat 0.9 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters0.9 = levels(gg_df$hvg_seurat_clusters0.9),
  label = levels(gg_df$hvg_seurat_clusters0.9))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters0.9) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2",
  col = "hvg_seurat_clusters0.9", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Subset HVG Seurat 0.9 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters0.9 = levels(gg_df$seurat_clusters0.9), label = levels(gg_df$seurat_clusters0.9))
label_df2 <- gg_df %>%
  group_by(seurat_clusters0.9) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "seurat_clusters0.9", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset Seurat 0.9 Resolution Clusters Subset HVG PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters0.9 = levels(gg_df$hvg_seurat_clusters0.9),
  label = levels(gg_df$hvg_seurat_clusters0.9))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters0.9) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "hvg_seurat_clusters0.9", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 0.9 Resolution Clusters Subset HVG PPK UMAP")
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat1.0.rds")
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {
  seurat <- as.Seurat(sce, counts = "counts", data = NULL)
  seurat <- NormalizeData(seurat)
  seurat <- FindVariableFeatures(seurat, nfeatures = dim(sce)[1])
  seurat <- ScaleData(seurat, features = rownames(seurat))
  seurat <- FindNeighbors(seurat, reduction = "PCA", dims = 1:retain_pcs)
  seurat <- FindClusters(seurat, resolution = 1.0)
  saveRDS(seurat, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat_hvg1.0.rds")
if (file.exists(rds)) {
  seurat_hvg <- readRDS(rds)
} else {
  seurat_hvg <- as.Seurat(sce_hvg, counts = "counts", data = NULL)
  seurat_hvg <- NormalizeData(seurat_hvg)
  seurat_hvg <- FindVariableFeatures(seurat_hvg, nfeatures = dim(sce_hvg)[1])
  seurat_hvg <- ScaleData(seurat_hvg, features = rownames(seurat_hvg))
  seurat_hvg <- FindNeighbors(seurat_hvg, reduction = "PCA", dims = 1:hvg_retain_pcs)
  seurat_hvg <- FindClusters(seurat_hvg, resolution = 1.0)
  saveRDS(seurat_hvg, rds)}
```

```{r}
gg_df$seurat_clusters1.0 <- seurat$seurat_clusters
gg_df$hvg_seurat_clusters1.0 <- seurat_hvg$seurat_clusters
rm(seurat, seurat_hvg)

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters1.0 = levels(gg_df$seurat_clusters1.0), label = levels(gg_df$seurat_clusters1.0))
label_df2 <- gg_df %>%
  group_by(seurat_clusters1.0) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2", col = "seurat_clusters1.0", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Subset Seurat 1.0 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters1.0 = levels(gg_df$hvg_seurat_clusters1.0),
  label = levels(gg_df$hvg_seurat_clusters1.0))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters1.0) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2",
  col = "hvg_seurat_clusters1.0", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 1.0 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters1.0 = levels(gg_df$seurat_clusters1.0), label = levels(gg_df$seurat_clusters1.0))
label_df2 <- gg_df %>%
  group_by(seurat_clusters1.0) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "seurat_clusters1.0", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset Seurat 1.0 Resolution Clusters Subset HVG PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters1.0 = levels(gg_df$hvg_seurat_clusters1.0),
  label = levels(gg_df$hvg_seurat_clusters1.0))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters1.0) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "hvg_seurat_clusters1.0", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 1.0 Resolution Clusters Subset HVG PPK UMAP")
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat1.1.rds")
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {
  seurat <- as.Seurat(sce, counts = "counts", data = NULL)
  seurat <- NormalizeData(seurat)
  seurat <- FindVariableFeatures(seurat, nfeatures = dim(sce)[1])
  seurat <- ScaleData(seurat, features = rownames(seurat))
  seurat <- FindNeighbors(seurat, reduction = "PCA", dims = 1:retain_pcs)
  seurat <- FindClusters(seurat, resolution = 1.1)
  saveRDS(seurat, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat_hvg1.1.rds")
if (file.exists(rds)) {
  seurat_hvg <- readRDS(rds)
} else {
  seurat_hvg <- as.Seurat(sce_hvg, counts = "counts", data = NULL)
  seurat_hvg <- NormalizeData(seurat_hvg)
  seurat_hvg <- FindVariableFeatures(seurat_hvg, nfeatures = dim(sce_hvg)[1])
  seurat_hvg <- ScaleData(seurat_hvg, features = rownames(seurat_hvg))
  seurat_hvg <- FindNeighbors(seurat_hvg, reduction = "PCA", dims = 1:hvg_retain_pcs)
  seurat_hvg <- FindClusters(seurat_hvg, resolution = 1.1)
  saveRDS(seurat_hvg, rds)}
```

```{r}
gg_df$seurat_clusters1.1 <- seurat$seurat_clusters
gg_df$hvg_seurat_clusters1.1 <- seurat_hvg$seurat_clusters
rm(seurat, seurat_hvg)

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters1.1 = levels(gg_df$seurat_clusters1.1), label = levels(gg_df$seurat_clusters1.1))
label_df2 <- gg_df %>%
  group_by(seurat_clusters1.1) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2", col = "seurat_clusters1.1", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Subset Seurat 1.1 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters1.1 = levels(gg_df$hvg_seurat_clusters1.1),
  label = levels(gg_df$hvg_seurat_clusters1.1))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters1.1) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2",
  col = "hvg_seurat_clusters1.1", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 1.1 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters1.1 = levels(gg_df$seurat_clusters1.1), label = levels(gg_df$seurat_clusters1.1))
label_df2 <- gg_df %>%
  group_by(seurat_clusters1.1) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "seurat_clusters1.1", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset Seurat 1.1 Resolution Clusters Subset HVG PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters1.1 = levels(gg_df$hvg_seurat_clusters1.1),
  label = levels(gg_df$hvg_seurat_clusters1.1))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters1.1) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "hvg_seurat_clusters1.1", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 1.1 Resolution Clusters Subset HVG PPK UMAP")
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat1.2.rds")
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {
  seurat <- as.Seurat(sce, counts = "counts", data = NULL)
  seurat <- NormalizeData(seurat)
  seurat <- FindVariableFeatures(seurat, nfeatures = dim(sce)[1])
  seurat <- ScaleData(seurat, features = rownames(seurat))
  seurat <- FindNeighbors(seurat, reduction = "PCA", dims = 1:retain_pcs)
  seurat <- FindClusters(seurat, resolution = 1.2)
  saveRDS(seurat, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat_hvg1.2.rds")
if (file.exists(rds)) {
  seurat_hvg <- readRDS(rds)
} else {
  seurat_hvg <- as.Seurat(sce_hvg, counts = "counts", data = NULL)
  seurat_hvg <- NormalizeData(seurat_hvg)
  seurat_hvg <- FindVariableFeatures(seurat_hvg, nfeatures = dim(sce_hvg)[1])
  seurat_hvg <- ScaleData(seurat_hvg, features = rownames(seurat_hvg))
  seurat_hvg <- FindNeighbors(seurat_hvg, reduction = "PCA", dims = 1:hvg_retain_pcs)
  seurat_hvg <- FindClusters(seurat_hvg, resolution = 1.2)
  saveRDS(seurat_hvg, rds)}
```

```{r}
gg_df$seurat_clusters1.2 <- seurat$seurat_clusters
gg_df$hvg_seurat_clusters1.2 <- seurat_hvg$seurat_clusters

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters1.2 = levels(gg_df$seurat_clusters1.2), label = levels(gg_df$seurat_clusters1.2))
label_df2 <- gg_df %>%
  group_by(seurat_clusters1.2) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2", col = "seurat_clusters1.2", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Subset Seurat 1.2 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters1.2 = levels(gg_df$hvg_seurat_clusters1.2),
  label = levels(gg_df$hvg_seurat_clusters1.2))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters1.2) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2",
  col = "hvg_seurat_clusters1.2", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 1.2 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters1.2 = levels(gg_df$seurat_clusters1.2), label = levels(gg_df$seurat_clusters1.2))
label_df2 <- gg_df %>%
  group_by(seurat_clusters1.2) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "seurat_clusters1.2", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset Seurat 1.2 Resolution Clusters Subset HVG PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters1.2 = levels(gg_df$hvg_seurat_clusters1.2),
  label = levels(gg_df$hvg_seurat_clusters1.2))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters1.2) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "hvg_seurat_clusters1.2", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 1.2 Resolution Clusters Subset HVG PPK UMAP")
```

The results appear to be very reasonable, though the putative intermediate cluster in the non-HVG UMAPs could not be classified as a distinct cluster, so we conclude for now that they seperate out due to technical noise.
We also note that the minimum clusters produced are 5.
It would also be useful to examine 2 and 3 clusters on this particular subset, therefore, we lower the `resolution` until these are produced.

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat0.14.rds")
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {
  seurat <- as.Seurat(sce, counts = "counts", data = NULL)
  seurat <- NormalizeData(seurat)
  seurat <- FindVariableFeatures(seurat, nfeatures = dim(sce)[1])
  seurat <- ScaleData(seurat, features = rownames(seurat))
  seurat <- FindNeighbors(seurat, reduction = "PCA", dims = 1:retain_pcs)
  seurat <- FindClusters(seurat, resolution = 0.14)
  saveRDS(seurat, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat_hvg0.19.rds")
if (file.exists(rds)) {
  seurat_hvg <- readRDS(rds)
} else {
  seurat_hvg <- as.Seurat(sce_hvg, counts = "counts", data = NULL)
  seurat_hvg <- NormalizeData(seurat_hvg)
  seurat_hvg <- FindVariableFeatures(seurat_hvg, nfeatures = dim(sce_hvg)[1])
  seurat_hvg <- ScaleData(seurat_hvg, features = rownames(seurat_hvg))
  seurat_hvg <- FindNeighbors(seurat_hvg, reduction = "PCA", dims = 1:hvg_retain_pcs)
  seurat_hvg <- FindClusters(seurat_hvg, resolution = 0.19)
  saveRDS(seurat_hvg, rds)}
```

```{r}
gg_df$seurat_clusters0.14 <- seurat$seurat_clusters
gg_df$hvg_seurat_clusters0.19 <- seurat_hvg$seurat_clusters

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters0.14 = levels(gg_df$seurat_clusters0.14), label = levels(gg_df$seurat_clusters0.14))
label_df2 <- gg_df %>%
  group_by(seurat_clusters0.14) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2", col = "seurat_clusters0.14", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Subset Seurat 0.14 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters0.19 = levels(gg_df$hvg_seurat_clusters0.19),
  label = levels(gg_df$hvg_seurat_clusters0.19))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters0.19) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2",
  col = "hvg_seurat_clusters0.19", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 0.19 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters0.14 = levels(gg_df$seurat_clusters0.14), label = levels(gg_df$seurat_clusters0.14))
label_df2 <- gg_df %>%
  group_by(seurat_clusters0.14) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "seurat_clusters0.14", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset Seurat 0.14 Resolution Clusters Subset HVG PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters0.19 = levels(gg_df$hvg_seurat_clusters0.19),
  label = levels(gg_df$hvg_seurat_clusters0.19))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters0.19) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "hvg_seurat_clusters0.19", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 0.19 Resolution Clusters Subset HVG PPK UMAP")
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat0.08.rds")
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {
  seurat <- as.Seurat(sce, counts = "counts", data = NULL)
  seurat <- NormalizeData(seurat)
  seurat <- FindVariableFeatures(seurat, nfeatures = dim(sce)[1])
  seurat <- ScaleData(seurat, features = rownames(seurat))
  seurat <- FindNeighbors(seurat, reduction = "PCA", dims = 1:retain_pcs)
  seurat <- FindClusters(seurat, resolution = 0.08)
  saveRDS(seurat, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "seurat_hvg0.15.rds")
if (file.exists(rds)) {
  seurat_hvg <- readRDS(rds)
} else {
  seurat_hvg <- as.Seurat(sce_hvg, counts = "counts", data = NULL)
  seurat_hvg <- NormalizeData(seurat_hvg)
  seurat_hvg <- FindVariableFeatures(seurat_hvg, nfeatures = dim(sce_hvg)[1])
  seurat_hvg <- ScaleData(seurat_hvg, features = rownames(seurat_hvg))
  seurat_hvg <- FindNeighbors(seurat_hvg, reduction = "PCA", dims = 1:hvg_retain_pcs)
  seurat_hvg <- FindClusters(seurat_hvg, resolution = 0.15)
  saveRDS(seurat_hvg, rds)}
```

```{r}
gg_df$seurat_clusters0.08 <- seurat$seurat_clusters
gg_df$hvg_seurat_clusters0.15 <- seurat_hvg$seurat_clusters

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters0.08 = levels(gg_df$seurat_clusters0.08), label = levels(gg_df$seurat_clusters0.08))
label_df2 <- gg_df %>%
  group_by(seurat_clusters0.08) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2", col = "seurat_clusters0.08", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("Subset Seurat 0.08 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters0.15 = levels(gg_df$hvg_seurat_clusters0.15),
  label = levels(gg_df$hvg_seurat_clusters0.15))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters0.15) %>%
  summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2",
  col = "hvg_seurat_clusters0.15", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 0.15 Resolution Clusters Subset PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters0.08 = levels(gg_df$seurat_clusters0.08), label = levels(gg_df$seurat_clusters0.08))
label_df2 <- gg_df %>%
  group_by(seurat_clusters0.08) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "seurat_clusters0.08", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset Seurat 0.08 Resolution Clusters Subset HVG PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  hvg_seurat_clusters0.15 = levels(gg_df$hvg_seurat_clusters0.15),
  label = levels(gg_df$hvg_seurat_clusters0.15))
label_df2 <- gg_df %>%
  group_by(hvg_seurat_clusters0.15) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "hvg_seurat_clusters0.15", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset HVG Seurat 0.15 Resolution Clusters Subset HVG PPK UMAP")
```

Interestingly, we see that though the outlier cluster seems to have characteristics closer to EX neurons, it is ambiguous enough that when using all genes in Seurat, it was clustered into the IN group.
We note that this cluster was originally identified as an ASC cell in @habib_massively_2017, though it was also not separated out in their tSNE.
We continue to investigate this intriguing cluster while moving two-way analysis between EX/IN neurons to a subset excluding this group.

#### SC3

We saw that Seurat produced a maximum of 10 clusters when iterating through a reasonable range of `resolution` values.
Therefore, we decide to compute 2-12 clusters with SC3, giving some margin of error to account for algorithmic differences.

```{r}
num_clust <- c(2:12)
counts(sce) <- as.matrix(counts(sce))
logcounts(sce) <- as.matrix(logcounts(sce))
counts(sce_hvg) <- as.matrix(counts(sce_hvg))
logcounts(sce_hvg) <- as.matrix(logcounts(sce_hvg))
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "sc3.rds")
if (file.exists(rds)) {
  sc3 <- readRDS(rds)
} else {
  sc3 <- sc3(sce, ks = num_clust)
  saveRDS(sc3, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "hvg_sc3.rds")
if (file.exists(rds)) {
  hvg_sc3 <- readRDS(rds)
} else {
  hvg_sc3 <- sc3(sce_hvg, ks = num_clust)
  saveRDS(hvg_sc3, rds)}
```

```{r}
sce <- sc3
sce_hvg <- hvg_sc3
hvg_gg_df <- gg_df
rm(sc3, hvg_sc3)

for (i in num_clust) {
  sc3_label <- paste0("sc3_", i, "_clusters")
  gg_df[[sc3_label]] <- as.factor(colData(sce)[ , sc3_label])}
clusts <- c(paste0("sc3_", num_clust, "_clusters"))
clusts_nice <- c(paste0("SC3 ", num_clust, " Clusters"))

for (i in 1:length(clusts)) {
  # Setup for automatic placement of cluster labels.
  label_df <- data.frame(tmp = levels(gg_df[[clusts[i]]]), label = levels(gg_df[[clusts[i]]]))
  names(label_df)[names(label_df) == "tmp"] <- clusts[i]
  label_df2 <- gg_df %>%
    group_by_at(clusts[i]) %>%
    summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
    left_join(label_df)

  print(dim_red_plot(
    data = gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2", col = clusts[i], type = "cat") +
      xlab("UMAP 1") + ylab("UMAP 2") +
      ggtitle(paste("Subset", clusts_nice[i], "Subset PPK UMAP")))

  # Setup for automatic placement of cluster labels.
  label_df2 <- gg_df %>%
    group_by_at(clusts[i]) %>%
    summarize(
      sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1),
      sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
    left_join(label_df)

  print(dim_red_plot(
    data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
    col = clusts[i], type = "cat") +
      xlab("UMAP 1") + ylab("UMAP 2") +
      ggtitle(paste("Subset", clusts_nice[i], "Subset HVG PPK UMAP")))}

for (i in num_clust) {
  sc3_label <- paste0("sc3_", i, "_clusters")
  hvg_gg_df[[sc3_label]] <- factor(colData(sce_hvg)[ , sc3_label])}
clusts <- c(paste0("sc3_", num_clust, "_clusters"))
clusts_nice <- c(paste0("SC3 ", num_clust, " Clusters"))

for (i in 1:length(clusts)) {
  # Setup for automatic placement of cluster labels.
  label_df <- data.frame(tmp = levels(hvg_gg_df[[clusts[i]]]), label = levels(hvg_gg_df[[clusts[i]]]))
  names(label_df)[names(label_df) == "tmp"] <- clusts[i]
  label_df2 <- hvg_gg_df %>%
    group_by_at(clusts[i]) %>%
    summarize(sub_ppk_umap1 = median(sub_ppk_umap1), sub_ppk_umap2 = median(sub_ppk_umap2)) %>%
    left_join(label_df)

  print(dim_red_plot(
    data = hvg_gg_df, x = "sub_ppk_umap1", y = "sub_ppk_umap2", col = clusts[i], type = "cat") +
      xlab("UMAP 1") + ylab("UMAP 2") +
      ggtitle(paste("Subset HVG", clusts_nice[i], "Subset PPK UMAP")))

  # Setup for automatic placement of cluster labels.
  label_df2 <- hvg_gg_df %>%
    group_by_at(clusts[i]) %>%
    summarize(
      sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1),
      sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
    left_join(label_df)

  print(dim_red_plot(
    data = hvg_gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
    col = clusts[i], type = "cat") +
      xlab("UMAP 1") + ylab("UMAP 2") +
      ggtitle(paste("Subset HVG", clusts_nice[i], "Subset HVG PPK UMAP")))}
```

We see that SC3 generally produces a less clear separation of cells.
It is suspected that this is due to the outlier cluster or selection of parameters.
Though the `gene_filter` was toggled, without appreciable difference, other parameters were not tested due to the long runtime of SC3.
We also note that SC3 did not identify the outlier cluster as a distinct population at any point and generally, cells in the cluster were partitioned into EX groups.
The interpretation of these findings so far are unclear.

#### Labeling

The most usable figures that we apply labels to are:

- Subset Seurat 0.14 Resolution Clusters
- Subset Seurat 0.6 Resolution Clusters
- Subset Seurat 1.2 Resolution ClusersP
- Subset SC3 3 Clusters
- Subset HVG SC3 6 Clusters Subset
- Subset SC3 9 Clusters Subset

```{r}
gg_df$seurat_clusters0.14_name <- NA
seurat_clusters0.14 <- as.integer(gg_df$seurat_clusters0.14)
for (i in 1:length(seurat_clusters0.14)) {
  if (seurat_clusters0.14[i] == 1) {
    gg_df[i, ncol(gg_df)] <- "exPFC"
  } else if (seurat_clusters0.14[i] == 2) {
    gg_df[i, ncol(gg_df)] <- "GABA"
  } else if (seurat_clusters0.14[i] == 3) {
    gg_df[i, ncol(gg_df)] <- "New1"}}

gg_df$seurat_clusters0.6_name <- NA
seurat_clusters0.6 <- as.integer(gg_df$seurat_clusters0.6)
for (i in 1:length(seurat_clusters0.6)) {
  if (seurat_clusters0.6[i] == 1) {
    gg_df[i, ncol(gg_df)] <- "exPFC1.2"
  } else if (seurat_clusters0.6[i] == 2) {
    gg_df[i, ncol(gg_df)] <- "exPFC1.1"
  } else if (seurat_clusters0.6[i] == 3) {
    gg_df[i, ncol(gg_df)] <- "GABA2.1"
  } else if (seurat_clusters0.6[i] == 4) {
    gg_df[i, ncol(gg_df)] <- "GABA1.1"
  } else if (seurat_clusters0.6[i] == 5) {
    gg_df[i, ncol(gg_df)] <- "GABA2.2"
  } else if (seurat_clusters0.6[i] == 6) {
    gg_df[i, ncol(gg_df)] <- "GABA1.2"
  } else if (seurat_clusters0.6[i] == 7) {
    gg_df[i, ncol(gg_df)] <- "New1"}}

gg_df$seurat_clusters1.2_name <- NA
seurat_clusters1.2 <- as.integer(gg_df$seurat_clusters1.2)
for (i in 1:length(seurat_clusters1.2)) {
  if (seurat_clusters1.2[i] == 1) {
    gg_df[i, ncol(gg_df)] <- "exPFC1.2"
  } else if (seurat_clusters1.2[i] == 2) {
    gg_df[i, ncol(gg_df)] <- "exPFC1.1"
  } else if (seurat_clusters1.2[i] == 3) {
    gg_df[i, ncol(gg_df)] <- "exPFC2"
  } else if (seurat_clusters1.2[i] == 4) {
    gg_df[i, ncol(gg_df)] <- "exPFC1.3"
  } else if (seurat_clusters1.2[i] == 5) {
    gg_df[i, ncol(gg_df)] <- "GABA2.1"
  } else if (seurat_clusters1.2[i] == 6) {
    gg_df[i, ncol(gg_df)] <- "GABA1.1"
  } else if (seurat_clusters1.2[i] == 7) {
    gg_df[i, ncol(gg_df)] <- "GABA2.2"
  } else if (seurat_clusters1.2[i] == 8) {
    gg_df[i, ncol(gg_df)] <- "GABA1.2"
  } else if (seurat_clusters1.2[i] == 9) {
    gg_df[i, ncol(gg_df)] <- "exPFC1.4"
  } else if (seurat_clusters1.2[i] == 10) {
    gg_df[i, ncol(gg_df)] <- "New1"}}

gg_df$sc3_3_clusters_name <- NA
sc3_3_clusters <- as.integer(gg_df$sc3_3_clusters)
for (i in 1:length(sc3_3_clusters)) {
  if (sc3_3_clusters[i] == 1) {
    gg_df[i, ncol(gg_df)] <- "exPFC"
  } else if (sc3_3_clusters[i] == 2) {
    gg_df[i, ncol(gg_df)] <- "New4"
  } else if (sc3_3_clusters[i] == 3) {
    gg_df[i, ncol(gg_df)] <- "GABA"}}

gg_df$sc3_6_clusters_name <- NA
sc3_6_clusters <- as.integer(hvg_gg_df$sc3_6_clusters)
for (i in 1:length(sc3_6_clusters)) {
  if (sc3_6_clusters[i] == 1) {
    gg_df[i, ncol(gg_df)] <- "exPFC1.2"
  } else if (sc3_6_clusters[i] == 2) {
    gg_df[i, ncol(gg_df)] <- "exPFC1.1"
  } else if (sc3_6_clusters[i] == 3) {
    gg_df[i, ncol(gg_df)] <- "GABA2.1"
  } else if (sc3_6_clusters[i] == 4) {
    gg_df[i, ncol(gg_df)] <- "GABA1.1"
  } else if (sc3_6_clusters[i] == 5) {
    gg_df[i, ncol(gg_df)] <- "GABA2.2"
  } else if (sc3_6_clusters[i] == 6) {
    gg_df[i, ncol(gg_df)] <- "GABA1.2"}}

gg_df$sc3_9_clusters_name <- NA
sc3_9_clusters <- as.integer(gg_df$sc3_9_clusters)
for (i in 1:length(sc3_9_clusters)) {
  if (sc3_9_clusters[i] == 1) {
    gg_df[i, ncol(gg_df)] <- "exPFC2"
  } else if (sc3_9_clusters[i] == 2) {
    gg_df[i, ncol(gg_df)] <- "exPFC1.2"
  } else if (sc3_9_clusters[i] == 3) {
    gg_df[i, ncol(gg_df)] <- "exPFC1.3"
  } else if (sc3_9_clusters[i] == 4) {
    gg_df[i, ncol(gg_df)] <- "GABA2.1"
  } else if (sc3_9_clusters[i] == 5) {
    gg_df[i, ncol(gg_df)] <- "GABA1.1"
  } else if (sc3_9_clusters[i] == 6) {
    gg_df[i, ncol(gg_df)] <- "exPFC1.1"
  } else if (sc3_9_clusters[i] == 7) {
    gg_df[i, ncol(gg_df)] <- "GABA2.2"
  } else if (sc3_9_clusters[i] == 8) {
    gg_df[i, ncol(gg_df)] <- "exPFC1.4"
  } else if (sc3_9_clusters[i] == 9) {
    gg_df[i, ncol(gg_df)] <- "GABA1.2"}}

gg_df$seurat_clusters0.14_name <- factor(gg_df$seurat_clusters0.14_name)
gg_df$seurat_clusters0.6_name <- factor(gg_df$seurat_clusters0.6_name)
gg_df$seurat_clusters1.2_name <- factor(gg_df$seurat_clusters1.2_name)
gg_df$sc3_3_clusters_name <- factor(gg_df$sc3_3_clusters_name)
gg_df$sc3_6_clusters_name <- factor(gg_df$sc3_6_clusters_name)
gg_df$sc3_9_clusters_name <- factor(gg_df$sc3_9_clusters_name)
colData(sce) <- cbind(
  colData(sce),
  seurat_clusters0.14 = gg_df$seurat_clusters0.14,
  seurat_clusters0.14_name = gg_df$seurat_clusters0.14_name,
  seurat_clusters0.6 = gg_df$seurat_clusters0.6,
  seurat_clusters0.6_name = gg_df$seurat_clusters0.6_name,
  seurat_clusters1.2 = gg_df$seurat_clusters1.2,
  seurat_clusters1.2_name = gg_df$seurat_clusters1.2_name,
  sc3_3_clusters = gg_df$sc3_3_clusters,
  sc3_3_clusters_name = gg_df$sc3_3_clusters_name,
  sc3_6_clusters = hvg_gg_df$sc3_6_clusters,
  sc3_6_clusters_name = gg_df$sc3_6_clusters_name,
  sc3_9_clusters = gg_df$sc3_9_clusters,
  sc3_9_clusters_name = gg_df$sc3_9_clusters_name)
colData(sce_hvg) <- cbind(
  colData(sce_hvg),
  seurat_clusters0.14 = gg_df$seurat_clusters0.14,
  seurat_clusters0.14_name = gg_df$seurat_clusters0.14_name,
  seurat_clusters0.6 = gg_df$seurat_clusters0.6,
  seurat_clusters0.6_name = gg_df$seurat_clusters0.6_name,
  seurat_clusters1.2 = gg_df$seurat_clusters1.2,
  seurat_clusters1.2_name = gg_df$seurat_clusters1.2_name,
  sc3_3_clusters = gg_df$sc3_3_clusters,
  sc3_3_clusters_name = gg_df$sc3_3_clusters_name,
  sc3_6_clusters = hvg_gg_df$sc3_6_clusters,
  sc3_6_clusters_name = gg_df$sc3_6_clusters_name,
  sc3_9_clusters = gg_df$sc3_9_clusters,
  sc3_9_clusters_name = gg_df$sc3_9_clusters_name)

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters0.14_name = levels(gg_df$seurat_clusters0.14_name),
  label = levels(gg_df$seurat_clusters0.14_name))
label_df2 <- gg_df %>%
  group_by(seurat_clusters0.14_name) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "seurat_clusters0.14_name", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset Seurat 0.14 Resolution Clusters Subset HVG PPK UMAP")


# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters0.6_name = levels(gg_df$seurat_clusters0.6_name),
  label = levels(gg_df$seurat_clusters0.6_name))
label_df2 <- gg_df %>%
  group_by(seurat_clusters0.6_name) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "seurat_clusters0.6_name", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset Seurat 0.6 Resolution Clusters Subset HVG PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  seurat_clusters1.2_name = levels(gg_df$seurat_clusters1.2_name),
  label = levels(gg_df$seurat_clusters1.2_name))
label_df2 <- gg_df %>%
  group_by(seurat_clusters1.2_name) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "seurat_clusters1.2_name", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") +
  ggtitle("Subset Seurat 1.2 Resolution Clusters Subset HVG PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  sc3_3_clusters_name = levels(gg_df$sc3_3_clusters_name), label = levels(gg_df$sc3_3_clusters_name))
label_df2 <- gg_df %>%
  group_by(sc3_3_clusters_name) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "sc3_3_clusters_name", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("SC3 3 Clusters Subset HVG PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  sc3_6_clusters_name = levels(gg_df$sc3_6_clusters_name), label = levels(gg_df$sc3_6_clusters_name))
label_df2 <- gg_df %>%
  group_by(sc3_6_clusters_name) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "sc3_6_clusters_name", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("SC3 6 Clusters Subset HVG PPK UMAP")

# Setup for automatic placement of cluster labels.
label_df <- data.frame(
  sc3_9_clusters_name = levels(gg_df$sc3_9_clusters_name), label = levels(gg_df$sc3_9_clusters_name))
label_df2 <- gg_df %>%
  group_by(sc3_9_clusters_name) %>%
  summarize(
    sub_hvg_ppk_umap1 = median(sub_hvg_ppk_umap1), sub_hvg_ppk_umap2 = median(sub_hvg_ppk_umap2)) %>%
  left_join(label_df)

dim_red_plot(
  data = gg_df, x = "sub_hvg_ppk_umap1", y = "sub_hvg_ppk_umap2",
  col = "sc3_9_clusters_name", type = "cat") +
  xlab("UMAP 1") + ylab("UMAP 2") + ggtitle("SC3 9 Clusters Subset HVG PPK UMAP")
```

### Differential Expression

In this section, we calculate differentially expressed genes (DEGs) and identify marker genes for each cluster.

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "sc3_calc_biology.rds")
if (file.exists(rds)) {
  sc3_calc_biology <- readRDS(rds)
} else {
  set.seed(1)
  sc3_calc_biology <- sc3_calc_biology(sce, ks = c(3, 9))
  saveRDS(sc3_calc_biology, rds)}
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache2", "hvg_sc3_calc_biology.rds")
if (file.exists(rds)) {
  hvg_sc3_calc_biology <- readRDS(rds)
} else {
  set.seed(1)
  hvg_sc3_calc_biology <- sc3_calc_biology(sce_hvg, ks = 6)
  saveRDS(hvg_sc3_calc_biology, rds)}
```

```{r}
sce <- sc3_calc_biology
sce_hvg <- hvg_sc3_calc_biology
labeled_clusts <- c(3, 6, 9)
pdata <- c(
  "sc3_3_clusters_name", "sc3_6_clusters_name", "sc3_9_clusters_name",
  "seurat_clusters0.14_name", "seurat_clusters0.6_name", "seurat_clusters1.2_name")
rm(sc3_calc_biology, hvg_sc3_calc_biology)
```

#### DEG

We calculate DEGs using a non-parametric Kruskal-Wallis test.
The figure displays the top 50 DEGs where $p < 0.01$.

```{r, fig.height = 14}
sc3_plot_de_genes(sce, k = 9, show_pdata = pdata)
```

#### Marker Genes

To assign marker genes, a binary classifier is constructed based on the mean cluster expression value.
Classifier prediction is then calculated using the gene expression ranks.
Genes with $AUROC > 0.7$ and $p < 0.01$ are selected and the top 10 are visualized in the following figure.
Note that $AUROC > 0.85$ is the default, but few to no marker genes could be detected with this threshold.

```{r, fig.height = 20}
for (i in labeled_clusts) {
  if (i == 6) {
    print(sc3_plot_markers(sce_hvg, k = i, auroc = 0.7, show_pdata = pdata))
  } else {
    print(sc3_plot_markers(sce, k = i, auroc = 0.7, show_pdata = pdata))}}
```

We can also view a data table containing all the marker genes.

```{r}
markers <- as.data.frame(organise_marker_genes(sce, k = 3, p_val = 0.01, auroc = 0.7))
markers$sc3_3_clusters_name <- NA
sc3_3_clusters <- as.integer(markers$sc3_3_markers_clusts)
for (i in 1:length(sc3_3_clusters)) {
  if (sc3_3_clusters[i] == 1) {
    markers[i, ncol(markers)] <- "exPFC"
  } else if (sc3_3_clusters[i] == 2) {
    markers[i, ncol(markers)] <- "New4"
  } else if (sc3_3_clusters[i] == 3) {
    markers[i, ncol(markers)] <- "GABA"}}
datatable_custom(as.data.table(markers))

markers <- as.data.frame(organise_marker_genes(sce_hvg, k = 6, p_val = 0.01, auroc = 0.7))
markers$sc3_6_clusters_name <- NA
sc3_6_clusters <- as.integer(markers$sc3_6_markers_clusts)
for (i in 1:length(sc3_6_clusters)) {
  if (sc3_6_clusters[i] == 1) {
    markers[i, ncol(markers)] <- "exPFC1.2"
  } else if (sc3_6_clusters[i] == 2) {
    markers[i, ncol(markers)] <- "exPFC1.1"
  } else if (sc3_6_clusters[i] == 3) {
    markers[i, ncol(markers)] <- "GABA2.1"
  } else if (sc3_6_clusters[i] == 4) {
    markers[i, ncol(markers)] <- "GABA1.1"
  } else if (sc3_6_clusters[i] == 5) {
    markers[i, ncol(markers)] <- "GABA2.2"
  } else if (sc3_6_clusters[i] == 6) {
    markers[i, ncol(markers)] <- "GABA1.2"}}
datatable_custom(as.data.table(markers))

markers <- as.data.frame(organise_marker_genes(sce, k = 9, p_val = 0.01, auroc = 0.7))
markers$sc3_9_clusters_name <- NA
sc3_9_clusters <- as.integer(markers$sc3_9_markers_clusts)
for (i in 1:length(sc3_9_clusters)) {
  if (sc3_9_clusters[i] == 1) {
    markers[i, ncol(markers)] <- "exPFC2"
  } else if (sc3_9_clusters[i] == 2) {
    markers[i, ncol(markers)] <- "exPFC1.2"
  } else if (sc3_9_clusters[i] == 3) {
    markers[i, ncol(markers)] <- "exPFC1.3"
  } else if (sc3_9_clusters[i] == 4) {
    markers[i, ncol(markers)] <- "GABA2.1"
  } else if (sc3_9_clusters[i] == 5) {
    markers[i, ncol(markers)] <- "GABA1.1"
  } else if (sc3_9_clusters[i] == 6) {
    markers[i, ncol(markers)] <- "exPFC1.1"
  } else if (sc3_9_clusters[i] == 7) {
    markers[i, ncol(markers)] <- "GABA2.2"
  } else if (sc3_9_clusters[i] == 8) {
    markers[i, ncol(markers)] <- "exPFC1.4"
  } else if (sc3_9_clusters[i] == 9) {
    markers[i, ncol(markers)] <- "GABA1.2"}}
datatable_custom(as.data.table(markers))
```

# References

This is the concluding section of the document.
Here we write relevant results to disk, output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
